library(shiny)
library(shinyWidgets)
library(ggrastr)
library(plyr)
library(ggplot2)
library(ggsci)
library(umap)
library(tidyr)
library(dbscan)
library(readr)
library(dplyr)
library(stringr)
library(ComplexHeatmap)
library(RColorBrewer)
library(gt)
library(gtsummary)
library(flextable)
library(Rediscover)
library(survival)
library(gridExtra)
library(survminer)
library(tranSurv)
library(DT)
library(ggsci)
library(scales)
library(patchwork)
library(sjPlot)
library(sjlabelled)
library(forcats)
library(markdown)
library(PropCIs)
library(shinythemes)
library(httr)
library(drawProteins)
library(ggrepel)
library(rms)
library(blorr)
library(dcurves)
library(Matching)
library(survRM2)
library(shinydashboard)
library(pROC)
library(tidymodels)
library(withr)
library(rpart)
library(ranger)
library(bonsai)
library(probably)
library(discrim)
library(flexsurv)
library(data.table)
library(parglm)
library(speedglm)
library(ggbeeswarm)
library(tidyquant)
library(ggpmisc)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Hsapiens.UCSC.hg19)
library(Biostrings)
library(mutSignatures)
library(readr)
library(corpcor)
library(maftools)
library(viridis)
library(epiR)
library(stringdist)
library(patchwork)
library(ggplotify)
library(ggseqlogo)
library(gridGraphics)
library(ggsignif)
library(cowplot)
library(lightgbm)
library(randomForest)
library(caret)
library(doParallel)
library(ggpubr)
library(forestplot)
library(lubridate)
library(grid)
library(gtable)

tidymodels_prefer()
options(pillar.advice = FALSE, pillar.min_title_chars = Inf)

nietzsche = read.csv(header = TRUE,
                     file("source/nietzsche.csv",
                          encoding='UTF-8-BOM'))$text
CH_GENE_NatComms =read_tsv("source/driver_discovery_table.tsv", show_col_types = FALSE)$Hugo_Symbol

Min_count = 10
Min_count_F1 = 10
Penal = 1
Toler = 10^-14
GENE_NO_THRESHOLD = 30

DDPS = file.exists("DDPS")

ht_opt$message = FALSE
options(shiny.maxRequestSize=16*1024^3)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_double)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_factor)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_integer)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_logical)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_numeric)
conflicted::conflicts_prefer(.quiet = T, patchwork::area)
conflicted::conflicts_prefer(.quiet = T, dplyr::arrange)
conflicted::conflicts_prefer(.quiet = T, sjlabelled::as_factor)
conflicted::conflicts_prefer(.quiet = T, gtsummary::as_flextable)
conflicted::conflicts_prefer(.quiet = T, sjlabelled::as_label)
conflicted::conflicts_prefer(.quiet = T, flextable::border)
conflicted::conflicts_prefer(.quiet = T, shinydashboard::box)
conflicted::conflicts_prefer(.quiet = T, readr::col_factor)
conflicted::conflicts_prefer(.quiet = T, purrr::compact)
conflicted::conflicts_prefer(.quiet = T, flextable::compose)
conflicted::conflicts_prefer(.quiet = T, gtsummary::continuous_summary)
conflicted::conflicts_prefer(.quiet = T, dplyr::count)
conflicted::conflicts_prefer(.quiet = T, DT::dataTableOutput)
conflicted::conflicts_prefer(.quiet = T, dplyr::desc)
conflicted::conflicts_prefer(.quiet = T, purrr::discard)
conflicted::conflicts_prefer(.quiet = T, dplyr::failwith)
conflicted::conflicts_prefer(.quiet = T, stringr::fixed)
conflicted::conflicts_prefer(.quiet = T, flextable::font)
conflicted::conflicts_prefer(.quiet = T, gt::html)
conflicted::conflicts_prefer(.quiet = T, dplyr::id)
conflicted::conflicts_prefer(.quiet = T, plyr::is.discrete)
conflicted::conflicts_prefer(.quiet = T, dplyr::lag)
conflicted::conflicts_prefer(.quiet = T, dplyr::mutate)
conflicted::conflicts_prefer(.quiet = T, survival::myeloma)
conflicted::conflicts_prefer(.quiet = T, shiny::observe)
conflicted::conflicts_prefer(.quiet = T, tidyr::pack)
conflicted::conflicts_prefer(.quiet = T, recipes::prepare)
conflicted::conflicts_prefer(.quiet = T, dials::prune)
conflicted::conflicts_prefer(.quiet = T, dplyr::rename)
conflicted::conflicts_prefer(.quiet = T, DT::renderDataTable)
conflicted::conflicts_prefer(.quiet = T, flextable::rotate)
conflicted::conflicts_prefer(.quiet = T, dplyr::select)
conflicted::conflicts_prefer(.quiet = T, dials::smoothness)
conflicted::conflicts_prefer(.quiet = T, yardstick::spec)
conflicted::conflicts_prefer(.quiet = T, dplyr::src)
conflicted::conflicts_prefer(.quiet = T, recipes::step)
conflicted::conflicts_prefer(.quiet = T, dplyr::summarise)
conflicted::conflicts_prefer(.quiet = T, dplyr::summarize)
conflicted::conflicts_prefer(.quiet = T, parsnip::translate)
conflicted::conflicts_prefer(.quiet = T, tidyr::unpack)
conflicted::conflicts_prefer(.quiet = T, recipes::update)
conflicted::conflicts_prefer(.quiet = T, rms::validate)
conflicted::conflicts_prefer(.quiet = T, dplyr::between)
conflicted::conflicts_prefer(.quiet = T, dplyr::first)
conflicted::conflicts_prefer(.quiet = T, dplyr::n_distinct)
conflicted::conflicts_prefer(.quiet = T, dplyr::last)
conflicted::conflicts_prefer(.quiet = T, purrr::transpose)
conflicted::conflicts_prefer(.quiet = T, base::as.data.frame)
conflicted::conflicts_prefer(.quiet = T, base::as.factor)
conflicted::conflicts_prefer(.quiet = T, graphics::legend)
conflicted::conflicts_prefer(.quiet = T, dplyr::union)
conflicted::conflicts_prefer(.quiet = T, dplyr::setdiff)
conflicted::conflicts_prefer(.quiet = T, dplyr::intersect)
conflicted::conflicts_prefer(.quiet = T, ggplot2::annotate)

Abs = function(x){
  x = as.numeric(x)
  x[is.na(x)]=1
  return(x)
}

fun_zero <- function(a, b){
  if(length(a) != length(b)){
    if(length(a) / length(b) == as.integer(length(a) / length(b))){
      b = rep(b,length(a) / length(b))
    }
  }
  return(ifelse(b == 0, 0, a / b))
}

gg_empty = function(){
  g = ggplot()
  g = g + geom_blank()
  g = g + ggtitle("")
  g = g + theme_void()
  return(g)
}

shannon.entropy = function(x, x_length){ 
  x = x[(!is.na(x))]
  x = x[(x != 0)]
  if(sum(x) <= 0) {
    return(log(x_length, base=2)) 
  } else {
    p = x/sum(x)
    score = -sum(p * log(p, base=2))
    return(score)
  }
} 

odds.ratio <- function(a, b, c, d, correct=FALSE){
  cl <- function(x){
    or*exp(c(1,-1)*qnorm(x)*sqrt(1/a+1/b+1/c+1/d))
  }
  if (correct || a*b*c*d==0) {
    a <- a+0.5
    b <- b+0.5
    c <- c+0.5
    d <- d+0.5
  }
  or <- a*d/(b*c)
  conf <- rbind(cl90=cl(0.05), cl95=cl(0.025), cl99=cl(0.005), cl999=cl(0.0005))
  conf <- data.frame(conf)
  colnames(conf) <- paste(c("lower","upper"), " limit" , sep="")
  rownames(conf) <- paste(c(90, 95, 99, 99.9), "%CI" , sep="")
  list(or=or, conf=conf)
}

surv_curv_CTx <- function(fit, data, title, legend_, diff_0){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from chemotherapy induction, risk-set adjusted (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    surv.scale = "percent",
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, min(365.25*5, max(data$time_palliative_final)) * 1.05),
    xscale = "d_m",
    break.x.by = 12 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
    legend.labs = legend_
  )
  if(is.null(diff_0)){
    tmp = summary(fit)$table
    legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
    g$plot = g$plot +
      labs(title = title,
           subtitle = paste0("Median OS, ", legends, " months"))
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    data_tmp = tidy(diff_0, exponentiate=TRUE, conf.int=TRUE)
    title_HR = NULL
    if(length(data_tmp$term) > 0){
      for(i in 1:length(data_tmp$term)){
        title_HR = c(title_HR,
                     paste0(format(data_tmp$estimate[i],digits=3), " (",
                            format(data_tmp$conf.low[i],digits=3), "-",
                            format(data_tmp$conf.high[i],digits=3), ") ",
                            "p=", format(data_tmp$p.value[i],digits=3)))
      }
    }
    g$plot = g$plot +
      labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
           subtitle = paste0("Median OS, ", legends, " months"))
  }
  g$table <- g$table + theme(plot.title = element_blank(),
                             plot.subtitle = element_blank())
  g$plot <- g$plot + theme(axis.title.x = element_blank(),
                           axis.text.x = element_blank(),
                           axis.ticks.x = element_blank())
  merged_plot <- wrap_elements(g$plot / g$table + plot_layout(heights = c(6,1)))
  return(merged_plot)
}

surv_curv_CTx_naive <- function(fit, data, title, legend_, diff_0){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from chemotherapy induction (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    surv.scale = "percent",
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, min(365.25*10, max(data$time_palliative_final)) * 1.05),
    xscale = "d_m",
    break.x.by = 12 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
    legend.labs = legend_
  )
  if(is.null(diff_0)){
    tmp = summary(fit)$table
    legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
    g$plot = g$plot +
      labs(title = title,
           subtitle = paste0("Median OS, ", legends, " months"))
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    data_tmp = tidy(diff_0, exponentiate=TRUE, conf.int=TRUE)
    title_HR = NULL
    if(length(data_tmp$term) > 0){
      for(i in 1:length(data_tmp$term)){
        title_HR = c(title_HR,
                     paste0(format(data_tmp$estimate[i],digits=3), " (",
                            format(data_tmp$conf.low[i],digits=3), "-",
                            format(data_tmp$conf.high[i],digits=3), ") ",
                            "p=", format(data_tmp$p.value[i],digits=3)))
      }
    }
    g$plot = g$plot +
      labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
           subtitle = paste0("Median OS, ", legends, " months"))
  }
  g$table <- g$table + theme(plot.title = element_blank(),
                             plot.subtitle = element_blank())
  g$plot <- g$plot + theme(axis.title.x = element_blank(),
                           axis.text.x = element_blank(),
                           axis.ticks.x = element_blank())
  merged_plot <- wrap_elements(g$plot / g$table + plot_layout(heights = c(6,1)))
  return(merged_plot)
}

surv_curv_CGP <- function(fit, data, title, legend_, diff_0){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from CGP test date (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    surv.scale = "percent",
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, min(365.25*5, max(data$time_enroll_final)) * 1.05),
    xscale = "d_m",
    break.x.by = 12 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
    legend.labs = legend_
  )
  if(is.null(diff_0)){
    tmp = summary(fit)$table
    legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
    g$plot = g$plot +
      labs(title = title,
           subtitle = paste0("Median OS, ", legends, " months"))
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    data_tmp = tidy(diff_0, exponentiate=TRUE, conf.int=TRUE)
    title_HR = NULL
    if(length(data_tmp$term) > 0){
      for(i in 1:length(data_tmp$term)){
        title_HR = c(title_HR,
                     paste0(format(data_tmp$estimate[i],digits=3), " (",
                            format(data_tmp$conf.low[i],digits=3), "-",
                            format(data_tmp$conf.high[i],digits=3), ") ",
                            "p=", format(data_tmp$p.value[i],digits=3)))
      }
    }
    g$plot = g$plot +
      labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
           subtitle = paste0("Median OS, ", legends, " months"))
  }
  g$table <- g$table + theme(plot.title = element_blank(),
                             plot.subtitle = element_blank())
  g$plot <- g$plot + theme(axis.title.x = element_blank(),
                           axis.text.x = element_blank(),
                           axis.ticks.x = element_blank())
  merged_plot <- wrap_elements(g$plot / g$table + plot_layout(heights = c(6,1)))
  return(merged_plot)
}

surv_curv_drug <- function(fit, data, title, diff_0, diff_1, diff_2=NULL){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from drug initiation (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    surv.scale = "percent",
    conf.int = FALSE,
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xscale = "d_m",
    xlim = c(0, 3*365.35),
    break.x.by = 3 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
  )
  if(is.null(diff_0)){
    if(is.null(diff_2)){
      tmp = summary(fit)$table
      legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
      g$plot = g$plot +
        labs(title = title,
             subtitle = paste0("Median OS, ", legends, " months"))
      g$table <- g$table + theme(plot.title = element_blank(),
                                 plot.subtitle = element_blank())
    } else {
      tmp = summary(fit)$table
      legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
      data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
      title_HR = NULL
      if(length(data_tmp$term) > 0){
        for(i in 1:length(data_tmp$term)){
          title_HR = c(title_HR,
                       paste0(format(data_tmp$estimate[i],digits=3), " (",
                              format(data_tmp$conf.low[i],digits=3), "-",
                              format(data_tmp$conf.high[i],digits=3), ") ",
                              "p=", format(data_tmp$p.value[i],digits=3)))
        }
      }
      g$plot = g$plot +
        labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
             subtitle = paste0("Median OS, ", legends, " months"))
      g$table <- g$table + theme(plot.title = element_blank(),
                                 plot.subtitle = element_blank())
    }
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
    title_HR = NULL
    if(length(data_tmp$term) > 0){
      for(i in 1:length(data_tmp$term)){
        title_HR = c(title_HR,
                     paste0(format(data_tmp$estimate[i],digits=3), " (",
                            format(data_tmp$conf.low[i],digits=3), "-",
                            format(data_tmp$conf.high[i],digits=3), ") ",
                            "p=", format(data_tmp$p.value[i],digits=3)))
      }
    }
    g$plot = g$plot +
      labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
           subtitle = paste0("Median OS, ", legends, " months"))
    g$plot = g$plot +
      labs(title = paste0(title, ": log-rank, p=", format(
        1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
        "/Wilcoxon, p=", format(
          1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)),
        subtitle = paste0("Median OS, ", legends," months, hazard ratio=", paste(title_HR, collapse = "/")))
    g$table <- g$table + theme(plot.title = element_blank(),
                               plot.subtitle = element_blank())
  }
  g$plot <- g$plot + theme(axis.title.x = element_blank(),
                           axis.text.x = element_blank(),
                           axis.ticks.x = element_blank())
  merged_plot <- wrap_elements(g$plot / g$table + plot_layout(heights = c(6,1)))
  return(merged_plot)
}

surv_curv_ICI <- function(fit, data, title, diff_0, diff_1, diff_2=NULL){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from drug initiation (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    surv.scale = "percent",
    conf.int = FALSE,
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xscale = "d_m",
    xlim = c(0, 6*365.35),
    break.x.by = 6 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
  )
  if(is.null(diff_0)){
    if(is.null(diff_2)){
      tmp = summary(fit)$table
      legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
      g$plot = g$plot +
        labs(title = title,
             subtitle = paste0("Median OS, ", legends, " months"))
      g$table <- g$table + theme(plot.title = element_blank(),
                                 plot.subtitle = element_blank())
    } else {
      tmp = summary(fit)$table
      legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
      data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
      title_HR = NULL
      if(length(data_tmp$term) > 0){
        for(i in 1:length(data_tmp$term)){
          title_HR = c(title_HR,
                       paste0(format(data_tmp$estimate[i],digits=3), " (",
                              format(data_tmp$conf.low[i],digits=3), "-",
                              format(data_tmp$conf.high[i],digits=3), ") ",
                              "p=", format(data_tmp$p.value[i],digits=3)))
        }
      }
      g$plot = g$plot +
        labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
             subtitle = paste0("Median OS, ", legends, " months"))
      g$table <- g$table + theme(plot.title = element_blank(),
                                 plot.subtitle = element_blank())
    }
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
    title_HR = NULL
    if(length(data_tmp$term) > 0){
      for(i in 1:length(data_tmp$term)){
        title_HR = c(title_HR,
                     paste0(format(data_tmp$estimate[i],digits=3), " (",
                            format(data_tmp$conf.low[i],digits=3), "-",
                            format(data_tmp$conf.high[i],digits=3), ") ",
                            "p=", format(data_tmp$p.value[i],digits=3)))
      }
    }
    g$plot = g$plot +
      labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
           subtitle = paste0("Median OS, ", legends, " months"))
    g$plot = g$plot +
      labs(title = paste0(title, ": log-rank, p=", format(
        1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
        "/Wilcoxon, p=", format(
          1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)),
        subtitle = paste0("Median OS, ", legends," months, hazard ratio=", paste(title_HR, collapse = "/")))
    g$table <- g$table + theme(plot.title = element_blank(),
                               plot.subtitle = element_blank())
  }
  g$plot <- g$plot + theme(axis.title.x = element_blank(),
                           axis.text.x = element_blank(),
                           axis.ticks.x = element_blank())
  merged_plot <- wrap_elements(g$plot / g$table + plot_layout(heights = c(6,1)))
  return(merged_plot)
}

# Define UI for application that draws a histogram
ui <- dashboardPage(skin = "black",
  dashboardHeader(tags$li(class = "dropdown",
                          tags$table(style="80%;align:center;border-collapse:seperate;border-spacing:20px;"),
                          tags$style(".main-header {max-height: 30px}"),
                          tags$style(".main-header .logo {height: 30px;}"),
                          tags$style(".sidebar-toggle {height: 30px; padding-top: 1px !important;}"),
                          tags$style(".navbar {min-height:30px !important}")),
                  title = HTML(
                    "<div style = 'background-color:FFFFFF; vertical-align:top'>
       <a href='https://github.com/MANO-B/BASTET'><img src = 'BASTET.png', align = 'left', height = '30px'></a> 
       </div><h6>BASTET version 1.15.1</h6>")
  ),
  dashboardSidebar(
    tags$style(".sidebar {height: calc(100vh - 50px); overflow-y: scroll; scrollbar-width: none;.left-side}, .main-sidebar {padding-top: 30px}"),
    sidebarMenu(style = "white-space: normal;",
      h3("Settings"),
      # menuItem("Input C-CAT files", tabName = "InputC-CATfiles", icon = icon("dashboard")),
      menuItem("Setting", tabName = "Setting", icon = icon("dashboard")),
      menuItem("Analysis", tabName = "Analysis", icon = icon("th")),
      hr(),
      h3("Results"),
      menuItem("Case summary", icon = icon("th"),
               hr(),
               p("Tables"),
               menuSubItem("Summarized by mutation pattern", tabName = "Summarizedbymutationpattern", icon = icon("angle-right")),
               menuSubItem("Summarized by histology", tabName = "Summarizedbyhistology", icon = icon("angle-right")),
               menuSubItem("Summarized by panel", tabName = "Summarizedbypanel", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Inter-panel comparison", icon = icon("th"),
               hr(),
               p("Figures"),
               menuSubItem("Main figure 1", tabName = "Inter_panel", icon = icon("angle-right")),
               menuSubItem("Main figure 2", tabName = "Inter_panel1", icon = icon("angle-right")),
               menuSubItem("Main figure 3", tabName = "Inter_panel1-2", icon = icon("angle-right")),
               menuSubItem("Main figure 4", tabName = "Inter_panel1-3", icon = icon("angle-right")),
               menuSubItem("Main figure 5", tabName = "Inter_panel1-4", icon = icon("angle-right")),
               menuSubItem("Main figure 6", tabName = "Inter_panel1-5", icon = icon("angle-right")),
               menuSubItem("Main figure 7", tabName = "Inter_panel1-6", icon = icon("angle-right")),
               menuSubItem("Main figure 8", tabName = "Inter_panel1-7", icon = icon("angle-right")),
               menuSubItem("Main figure 9", tabName = "Inter_panel1-8", icon = icon("angle-right")),
               menuSubItem("Main figure 10", tabName = "Inter_panel1-9", icon = icon("angle-right")),
               menuSubItem("Main figure 11", tabName = "Inter_panel1-10", icon = icon("angle-right")),
               menuSubItem("Main figure 12", tabName = "Inter_panel1-11", icon = icon("angle-right")),
               menuSubItem("Main figure 13", tabName = "Inter_panel1-12", icon = icon("angle-right")),
               menuSubItem("Main figure 14", tabName = "Inter_panel1-13", icon = icon("angle-right")),
               menuSubItem("Main figure 15", tabName = "Inter_panel1-14", icon = icon("angle-right")),
               menuSubItem("Main figure 16", tabName = "Inter_panel1-15", icon = icon("angle-right")),
               menuSubItem("Nomogram", tabName = "Inter_panel2", icon = icon("angle-right")),
               menuSubItem("Mutational signature, C-CAT", tabName = "Inter_panel3", icon = icon("angle-right")),
               menuSubItem("Cosmic signature, C-CAT", tabName = "Inter_panel4", icon = icon("angle-right")),
               menuSubItem("Oncoplot", tabName = "Inter_panel5", icon = icon("angle-right")),
               menuSubItem("Mutational signature, DNMT3A", tabName = "Inter_panel6", icon = icon("angle-right")),
               menuSubItem("Cosmic signature, DNMT3A", tabName = "Inter_panel7", icon = icon("angle-right")),
               menuSubItem("Mutational signaturem GENIE", tabName = "Inter_panel8", icon = icon("angle-right")),
               menuSubItem("Cosmic signature, GENIE", tabName = "Inter_panel9", icon = icon("angle-right")),
               menuSubItem("Mutational signature, DNMT3A Y735C", tabName = "Inter_panel10", icon = icon("angle-right")),
               menuSubItem("Cosmic signature, DNMT3A Y735C", tabName = "Inter_panel11", icon = icon("angle-right")),
               p("Tables"),
               menuSubItem("Summarized by patient", tabName = "SummarizedbyCHvariant", icon = icon("angle-right")),
               menuSubItem("Summarized by hotspot type", tabName = "Summarizedbytype", icon = icon("angle-right")),
               menuSubItem("Summarized by hotspot type in detail", tabName = "Summarizedbytype2", icon = icon("angle-right")),
               menuSubItem("Table of comparison data", tabName = "Table_inter_panel1", icon = icon("angle-right")),
               menuSubItem("Download all data", tabName = "Table_inter_panel2", icon = icon("angle-right")),
               menuSubItem("Table of comparison data", tabName = "Table_inter_panel3", icon = icon("angle-right")),
               menuSubItem("Odds ratio table", tabName = "Table_inter_panel4", icon = icon("angle-right")),
               menuSubItem("comparison, CCAT and GENIE", tabName = "Table_inter_panel5", icon = icon("angle-right")),
               menuSubItem("DNMT3A mutations with liquid-only muts", tabName = "DNMT3A_odds", icon = icon("angle-right")),
               hr()
      ),
      #hr(),
      #h3("Instruction"),
      #menuItem("About BASTET", tabName = "Instruction", icon = icon("th")),
      #menuItem("Tips", tabName = "Tips", icon = icon("th")),
      hr()
    )
  ),
  ## Body content
  dashboardBody(style='overflow-x: scroll;overflow-y: scroll;',
                tags$head(tags$style(HTML('
      .main-header .sidebar-toggle:before {
        content: "\\e068";}'))),
    tabItems(
      tabItem(tabName = "InputC-CATfiles",
              fluidRow(
                column(4,
                       strong("Required: C-CAT data files (zipped files are acceptable)"),
                       hr(),
                       fileInput(inputId = "clinical_files",
                                 label = "Choose case CSV Files",
                                 multiple = TRUE
                       ),
                       hr(),
                       fileInput(inputId = "report_files",
                                 label = "Choose report CSV Files",
                                 multiple = TRUE
                       ),
                       br(),
                       h6("Files need to be loaded only for the first analysis; loading is not required for the second and subsequent analyses. Go to Setting tab and click Start file loading/analysis settings.")
                ),
              ),
              hr(),
              h6("BASTET; clonAl haemaTopoiesis varianT prEdiction Based on liquid Sequencing for C-CAT database."),
              a(href="https://github.com/MANO-B/BASTET",h6("https://github.com/MANO-B/BASTET"))
      ),
      tabItem(tabName = "Setting",
              actionButton("start_setting", "Start file loading/analysis settings"),
              br(),
              hr(),
              fluidRow(
                column(3,strong("Filter on histology"),
                       hr(),
                       htmlOutput("select_histology"),
                       htmlOutput("select_histology_group_1"),
                       htmlOutput("select_histology_group_1_name"),
                       htmlOutput("select_histology_group_2"),
                       htmlOutput("select_histology_group_2_name"),
                       htmlOutput("select_histology_group_3"),
                       htmlOutput("select_histology_group_3_name"),
                       htmlOutput("select_histology_group_4"),
                       htmlOutput("select_histology_group_4_name"),
                       htmlOutput("select_minimum_pts"),
                       htmlOutput("select_histology_detail")
                ),
                column(3,strong("Filters for clinical information"),
                       hr(),
                       htmlOutput("select_sex"),
                       h5("女: Female, 男: Male, 未入力・不明: Unknown"),
                       br(),
                       htmlOutput("select_panel"),
                       htmlOutput("select_age"),
                       htmlOutput("select_mid_age"),
                       htmlOutput("select_PS"),
                       h5("不明: Unknown"),
                       br(),
                       htmlOutput("select_smoking"),
                       h5("あり: Yes, なし: No, 不明: Unknown"),
                       br(),
                       htmlOutput("select_year")
                ),
                column(3,strong("Filters on genes"),
                       hr(),
                       htmlOutput("select_gene"),
                       htmlOutput("select_gene_group_1"),
                       htmlOutput("select_gene_group_2"),
                       htmlOutput("select_gene_group_analysis"),
                       htmlOutput("select_lolliplot_gene"),
                       htmlOutput("select_lolliplot_no"),
                       br(),
                       br(),
                       strong("Filter on mutation types"),
                       hr(),
                       strong("For detailed study of mutations of a gene"),
                       br(),
                       htmlOutput("select_special_gene"),
                       htmlOutput("select_special_gene_mutation_1"),
                       htmlOutput("select_special_gene_mutation_1_name"),
                       htmlOutput("select_special_gene_mutation_2"),
                       htmlOutput("select_special_gene_mutation_2_name"),
                       htmlOutput("select_special_gene_annotation"),
                       htmlOutput("select_special_gene_independent")
                ),
                column(3,strong("Other Settings"),
                       hr(),
                       htmlOutput("select_gene_no"),
                       htmlOutput("select_oncoprint_option"),
                       htmlOutput("select_patho"),
                       htmlOutput("select_clonal_hematopoiesis_genes"),
                       htmlOutput("select_target_line")
                ),
              )
      ),
      tabItem(tabName = "Analysis",
              fluidRow(
                column(4,
                       strong("Standard Analysis"),
                       hr(),
                       actionButton("summary_base", "Case summary"),
                       br(),
                       br(),
                       actionButton("inter_panel", "Inter-panel comparison"),
                       br()
                )
              ),
              hr(),
              h5("Figures in results are downloadable as png files.")
      ),
      tabItem(tabName = "Summarizedbymutationpattern",
              gt_output("table_summary_1"),
                       hr(),
                       h6(paste("Table. Characteristics for selected patients.",
                                "The present, retrospective cohort study was performed with clinicogenomic,",
                                "real-world data on the patients who were registered in the C-CAT database",
                                "from June 1, 2019. The patients were registered by hospitals throughout Japan",
                                "and provided written informed consent to the secondary use of their clinicogenomic",
                                "data for research."))
      ),
      tabItem(tabName = "Summarizedbyhistology",
              gt_output("table_summary_2"),
              hr(),
              DT::dataTableOutput("table_summary_3"),
              hr(),
              h6("Table. Characteristics for selected patients.")
      ),
      tabItem(tabName = "Summarizedbypanel",
              gt_output("table_summary_4"),
              hr(),
              h6("Table. Characteristics for selected patients.")
      ),
      tabItem(tabName = "Inter_panel",
              plotOutput('figure_inter_panel_a', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_a", "Download the figure as a pdf file"),
              hr()
      ),
      tabItem(tabName = "Inter_panel1",
              plotOutput('figure_inter_panel_b', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_b", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-2",
              plotOutput('figure_inter_panel_c', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_c", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-3",
              plotOutput('figure_inter_panel_d', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_d", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-4",
              plotOutput('figure_inter_panel_e', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_e", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-5",
              plotOutput('figure_inter_panel_f', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_f", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-6",
              plotOutput('figure_inter_panel_h', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_h", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-7",
              plotOutput('figure_inter_panel_i', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_i", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-8",
              plotOutput('figure_inter_panel_j', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_j", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-9",
              plotOutput('figure_inter_panel_k', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_k", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-10",
              plotOutput('figure_inter_panel_l', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_l", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-11",
              plotOutput('figure_inter_panel_m', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_m", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-12",
              plotOutput('figure_inter_panel_n', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_n", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-13",
              plotOutput('figure_inter_panel_o', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_o", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-14",
              plotOutput('figure_inter_panel_p', 
                         height = "4000px",
                         width = "4000px"),
              downloadButton("figure_inter_panel_download_p", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel1-15",
              plotOutput('figure_inter_panel_g', 
                         height = "4000px",
                         width = "1600px"),
              downloadButton("figure_inter_panel_download_g", "Download the figure as a pdf file")
      ),
      tabItem(tabName = "Inter_panel2",
              plotOutput('figure_inter_panel2', 
                         height = "2000px",
                         width = "1000px"),
              hr(),
              plotOutput('figure_inter_panel2_2', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              verbatimTextOutput("text_inter_panel"),
              hr(),
              h6("Figure. The nomogram was created with the lrm function of the rms package for R with a setting of penalty=1. Nagelkerke R2 was calculated with blorr package for R. Possible sampling bias was corrected with 500-time bootstrap sampling and then concordance index was estimated.")
      ),
      tabItem(tabName = "Inter_panel3",
              plotOutput('figure_signature',
                         height = "1500px",
                         width = "1500px"),
              hr()      ),
      tabItem(tabName = "Inter_panel4",
              plotOutput('figure_signature2',
                         height = "3000px",
                         width = "1500px"),
              hr()      ),
      tabItem(tabName = "Inter_panel5",
              h6("Oncoplot for liquid-favor hotspot variants"),
              plotOutput('figure_oncoplot1',
                         height = "1500px",
                         width = "1500px"),
              h6("Oncoplot for liquid-favor hotspot variants"),
              h6("FoundationOne liquid, especially for CH genes"),
              plotOutput('figure_oncoplot2',
                         height = "1000px",
                         width = "1500px"),
              h6("Mutually exclusive/co-occurence for liquid-favor hotspot variants"),
              plotOutput('figure_oncoplot3',
                         height = "1000px",
                         width = "1000px"),
              DT::dataTableOutput("figure_oncoplot3_data"),
              h6("Mutually exclusive/co-occurence for liquid-favor hotspot variants"),
              h6("FoundationOne liquid, especially for DNMT3A variants and mutation patterns"),
              plotOutput('figure_oncoplot4',
                         height = "1000px",
                         width = "1000px"),
              DT::dataTableOutput("figure_oncoplot4_data"),
              h6("Oncoplot for blood-favor hotspot variants"),
              h6("GENIE blood cancer, especially for 5 CH genes"),
              plotOutput('figure_oncoplot5',
                         height = "1000px",
                         width = "1500px"),
              h6("Mutually exclusive/co-occurence for blood-favor hotspot variants"),
              h6("GENIE blood cancer, especially for 5 CH genes"),
              plotOutput('figure_oncoplot6',
                         height = "1000px",
                         width = "1000px"),
              DT::dataTableOutput("figure_oncoplot6_data"),
              hr()      ),
      tabItem(tabName = "Inter_panel6",
              plotOutput('figure_signature_DNMT3A',
                         height = "1500px",
                         width = "1500px"),
              hr(),
              DT::dataTableOutput("table_summary_CCAT")
              ),
      tabItem(tabName = "Inter_panel7",
              plotOutput('figure_signature2_DNMT3A',
                         height = "3000px",
                         width = "1500px"),
              hr()      ),
      tabItem(tabName = "Inter_panel8",
              plotOutput('figure_signature_DNMT3A_GENIE',
                         height = "1500px",
                         width = "1500px"),
              hr(),
              DT::dataTableOutput("table_summary_GENIE")      ),
      tabItem(tabName = "Inter_panel9",
              plotOutput('figure_signature2_DNMT3A_GENIE',
                         height = "3000px",
                         width = "1500px"),
              hr()
              ),
      tabItem(tabName = "Inter_panel10",
              plotOutput('figure_signature_DNMT3A_2',
                         height = "1500px",
                         width = "1500px"),
              hr(),
              DT::dataTableOutput("table_summary_CCAT")
      ),
      tabItem(tabName = "Inter_panel11",
              plotOutput('figure_signature2_DNMT3A_2',
                         height = "3000px",
                         width = "1500px"),
              hr()      ),
      tabItem(tabName = "Table_inter_panel1",
              downloadButton('table_all_variants',"Download blood-favor data in C-CAT"),
              downloadButton('table_all_variants_GENIE',"Download blood-favor data in GENIE"),
              #DT::dataTableOutput("table_all_variants"),
              hr()
      ),
      tabItem(tabName = "Table_inter_panel2",
              downloadButton('table_volcano_variant',"Download the volcano plot data"),
              #DT::dataTableOutput("table_volcano_variant"),
              hr()
      ),
      tabItem(tabName = "Table_inter_panel3",
              DT::dataTableOutput("table_inter_panel1"),
              hr()
      ),
      tabItem(tabName = "Table_inter_panel4",
              gt_output("table_inter_panel2"),
              hr()
      ),
      tabItem(tabName = "Table_inter_panel5",
              DT::dataTableOutput("comparison_CCAT_GENIE"),
      ),
      tabItem(tabName = "Summarizedbytype",
              gt_output("table_inter_panel3"),
              hr(),
              h6("Table. Characteristics for variants.")
      ),
      tabItem(tabName = "Summarizedbytype2",
              gt_output("table_inter_panel5"),
              hr(),
              h6("Table. Characteristics for variants.")
      ),
      tabItem(tabName = "SummarizedbyCHvariant",
              gt_output("table_inter_panel4"),
              hr(),
              gt_output("table_inter_panel4_detail"),
              hr(),
              gt_output("table_inter_panel_TICH"),
              hr(),
              gt_output("table_inter_panel_TOCH"),
              hr(),
              gt_output("table_inter_panel_DDR_TICH"),
              hr(),
              gt_output("table_inter_panel_DDR_TOCH"),
              hr(),
              gt_output("table_inter_panel_Epi_TICH"),
              hr(),
              gt_output("table_inter_panel_Epi_TOCH"),
              hr(),
              gt_output("table_inter_panel_DDR_SV"),
              hr(),
              gt_output("table_inter_panel_DDR_LV"),
              hr(),
              gt_output("table_inter_panel_Epi_SV"),
              hr(),
              gt_output("table_inter_panel_Epi_LV"),
              hr(),
              gt_output("table_inter_panel_TICH_2"),
              hr(),
              gt_output("table_inter_panel_TOCH_2"),
              hr(),
              gt_output("table_inter_panel_DDR_TICH_2"),
              hr(),
              gt_output("table_inter_panel_DDR_TOCH_2"),
              hr(),
              gt_output("table_inter_panel_Epi_TICH_2"),
              hr(),
              gt_output("table_inter_panel_Epi_TOCH_2"),
              hr(),
              gt_output("table_inter_panel_DDR_SV_2"),
              hr(),
              gt_output("table_inter_panel_DDR_LV_2"),
              hr(),
              gt_output("table_inter_panel_Epi_SV_2"),
              hr(),
              gt_output("table_inter_panel_Epi_LV_2"),
              hr(),
              h6("Table. Characteristics for patients.")
      ),

      tabItem(tabName = "Oncoprint",
              plotOutput('figure_oncoprint', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Recurrent oncogenic mutations in selected cases. The 30 genes with the highest frequency of oncogenic mutations are shown. Mutational landscapes were created using ComplexHeatmap package for R.")
      ),
      tabItem(tabName = "Lolliplotfortheselectedgene",
              plotOutput('figure_lolliplot2', 
                                  height = "500px",
                                  width = "1500px"),
                       hr(),
                       h6("Figure. Frequency of oncogenic mutations in the selected gene. The most frequent oncogenic mutations are shown with amino acid change."),
                       h6("Mutplot by Zhang W, PMID:31091262. If error occurs, correct 'source/UniPlot.txt'."),
                       h6("Protein structure source: Uniprot"),
                       HTML("<p>Github for Mutplot. <a href='https://github.com/VivianBailey/Mutplot'>Link for the website</a></p>")
      ),
      tabItem(tabName = "DNMT3A_odds",
              DT::dataTableOutput("table_DNMT3A_odds"),
              hr(),
              DT::dataTableOutput("table_DNMT3A_odds_blood"),
              h6("Table. DNMT3A mutations co-occur with liquid-only mutations in C-CAT FoundationOne Liquid samples.")
      )
    )
  )
)

  
# Define server logic required to draw a histogram
server <- function(input, output, session) {
  Data_case_raw =  reactive({
    withProgress(message = "Clinical data loading.", {
      if(file.exists("source/clinical_data.rda")){
        load(file="source/clinical_data.rda")
      } else {
        Encode = guess_encoding(input$clinical_files[[1, 'datapath']])[[1]][[1]]
        if(Encode == "Shift_JIS"){
          Encode = "CP932"
        }
        clin_tmp <- read_csv(col_names = TRUE,
                             file = input$clinical_files$datapath,
                             locale = locale(encoding=Encode),
                             num_threads=max(1, parallel::detectCores() - 1, na.rm = TRUE),
                             progress=FALSE,
                             show_col_types=FALSE,
                             name_repair=make.names,
                             col_select = c(
                               C.CAT調査結果.基本項目.ハッシュID,
                               症例.基本情報.年齢,
                               症例.背景情報.病理診断名,
                               症例.背景情報.臨床診断名,
                               症例.検体情報.病理診断名,
                               症例.基本情報.性別.名称.,
                               症例.基本情報.がん種.OncoTree.,
                               症例.基本情報.がん種.OncoTree..名称.,
                               症例.基本情報.がん種.OncoTree.LEVEL1.,
                               症例.検体情報.パネル.名称.,
                               症例.背景情報.ECOG.PS.名称.,
                               症例.背景情報.家族歴有無.名称.,
                               症例.背景情報.喫煙歴有無.名称.,
                               症例.背景情報.喫煙年数,
                               #症例.背景情報.喫煙１日本数,
                               症例.背景情報.アルコール多飲有無.名称.,
                               症例.背景情報.重複がん有無.異なる臓器..名称.,
                               症例.背景情報.多発がん有無.同一臓器..名称.,
                               症例.がん種情報.登録時転移有無.名称.,
                               症例.がん種情報.登録時転移部位.名称.,
                               症例.検体情報.腫瘍細胞含有割合,
                               症例.検体情報.検体採取部位.名称.,
                               症例.検体情報.検体採取日.腫瘍組織.,
                               症例.EP前レジメン情報.治療ライン.名称.,
                               症例.EP前レジメン情報.実施目的.名称.,
                               症例.EP前レジメン情報.化学療法レジメン名称,
                               症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                               症例.EP前レジメン情報.投与開始日,
                               症例.EP前レジメン情報.投与終了日,
                               症例.EP前レジメン情報.終了理由.名称.,
                               症例.EP前レジメン情報.レジメン継続区分.名称.,
                               症例.EP前レジメン情報.最良総合効果.名称.,
                               症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                               症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                               症例.EP後レジメン情報.治療ライン,
                               症例.EP後レジメン情報.治療ライン.名称.,
                               症例.EP後レジメン情報.化学療法レジメン名称,
                               症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                               症例.EP後レジメン情報.投与開始日,
                               症例.EP後レジメン情報.投与終了日,
                               症例.EP後レジメン情報.エキスパートパネル開催日,
                               症例.EP後レジメン情報.終了理由.名称.,
                               症例.EP後レジメン情報.レジメン継続区分.名称.,
                               症例.EP後レジメン情報.最良総合効果.名称.,
                               症例.転帰情報.転帰.名称.,
                               症例.背景情報.診断日,
                               症例.管理情報.登録日,
                               症例.転帰情報.最終生存確認日,
                               症例.転帰情報.死亡日,
                               # 症例.がん種情報.固形がん.マイクロサテライト不安定性.名称.,
                               # 症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称.,
                               # 症例.がん種情報.肺.EGFR.名称.,
                               # 症例.がん種情報.肺.ALK融合.名称.,
                               # 症例.がん種情報.肺.ROS1.名称.,
                               # 症例.がん種情報.肺.BRAF.V600..名称.,
                               # 症例.がん種情報.肺.MET遺伝子エクソン14スキッピング変異.名称.,
                               # 症例.がん種情報.肺.KRAS.G12C遺伝子変異.名称.,
                               # 症例.がん種情報.肺.RET融合遺伝子.名称.,
                               # 症例.がん種情報.肺.PD.L1.IHC..名称.,
                               # 症例.がん種情報.乳.HER2.IHC..名称.,
                               # 症例.がん種情報.乳.HER2.FISH..名称.,
                               # 症例.がん種情報.乳.ER.名称.,
                               # 症例.がん種情報.乳.PgR.名称.,
                               # 症例.がん種情報.食道.胃.腸.HER2.名称.,
                               # 症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称.
                             ),
                             col_types = cols(
                               C.CAT調査結果.基本項目.ハッシュID = col_character(),
                               症例.基本情報.年齢 = col_integer(),
                               症例.背景情報.病理診断名 = col_character(),
                               症例.背景情報.臨床診断名 = col_character(),
                               症例.検体情報.病理診断名 = col_character(),
                               症例.基本情報.性別.名称. = col_character(),
                               症例.基本情報.がん種.OncoTree. = col_character(),
                               症例.基本情報.がん種.OncoTree..名称. = col_character(),
                               症例.基本情報.がん種.OncoTree.LEVEL1. = col_character(),
                               症例.検体情報.パネル.名称. = col_character(),
                               症例.背景情報.ECOG.PS.名称. = col_character(),
                               症例.背景情報.家族歴有無.名称. = col_character(),
                               症例.背景情報.喫煙歴有無.名称. = col_character(),
                               症例.背景情報.喫煙年数 = col_integer(),
                               #症例.背景情報.喫煙１日本数 = col_integer(),
                               症例.背景情報.アルコール多飲有無.名称. = col_character(),
                               症例.背景情報.重複がん有無.異なる臓器..名称. = col_character(),
                               症例.背景情報.多発がん有無.同一臓器..名称. = col_character(),
                               症例.がん種情報.登録時転移有無.名称. = col_character(),
                               症例.がん種情報.登録時転移部位.名称. = col_character(),
                               症例.検体情報.腫瘍細胞含有割合 = col_integer(),
                               症例.検体情報.検体採取部位.名称. = col_character(),
                               症例.検体情報.検体採取日.腫瘍組織. = col_character(),
                               症例.EP前レジメン情報.治療ライン.名称. = col_character(),
                               症例.EP前レジメン情報.実施目的.名称. = col_character(),
                               症例.EP前レジメン情報.化学療法レジメン名称 = col_character(),
                               症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = col_character(),
                               症例.EP前レジメン情報.投与開始日 = col_character(),
                               症例.EP前レジメン情報.投与終了日 = col_character(),
                               症例.EP前レジメン情報.終了理由.名称. = col_character(),
                               症例.EP前レジメン情報.レジメン継続区分.名称. = col_character(),
                               症例.EP前レジメン情報.最良総合効果.名称. = col_character(),
                               症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. = col_character(),
                               症例.EP後レジメン情報.提示された治療薬を投与した.名称. = col_character(),
                               症例.EP後レジメン情報.治療ライン = col_integer(),
                               症例.EP後レジメン情報.治療ライン.名称. = col_character(),
                               症例.EP後レジメン情報.化学療法レジメン名称 = col_character(),
                               症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = col_character(),
                               症例.EP後レジメン情報.投与開始日 = col_character(),
                               症例.EP後レジメン情報.投与終了日 = col_character(),
                               症例.EP後レジメン情報.エキスパートパネル開催日 = col_character(),
                               症例.EP後レジメン情報.終了理由.名称. = col_character(),
                               症例.EP後レジメン情報.レジメン継続区分.名称. = col_character(),
                               症例.EP後レジメン情報.最良総合効果.名称. = col_character(),
                               症例.転帰情報.転帰.名称. = col_character(),
                               症例.背景情報.診断日 = col_character(),
                               症例.管理情報.登録日 = col_character(),
                               症例.転帰情報.最終生存確認日 = col_character(),
                               症例.転帰情報.死亡日 = col_character(),
                               .default = col_guess()
                             ))
        incProgress(1 / 4)
        #clin_tmp = as.data.frame(clin_tmp)
        clin_tmp = data.table(clin_tmp)
        clin_tmp$症例.背景情報.ECOG.PS.名称. = as.character(clin_tmp$症例.背景情報.ECOG.PS.名称.)
        clin_tmp[, 症例.基本情報.がん種.OncoTree..名称. := 
                   str_split_fixed(str_replace(症例.基本情報.がん種.OncoTree..名称., "\\)_", "\\)__"), "__", 2)[, 1]]
        
        # `case_when()` を `:=` で高速化
        clin_tmp[, 症例.基本情報.がん種.OncoTree..名称. := 
                   ifelse(str_detect(症例.基本情報.がん種.OncoTree..名称., paste0(" \\(", 症例.基本情報.がん種.OncoTree.LEVEL1., "\\)")),
                          症例.基本情報.がん種.OncoTree.LEVEL1., 症例.基本情報.がん種.OncoTree..名称.)]
        
        clin_tmp[, 症例.背景情報.家族歴有無.名称. := 
                   fifelse(is.na(症例.背景情報.家族歴有無.名称.) | 症例.背景情報.家族歴有無.名称. == "", "不明", 症例.背景情報.家族歴有無.名称.)]
        
        clin_tmp[, 症例.背景情報.ECOG.PS.名称. := 
                   fifelse(is.na(症例.背景情報.ECOG.PS.名称.) | 症例.背景情報.ECOG.PS.名称. == "", "不明", 症例.背景情報.ECOG.PS.名称.)]
        
        clin_tmp[, 症例.基本情報.性別.名称. := 
                   fifelse(is.na(症例.基本情報.性別.名称.) | 症例.基本情報.性別.名称. == "", "未入力・不明", 症例.基本情報.性別.名称.)]
        
        clin_tmp[, 症例.基本情報.がん種.OncoTree.LEVEL1. := 
                   fifelse(症例.基本情報.がん種.OncoTree.LEVEL1. == "WHO_BRAIN", "BRAIN", 症例.基本情報.がん種.OncoTree.LEVEL1.)]        
        if(input$histology_detail == "Yes, use oncotree 1st level"){
          clin_tmp$症例.基本情報.がん種.OncoTree..名称. = clin_tmp$症例.基本情報.がん種.OncoTree.LEVEL1.
          clin_tmp$症例.基本情報.がん種.OncoTree. = clin_tmp$症例.基本情報.がん種.OncoTree.LEVEL1.
        }
        if(!is.null(input$ID_histology)){
          ID_histology_list = data.frame(NULL)
          for(i in 1:length(input$ID_histology[,1])){
            ID_histology_list <- rbind(ID_histology_list,
                              read.csv(header = TRUE,
                                file(input$ID_histology[[i, 'datapath']],
                                     encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.基本情報.がん種.OncoTree..名称._tmp = unlist(lapply(list(clin_tmp$C.CAT調査結果.基本項目.ハッシュID), function(x) {
            as.vector(ID_histology_list$Histology[match(x, ID_histology_list$ID)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.基本情報.がん種.OncoTree..名称. = case_when(
              !is.na(症例.基本情報.がん種.OncoTree..名称._tmp) ~ 症例.基本情報.がん種.OncoTree..名称._tmp,
              TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
            )
          )
        }
        if(!is.null(input$regimen_rename)){
          RenList = data.frame(NULL)
          for(i in 1:length(input$regimen_rename[,1])){
            RenList <- rbind(RenList,
                             read.csv(header = TRUE,
                                      file(input$regimen_rename[[i, 'datapath']],
                                           encoding='UTF-8-BOM')))
          }
          for(i in 1:length(RenList$P1)){
            clin_tmp = clin_tmp %>% dplyr::mutate(
              症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "" &
                  症例.EP前レジメン情報.化学療法レジメン名称 == RenList$P1[i] ~ RenList$P2[i],
                TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
              ))
          }
        }
        if(!is.null(input$drug_rename)){
          drug_rename_list = data.frame(NULL)
          for(i in 1:length(input$drug_rename[,1])){
            drug_rename_list <- rbind(drug_rename_list,
                                       read.csv(header = TRUE,
                                         file(input$drug_rename[[i, 'datapath']],
                                              encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp = unlist(lapply(list(clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.), function(x) {
            as.vector(drug_rename_list$Rename[match(x, drug_rename_list$Drug)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
              !is.na(症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp) ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp,
              TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
            )
          )
        }
        clin_tmp = clin_tmp %>% dplyr::mutate(
          症例.基本情報.がん種.OncoTree..名称. = case_when(
            is.na(症例.基本情報.がん種.OncoTree..名称.) ~ 症例.基本情報.がん種.OncoTree.LEVEL1.,
            症例.基本情報.がん種.OncoTree..名称. == "" ~ 症例.基本情報.がん種.OncoTree.LEVEL1.,
            TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
          )
        )
        if(!is.null(input$histology_rename)){
          histology_rename_list = data.frame(NULL)
          for(i in 1:length(input$histology_rename[,1])){
            histology_rename_list <- rbind(histology_rename_list,
                                           read.csv(header = TRUE,
                                             file(input$histology_rename[[i, 'datapath']],
                                                  encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.基本情報.がん種.OncoTree..名称._tmp = unlist(lapply(list(clin_tmp$症例.基本情報.がん種.OncoTree..名称.), function(x) {
            as.vector(histology_rename_list$Rename[match(x, histology_rename_list$Histology)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.基本情報.がん種.OncoTree..名称. = case_when(
              !is.na(症例.基本情報.がん種.OncoTree..名称._tmp) ~ 症例.基本情報.がん種.OncoTree..名称._tmp,
              TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
            )
          )
        }
        replace_na_values <- list(
          "症例.背景情報.喫煙年数" = 0,
          "症例.検体情報.検体採取部位.名称." = "不明",
          "症例.背景情報.重複がん有無.異なる臓器..名称." = "不明",
          "症例.背景情報.多発がん有無.同一臓器..名称." = "不明",
          "症例.背景情報.喫煙歴有無.名称." = "",
          "症例.背景情報.ECOG.PS.名称." = "不明",
          "症例.検体情報.パネル.名称." = "",
          "症例.がん種情報.登録時転移部位.名称." = "",
          "症例.転帰情報.死亡日" = "",
          "症例.転帰情報.最終生存確認日" = "",
          "症例.管理情報.登録日" = "",
          "症例.EP前レジメン情報.投与開始日" = "",
          "症例.EP前レジメン情報.投与終了日" = "",
          "症例.検体情報.検体採取日.腫瘍組織." = "",
          "症例.EP後レジメン情報.投与開始日" = "",
          "症例.EP後レジメン情報.投与終了日" = "",
          "症例.EP後レジメン情報.エキスパートパネル開催日" = "",
          "症例.背景情報.診断日" = ""
        )
        for (col in names(replace_na_values)) {
          set(clin_tmp, which(is.na(clin_tmp[[col]])), col, replace_na_values[[col]])
        }
        set(clin_tmp, which(clin_tmp$症例.背景情報.喫煙年数 == 99), "症例.背景情報.喫煙年数", NA)
        date_columns <- c(
          "症例.管理情報.登録日", "症例.転帰情報.最終生存確認日", "症例.転帰情報.死亡日",
          "症例.EP前レジメン情報.投与開始日", "症例.EP後レジメン情報.投与開始日",
          "症例.EP前レジメン情報.投与終了日", "症例.EP後レジメン情報.投与終了日",
          "症例.EP後レジメン情報.エキスパートパネル開催日", "症例.背景情報.診断日",
          "症例.検体情報.検体採取日.腫瘍組織."
        )
        for (col in date_columns) {
          set(clin_tmp, , col, str_replace_all(clin_tmp[[col]], "/", "-"))
        }
        # 年齢の変換
        set(clin_tmp, , "症例.基本情報.年齢", as.integer(clin_tmp$症例.基本情報.年齢))
        incProgress(1 / 4)
        
        # がん種名称の前処理
        clin_tmp[, 症例.基本情報.がん種.OncoTree..名称. := 
                   str_split_fixed(str_replace(症例.基本情報.がん種.OncoTree..名称., "\\)_", "\\)__"), "__", 2)[, 1]]
        
        # ID_histology の適用
        if (!is.null(input$ID_histology)) {
          clin_tmp[, 症例.基本情報.がん種.OncoTree..名称. := 
                     fifelse(!is.na(ID_histology_list$Histology[match(症例.基本情報.がん種.OncoTree..名称., ID_histology_list$ID)]),
                             ID_histology_list$Histology[match(症例.基本情報.がん種.OncoTree..名称., ID_histology_list$ID)],
                             症例.基本情報.がん種.OncoTree..名称.)]
        }
        
        # がん種の欠損処理
        clin_tmp[, 症例.基本情報.がん種.OncoTree..名称. := 
                   fifelse(症例.基本情報.がん種.OncoTree..名称. == "", 症例.基本情報.がん種.OncoTree.LEVEL1., 症例.基本情報.がん種.OncoTree..名称.)]
        
        clin_tmp[, 症例.基本情報.がん種.OncoTree. := 
                   fifelse(症例.基本情報.がん種.OncoTree. == "", 症例.基本情報.がん種.OncoTree.LEVEL1., 症例.基本情報.がん種.OncoTree.)]
        
        # histology_rename の適用
        if (!is.null(input$histology_rename)) {
          clin_tmp[, 症例.基本情報.がん種.OncoTree..名称. := 
                     fifelse(!is.na(histology_rename_list$Rename[match(症例.基本情報.がん種.OncoTree..名称., histology_rename_list$Histology)]),
                             histology_rename_list$Rename[match(症例.基本情報.がん種.OncoTree..名称., histology_rename_list$Histology)],
                             症例.基本情報.がん種.OncoTree..名称.)]
        }
        
        # OncoTree の処理
        clin_tmp[, 症例.基本情報.がん種.OncoTree. := 
                   str_replace(str_replace(症例.基本情報.がん種.OncoTree..名称., "\\)", ""), ".*\\(", "")]
        
        # 背景情報の欠損処理
        clin_tmp[, `:=`(
          症例.背景情報.アルコール多飲有無.名称. = fifelse(is.na(症例.背景情報.アルコール多飲有無.名称.) | 症例.背景情報.アルコール多飲有無.名称. == "", "不明", 症例.背景情報.アルコール多飲有無.名称.),
          症例.背景情報.重複がん有無.異なる臓器..名称. = fifelse(症例.背景情報.重複がん有無.異なる臓器..名称. == "", "不明", 症例.背景情報.重複がん有無.異なる臓器..名称.),
          症例.背景情報.多発がん有無.同一臓器..名称. = fifelse(症例.背景情報.多発がん有無.同一臓器..名称. == "", "不明", 症例.背景情報.多発がん有無.同一臓器..名称.),
          症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. = fifelse(is.na(症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.) | 症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. == "", "いいえ", 症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.),
          症例.EP後レジメン情報.提示された治療薬を投与した.名称. = fifelse(is.na(症例.EP後レジメン情報.提示された治療薬を投与した.名称.) | 症例.EP後レジメン情報.提示された治療薬を投与した.名称. %in% c("", "不明"), "投与しなかった", 症例.EP後レジメン情報.提示された治療薬を投与した.名称.),
          症例.背景情報.喫煙歴有無.名称. = fifelse(症例.背景情報.喫煙歴有無.名称. == "", "不明", 症例.背景情報.喫煙歴有無.名称.)
        )]
        
        # 日付の処理を最適化
        date_columns <- c("症例.EP前レジメン情報.投与開始日", "症例.EP後レジメン情報.投与開始日",
                          "症例.EP前レジメン情報.投与終了日", "症例.EP後レジメン情報.投与終了日",
                          "症例.EP後レジメン情報.エキスパートパネル開催日")
        
        for (col in date_columns) {
          clin_tmp[, (col) := fifelse(get(col) < "1924-02-19" & get(col) > "1901-01-01",
                                      paste0("20", str_sub(get(col), 3, 10)),
                                      get(col))]
        }
        clin_tmp = clin_tmp %>%
          mutate(
            転移分類 = case_when(
              症例.がん種情報.登録時転移部位.名称. %in% c("リンパ節", "リンパ節/リンパ管") ~ "lymph",
              症例.がん種情報.登録時転移部位.名称. %in% c("脳", "中枢神経系") ~ "brain",
              症例.がん種情報.登録時転移部位.名称. == "肺" ~ "lung",
              症例.がん種情報.登録時転移部位.名称. %in% c("骨髄", "骨") ~ "bone",
              症例.がん種情報.登録時転移部位.名称. == "肝" ~ "liver",
              !症例.がん種情報.登録時転移部位.名称. %in% c("リンパ節", "リンパ節/リンパ管", "脳", "中枢神経系", "肺", "骨髄", "骨", "肝", "") ~ "other",
              TRUE ~ NA_character_
            )
          )
        
        # 各分類の ID を取得
        ID_list = clin_tmp %>%
          filter(!is.na(転移分類)) %>%
          group_by(転移分類) %>%
          summarise(ID = list(C.CAT調査結果.基本項目.ハッシュID), .groups = "drop")
        
        # それぞれの転移部位の ID を抽出
        ID_lymph_met = ID_list$ID[ID_list$転移分類 == "lymph"]
        ID_brain_met = ID_list$ID[ID_list$転移分類 == "brain"]
        ID_lung_met  = ID_list$ID[ID_list$転移分類 == "lung"]
        ID_bone_met  = ID_list$ID[ID_list$転移分類 == "bone"]
        ID_liver_met = ID_list$ID[ID_list$転移分類 == "liver"]
        ID_other_met = ID_list$ID[ID_list$転移分類 == "other"]
        # ID_not_brain_bone_liver_met = (clin_tmp %>% dplyr::filter(!症例.がん種情報.登録時転移部位.名称. %in%
        #                                              c("脳", "中枢神経系", "骨髄", "骨", "肝", "")))$C.CAT調査結果.基本項目.ハッシュID
        met_list <- list(
          Lymph_met = ID_lymph_met,
          Brain_met = ID_brain_met,
          Lung_met = ID_lung_met,
          Bone_met = ID_bone_met,
          Liver_met = ID_liver_met,
          Other_met = ID_other_met
        )
        
        # 各転移フラグ列を作って結合
        clin_tmp <- bind_cols(
          clin_tmp,
          map_dfc(names(met_list), function(met_name) {
            tibble(!!met_name := ifelse(
              clin_tmp$C.CAT調査結果.基本項目.ハッシュID %in% unlist(met_list[[met_name]]),
              "Yes", "No"
            ))
          })
        )

        RenList = read.csv(file("source/drug_rename.csv",
                                encoding='UTF-8-BOM'), header = T)
        for(i in 1:length(RenList$P1)){
          clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., RenList$P1[i], RenList$P2[i])
          clin_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(clin_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., RenList$P1[i], RenList$P2[i])
        }
        
        if(!is.null(input$regimen_rename)){
          RenList = data.frame(NULL)
          for(i in 1:length(input$regimen_rename[,1])){
            RenList <- rbind(RenList,
                             read.csv(header = TRUE,
                                      file(input$regimen_rename[[i, 'datapath']],
                                           encoding='UTF-8-BOM')))
          }
          for(i in 1:length(RenList$P1)){
            clin_tmp = clin_tmp %>% dplyr::mutate(
              症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "" &
                  症例.EP前レジメン情報.化学療法レジメン名称 == RenList$P1[i] ~ RenList$P2[i],
                TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
              ))
          }
        }
        if(!is.null(input$drug_rename)){
          drug_rename_list = data.frame(NULL)
          for(i in 1:length(input$drug_rename[,1])){
            drug_rename_list <- rbind(drug_rename_list,
                                      read.csv(header = TRUE,
                                               file(input$drug_rename[[i, 'datapath']],
                                                    encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp = unlist(lapply(list(clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.), function(x) {
            as.vector(drug_rename_list$Rename[match(x, drug_rename_list$Drug)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
              !is.na(症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp) ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp,
              TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
            )
          )
        }
        incProgress(1 / 4)
        # clin_tmp = clin_tmp %>%
        #   dplyr::mutate(HER2_IHC = case_when(
        #     症例.がん種情報.乳.HER2.IHC..名称. == "陽性（3+）" |
        #       症例.がん種情報.食道.胃.腸.HER2.名称. == "陽性（3+）" |
        #       症例.がん種情報.乳.HER2.IHC..名称. == "境界域（2+）" & 症例.がん種情報.乳.HER2.FISH..名称. == "陽性" |
        #       症例.がん種情報.食道.胃.腸.HER2.名称. == "境界域（2+）" & 症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称. == "陽性" ~ "Positive",
        #     症例.がん種情報.乳.HER2.IHC..名称. %in% c("陰性",  "陰性（1+）") |
        #       症例.がん種情報.食道.胃.腸.HER2.名称. %in% c("陰性",  "陰性（1+）") |
        #       症例.がん種情報.乳.HER2.IHC..名称. == "境界域（2+）" & 症例.がん種情報.乳.HER2.FISH..名称. %in% c("equivocal", "陰性") |
        #       症例.がん種情報.食道.胃.腸.HER2.名称. == "境界域（2+）" & 症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称. %in% c("equivocal", "陰性") ~ "Negative",
        #     TRUE ~ "Unknown"
        #   ),
        #   MSI_PCR = case_when(
        #     症例.がん種情報.固形がん.マイクロサテライト不安定性.名称. == "陽性" ~ "Positive",
        #     症例.がん種情報.固形がん.マイクロサテライト不安定性.名称. == "陰性" ~ "Negative",
        #     TRUE ~ "Unknown"
        #   ),
        #   MMR_IHC = case_when(
        #     症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称. == "dMMR(欠損)" ~ "dMMR",
        #     症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称. == "pMMR(正常)" ~ "pMMR",
        #     TRUE ~ "Unknown"
        #   )
        # )
        clin_tmp <- clin_tmp %>%
          dplyr::mutate(
            EP_option = case_when(
              症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. == "はい" ~ 1,
              TRUE ~ 0
            ),
            EP_treat = case_when(
              症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した" ~ 1,
              TRUE ~ 0
            )
          )
        clin_tmp = clin_tmp %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        症例.基本情報.年齢,
                        Lymph_met,
                        Brain_met,
                        Lung_met,
                        Bone_met,
                        Liver_met,
                        Other_met,
                        # Not_brain_bone_liver_met,
                        EP_option,
                        EP_treat,
                        症例.背景情報.病理診断名,
                        症例.背景情報.臨床診断名,
                        症例.検体情報.病理診断名,
                        症例.基本情報.性別.名称.,
                        症例.基本情報.がん種.OncoTree.,
                        症例.基本情報.がん種.OncoTree..名称.,
                        症例.基本情報.がん種.OncoTree.LEVEL1.,
                        症例.検体情報.パネル.名称.,
                        症例.背景情報.ECOG.PS.名称.,
                        症例.背景情報.家族歴有無.名称.,
                        症例.背景情報.喫煙歴有無.名称.,
                        症例.背景情報.喫煙年数,
                        #症例.背景情報.喫煙１日本数,
                        症例.背景情報.アルコール多飲有無.名称.,
                        症例.背景情報.重複がん有無.異なる臓器..名称.,
                        症例.背景情報.多発がん有無.同一臓器..名称.,
                        症例.がん種情報.登録時転移有無.名称.,
                        症例.がん種情報.登録時転移部位.名称.,
                        症例.検体情報.腫瘍細胞含有割合,
                        症例.検体情報.検体採取部位.名称.,
                        症例.検体情報.検体採取日.腫瘍組織.,
                        症例.EP前レジメン情報.治療ライン.名称.,
                        症例.EP前レジメン情報.実施目的.名称.,
                        症例.EP前レジメン情報.化学療法レジメン名称,
                        症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                        症例.EP前レジメン情報.投与開始日,
                        症例.EP前レジメン情報.投与終了日,
                        症例.EP前レジメン情報.終了理由.名称.,
                        症例.EP前レジメン情報.レジメン継続区分.名称.,
                        症例.EP前レジメン情報.最良総合効果.名称.,
                        症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                        症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                        症例.EP後レジメン情報.治療ライン,
                        症例.EP後レジメン情報.治療ライン.名称.,
                        症例.EP後レジメン情報.化学療法レジメン名称,
                        症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                        症例.EP後レジメン情報.投与開始日,
                        症例.EP後レジメン情報.投与終了日,
                        症例.EP後レジメン情報.終了理由.名称.,
                        症例.EP後レジメン情報.レジメン継続区分.名称.,
                        症例.EP後レジメン情報.最良総合効果.名称.,
                        症例.転帰情報.転帰.名称.,
                        症例.背景情報.診断日,
                        症例.管理情報.登録日,
                        症例.転帰情報.最終生存確認日,
                        症例.転帰情報.死亡日#,
          )
        incProgress(1 / 4)
        save(clin_tmp, file="source/clinical_data.rda")
      }
    })
    clin_tmp = clin_tmp %>% filter(C.CAT調査結果.基本項目.ハッシュID != "9AB8F8522917C3FB5675D30899E0A7ED56E691B93677068EFCCB8BCE8EE4A98C")
    return(clin_tmp)
  })

  Data_report_raw =  reactive({
    withProgress(message = "Mutation data loading.", {
      if(file.exists("source/variant_data.rda")){
        load(file="source/variant_data.rda")
      } else {
        Encode = guess_encoding(input$report_files[[1, 'datapath']])[[1]][[1]]
        if(Encode == "Shift_JIS"){
          Encode = "CP932"
        }
        clin_tmp <- read_csv(col_names = TRUE,
                             file = input$report_files$datapath,
                             locale = locale(encoding=Encode),
                             num_threads=max(1, parallel::detectCores() - 1, na.rm = TRUE),
                             progress=FALSE,
                             show_col_types=FALSE,
                             name_repair=make.names,
                             col_select = c(
                               C.CAT調査結果.基本項目.ハッシュID,
                               C.CAT調査結果.変異情報.変異種類,
                               C.CAT調査結果.変異情報.変異由来,
                               C.CAT調査結果.変異情報.マーカー,
                               C.CAT調査結果.変異情報.変異内容,
                               C.CAT調査結果.変異情報.マーカー詳細,
                               C.CAT調査結果.変異情報.変異アレル頻度,
                               C.CAT調査結果.変異情報.コピー数変化,
                               C.CAT調査結果.変異情報.クロモソーム番号,
                               C.CAT調査結果.変異情報.物理位置,
                               C.CAT調査結果.変異情報.リファレンス塩基,
                               C.CAT調査結果.変異情報.塩基変化,
                               C.CAT調査結果.変異情報.変異種類,
                               C.CAT調査結果.変異情報TMB.TMB,
                               C.CAT調査結果.エビデンス.エビデンスレベル,
                               C.CAT調査結果.エビデンス.薬剤,
                               C.CAT調査結果.変異情報.ToMMoアレル頻度
                             ),
                             col_types = cols(
                               C.CAT調査結果.基本項目.ハッシュID = col_character(),
                               C.CAT調査結果.変異情報.変異種類 = col_character(),
                               C.CAT調査結果.変異情報.変異由来 = col_character(),
                               C.CAT調査結果.変異情報.マーカー = col_character(),
                               C.CAT調査結果.変異情報.変異内容 = col_character(),
                               C.CAT調査結果.変異情報.マーカー詳細 = col_character(),
                               C.CAT調査結果.変異情報.変異アレル頻度 = col_character(),
                               C.CAT調査結果.変異情報.コピー数変化 = col_character(),
                               C.CAT調査結果.変異情報.クロモソーム番号 = col_character(),
                               C.CAT調査結果.変異情報.物理位置 = col_character(),
                               C.CAT調査結果.変異情報.リファレンス塩基 = col_character(),
                               C.CAT調査結果.変異情報.塩基変化 = col_character(),
                               C.CAT調査結果.変異情報.変異種類 = col_character(),
                               C.CAT調査結果.変異情報TMB.TMB = col_character(),
                               C.CAT調査結果.エビデンス.エビデンスレベル = col_character(),
                               C.CAT調査結果.エビデンス.薬剤 = col_character(),
                               C.CAT調査結果.変異情報.ToMMoアレル頻度 = col_integer(),
                              .default = col_guess()
                             ))
        incProgress(1 / 4)
        #clin_tmp = as.data.frame(clin_tmp)
        clin_tmp = data.table(clin_tmp)
        clin_tmp = clin_tmp %>%
          dplyr::filter(
            !str_detect(C.CAT調査結果.変異情報.マーカー, ",") &
              C.CAT調査結果.変異情報.マーカー != "" &
              C.CAT調査結果.変異情報.変異種類 != "expression"
          )
        clin_tmp = clin_tmp %>% dplyr::mutate(
          C.CAT調査結果.変異情報.変異種類 = case_when(
            C.CAT調査結果.変異情報.変異種類 == "rearrangement" &
              C.CAT調査結果.変異情報.変異内容 == "rearrangements" ~ "rearrangement",
            str_detect(C.CAT調査結果.変異情報.変異種類, "exon skipping") ~ "exon_skipping",
            str_detect(C.CAT調査結果.変異情報.変異内容, "exon skipping") ~ "exon_skipping",
            C.CAT調査結果.変異情報.変異種類 == "rearrangement" &
              C.CAT調査結果.変異情報.変異内容 == "long deletion" ~ "deletion",
            C.CAT調査結果.変異情報.変異種類 == "rearrangement" ~ C.CAT調査結果.変異情報.変異内容,
            C.CAT調査結果.変異情報.変異種類 == "copy_number_alteration" ~ C.CAT調査結果.変異情報.変異内容,
            C.CAT調査結果.変異情報.変異種類 == "small_scale_variant" &
             str_detect(C.CAT調査結果.変異情報.変異内容, "\\*") ~ "nonsense_frameshift",
            C.CAT調査結果.変異情報.変異種類 == "small_scale_variant"  ~ "small_scale_variant",
            C.CAT調査結果.変異情報.変異種類 == "other_biomarker" ~ "TMB_MSI_high",
            TRUE ~ C.CAT調査結果.変異情報.変異種類
          ))
        if(!is.null(input$reaanotation)){
          reaanotation_list = data.frame(NULL)
          for(i in 1:length(input$reaanotation[,1])){
            reaanotation_list <- rbind(reaanotation_list,
                                          read.csv(header = TRUE,
                                                   file(input$reaanotation[[i, 'datapath']],
                                                        encoding='UTF-8-BOM')))
          }
          clin_tmp$Gene_alt = paste0(clin_tmp$C.CAT調査結果.変異情報.マーカー, "_", clin_tmp$C.CAT調査結果.変異情報.変異内容)
          clin_tmp$C.CAT調査結果.エビデンス.エビデンスレベル_tmp = unlist(lapply(list(clin_tmp$Gene_alt), function(x) {
            as.vector(reaanotation_list$Level[match(x, reaanotation_list$Gene_alt)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            C.CAT調査結果.エビデンス.エビデンスレベル = case_when(
              !is.na(C.CAT調査結果.エビデンス.エビデンスレベル_tmp) ~ C.CAT調査結果.エビデンス.エビデンスレベル_tmp,
              TRUE ~ C.CAT調査結果.エビデンス.エビデンスレベル
            ))
          clin_tmp = clin_tmp %>% dplyr::select(-Gene_alt, -C.CAT調査結果.エビデンス.エビデンスレベル_tmp)
        }
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX2\\-1", "NKX2XXXX1")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX3\\-1", "NKX3XXXX1")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H1\\-2", "H1XXXX2")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-4", "H3XXXX4")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-5", "H3XXXX5")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-3B", "H3XXXX3B")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-3A", "H3XXXX3A")
        )
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "-", "_")
          )
        
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX2XXXX1", "NKX2_1")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX3XXXX1", "NKX3_1")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H1XXXX2", "H1_2")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX4", "H3_4")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX5", "H3_5")
          ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX3B", "H3_3B")
          ) %>%
          dplyr::mutate(
          C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX3A", "H3_3A")
        )
        incProgress(1 / 4)
        Data_MAF = clin_tmp
        Data_MAF$C.CAT調査結果.変異情報.ToMMoアレル頻度[is.na(Data_MAF$C.CAT調査結果.変異情報.ToMMoアレル頻度)] = 0
        Data_MAF$C.CAT調査結果.変異情報.物理位置2 =
          Data_MAF$C.CAT調査結果.変異情報.物理位置
        Data_MAF = Data_MAF %>%
          dplyr::select(C.CAT調査結果.変異情報.マーカー,
                        C.CAT調査結果.変異情報.クロモソーム番号,
                        C.CAT調査結果.変異情報.物理位置,
                        C.CAT調査結果.変異情報.物理位置2,
                        C.CAT調査結果.変異情報.リファレンス塩基,
                        C.CAT調査結果.変異情報.塩基変化,
                        C.CAT調査結果.変異情報.変異種類,
                        C.CAT調査結果.エビデンス.エビデンスレベル,
                        C.CAT調査結果.エビデンス.薬剤,
                        C.CAT調査結果.基本項目.ハッシュID,
                        C.CAT調査結果.変異情報.変異アレル頻度,
                        C.CAT調査結果.変異情報.変異内容,
                        C.CAT調査結果.変異情報TMB.TMB,
                        C.CAT調査結果.変異情報.ToMMoアレル頻度)
        colnames(Data_MAF) = c("Hugo_Symbol",
                               "Chromosome",
                               "Start_Position",
                               "End_Position",
                               "Reference_Allele",
                               "Tumor_Seq_Allele2",
                               "Variant_Classification",
                               "Evidence_level",
                               "Drug",
                               "Tumor_Sample_Barcode",
                               "VAF",
                               "amino.acid.change",
                               "TMB",
                               "MinorAF")
        Data_MAF = Data_MAF %>% dplyr::mutate(
          Variant_Type = case_when(
            Hugo_Symbol == "MSI" ~ "MSI",
            Hugo_Symbol == "TMB" ~ "TMB",
            nchar(Reference_Allele) == 1 &
              nchar(Tumor_Seq_Allele2) == 1 ~ "SNV",
            nchar(Reference_Allele) == 2 &
              nchar(Tumor_Seq_Allele2) == 2 ~ "DNP",
            nchar(Reference_Allele) == 3 &
              nchar(Tumor_Seq_Allele2) == 3 ~ "TNP",
            nchar(Reference_Allele) == 1 &
              Reference_Allele != "-" &
              nchar(Tumor_Seq_Allele2) > 1 ~ "INS",
            nchar(Reference_Allele) > 1 &
              Tumor_Seq_Allele2 != "-" ~ Variant_Classification,
            TRUE ~ "Other"
          ))
        incProgress(1 / 4)
        TMB_tmp = str_split(Data_MAF$TMB, "Muts", simplify = TRUE)[,1]
        TMB_tmp = str_split(TMB_tmp, "mutations", simplify = TRUE)[,1]
        Data_MAF$TMB = as.numeric(TMB_tmp)
        if(length(Data_MAF[is.na(Data_MAF$TMB),]$TMB) > 0){
          Data_MAF[is.na(Data_MAF$TMB),]$TMB = 0
        }
        if(length(Data_MAF[is.na(Data_MAF$Evidence_level),]$Evidence_level) > 0){
          Data_MAF[is.na(Data_MAF$Evidence_level),]$Evidence_level = ""
        }
        Data_report_TMB = Data_MAF %>%
          dplyr::filter(Hugo_Symbol == "TMB") %>%
          dplyr::distinct(Tumor_Sample_Barcode, TMB, .keep_all = T) %>%
          dplyr::arrange(Tumor_Sample_Barcode) %>%
          dplyr::mutate(
            Evidence_level = case_when(
              TMB < 10 ~ "",
              TMB >= 10 ~ "F",
              TRUE ~ ""
            ),
            amino.acid.change = case_when(
              TMB < 10 ~ "",
              TMB >= 10 ~ "high",
              TRUE ~ ""
            )
          )
        Data_MAF = Data_MAF %>%
          dplyr::arrange(Tumor_Sample_Barcode) %>%
          dplyr::filter(Hugo_Symbol != "TMB") %>%
          dplyr::mutate(Evidence_level = case_when(
            Hugo_Symbol == "MSI" & amino.acid.change == "high" & Evidence_level == "" ~ "F",
            TRUE ~ Evidence_level
          ))
        Data_MAF = rbind(Data_MAF, Data_report_TMB)
        incProgress(1 / 4)
        save(Data_MAF, file="source/variant_data.rda")
      }
    })
    Data_MAF = Data_MAF %>% filter(Tumor_Sample_Barcode != "9AB8F8522917C3FB5675D30899E0A7ED56E691B93677068EFCCB8BCE8EE4A98C")
    return(Data_MAF)
  })

  Data_case =  reactive({
    clin_tmp = Data_case_raw()
    if(!is.null(input$histology_group_1)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_1 ~ str_replace(str_replace(input$histology_group_1_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_1 ~ input$histology_group_1_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
        
      )
    }
    if(!is.null(input$histology_group_2)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_2 ~ str_replace(str_replace(input$histology_group_2_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_2 ~ input$histology_group_2_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
      )
    }
    if(!is.null(input$histology_group_3)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_3 ~ str_replace(str_replace(input$histology_group_3_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_3 ~ input$histology_group_3_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
      )
    }
    if(!is.null(input$histology_group_4)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_4 ~ str_replace(str_replace(input$histology_group_4_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_4 ~ input$histology_group_4_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
      )
    }
    if(!is.null(input$minimum_pts)){
      Cancername = table((clin_tmp %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.がん種.OncoTree.)
                          )$症例.基本情報.がん種.OncoTree.)
      Cancername = names(Cancername[Cancername >= input$minimum_pts])
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree. %in% Cancername ~ 症例.基本情報.がん種.OncoTree..名称.,
          TRUE ~ 症例.基本情報.がん種.OncoTree.LEVEL1.
      ))
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree. %in% Cancername ~ 症例.基本情報.がん種.OncoTree.,
          TRUE ~ 症例.基本情報.がん種.OncoTree.LEVEL1.
        ))
    }
    if(!is.null(input$histology)){
      clin_tmp = clin_tmp %>%
      dplyr::filter(
        症例.基本情報.がん種.OncoTree..名称. %in% input$histology
      )
    }
    if(!is.null(input$panel)){
      clin_tmp = clin_tmp %>%
        dplyr::filter(
          症例.検体情報.パネル.名称. %in% input$panel
        )
    }
    if(!is.null(input$PS)){
      clin_tmp = clin_tmp %>%
        dplyr::filter(
          症例.背景情報.ECOG.PS.名称. %in% input$PS
        )
    }
    if(!is.null(input$sex)){
      clin_tmp = clin_tmp %>%
        dplyr::filter(
          症例.基本情報.性別.名称. %in% input$sex
        )
    }
    if(!is.null(input$smoking)){
      clin_tmp = clin_tmp %>%
        dplyr::filter(
          症例.背景情報.喫煙歴有無.名称. %in% input$smoking
        )
    }
    if(!is.null(input$year)){
      if(!sum(as.integer(input$year)) == sum(unique(as.POSIXlt(as.Date(clin_tmp$症例.管理情報.登録日))$year + 1900))){
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            症例.管理情報.登録日 = as.Date(症例.管理情報.登録日)
          ) %>%
          dplyr::filter(
            !is.null(症例.管理情報.登録日)
          ) %>%
          dplyr::filter(
            (as.POSIXlt(症例.管理情報.登録日)$year + 1900) %in% as.integer(input$year)
          )
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            症例.管理情報.登録日 = as.character(症例.管理情報.登録日)
          )
      }
    }
    if(!is.null(input$age)){
      if(any(input$age != c(min(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE),
                            max(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE)))){
        clin_tmp$症例.基本情報.年齢 = tidyr::replace_na(clin_tmp$症例.基本情報.年齢, -1) 
        clin_tmp = clin_tmp %>% dplyr::filter(
          症例.基本情報.年齢 >= input$age[1] &
            症例.基本情報.年齢 <= input$age[2]
        )
      }
    }
    if(!is.null(input$mid_age)){
      clin_tmp = clin_tmp %>% dplyr::mutate(YoungOld = case_when(
        症例.基本情報.年齢 < input$mid_age ~ "Younger",
        TRUE ~ "Older"))
    }
    clin_tmp = clin_tmp %>% dplyr::distinct() %>% dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
    return(clin_tmp)
  })

  Data_report =  reactive({
    Data_MAF = Data_report_raw()
    if(!is.null(input$special_gene_annotation)){
      if(input$special_gene_annotation == "This gene is also analyzed for VUS and Benign" &
         !is.null(input$special_gene)){
        Data_MAF = Data_MAF %>% dplyr::mutate(
          Evidence_level = case_when(
            Hugo_Symbol == input$special_gene &
              Evidence_level == "" ~ "F",
            TRUE ~ Evidence_level
          )
        )
      }
    }
    if(!is.null(input$special_gene_independent)){
      if(input$special_gene_independent == "Treat as independent genes in analyses" &
       !is.null(input$special_gene)){
        Data_MAF = Data_MAF %>% dplyr::mutate(
          Hugo_Symbol = case_when(
            Hugo_Symbol == input$special_gene &
              amino.acid.change %in% input$special_gene_mutation_1 ~ paste0(Hugo_Symbol, "_", input$special_gene_mutation_1_name),
            Hugo_Symbol == input$special_gene &
              amino.acid.change %in% input$special_gene_mutation_2 ~ paste0(Hugo_Symbol, "_", input$special_gene_mutation_2_name),
            Hugo_Symbol == input$special_gene ~ paste0(Hugo_Symbol, "_NOS"),
            TRUE ~ Hugo_Symbol
          )
        )
      }
    }
    if(!is.null(input$gene_group_analysis)){
      if(input$gene_group_analysis == "Only cases with mutations in the gene set are analyzed" &
       !is.null(input$gene)){
        if(input$patho == "Only pathogenic muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene & Evidence_level == "F"
          ))$Tumor_Sample_Barcode)
        } else if(input$patho == "All muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene
          ))$Tumor_Sample_Barcode)
        }
        Data_MAF = Data_MAF %>%
          dplyr::filter(Tumor_Sample_Barcode %in%
                          ID_geneset)
      }
      if(input$gene_group_analysis == "Only cases without mutations in the gene set are analyzed" &
         !is.null(input$gene)){
        if(input$patho == "Only pathogenic muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene & Evidence_level == "F"
          ))$Tumor_Sample_Barcode)
        } else if(input$patho == "All muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene
          ))$Tumor_Sample_Barcode)
        }
        Data_MAF = Data_MAF %>%
          dplyr::filter(!Tumor_Sample_Barcode %in% ID_geneset)
        ALL_ID =  (Data_case_raw() %>% dplyr::filter(
          !C.CAT調査結果.基本項目.ハッシュID %in% ID_geneset
        ))$C.CAT調査結果.基本項目.ハッシュID
        DEFF_ID = ALL_ID[!ALL_ID %in% Data_MAF$Tumor_Sample_Barcode]
        if(length(DEFF_ID)>0){
          TEST_ALL = NULL
          TEST = (Data_MAF %>% dplyr::filter(
            Hugo_Symbol == "TMB"
          ))[1,]
          TEST$TMB=0
          TEST$Evidence_level = ""
          TEST$Drug = ""
          for(i in 1:length(DEFF_ID)){
            TEST$Tumor_Sample_Barcode = DEFF_ID[i]
            TEST_ALL = rbind(TEST_ALL, TEST)
          }
          Data_MAF = rbind(Data_MAF, TEST_ALL)
        }
      }
    }
    Data_MAF = Data_MAF %>% dplyr::distinct() %>% dplyr::arrange(Tumor_Sample_Barcode)
    return(Data_MAF)}
  )
  
  observeEvent(input$start_setting, {
    withProgress(message = sample(nietzsche)[1], {
      output$select_histology = renderUI({ 
        pickerInput("histology", "Filter by histology",
                    choices = sort(unique(Data_case_raw()$症例.基本情報.がん種.OncoTree..名称.)), 
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    selected = sort(unique(Data_case_raw()$症例.基本情報.がん種.OncoTree..名称.)), multiple = TRUE)
      })
      output$select_histology_group_1 = renderUI({ 
        pickerInput("histology_group_1", "Histology type 1 to be analyzed (if none, not selected)",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology, multiple = TRUE)
      })
      incProgress(1 / 8)
      output$select_histology_group_1_name = renderUI({ 
        pickerInput("histology_group_1_name", "Name for histology type 1",
                    options = list(`actions-box` = TRUE),
                    input$histology_group_1, multiple = FALSE)
      })
      output$select_histology_group_2 = renderUI({ 
        pickerInput("histology_group_2", "Histology type 2 to be analyzed",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology[-which(input$histology %in%
                                             input$histology_group_1)], multiple = TRUE)
      })
      output$select_histology_group_2_name = renderUI({ 
        pickerInput("histology_group_2_name", "Name for histology type 2",
                    input$histology_group_2, multiple = FALSE)
      })
      incProgress(1 / 8)
      output$select_histology_group_3 = renderUI({ 
        pickerInput("histology_group_3", "Histology type 3 to be analyzed",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology[-which(input$histology %in%
                                             c(input$histology_group_1,
                                               input$histology_group_2))], multiple = TRUE)
      })
      incProgress(1 / 8)
      output$select_histology_group_3_name = renderUI({ 
        pickerInput("histology_group_3_name", "Name for histology type 3",
                    options = list(`actions-box` = TRUE),
                    input$histology_group_3, multiple = FALSE)
      })
      output$select_histology_group_4 = renderUI({ 
        pickerInput("histology_group_4", "Histology type 4 to be analyzed",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology[-which(input$histology %in%
                                             c(input$histology_group_1,
                                               input$histology_group_2,
                                               input$histology_group_3))], multiple = TRUE)
      })
      output$select_histology_group_4_name = renderUI({ 
        pickerInput("histology_group_4_name", "Name for histology type 4",
                    options = list(`actions-box` = TRUE),
                    input$histology_group_4, multiple = FALSE)
      })
      output$select_sex = renderUI({ 
        pickerInput("sex", "Filter by sex",
                    choices = sort(unique(Data_case_raw()$症例.基本情報.性別.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.基本情報.性別.名称.)), multiple = TRUE)
      })
      output$select_minimum_pts = renderUI({ 
        numericInput("minimum_pts", "Minimum patients for each histology",
                     value = 1,
                     min = 1,
                     max = length((Data_case_raw() %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID))$C.CAT調査結果.基本項目.ハッシュID),
                     step = 1)
      })
      incProgress(1 / 8)
      output$select_panel = renderUI({ 
        pickerInput("panel", "Filter by panel",
                    choices = sort(unique(Data_case_raw()$症例.検体情報.パネル.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.検体情報.パネル.名称.)), multiple = TRUE)
      })
      output$select_year = renderUI({ 
        pickerInput("year", "Filter by test-year",
                    choices = sort(unique(as.POSIXlt(as.Date(Data_case_raw()$症例.管理情報.登録日))$year + 1900)),
                    selected = sort(unique(as.POSIXlt(as.Date(Data_case_raw()$症例.管理情報.登録日))$year + 1900)), multiple = TRUE)
      })
      output$select_age = renderUI({ 
        sliderInput(inputId = "age", label = "Age for analysis",
                    value = c(
                      min(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE),
                      max(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE)),
                    min = min(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE),
                    max = max(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE),
                    step = 1)
      })
      output$select_mid_age = renderUI({
        tmp = Data_case_raw() %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.年齢)
        sliderInput(inputId = "mid_age", label = "Threshold age for oncoprint",
                    #value = median(tmp$症例.基本情報.年齢, na.rm = TRUE),
                    value = 60,
                    min = min(tmp$症例.基本情報.年齢, na.rm = TRUE),
                    max = max(tmp$症例.基本情報.年齢, na.rm = TRUE),
                    step = 1)
      })
      incProgress(1 / 8)
      output$select_PS = renderUI({ 
        pickerInput("PS", "Filter by performance status",
                    choices = sort(unique(Data_case_raw()$症例.背景情報.ECOG.PS.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.背景情報.ECOG.PS.名称.)), multiple = TRUE)
      })
      output$select_smoking = renderUI({ 
        pickerInput("smoking", "Filter by smoking status",
                    choices = sort(unique(Data_case_raw()$症例.背景情報.喫煙歴有無.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.背景情報.喫煙歴有無.名称.)), multiple = TRUE)
      })
      output$select_gene = renderUI({
        candidate = sort(unique(c(Data_report_raw()$Hugo_Symbol,
                      paste0(input$special_gene, "_", input$special_gene_mutation_1_name),
                      paste0(input$special_gene, "_", input$special_gene_mutation_2_name),
                      paste0(input$special_gene, "_NOS"))))
        candidate = candidate[!candidate %in% c("", "_", "_NOS", paste0(input$special_gene, "_"))]
        candidate = candidate[!is.na(candidate)]
        pickerInput("gene", "Genes of interest (if any)",
                    candidate,
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                     selected = unique(names(sort(table(Data_report_raw()$Hugo_Symbol),
                                                  decreasing = T)))[1],
                    multiple = TRUE)
      })
      output$select_lolliplot_gene = renderUI({ 
        pickerInput("lolliplot_gene", "Gene for lolliplot (if any)",
                    options = list(`live-search`=TRUE),
                    choices = c("", sort(unique(Data_report_raw()$Hugo_Symbol))), multiple = FALSE)
      })
      output$select_lolliplot_no = renderUI({
        tmp = Data_case_raw() %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.年齢)
        sliderInput(inputId = "lolliplot_no", label = "Threshold mutation count for lolliplot",
                    value = 30,
                    min = 1,
                    max = 100,
                    step = 1)
      })
      
      
      output$select_gene_group_1 = renderUI({ 
        pickerInput("gene_group_1", "Gene-set 1 of interest (if any)",
                    input$gene,
                    options = list(`actions-box` = TRUE),
                    multiple = TRUE)
      })
      output$select_gene_group_2 = renderUI({ 
        pickerInput("gene_group_2", "Gene-set 2 of interest (if any)",
                    input$gene[-which(input$gene %in% input$gene_group_1)],
                    options = list(`actions-box` = TRUE),
                    multiple = TRUE)
      })
      incProgress(1 / 8)
      output$select_special_gene = renderUI({
        pickerInput("special_gene", "Gene to analyze (if any)", c("", sort(unique(Data_report_raw()$Hugo_Symbol))), multiple = FALSE)
      })
      output$select_special_gene_mutation_1 = renderUI({
        if(!is.null(input$special_gene)){
          pickerInput("special_gene_mutation_1", "Variants 1", sort(unique(
            (Data_report_raw() %>% dplyr::filter(Hugo_Symbol == input$special_gene | (input$special_gene != "" & str_detect(Hugo_Symbol, paste0(input$special_gene, "_")))))$amino.acid.change)),
            options = list(`actions-box` = TRUE),
            multiple = TRUE)
        } else {
          pickerInput("special_gene_mutation_1", "Variants 1", NULL,
                      options = list(`actions-box` = TRUE),
                      multiple = TRUE)
        }
      })
      output$select_special_gene_mutation_1_name = renderUI({
        if(!is.null(input$special_gene_mutation_1)){
          textInput(inputId = "special_gene_mutation_1_name", label = "Name for variants 1",
                  value = input$special_gene_mutation_1[[1]])
        } else {
          textInput(inputId = "special_gene_mutation_1_name", label = "Name for variants 1",
                    value = NULL)
        }
      })
      output$select_special_gene_mutation_2 = renderUI({
        if(!is.null(input$special_gene)){
          tmp_mut = sort(unique(
            (Data_report_raw() %>% dplyr::filter(Hugo_Symbol == input$special_gene | (input$special_gene != "" & str_detect(Hugo_Symbol, paste0(input$special_gene, "_")))))$amino.acid.change))
          pickerInput("special_gene_mutation_2", "Variants 2", tmp_mut[!tmp_mut %in% input$special_gene_mutation_1],
                      options = list(`actions-box` = TRUE),
                      multiple = TRUE)
        } else{
          pickerInput("special_gene_mutation_2", "Variants 2", NULL,
                      options = list(`actions-box` = TRUE),
                      multiple = TRUE)
        }
      })
      output$select_special_gene_mutation_2_name = renderUI({
        if(!is.null(input$special_gene_mutation_2)){
          textInput(inputId = "special_gene_mutation_2_name", label = "Name for variants 2",
                  value = input$special_gene_mutation_2[[1]])
        } else{
          textInput(inputId = "special_gene_mutation_2_name", label = "Name for variants 2",
                    value = NULL)
        }
      })
      output$select_gene_no = renderUI({ 
        numericInput(inputId = "gene_no", label = "Gene number for oncoprint",
                     value = min(GENE_NO_THRESHOLD, length(sort(unique(Data_report_raw()$Hugo_Symbol)))),
                     min = max(1, length(input$gene)),
                     max = length(sort(unique(Data_report_raw()$Hugo_Symbol))),
                     step = 1)
      })
      incProgress(1 / 8)
      output$select_patho = renderUI({ 
        radioButtons(
          "patho", "Variants for analysis",
          choices = c("Only pathogenic muts", "All muts"),
          #selected = "All muts")
          selected = "All muts")
      })
      output$select_oncoprint_option = renderUI({ 
        radioButtons(
          "oncoprint_option", "Oncoprint setting",
          choices = c("Sort by mutation frequency", "Sort by pathology", "Sort by age", "Sort by metastasis status at CGP"),
          selected = "Sort by mutation frequency")
      })
      output$select_target_line = renderUI({ 
        pickerInput("target_line", "CTx lines to analyze",
                    choices = c("1","2","3","4", "5~"), 
                    selected =  c("1","2","3","4", "5~"),
                    options = list(`actions-box` = TRUE),
                    multiple = TRUE)
      })
      output$select_gene_group_analysis = renderUI({ 
        radioButtons(
          "gene_group_analysis", "Case selection based on the mutations",
          choices = c("Only cases with mutations in the gene set are analyzed",
                      "Only cases without mutations in the gene set are analyzed",
                      "All cases"),
          selected = "All cases")
      })
      output$select_special_gene_annotation = renderUI({ 
        radioButtons(
          "special_gene_annotation", "Pathological significance of the genes",
          choices = c("Only pathological mutations are included in the analysis",
                      "This gene is also analyzed for VUS and Benign"),
          selected = "Only pathological mutations are included in the analysis")
      })
      output$select_special_gene_independent = renderUI({ 
        radioButtons(
          "special_gene_independent", "Treat specified variants independently?",
          choices = c("Treat as independent genes in analyses",
                      "No"),
          selected = "Treat as independent genes in analyses")
      })

      
      incProgress(1 / 8)
    })
  })
  
  output$select_histology_detail = renderUI({ 
    radioButtons(
      "histology_detail", "Analyze without detailed histology",
      choices = c("Yes, use oncotree 1st level",
                  "No"),
      selected = "Yes, use oncotree 1st level")
  })

  output$select_clonal_hematopoiesis_genes = renderUI({
    pickerInput(
      "clonal_hematopoiesis_genes", "Select clonal hematopoiesis genes",
      choices = CH_GENE_NatComms,
      selected = CH_GENE_NatComms,
      # selected = c("DNMT3A", "TET2", "ASXL1", "TP53", "CHEK2", "ATM", "SF3B1", "U2AF1"),
      options = list(`actions-box` = TRUE, `live-search`=TRUE),
      multiple = TRUE
    )
  })

  observeEvent(input$summary_base, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in%
         c("Only cases with mutations in the gene set are analyzed",
           "Only cases without mutations in the gene set are analyzed")){
      Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      incProgress(1 / 8)

      Data_summary = Data_case_target %>%
        # dplyr::mutate( # treatment only palliative. remove for all treatment.
        #   症例.EP前レジメン情報.投与開始日 = case_when(
        #     症例.EP前レジメン情報.実施目的.名称. %in% c("緩和","その他") ~ 症例.EP前レジメン情報.投与開始日,
        #     TRUE ~ NA
        #   )
        # ) %>%
        dplyr::group_by(C.CAT調査結果.基本項目.ハッシュID) %>%
        dplyr::mutate(
          最終投与開始日_date = min(症例.EP前レジメン情報.投与開始日, na.rm = TRUE)
        ) %>%
        dplyr::mutate(
          症例.EP前レジメン情報.投与開始日 = ifelse(is.infinite(最終投与開始日_date),
                                      NA_character_,
                                      最終投与開始日_date)
        ) %>%
        dplyr::slice(1) %>%
        dplyr::ungroup() %>%
        dplyr::select(-最終投与開始日_date) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
        dplyr::select(
          C.CAT調査結果.基本項目.ハッシュID,
          症例.基本情報.がん種.OncoTree.,
          症例.基本情報.がん種.OncoTree..名称.,
          # 症例.基本情報.がん種.OncoTree.LEVEL1.,
          症例.基本情報.性別.名称.,
          症例.基本情報.年齢,
          症例.背景情報.診断日,
          症例.管理情報.登録日,
          症例.背景情報.家族歴有無.名称.,
          症例.背景情報.喫煙歴有無.名称.,
          症例.背景情報.喫煙年数,
          症例.背景情報.アルコール多飲有無.名称.,
          症例.背景情報.ECOG.PS.名称.,
          症例.背景情報.重複がん有無.異なる臓器..名称.,
          症例.背景情報.多発がん有無.同一臓器..名称.,
          症例.がん種情報.登録時転移有無.名称.,
          症例.検体情報.パネル.名称.,
          症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
          症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
          症例.EP前レジメン情報.投与開始日,
          症例.検体情報.検体採取日.腫瘍組織.,
          Lymph_met,
          Brain_met,
          Lung_met,
          Bone_met,
          Liver_met,
          Other_met,
          # Not_brain_bone_liver_met,
          YoungOld)
      incProgress(1 / 8)
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
      if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }
      incProgress(1 / 8)

      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      } else {
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
      }
      incProgress(1 / 8)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change, .keep_all = T) %>%
        dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
      incProgress(1 / 8)
      if(!input$special_gene == ""){
        ID_special_gene = (Data_MAF_target %>%
          dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
        ID_special_gene_mutation_1 = (Data_MAF_target %>%
                             dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                             amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
        ID_special_gene_mutation_2 = (Data_MAF_target %>%
                             dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                             amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
        Data_summary = Data_summary %>% dplyr::mutate(
          separation_value = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_1 ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
            C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_2 ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
            C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene ~ paste0("Other mut in ", input$special_gene),
            TRUE ~ paste0("No mut in ", input$special_gene)
          )
        )
      } else {
        Data_summary = Data_summary %>% dplyr::mutate(
          separation_value = "No mutation pattern")
      }
      incProgress(1 / 8)
      Data_summary_ID = Data_summary$C.CAT調査結果.基本項目.ハッシュID
      Data_summary = Data_summary %>% dplyr::select(
        -C.CAT調査結果.基本項目.ハッシュID)

      Data_summary$Lymph_met = as.character(Data_summary$Lymph_met)
      Data_summary$Brain_met = as.character(Data_summary$Brain_met)
      Data_summary$Lung_met = as.character(Data_summary$Lung_met)
      Data_summary$Bone_met = as.character(Data_summary$Bone_met)
      Data_summary$Liver_met = as.character(Data_summary$Liver_met)
      Data_summary$Other_met = as.character(Data_summary$Other_met)
      # Data_summary$Not_brain_bone_liver_met = as.character(Data_summary$Not_brain_bone_liver_met)
      
      Data_summary$CTx_start_to_tumor_sampling_days = 
        as.integer((as.Date(Data_summary$症例.検体情報.検体採取日.腫瘍組織.) - as.Date(Data_summary$症例.EP前レジメン情報.投与開始日)))
      Data_summary$CTx_start_to_liquid_sampling_days = 
        as.integer((as.Date(Data_summary$症例.管理情報.登録日) - as.Date(Data_summary$症例.EP前レジメン情報.投与開始日)))
      
      Data_summary = Data_summary %>%
        dplyr::mutate(CTx_start_to_tumor_sampling_days = case_when(
          症例.検体情報.パネル.名称. %in% c('F1Liquid CDx', 'Guardant360 CDx') ~ NA,
          CTx_start_to_tumor_sampling_days < 0 ~ NA,
          TRUE ~ CTx_start_to_tumor_sampling_days
        ), CTx_start_to_liquid_sampling_days = case_when(
          !症例.検体情報.パネル.名称. %in% c('F1Liquid CDx', 'Guardant360 CDx') ~ NA,
          CTx_start_to_liquid_sampling_days < 0 ~ NA,
          TRUE ~ CTx_start_to_liquid_sampling_days
        )
      )
      
      Data_summary$症例.基本情報.年齢.診断時 = 
        Data_summary$症例.基本情報.年齢 - 
        ceiling(as.integer(as.Date(Data_summary$症例.管理情報.登録日) - as.Date(Data_summary$症例.背景情報.診断日))/365.25)
      Data_summary$症例.基本情報.年齢.診断時[Data_summary$症例.基本情報.年齢.診断時 < 0] = NA_integer_
      Data_summary = Data_summary %>% dplyr::select(
        -症例.背景情報.診断日,
        -症例.管理情報.登録日,
        -症例.検体情報.検体採取日.腫瘍組織.,
        -症例.EP前レジメン情報.投与開始日
      )
      
      Data_summary = Data_summary %>%
        dplyr::mutate(
          YoungOld = case_when(
            YoungOld == "Younger" ~ paste0("Younger than ", input$mid_age),
            TRUE ~ paste0(input$mid_age, " and older")
          ))
      colnames_tmp = colnames(Data_summary)
      colnames_tmp[colnames_tmp == "CTx_start_to_tumor_sampling_days"] = "CTx duration from 1L-CTx to sampling (tumor), days"
      colnames_tmp[colnames_tmp == "CTx_start_to_liquid_sampling_days"] = "CTx duration from 1L-CTx to sampling (liquid), days"
      colnames_tmp[colnames_tmp == "症例.基本情報.がん種.OncoTree."] = "Diagnosis"
      colnames_tmp[colnames_tmp == "症例.基本情報.がん種.OncoTree..名称."] = "Diagnosis (OncoTree)"
      colnames_tmp[colnames_tmp == "症例.基本情報.性別.名称."] = "Sex"
      colnames_tmp[colnames_tmp == "症例.基本情報.年齢"] = "Age at CGP"
      colnames_tmp[colnames_tmp == "症例.基本情報.年齢.診断時"] = "Age at diagnosis"
      colnames_tmp[colnames_tmp == "症例.背景情報.家族歴有無.名称."] = "Family cancer history"
      colnames_tmp[colnames_tmp == "症例.背景情報.喫煙歴有無.名称."] = "Smoking history"
      colnames_tmp[colnames_tmp == "症例.背景情報.喫煙年数"] = "Years of smoking"
      colnames_tmp[colnames_tmp == "症例.背景情報.アルコール多飲有無.名称."] = "Alcoholic history"
      colnames_tmp[colnames_tmp == "症例.背景情報.ECOG.PS.名称."] = "Performance status"
      colnames_tmp[colnames_tmp == "症例.背景情報.重複がん有無.異なる臓器..名称."] = "Double cancer in a different organ"
      colnames_tmp[colnames_tmp == "症例.背景情報.多発がん有無.同一臓器..名称."] = "Multiple tumor nodules in the organ"
      colnames_tmp[colnames_tmp == "症例.がん種情報.登録時転移有無.名称."] = "Any distant metastasis"
      colnames_tmp[colnames_tmp == "症例.検体情報.パネル.名称."] = "Cancer gene panel"
      colnames_tmp[colnames_tmp == "症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称."] = "Treatment recommended by expert panel"
      colnames_tmp[colnames_tmp == "症例.EP後レジメン情報.提示された治療薬を投与した.名称."] = "Indicated treatment administered"
      colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
      colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
      colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
      colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
      colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
      colnames_tmp[colnames_tmp == "Other_met"] = "Other organ metastasis"
      colnames_tmp[colnames_tmp == "YoungOld"] = paste0(input$mid_age, " and older")

      colnames(Data_summary) = colnames_tmp

      Data_summary = Data_summary %>%
        dplyr::mutate(
          Sex = case_when(
            Sex == "女" ~ "Female",
            Sex == "未入力・不明" ~ "Unknown",
            Sex == "男" ~ "Male"
          ),
          `Family cancer history` = case_when(
            `Family cancer history` == "あり" ~ "Yes",
            `Family cancer history` == "不明" ~ "Unknown",
            `Family cancer history` == "なし" ~ "No"
          ),
          `Smoking history` = case_when(
            `Smoking history` == "あり" ~ "Yes",
            `Smoking history` == "不明" ~ "Unknown",
            `Smoking history` == "なし" ~ "No"
          ),
          `Alcoholic history` = case_when(
            `Alcoholic history` == "あり" ~ "Yes",
            `Alcoholic history` == "不明" ~ "Unknown",
            `Alcoholic history` == "なし" ~ "No"
          ),
          `Performance status` = case_when(
            `Performance status` == "不明" ~ "Unknown",
            TRUE ~ `Performance status`
          ),
          `Double cancer in a different organ` = case_when(
            `Double cancer in a different organ` == "あり" ~ "Yes",
            `Double cancer in a different organ` == "不明" ~ "Unknown",
            `Double cancer in a different organ` == "なし" ~ "No"
          ),
          `Multiple tumor nodules in the organ` = case_when(
            `Multiple tumor nodules in the organ` == "あり" ~ "Yes",
            `Multiple tumor nodules in the organ` == "不明" ~ "Unknown",
            `Multiple tumor nodules in the organ` == "なし" ~ "No"
          ),
          `Any distant metastasis` = case_when(
            `Any distant metastasis` == "あり" ~ "Yes",
            `Any distant metastasis` == "不明" ~ "Unknown",
            `Any distant metastasis` == "なし" ~ "No"
          ),
          `Treatment recommended by expert panel` = case_when(
            `Treatment recommended by expert panel` == "いいえ" ~ "No",
            `Treatment recommended by expert panel` == "はい" ~ "Yes"
          ),
          `Indicated treatment administered` = case_when(
            `Indicated treatment administered` == "投与した" ~ "Yes",
            `Indicated treatment administered` == "投与しなかった" ~ "No"
          ),
          Age_decade = case_when(
            `Age at CGP` < 50 ~ "~49",
            `Age at CGP` < 60 ~ "50~59",
            `Age at CGP` < 70 ~ "60~69",
            TRUE ~ "70~"
          )
        )
      output$table_summary_1 <- render_gt({
        Data_summary %>% dplyr::select(
          -`Diagnosis`) %>%
          tbl_summary(
            by = "separation_value",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
                          type = list(all_continuous() ~ "continuous2",
                                      all_dichotomous() ~ "categorical"),
                          digits = all_continuous() ~ 2,
                          missing = "ifany") %>%
          modify_caption("**Patient Characteristics** (N = {N})") %>%
          add_n() %>%
          bold_labels() %>% as_gt()
      })
      table_summary <- Data_summary %>% dplyr::select(
        -`Diagnosis`) %>%
        tbl_summary(
          by = "separation_value",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Patient Characteristics** (N = {N})") %>%
        add_n() %>%
        bold_labels()
      as_hux_xlsx(table_summary,file = "source/Table_odds_1.xlsx")
      incProgress(1 / 8)
      output$table_summary_2 <- render_gt({
        Data_summary %>% dplyr::select(
          -"Diagnosis (OncoTree)",
          -separation_value) %>%
          tbl_summary(
            by = "Diagnosis",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Patient Characteristics, by diagnosis** (N = {N})") %>%
          add_n() %>%
          bold_labels() %>% as_gt()
      })
      table_summary <- Data_summary %>% dplyr::select(
        -"Diagnosis (OncoTree)",
        -separation_value) %>%
        tbl_summary(
          by = "Diagnosis",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Patient Characteristics, by diagnosis** (N = {N})") %>%
        add_n() %>%
        bold_labels()
      # gtsummary オブジェクトを flextable オブジェクトに変換
      as_hux_xlsx(table_summary,file = "source/Table_odds_2.xlsx")

      output$table_summary_4 <- render_gt({
        Data_summary %>% dplyr::select(
          -`Diagnosis`,
          -separation_value) %>%
          tbl_summary(
            by = "Cancer gene panel",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Patient Characteristics, by diagnosis** (N = {N})") %>%
          add_n() %>%
          bold_labels() %>% as_gt()
      })
      table_summary <- Data_summary %>% dplyr::select(
        -`Diagnosis`,
        -separation_value) %>%
        tbl_summary(
          by = "Cancer gene panel",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Patient Characteristics, by diagnosis** (N = {N})") %>%
        add_n() %>%
        bold_labels()
      as_hux_xlsx(table_summary,file = "source/Table_odds_3.xlsx")

      Data_summary_table = Data_summary
      Data_summary_table$ID = Data_summary_ID
      if(!DDPS){
        output$table_summary_3 = DT::renderDataTable(Data_summary_table,
                                                 server = FALSE,
                                                 filter = 'top', 
                                                 extensions = c('Buttons'), 
                                                 options = list(pageLength = 100, 
                                                                scrollX = TRUE,
                                                                scrollY = "1000px",
                                                                scrollCollapse = TRUE,
                                                                dom="Blfrtip",
                                                                buttons = c('csv', 'excel', 'copy')))
      }
      incProgress(1 / 8)
    })
  })
  
  observeEvent(input$inter_panel, {
    withProgress(message = sample(nietzsche)[1], {
      if(file.exists("source/Data_summary.rda")){
        load(file="source/Data_summary.rda")
      } else {
        Data_case_target = Data_case()
        Data_platinum_pre = Data_case_target %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
          dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "platin")) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
        Data_case_target$Platinum = ifelse(Data_case_target$C.CAT調査結果.基本項目.ハッシュID %in% Data_platinum_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
        Data_ICI_pre = Data_case_target %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                        症例.EP前レジメン情報.最良総合効果.名称.,
                        症例.EP前レジメン情報.実施目的.名称.,
                        症例.管理情報.登録日,
                        症例.EP前レジメン情報.投与開始日,
                        症例.EP前レジメン情報.投与終了日,
                        症例.EP前レジメン情報.レジメン継続区分.名称.) %>%
          dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in% c("緩和","その他") &
            症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% c(#"Nivolumab,Ipilimumab",
                            #"Nivolumab",
                            #"Ipilimumab",
                            "Pembrolizumab"#,
                            #"Atezolizumab,Paclitaxel",
                            #"Atezolizumab"
                            )) %>%
          dplyr::select(-症例.EP前レジメン情報.実施目的.名称.) %>%
          dplyr::distinct()
        colnames(Data_ICI_pre) = c("C.CAT調査結果.基本項目.ハッシュID", "drug", "RECIST", "entry_date", "ICI_start_data", "ICI_end_date", "ongoing")
        Data_ICI_pre = Data_ICI_pre %>%
          dplyr::filter(ICI_start_data != "") %>%
          dplyr::mutate(
            time_on_treatment = case_when(
              ICI_end_date != "" ~ as.integer(as.Date(ICI_end_date) - as.Date(ICI_start_data)),
              TRUE ~ as.integer(as.Date(entry_date) - as.Date(ICI_start_data))),
            ICI_censor  = case_when(
              !is.na(ongoing) ~ 0,
              TRUE ~ 1)
          ) %>%
          dplyr::select(-entry_date, -ongoing)
        Data_ICI_post = Data_case_target %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                        症例.EP後レジメン情報.最良総合効果.名称.,
                        症例.EP後レジメン情報.投与開始日,
                        症例.EP後レジメン情報.投与終了日) %>%
          dplyr::filter(症例.EP後レジメン情報.薬剤名.YJ一般名.EN. %in% c(#"Nivolumab,Ipilimumab",
                            #"Nivolumab",
                            #"Ipilimumab",
                            "Pembrolizumab"#,
                            #"Atezolizumab,Paclitaxel",
                            #"Atezolizumab"
                          )) %>%
          dplyr::distinct()
        colnames(Data_ICI_post) = c("C.CAT調査結果.基本項目.ハッシュID", "drug", "RECIST", "ICI_start_data", "ICI_end_date")
        Data_ICI_post = Data_ICI_post %>%
          dplyr::filter(ICI_start_data != "" & ICI_end_date != "") %>%
          dplyr::mutate(time_on_treatment = as.integer(as.Date(ICI_end_date) - as.Date(ICI_start_data)))
        Data_ICI_post$ICI_censor = 1
        Data_ICI = rbind(Data_ICI_pre, Data_ICI_post)
        Data_ICI_time = Data_ICI %>%
          dplyr::filter(ICI_start_data != "" & ICI_end_date != "") %>%
          dplyr::arrange(ICI_start_data) %>%
          dplyr::filter(time_on_treatment>=0) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=T) %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_on_treatment, ICI_censor)
        Data_case_target = left_join(Data_case_target, Data_ICI_time, by="C.CAT調査結果.基本項目.ハッシュID")
        Data_ICI$RECIST[is.na(Data_ICI$RECIST)] = "NE"
        Data_ICI$RECIST = factor(Data_ICI$RECIST, levels = c("CR","PR","SD","PD","NE"))
        Data_ICI = Data_ICI %>%
          dplyr::filter(ICI_start_data != "") %>%
          dplyr::arrange(ICI_start_data) %>%
          dplyr::filter(RECIST != "NE") %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=T) %>%
          dplyr::mutate(ICI_Effect = case_when(
            RECIST %in% c("CR","PR") ~ 1,
            TRUE ~ 0
          )) %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, ICI_Effect)
        Data_case_target = left_join(Data_case_target, Data_ICI, by="C.CAT調査結果.基本項目.ハッシュID")
        Data_case_target$症例.背景情報.喫煙年数[is.na(Data_case_target$症例.背景情報.喫煙年数)] = 0
        Data_case_target$症例.背景情報.喫煙年数[Data_case_target$症例.背景情報.喫煙年数==99] = NA
        if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
          Data_case_target = Data_case_target %>%
            dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                            Data_report()$Tumor_Sample_Barcode)
        }
        incProgress(1 / 8)
        setProgress(detail = "2826")
        Data_survival = Data_case_target %>%
          dplyr::mutate(final_observe = case_when(
            症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
            症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
            症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
            TRUE ~ 症例.管理情報.登録日))
        Data_survival = Data_survival %>%
          dplyr::arrange(desc(final_observe)) %>%
          dplyr::mutate(censor = case_when(
            症例.転帰情報.死亡日 != "" ~ 1,
            TRUE ~ 0)) %>%
          dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
        Data_survival = Data_survival %>%
          dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
          dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
        Data_survival_tmp = Data_survival %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
        Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
        Data_survival = Data_survival %>%
          dplyr::mutate(final_observe = as.Date(Data_survival$final_observe),
                        症例.管理情報.登録日 =
                          as.Date(Data_survival$症例.管理情報.登録日),
                        症例.EP前レジメン情報.投与開始日 =
                          as.Date(症例.EP前レジメン情報.投与開始日))
        Data_survival = Data_survival %>%
          dplyr::mutate(time_enroll_final =
                          as.integer(difftime(Data_survival$final_observe,
                                              Data_survival$症例.管理情報.登録日,
                                              units = "days")),
                        time_palliative_final =
                          as.integer(difftime(Data_survival$final_observe,
                                              Data_survival$症例.EP前レジメン情報.投与開始日,
                                              units = "days")))
        Data_survival = Data_survival %>%
          dplyr::mutate(time_palliative_enroll =
                          Data_survival$time_palliative_final -
                          Data_survival$time_enroll_final)
        Data_survival_tmp = Data_survival %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, final_observe) %>%
          dplyr::arrange(desc(final_observe)) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
        Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
        Data_survival_tmp = Data_survival %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_enroll_final) %>%
          dplyr::arrange(desc(time_enroll_final)) %>%
          dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final)) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
        Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
        Data_survival_tmp = Data_survival %>% # CTx only palliative or all CTx
          # dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
          #                 c("その他", "緩和")) %>%
          dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_palliative_enroll, time_palliative_final) %>%
          dplyr::filter(time_palliative_enroll > 0 & time_palliative_enroll < 10000 & !is.na(time_palliative_enroll) & is.finite(time_palliative_enroll)) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
        Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
        Data_summary = Data_case_target
        Data_summary$症例.管理情報.登録日 = as.Date(Data_summary$症例.管理情報.登録日)
        Data_summary$症例.EP前レジメン情報.投与開始日 = as.Date(Data_summary$症例.EP前レジメン情報.投与開始日)
        Data_summary$症例.EP前レジメン情報.治療ライン.名称.[Data_summary$症例.EP前レジメン情報.治療ライン.名称. == "不明"] = "０次治療"
        Data_summary$症例.EP前レジメン情報.治療ライン.名称.[is.na(Data_summary$症例.EP前レジメン情報.治療ライン.名称.)] = "0"
        CTx_start_ID = Data_summary %>% dplyr::filter(症例.EP前レジメン情報.治療ライン.名称. == "１次治療")
        CTx_start_ID = CTx_start_ID %>% dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                                                      症例.管理情報.登録日,
                                                      症例.EP前レジメン情報.投与開始日)
        CTx_start_ID = CTx_start_ID %>%
          dplyr::mutate(CTx_length =
                          round(as.integer(difftime(CTx_start_ID$症例.管理情報.登録日,
                                                    CTx_start_ID$症例.EP前レジメン情報.投与開始日,
                                                    units = "days"))/365.25,digits = 2))
        CTx_start_ID = CTx_start_ID %>% dplyr::select(-症例.管理情報.登録日,
                                                      -症例.EP前レジメン情報.投与開始日)
        CTx_start_ID = CTx_start_ID %>% dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID, desc(CTx_length))
        CTx_start_ID = CTx_start_ID %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T)
        Data_summary = Data_summary %>% dplyr::left_join(CTx_start_ID, by = "C.CAT調査結果.基本項目.ハッシュID")
        Data_summary$CTx_length[is.na(Data_summary$CTx_length)] = 0
        Data_summary$CTx_length[Data_summary$CTx_length < 0] = 0
        Data_summary = Data_summary %>%
          dplyr::arrange(desc(症例.EP前レジメン情報.治療ライン.名称.),.keep_all=TRUE) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
          dplyr::select(
            C.CAT調査結果.基本項目.ハッシュID,
            症例.基本情報.年齢,
            症例.基本情報.がん種.OncoTree.,
            症例.検体情報.パネル.名称.,
            Lymph_met,
            Brain_met,
            Lung_met,
            Bone_met,
            Liver_met,
            症例.基本情報.性別.名称.,
            症例.背景情報.喫煙歴有無.名称.,
            症例.背景情報.喫煙年数,
            症例.背景情報.アルコール多飲有無.名称.,
            症例.背景情報.重複がん有無.異なる臓器..名称.,
            症例.背景情報.多発がん有無.同一臓器..名称.,
            症例.EP前レジメン情報.治療ライン.名称.,
            CTx_length,
            ICI_Effect,
            Platinum,
            time_on_treatment,
            ICI_censor,
            time_enroll_final,
            time_palliative_enroll,
            time_palliative_final,
            censor
        )
        colnames(Data_summary) = c("Tumor_Sample_Barcode",
                                   "Age",
                                   "Cancername",
                                   "Panel",
                                   "Lymph_met",
                                   "Brain_met",
                                   "Lung_met",
                                   "Bone_met",
                                   "Liver_met",
                                   "Sex",
                                   "Smoking_history",
                                   "Smoking_years",
                                   "Alcoholic_history",
                                   "Double_cancer",
                                   "Multiple_nodules",
                                   "Lines",
                                   "CTx_length",
                                   "ICI_Effect",
                                   "Platinum",
                                   "time_on_treatment",
                                   "ICI_censor",
                                   "time_enroll_final",
                                   "time_palliative_enroll",
                                   "time_palliative_final",
                                   "censor"
                                   )
        CTx_unknown_ID = Data_summary$Tumor_Sample_Barcode[Data_summary$Lines == "０次治療"]
        CTx_unknown_ID = data.frame(CTx_unknown_ID)
        colnames(CTx_unknown_ID) = "Tumor_Sample_Barcode"
        if(length(CTx_unknown_ID$Tumor_Sample_Barcode) > 0){
          CTx_unknown_ID$courses = 0
          for(i in 1:length(CTx_unknown_ID$Tumor_Sample_Barcode)){
            Data_case_target_tmp = Data_case_target %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% CTx_unknown_ID$Tumor_Sample_Barcode[i])
            CTx_unknown_ID$courses[i] = max(length(unique(Data_case_target_tmp$症例.EP前レジメン情報.化学療法レジメン名称)),
                                            length(unique(Data_case_target_tmp$症例.EP後レジメン情報.投与開始日)))
          }
        }
        Data_summary = Data_summary %>% dplyr::left_join(CTx_unknown_ID, by = "Tumor_Sample_Barcode")
        Data_summary$courses = as.integer(Data_summary$courses)
        Data_summary$Sex[Data_summary$Sex == "男"] = "Male"
        Data_summary$Sex[Data_summary$Sex == "未入力・不明"] = "Unknown"
        Data_summary$Sex[Data_summary$Sex == "女"] = "Female"
        Data_summary$Smoking_history[Data_summary$Smoking_history == "あり"] = "Yes"
        Data_summary$Smoking_history[Data_summary$Smoking_history == "なし"] = "No"
        Data_summary$Smoking_history[Data_summary$Smoking_history == "不明"] = "Unknown"
        Data_summary$Alcoholic_history[Data_summary$Alcoholic_history == "あり"] = "Yes"
        Data_summary$Alcoholic_history[Data_summary$Alcoholic_history == "なし"] = "No"
        Data_summary$Alcoholic_history[Data_summary$Alcoholic_history == "不明"] = "Unknown"
        Data_summary$Double_cancer[Data_summary$Double_cancer == "あり"] = "Yes"
        Data_summary$Double_cancer[Data_summary$Double_cancer == "なし"] = "No"
        Data_summary$Double_cancer[Data_summary$Double_cancer == "不明"] = "Unknown"
        Data_summary$Multiple_nodules[Data_summary$Multiple_nodules == "あり"] = "Yes"
        Data_summary$Multiple_nodules[Data_summary$Multiple_nodules == "なし"] = "No"
        Data_summary$Multiple_nodules[Data_summary$Multiple_nodules == "不明"] = "Unknown"
        Data_summary = Data_summary %>%
          dplyr::mutate(Lines_pre_CGP = case_when(
            Lines == "５次治療以降" ~ 5,
            Lines == "４次治療" ~ 4,
            Lines == "３次治療" ~ 3,
            Lines == "２次治療" ~ 2,
            Lines == "１次治療" ~ 1,
            Lines == "０次治療" ~ courses,
            Lines == "0" ~ 0
          ))
        Data_summary = Data_summary %>%
          dplyr::mutate(Lines_pre_CGP_2_or_more = case_when(
            Lines_pre_CGP >= 2 ~ 1,
            TRUE ~ 0
          ))
        save(Data_summary, file="source/Data_summary.rda")
      }
      incProgress(1 / 8)
      setProgress(detail = "3007")
      Data_MAF = Data_report() %>%
        dplyr::select(Tumor_Sample_Barcode,
                      Hugo_Symbol,
                      Chromosome,
                      Start_Position,
                      End_Position,
                      Reference_Allele,
                      Tumor_Seq_Allele2,
                      Evidence_level,
                      VAF,
                      Variant_Classification,
                      amino.acid.change,
                      TMB,
                      MinorAF)
      Data_MAF$Reference_Allele = str_replace_all(Data_MAF$Reference_Allele, "\\(ABBREVIATED\\)", "")
      Data_MAF$Tumor_Seq_Allele2 = str_replace_all(Data_MAF$Tumor_Seq_Allele2, "\\(ABBREVIATED\\)", "")
      Data_MAF$Start_Position = as.integer(Data_MAF$Start_Position)
      Data_MAF$End_Position = as.integer(Data_MAF$End_Position)
      Data_MAF_unique = Data_MAF %>%
        dplyr::select(Tumor_Sample_Barcode,TMB) %>%
        dplyr::distinct()
      Data_MAF = Data_MAF %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","F") &
            Variant_Classification %in% c("small_scale_variant", "nonsense_frameshift")
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        Reference_Allele,
                        Tumor_Seq_Allele2,
                        .keep_all = TRUE)
      Data_MAF$unique_position = paste0(Data_MAF$Chromosome, "_",
                                        Data_MAF$Start_Position, "_",
                                        Data_MAF$End_Position, "_",
                                        Data_MAF$Reference_Allele, "_",
                                        Data_MAF$Tumor_Seq_Allele2
                                       )
      Data_MAF$name = paste0(Data_MAF$Hugo_Symbol, "_", Data_MAF$amino.acid.change)
      Data_MAF = Data_MAF %>%
        dplyr::mutate(unique_alt = case_when(
          str_detect(amino.acid.change, "\\*") &
            str_sub(amino.acid.change, 1,1) != "*" &
            (nchar(Reference_Allele) != 1 |
             nchar(Tumor_Seq_Allele2) != 1) ~ paste0(Hugo_Symbol, "_fs_NOS"),
          str_detect(amino.acid.change, "\\*") &
            str_sub(amino.acid.change, 1,1) != "*" &
            Reference_Allele == "C" &
            Tumor_Seq_Allele2 == "T" ~ paste0(Hugo_Symbol, "_trunc_CtoT"),
          str_detect(amino.acid.change, "\\*") &
            str_sub(amino.acid.change, 1,1) != "*" ~ paste0(Hugo_Symbol, "_trunc_NOS"),
          TRUE ~ name
        ))
      Data_MAF = Data_MAF %>%
        dplyr::mutate(CtoT_mutation = case_when(
          Reference_Allele == "C" & Tumor_Seq_Allele2 == "T" |
            Reference_Allele == "G" & Tumor_Seq_Allele2 == "A" ~ "Yes",
          TRUE ~ "No"
        ))
      Data_MAF = Data_MAF %>% left_join(Data_summary, by="Tumor_Sample_Barcode")
      incProgress(1 / 8)
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_summary$Tumor_Sample_Barcode)
      Data_MAF_target$VAF = as.numeric(str_replace(Data_MAF_target$VAF, "\\ \\(.*\\)", ""))

      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx"))$unique_alt))
      F1L_recurrence_table_tmp = data.frame(
        unique_alt = names(F1L_recurrence_table),
        No_found_in_F1L = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "unique_alt")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx"))$unique_alt))
      F1_recurrence_table_tmp = data.frame(
        unique_alt = names(F1_recurrence_table),
        No_found_in_F1 = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "unique_alt")
      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx"))$Hugo_Symbol))
      F1L_recurrence_table_tmp = data.frame(
        Hugo_Symbol = names(F1L_recurrence_table),
        No_found_in_F1L_gene = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "Hugo_Symbol")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx"))$Hugo_Symbol))
      F1_recurrence_table_tmp = data.frame(
        Hugo_Symbol = names(F1_recurrence_table),
        No_found_in_F1_gene = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "Hugo_Symbol")
      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx"))$unique_position))
      F1L_recurrence_table_tmp = data.frame(
        unique_position = names(F1L_recurrence_table),
        No_found_in_F1L_position = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "unique_position")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx"))$unique_position))
      F1_recurrence_table_tmp = data.frame(
        unique_position = names(F1_recurrence_table),
        No_found_in_F1_position = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "unique_position")
      Data_MAF_target$No_found_in_F1L[is.na(Data_MAF_target$No_found_in_F1L)] = 0
      Data_MAF_target$No_found_in_F1[is.na(Data_MAF_target$No_found_in_F1)] = 0
      Data_MAF_target$No_found_in_F1L_gene[is.na(Data_MAF_target$No_found_in_F1L_gene)] = 0
      Data_MAF_target$No_found_in_F1_gene[is.na(Data_MAF_target$No_found_in_F1_gene)] = 0
      Data_MAF_target$No_found_in_F1L_position[is.na(Data_MAF_target$No_found_in_F1L_position)] = 0
      Data_MAF_target$No_found_in_F1_position[is.na(Data_MAF_target$No_found_in_F1_position)] = 0
      Data_MAF_target$F1_patients = length((Data_summary %>%
                                              dplyr::filter(Panel == "FoundationOne CDx") %>%
                                              dplyr::distinct(Tumor_Sample_Barcode))$Tumor_Sample_Barcode)
      Data_MAF_target$F1L_patients = length((Data_summary %>%
                                              dplyr::filter(Panel == "F1Liquid CDx") %>%
                                              dplyr::distinct(Tumor_Sample_Barcode))$Tumor_Sample_Barcode)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(unique_alt = case_when(
          No_found_in_F1L_position >= Min_count ~ name,
          TRUE ~ unique_alt
        ))
      if(file.exists("source/Data_MAF_target_table_1.rda")){
        load(file="source/Data_MAF_target_table_1.rda")
        Data_tmp = Data_MAF_target_table_1
      } else{
        Data_tmp <- as.data.table(Data_MAF_target)[Panel == "F1Liquid CDx" & No_found_in_F1L >= 1, .(unique_position, Hugo_Symbol)]
        Data_tmp <- unique(Data_tmp)
        Data_MAF_F1L <- as.data.table(Data_MAF_target)[Panel == "F1Liquid CDx" & No_found_in_F1L >= 1, .(unique_position, Hugo_Symbol, VAF)]
        peak_density_calc <- function(VAF) {
          VAF_filtered <- VAF[VAF > 0.2 & VAF < 0.8]
          if (length(VAF_filtered) == 0) return(list(0, 0, 0, 0))  # list() で返す
          tmp_density <- density(VAF_filtered, na.rm = TRUE, width = 0.1, n = 128)
          peak_VAF <- max(tmp_density$y)
          peak_density_low <- min(tmp_density$x[tmp_density$y > (peak_VAF / 2)])
          peak_density_high <- max(tmp_density$x[tmp_density$y > (peak_VAF / 2)])
          peak_count <- sum(VAF_filtered > peak_density_low & VAF_filtered < peak_density_high)
          return(list(peak_VAF, peak_density_low, peak_density_high, peak_count))  # 必ず list で返す
        }
        Data_tmp <- Data_MAF_F1L[, .(
          median_VAF = median(VAF, na.rm = TRUE),
          min_VAF = -quantile(-VAF, na.rm = TRUE, probs = 0.9, type = 3)[[1]],
          max_VAF = quantile(VAF, na.rm = TRUE, probs = 0.9, type = 3)[[1]],
          peak_values = list(peak_density_calc(VAF))  # list() で格納
        ), by = unique_position]
        peak_dt <- rbindlist(lapply(Data_tmp$peak_values, as.list))  # list を data.table に変換
        setnames(peak_dt, c("peak_VAF", "peak_density_low", "peak_density_high", "peak_count"))
        Data_tmp[, c("peak_VAF", "peak_density_low", "peak_density_high", "peak_count") := peak_dt]
        Data_tmp[, peak_values := NULL]
        Data_MAF_target_table_1 = Data_tmp
        save(Data_MAF_target_table_1, file="source/Data_MAF_target_table_1.rda")
      }
      Data_tmp$peak_density_width = Data_tmp$peak_density_high - Data_tmp$peak_density_low
      F1L_non_germline_non_error = (Data_tmp %>% dplyr::filter((median_VAF <= 0.25 | min_VAF <= 0.15 | max_VAF >= 0.75) & (median_VAF <= 0.4 | max_VAF <= 0.95)))$unique_position
      F1L_germline_error = (Data_tmp %>% dplyr::filter(median_VAF > 0.25 & min_VAF > 0.15 & max_VAF < 0.75 | median_VAF > 0.4 & max_VAF > 0.95))$unique_position
      
      Data_tmp = Data_tmp %>% dplyr::filter((unique_position %in% F1L_non_germline_non_error) &
                                              peak_VAF > 8 &
                                              peak_density_width < 0.15 &
                                              #peak_count > 3 &
                                              Data_tmp$peak_density_low > 0.35 &
                                              Data_tmp$peak_density_high < 0.65)

      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(
          non_germline_non_error = case_when(
            unique_position %in% F1L_non_germline_non_error ~ 1,
            unique_position %in% F1L_germline_error ~ 2,
            #(No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) ~ 2,
            TRUE ~ 0
          )
        )
      Data_MAF_target <- Data_MAF_target %>%
        left_join(Data_tmp, by = "unique_position") %>%
        mutate(
          non_germline_non_error = ifelse(
            !is.na(VAF) & !is.na(peak_density_low) & !is.na(peak_density_high) & 
              VAF > peak_density_low & VAF < peak_density_high,
            3,
            non_germline_non_error
          )
        ) %>%
        select(-peak_VAF, -peak_density_low, -peak_density_high, -peak_count, -peak_density_width, -median_VAF, -min_VAF, -max_VAF)
      
      # Data_tmp = Data_MAF_target %>%
      #   dplyr::filter(Panel == "F1Liquid CDx" & No_found_in_F1L >= 1 & non_germline_non_error == 1) %>%
      #   dplyr::select(unique_position, Hugo_Symbol) %>%
      #   dplyr::distinct()
      # Data_MAF_F1L = Data_MAF_target %>%
      #   dplyr::filter(Panel == "F1Liquid CDx" & No_found_in_F1L >= 1 & non_germline_non_error == 1) %>%
      #   dplyr::select(unique_position, Hugo_Symbol, VAF)
      # Data_tmp$median_VAF = 0
      # Data_tmp$min_VAF = 0
      # #Data_tmp$high_VAF = 0
      # Data_tmp$max_VAF = 0
      # for(i in 1:length(Data_tmp$unique_position)){
      #   Data_MAF_target_tmp2 = Data_MAF_F1L %>% dplyr::filter(unique_position == Data_tmp$unique_position[i])
      #   Data_tmp$median_VAF[i] = median(Data_MAF_target_tmp2$VAF,na.rm = T)
      #   Data_tmp$min_VAF[i] = -quantile(-Data_MAF_target_tmp2$VAF,na.rm = T,probs = 0.9, type = 3)[[1]]
      #   #Data_tmp$high_VAF[i] = quantile(Data_MAF_target_tmp2$VAF,na.rm = T,probs = 0.8)
      #   Data_tmp$max_VAF[i] = quantile(Data_MAF_target_tmp2$VAF,na.rm = T,probs = 0.9, type = 3)[[1]]
      # }

      Data_tmp <- unique(Data_MAF_target[
        Panel == "F1Liquid CDx" & No_found_in_F1L >= 1 & non_germline_non_error == 1, 
        .(unique_position, Hugo_Symbol)
      ])
      
      Data_MAF_F1L <- Data_MAF_target[
        Panel == "F1Liquid CDx" & No_found_in_F1L >= 1 & non_germline_non_error == 1, 
        .(unique_position, Hugo_Symbol, VAF)
      ]
      
      # 統計量を計算
      Data_tmp <- Data_MAF_F1L[, .(
        median_VAF = median(VAF, na.rm = TRUE),
        min_VAF = -quantile(-VAF, probs = 0.9, type = 3, na.rm = TRUE),
        max_VAF = quantile(VAF, probs = 0.9, type = 3, na.rm = TRUE)
      ), by = unique_position][Data_tmp, on = "unique_position"]
      F1L_germline_error2 = (Data_tmp %>% dplyr::filter(median_VAF > 0.25 & min_VAF > 0.15 & max_VAF < 0.75 | median_VAF > 0.4 & max_VAF > 0.95))$unique_position
      F1L_germline_error2 = unique(c(F1L_germline_error2,
                                     "5_80654913_80654913_G_GGCCGCAGCG",
                                     "5_80654916_80654916_C_CGCAGCGCCCGCAGCGCCC",
                                     "11_118436670_118436670_C_T",
                                     "17_12020969_12020969_C_T"
      ))
      F1L_germline_error = Data_tmp %>% 
        dplyr::arrange(desc(max_VAF)) %>%
        dplyr::distinct(Hugo_Symbol, .keep_all = T) %>% 
        dplyr::select(Hugo_Symbol, max_VAF)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(
          non_germline_non_error = case_when(
            unique_position %in% F1L_germline_error2 ~ 2,
            TRUE ~ non_germline_non_error
          )
        )

      Data_MAF_target = Data_MAF_target %>% dplyr::select(
        -No_found_in_F1L, -No_found_in_F1, -No_found_in_F1L_gene, -No_found_in_F1_gene, -No_found_in_F1_position, -No_found_in_F1L_position
      )
      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error <= 1))$unique_alt))
      F1L_recurrence_table_tmp = data.frame(
        unique_alt = names(F1L_recurrence_table),
        No_found_in_F1L = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "unique_alt")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx" & non_germline_non_error <= 1))$unique_alt))
      F1_recurrence_table_tmp = data.frame(
        unique_alt = names(F1_recurrence_table),
        No_found_in_F1 = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "unique_alt")
      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error <= 1))$unique_position))
      F1L_recurrence_table_tmp = data.frame(
        unique_position = names(F1L_recurrence_table),
        No_found_in_F1L_position = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "unique_position")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx" & non_germline_non_error <= 1))$unique_position))
      F1_recurrence_table_tmp = data.frame(
        unique_position = names(F1_recurrence_table),
        No_found_in_F1_position = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "unique_position")

      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error <= 1))$Hugo_Symbol))
      F1L_recurrence_table_tmp = data.frame(
        Hugo_Symbol = names(F1L_recurrence_table),
        No_found_in_F1L_gene = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "Hugo_Symbol")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx" & non_germline_non_error <= 1))$Hugo_Symbol))
      F1_recurrence_table_tmp = data.frame(
        Hugo_Symbol = names(F1_recurrence_table),
        No_found_in_F1_gene = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "Hugo_Symbol")
      Data_MAF_target$No_found_in_F1L[is.na(Data_MAF_target$No_found_in_F1L)] = 0
      Data_MAF_target$No_found_in_F1[is.na(Data_MAF_target$No_found_in_F1)] = 0
      Data_MAF_target$No_found_in_F1L_position[is.na(Data_MAF_target$No_found_in_F1L_position)] = 0
      Data_MAF_target$No_found_in_F1_position[is.na(Data_MAF_target$No_found_in_F1_position)] = 0
      Data_MAF_target$No_found_in_F1L_gene[is.na(Data_MAF_target$No_found_in_F1L_gene)] = 0
      Data_MAF_target$No_found_in_F1_gene[is.na(Data_MAF_target$No_found_in_F1_gene)] = 0
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(
          non_germline_non_error = case_when(
            (No_found_in_F1L < Min_count & No_found_in_F1 < Min_count_F1) & non_germline_non_error == 1 ~ 0,
            TRUE ~ non_germline_non_error
          )
        )
      
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(unique_alt = case_when(
          # No_found_in_F1L_position >= Min_count & !str_detect(unique_alt, "fs") & !str_detect(unique_alt, "trunc") ~ name,
          No_found_in_F1L_position >= Min_count ~ name,
          TRUE ~ unique_alt
        ))

      Data_MAF_target = Data_MAF_target %>% dplyr::select(
        -No_found_in_F1L, -No_found_in_F1, -No_found_in_F1L_gene, -No_found_in_F1_gene, -No_found_in_F1L_position, -No_found_in_F1_position
      )
      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error <= 1))$unique_alt))
      F1L_recurrence_table_tmp = data.frame(
        unique_alt = names(F1L_recurrence_table),
        No_found_in_F1L = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "unique_alt")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx" & non_germline_non_error <= 1))$unique_alt))
      F1_recurrence_table_tmp = data.frame(
        unique_alt = names(F1_recurrence_table),
        No_found_in_F1 = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "unique_alt")
      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error <= 1))$unique_position))
      F1L_recurrence_table_tmp = data.frame(
        unique_position = names(F1L_recurrence_table),
        No_found_in_F1L_position = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "unique_position")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx" & non_germline_non_error <= 1))$unique_position))
      F1_recurrence_table_tmp = data.frame(
        unique_position = names(F1_recurrence_table),
        No_found_in_F1_position = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "unique_position")
      
      F1L_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error <= 1))$Hugo_Symbol))
      F1L_recurrence_table_tmp = data.frame(
        Hugo_Symbol = names(F1L_recurrence_table),
        No_found_in_F1L_gene = as.integer(F1L_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1L_recurrence_table_tmp, by = "Hugo_Symbol")
      F1_recurrence_table = sort(table((Data_MAF_target %>% dplyr::filter(Panel == "FoundationOne CDx" & non_germline_non_error <= 1))$Hugo_Symbol))
      F1_recurrence_table_tmp = data.frame(
        Hugo_Symbol = names(F1_recurrence_table),
        No_found_in_F1_gene = as.integer(F1_recurrence_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(F1_recurrence_table_tmp, by = "Hugo_Symbol")
      Data_MAF_target$No_found_in_F1L[is.na(Data_MAF_target$No_found_in_F1L)] = 0
      Data_MAF_target$No_found_in_F1[is.na(Data_MAF_target$No_found_in_F1)] = 0
      Data_MAF_target$No_found_in_F1L_position[is.na(Data_MAF_target$No_found_in_F1L_position)] = 0
      Data_MAF_target$No_found_in_F1_position[is.na(Data_MAF_target$No_found_in_F1_position)] = 0
      Data_MAF_target$No_found_in_F1L_gene[is.na(Data_MAF_target$No_found_in_F1L_gene)] = 0
      Data_MAF_target$No_found_in_F1_gene[is.na(Data_MAF_target$No_found_in_F1_gene)] = 0
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(
          Variant_detection_rate_F1 = round(No_found_in_F1 / F1_patients,digits = 5),
          Variant_detection_rate_F1L = round(No_found_in_F1L / F1L_patients,digits = 5),
          Position_detection_rate_F1 = round(No_found_in_F1_position / F1_patients,digits = 5),
          Position_detection_rate_F1L = round(No_found_in_F1L_position / F1L_patients,digits = 5),
          Gene_detection_rate_F1 = round(No_found_in_F1_gene / F1_patients,digits = 5),
          Gene_detection_rate_F1L = round(No_found_in_F1L_gene / F1L_patients,digits = 5),
        )
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(
          non_germline_non_error = case_when(
            (No_found_in_F1L < Min_count & No_found_in_F1 < Min_count_F1) & non_germline_non_error == 1 ~ 0,
            TRUE ~ non_germline_non_error
          )
        )
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(
          non_germline_non_error = case_when(
            MinorAF >= 0.0001 & non_germline_non_error == 0 & VAF > 0.25 ~ 5,
            TRUE ~ non_germline_non_error
          )
        )
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(
          non_germline_non_error = case_when(
            Panel == "FoundationOne CDx" & non_germline_non_error <= 1 & VAF > 0.5 ~ 6,
            TRUE ~ non_germline_non_error
          )
        )
      
      oncogenic_table = sort(table((Data_MAF_target %>% dplyr::filter(Evidence_level == "F"))$Tumor_Sample_Barcode))
      oncogenic_table_tmp = data.frame(
        Tumor_Sample_Barcode = names(oncogenic_table),
        oncogenic_variant_count = as.integer(oncogenic_table)
      )
      Data_MAF_target = Data_MAF_target %>% dplyr::left_join(oncogenic_table_tmp, by = "Tumor_Sample_Barcode")
      Data_MAF_target$oncogenic_variant_count[is.na(Data_MAF_target$oncogenic_variant_count)] = 0

      incProgress(1 / 8)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, Start_Position, Reference_Allele, Tumor_Seq_Allele2, .keep_all = T) %>%
        dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, Start_Position)
      Data_summary = Data_summary %>% left_join(Data_MAF_unique, by="Tumor_Sample_Barcode")
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(CH_genes = case_when(
          Hugo_Symbol %in% input$clonal_hematopoiesis_genes ~ 1,
          TRUE ~ 0
        ))
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(oncogenic_muts_more_than_2 = case_when(
          oncogenic_variant_count > 2 ~ 1,
          TRUE ~ 0
        ))
      plots = list()
      data_tmp = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx", 
                      Sex %in% c("Male", "Female")) %>%
        dplyr::distinct(Tumor_Sample_Barcode, .keep_all = T) %>%
        dplyr::count(Cancername, Sex) %>%
        dplyr::rename(Cancer = Cancername, Cases = n)
      # 全体のカウントでソートするための順序を決定（上下反転のため昇順）
      cancer_order = data_tmp %>%
        dplyr::group_by(Cancer) %>%
        dplyr::summarise(total = sum(Cases)) %>%
        dplyr::arrange(total) %>%
        dplyr::pull(Cancer)
      # すべてのがん種のリストを作成（欠損値対応）
      all_cancers = factor(cancer_order, levels = cancer_order)
      # Male用とFemale用にデータを分割し、すべてのがん種を含むように調整
      male_data = data.frame(Cancer = all_cancers) %>%
        left_join(data_tmp %>% filter(Sex == "Male") %>% select(Cancer, Cases), by = "Cancer") %>%
        mutate(Cases = ifelse(is.na(Cases), 0, Cases))
      female_data = data.frame(Cancer = all_cancers) %>%
        left_join(data_tmp %>% filter(Sex == "Female") %>% select(Cancer, Cases), by = "Cancer") %>%
        mutate(Cases = ifelse(is.na(Cases), 0, Cases))
      # Y軸の範囲を統一
      max_cases = max(data_tmp$Cases)
      gap_width = max_cases * 0.6  # 中央の空間の幅
      # 癌種に番号を割り当て
      cancer_numbers = data.frame(
        Cancer = all_cancers,
        x_num = 1:length(all_cancers)
      )
      # 男性・女性データを別々に処理（棒グラフの開始点を調整）
      male_data_plot = male_data %>% 
        filter(Cases > 0) %>%
        left_join(cancer_numbers, by = "Cancer") %>%
        mutate(Sex = "Male", 
               y_start = -gap_width/2,  # 男性の開始点（中央右端）
               y_end = -(gap_width/2 + Cases))  # 男性の終了点（左方向）
      female_data_plot = female_data %>% 
        filter(Cases > 0) %>%
        left_join(cancer_numbers, by = "Cancer") %>%
        mutate(Sex = "Female", 
               y_start = gap_width/2,  # 女性の開始点（中央左端）
               y_end = gap_width/2 + Cases)  # 女性の終了点（右方向）
      # 癌種名用データ
      label_data = cancer_numbers %>%
        mutate(y = 0)
      # グラフ作成
      g = ggplot() +
        # 男性の棒グラフ（中央から左へ）
        geom_rect(data = male_data_plot, 
                  aes(xmin = x_num - 0.4, xmax = x_num + 0.4,
                      ymin = y_end, ymax = y_start), 
                  fill = "steelblue", color = "black") +
        # 女性の棒グラフ（中央から右へ）
        geom_rect(data = female_data_plot, 
                  aes(xmin = x_num - 0.4, xmax = x_num + 0.4,
                      ymin = y_start, ymax = y_end), 
                  fill = "red", color = "black") +
        # 中央の癌種名
        geom_text(data = label_data, 
                  aes(x = x_num, y = y, label = Cancer), 
                  hjust = 0.5, vjust = 0.5, size = 3, fontface = "bold") +
        # 男性・女性それぞれのゼロ点に縦線
        geom_hline(yintercept = -gap_width/2, color = "black", linetype = "solid", size = 0.8) +
        geom_hline(yintercept = gap_width/2, color = "black", linetype = "solid", size = 0.8) +
        coord_flip() +
        scale_x_continuous(breaks = 1:length(all_cancers), labels = NULL, 
                           limits = c(0.5, length(all_cancers) + 0.5)) +
        scale_y_continuous(
          breaks = c(-(gap_width/2 + c(2000, 1500, 1000, 500)), -gap_width/2, 0, gap_width/2,
                     gap_width/2 + c(500, 1000, 1500, 2000)),
          labels = c("2000", "1500", "1000", "500", "0", "Cases", "0", "500", "1000", "1500", "2000"),
          limits = c(-(max_cases + gap_width), max_cases + gap_width)
        ) +
        theme_classic() +
        theme(
          axis.title = element_blank(),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          axis.line.y = element_blank(),
          plot.title = element_text(hjust = 0.5, size = 14, face = "bold")
        ) +
        labs(title = "Cases underwent FoundationOne Liquid CDx by Gender") +
        # MaleとFemaleのラベルを一番上のグラフのすぐ上に配置
        annotate("text", x = length(all_cancers) - 10.4, y = -gap_width, 
                 label = "Male", color = "steelblue", fontface = "bold", hjust = 0.5, size = 4) +
        annotate("text", x = length(all_cancers) - 10.4, y = gap_width, 
                 label = "Female", color = "red", fontface = "bold", hjust = 0.5, size = 4)
      plots[[1]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter((No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) & non_germline_non_error == 1 & Panel == "F1Liquid CDx")
      Data_tmp = data_tmp %>% dplyr::filter(CH_genes == 1)
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("Hotspot variants in CH-related genes, F1-Liquid CDx") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[4]] = g
      Data_tmp = data_tmp %>% dplyr::filter(CH_genes == 0)
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("Hotspot variants in non-CH-related genes, F1-Liquid CDx") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[5]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter((non_germline_non_error >= 2) & Panel == "F1Liquid CDx" & CH_genes == 1)
      g <- ggplot(data_tmp, aes(x = Age, y = VAF, color = CH_genes))
      g <- g + geom_point() +
        theme_bw() +
        ggtitle("Filtered variants in CH-related genes , F1-Liquid CDx")
      plots[[6]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter((non_germline_non_error >= 2) & Panel == "F1Liquid CDx" & CH_genes == 0)
      g <- ggplot(data_tmp, aes(x = Age, y = VAF, color = CH_genes))
      g <- g + geom_point() +
        theme_bw() +
        ggtitle("Filtered variants in non-CH-related genes , F1-Liquid CDx")
      plots[[7]] = g
      Data_MAF_target_save = Data_MAF_target
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" &
                        non_germline_non_error == 1 &
                        #Hugo_Symbol %in% CH_GENE_NatComms &
                        (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1)) %>%
        dplyr::select(unique_position, unique_alt, name, Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, amino.acid.change, Evidence_level, CH_genes, No_found_in_F1, No_found_in_F1L, No_found_in_F1_position, No_found_in_F1L_position, F1_patients, F1L_patients, Variant_detection_rate_F1, Variant_detection_rate_F1L, unique_position) %>%
        dplyr::distinct(Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      data_tmp$no_mut_F1 = data_tmp$F1_patients - data_tmp$No_found_in_F1
      data_tmp$no_mut_F1L = data_tmp$F1L_patients - data_tmp$No_found_in_F1L
      data_tmp_fisher = data_tmp %>% dplyr::select(No_found_in_F1L,No_found_in_F1,no_mut_F1L,no_mut_F1)
      data_tmp$p_value = 1
      data_tmp$odds_ratio = 1
      data_tmp$triplet_pre = ""
      data_tmp$triplet_post = ""
      data_tmp$CpG_TpG = 0
      data_tmp$CpC_TpC = 0
      data_tmp$SNV_check = nchar(data_tmp$Reference_Allele) * nchar(data_tmp$Tumor_Seq_Allele2)
      Name_list = NULL
      triplet_pre = DNAStringSet(rep("NNN",length(data_tmp$Hugo_Symbol)))
      triplet_post = DNAStringSet(rep("NNN",length(data_tmp$Hugo_Symbol)))
      Data_summary$favor_liquid = ifelse(Data_summary$Panel == "FoundationOne CDx", 0, 1)
      Data_summary_tmp = Data_summary %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::select(Tumor_Sample_Barcode, Age, Cancername, favor_liquid)
      #save(Data_MAF_target, file="source/Data_MAF_target.rda")
      #save(Data_summary, file="source/Data_summary.rda")
      if(file.exists("source/Data_MAF_target_table_4.rda")){
        load(file="source/Data_MAF_target_table_4.rda")
        load(file="source/Name_list_4.rda")
        load(file="source/Data_MAF_target_4.rda")
        data_tmp = Data_MAF_target_table_2
      } else{
        ref_genome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
        if(file.exists("source/Data_MAF_target_table_4_5.rda")){
          load(file="source/Data_MAF_target_table_4_5.rda")
          load(file="source/Name_list_4.rda")
          data_tmp = Data_MAF_target_table_2_5
        } else{
          for(i in 1:length(data_tmp$Hugo_Symbol)){
            mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
            Data_summary_tmp = Data_summary_tmp %>%
              dplyr::mutate(mutation = case_when(
                Tumor_Sample_Barcode %in% mut_ID ~ 1,
                TRUE ~ 0))
            if(data_tmp$SNV_check[i] == 1){
              triplet_pre[[i]] = ref_genome[[paste0("chr", data_tmp$Chromosome[i])]][(data_tmp$Start_Position[i] - 1):(data_tmp$Start_Position[i] + 1)]
            }
            if(data_tmp$odds_ratio[i] == 1 & data_tmp$p_value[i] == 1){
              if(str_detect(data_tmp$unique_alt[i], "_trunc_NOS") | str_detect(data_tmp$unique_alt[i], "fs_NOS")){
                Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
                data_tmp$odds_ratio[data_tmp$unique_alt == data_tmp$unique_alt[i]] = exp(summary(Data_ORR)$coefficients["mutation",1])
                if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
                  fisher_tmp = matrix(c(data_tmp_fisher[i,1][[1]],
                                        data_tmp_fisher[i,2][[1]],
                                        data_tmp_fisher[i,3][[1]],
                                        data_tmp_fisher[i,4][[1]]),
                                      ncol = 2
                  )
                  data_tmp$p_value[data_tmp$unique_alt == data_tmp$unique_alt[i]] = fisher.test(fisher_tmp,)$p.value
                } else{
                  data_tmp$p_value[data_tmp$unique_alt == data_tmp$unique_alt[i]] = summary(Data_ORR)$coefficients["mutation",4]
                }
              }
              if(str_detect(data_tmp$unique_alt[i], "_trunc_CtoT")){
                Name_list = c(Name_list, data_tmp$unique_alt[i])
              }else {
                Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
                data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
                data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
                if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
                  fisher_tmp = matrix(c(data_tmp_fisher[i,1][[1]],
                                        data_tmp_fisher[i,2][[1]],
                                        data_tmp_fisher[i,3][[1]],
                                        data_tmp_fisher[i,4][[1]]),
                                      ncol = 2
                  )
                  data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
                }
              }
            }
            if(i %% 10 == 0){
              print(paste(i, ",", length(data_tmp$Hugo_Symbol)))
            }
          }
          triplet_post = 
            xscat(subseq(triplet_pre,1,1),
                  DNAStringSet(data_tmp$Tumor_Seq_Allele2),
                  subseq(triplet_pre,3,3))
          ID_revcom = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,2) %in% DNAStringSet(c("A", "G"))
          triplet_pre[ID_revcom] =reverseComplement(triplet_pre[ID_revcom])
          triplet_post[ID_revcom] =reverseComplement(triplet_post[ID_revcom])
          ID_CpG = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,3) == DNAString("CG") &
            subseq(triplet_post,2,3) == DNAString("TG")
          data_tmp$CpG_TpG[ID_CpG] = 1
          ID_CpC = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,3) == DNAString("CC") &
            subseq(triplet_post,2,3) == DNAString("TC")
          data_tmp$CpC_TpC[ID_CpC] = 1
          
          data_tmp$triplet_pre = as.character(triplet_pre)
          data_tmp$triplet_post = as.character(triplet_post)
          
          Name_list = unique(Name_list)
          Data_MAF_target_table_2_5 = data_tmp
          save(Data_MAF_target_table_2_5, file="source/Data_MAF_target_table_4_5.rda")
          save(Name_list, file="source/Name_list_4.rda")
        }
        for(i in 1:length(Name_list)){
          hotspot_CpG = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpG_TpG == 1)
          hotspot_CpC = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpC_TpC == 1)
          hotspot_nonCpG = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpG_TpG == 0& CpC_TpC == 0)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1L_position)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1L_position)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1L_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1_position)
          Data_MAF_target$No_found_in_F1L[Data_MAF_target$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1L_position)
          Data_MAF_target$No_found_in_F1L[Data_MAF_target$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1L_position)
          Data_MAF_target$No_found_in_F1L[Data_MAF_target$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1L_position)
          Data_MAF_target$No_found_in_F1[Data_MAF_target$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1_position)
          Data_MAF_target$No_found_in_F1[Data_MAF_target$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1_position)
          Data_MAF_target$No_found_in_F1[Data_MAF_target$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1_position)
          data_tmp$unique_alt[data_tmp$unique_position %in% hotspot_CpC$unique_position] = paste0(Name_list[i],"_CpC")
          data_tmp$unique_alt[data_tmp$unique_position %in% hotspot_CpG$unique_position] = paste0(Name_list[i],"_CpG")
          Data_MAF_target$unique_alt[Data_MAF_target$unique_position %in% hotspot_CpG$unique_position] = paste0(Name_list[i],"_CpG")
          Data_MAF_target$unique_alt[Data_MAF_target$unique_position %in% hotspot_CpC$unique_position] = paste0(Name_list[i],"_CpC")
          mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == Name_list[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_nonCpG$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_nonCpG$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_nonCpG$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == Name_list[i]] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_nonCpG$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_nonCpG$No_found_in_F1_position),
                                    sum(hotspot_nonCpG$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_nonCpG$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_MAF_target$non_germline_non_error[Data_MAF_target$unique_alt == Name_list[i]] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == Name_list[i]] = 1
            data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = 1
          }
          mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == paste0(Name_list[i],"_CpG")))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_CpG$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_CpG$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_CpG$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_CpG$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_CpG$No_found_in_F1_position),
                                    sum(hotspot_CpG$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_CpG$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_MAF_target$non_germline_non_error[Data_MAF_target$unique_alt == paste0(Name_list[i],"_CpG")] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = 1
            data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = 1
          }
          mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == paste0(Name_list[i],"_CpC")))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_CpC$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_CpC$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_CpC$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_CpC$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_CpC$No_found_in_F1_position),
                                    sum(hotspot_CpC$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_CpC$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_MAF_target$non_germline_non_error[Data_MAF_target$unique_alt == paste0(Name_list[i],"_CpC")] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = 1
            data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = 1
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(Name_list)))
          }
        }
        Data_MAF_target_table_2 = data_tmp
        save(Data_MAF_target_table_2, file="source/Data_MAF_target_table_4.rda")
        save(Data_MAF_target, file="source/Data_MAF_target_4.rda")
      }
      data_tmp_save = data_tmp
      data_tmp = data_tmp %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::filter((No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1)) %>%
        dplyr::distinct(unique_alt, .keep_all = TRUE)
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^10] = 2^10
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-10] = 2^-10
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-30] = 10^-30
      data_tmp = data_tmp %>% dplyr::mutate(CtoT = case_when(str_sub(triplet_pre,1,1)!= "N" & str_sub(triplet_pre,2,2) == "C" & str_sub(triplet_post,2,2) == "T" ~ 1,TRUE ~ 0))
      data_tmp = data_tmp %>%
        dplyr::mutate(mut_pattern = case_when(
          CpG_TpG == 1 ~ "CpG>TpG",
          CpC_TpC == 1 ~ "CpC>TpC",
          CtoT == 1 ~ "C>T, NOS",
          SNV_check != 1 ~ "Indel",
          TRUE ~ "SNV, NOS"
        ))
      data_tmp$CH_genes = factor(data_tmp$CH_genes)
      data_tmp$mut_pattern = factor(data_tmp$mut_pattern)
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = ifelse(
                  (logfold > log2(threshold_OR) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     logfold < -log2(threshold_OR_tumor) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     logfold > log2(threshold_OR_OTHER) & CH_genes == 0 |
                     logfold < -log2(threshold_OR_tumor) & CH_genes == 0 |
                     logfold > log2(threshold_OR_TP53) & Hugo_Symbol == "TP53" |
                     logfold < -log2(threshold_OR_tumor) & Hugo_Symbol == "TP53") &
                    (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                       logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                       logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53"),
                  unique_alt, "" ),
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     logfold > log2(threshold_OR_OTHER) & CH_genes == 0 |
                     logfold > log2(threshold_OR_TP53) & Hugo_Symbol == "TP53") &
                    (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                       logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                       logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53"), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      data_tmp_odds = data_tmp %>% dplyr::select(unique_alt, odds_ratio_adj, p_value_adj, aelabel, pnt_col, pnt_col, logfold, logpval, mut_pattern)
      data_tmp_save = data_tmp_save %>% dplyr::left_join(data_tmp_odds, by = "unique_alt")
      data_tmp = data_tmp_save
      Data_triplet = data_tmp %>% dplyr::select(unique_position, triplet_pre, triplet_post, CpG_TpG, CpC_TpC, odds_ratio, p_value, odds_ratio_adj, p_value_adj, pnt_col, logfold, logpval, mut_pattern, SNV_check)
      Data_MAF_target_ = Data_MAF_target %>% left_join(Data_triplet, by=c('unique_position'))
      Data_MAF_target_all = Data_MAF_target_
      Data_tmp = Data_MAF_target_ %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(#unique_alt %in% CH_candidates_base &
          #Hugo_Symbol %in% CH_gene_entropy_3 &
          Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        #dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot of hotspot variants for favor F1 CDx or F1-Liquid CDx",
             subtitle = paste0(nrow(Data_tmp), " hotspot variants found in F1 or F1L >= 10 times, all genes"),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[365]] = g
      setProgress(detail = "3868")
      table_save_tmp_ = Data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant/enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp_, "source/Table_S5_all genes.csv")
      Data_MAF_target = Data_MAF_target_save
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" &
                        non_germline_non_error == 1 &
                        Hugo_Symbol %in% CH_GENE_NatComms &
                        (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1)) %>%
        dplyr::select(unique_position, unique_alt, name, Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, amino.acid.change, Evidence_level, CH_genes, No_found_in_F1, No_found_in_F1L, No_found_in_F1_position, No_found_in_F1L_position, F1_patients, F1L_patients, Variant_detection_rate_F1, Variant_detection_rate_F1L, unique_position) %>%
        dplyr::distinct(Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      data_tmp$no_mut_F1 = data_tmp$F1_patients - data_tmp$No_found_in_F1
      data_tmp$no_mut_F1L = data_tmp$F1L_patients - data_tmp$No_found_in_F1L
      data_tmp_fisher = data_tmp %>% dplyr::select(No_found_in_F1L,No_found_in_F1,no_mut_F1L,no_mut_F1)
      data_tmp$p_value = 1
      data_tmp$odds_ratio = 1
      data_tmp$triplet_pre = ""
      data_tmp$triplet_post = ""
      data_tmp$CpG_TpG = 0
      data_tmp$CpC_TpC = 0
      data_tmp$SNV_check = nchar(data_tmp$Reference_Allele) * nchar(data_tmp$Tumor_Seq_Allele2)
      Name_list = NULL
      triplet_pre = DNAStringSet(rep("NNN",length(data_tmp$Hugo_Symbol)))
      triplet_post = DNAStringSet(rep("NNN",length(data_tmp$Hugo_Symbol)))
      Data_summary$favor_liquid = ifelse(Data_summary$Panel == "FoundationOne CDx", 0, 1)
      Data_summary_tmp = Data_summary %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::select(Tumor_Sample_Barcode, Age, Cancername, favor_liquid)
      #save(Data_MAF_target, file="source/Data_MAF_target.rda")
      #save(Data_summary, file="source/Data_summary.rda")
      if(file.exists("source/Data_MAF_target_table_2.rda")){
        load(file="source/Data_MAF_target_table_2.rda")
        load(file="source/Name_list.rda")
        load(file="source/Data_MAF_target.rda")
        data_tmp = Data_MAF_target_table_2
      } else{
        ref_genome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
        if(file.exists("source/Data_MAF_target_table_2_5.rda")){
          load(file="source/Data_MAF_target_table_2_5.rda")
          load(file="source/Name_list.rda")
          data_tmp = Data_MAF_target_table_2_5
        } else{
          for(i in 1:length(data_tmp$Hugo_Symbol)){
            mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
            Data_summary_tmp = Data_summary_tmp %>%
              dplyr::mutate(mutation = case_when(
                Tumor_Sample_Barcode %in% mut_ID ~ 1,
                TRUE ~ 0))
            if(data_tmp$SNV_check[i] == 1){
              triplet_pre[[i]] = ref_genome[[paste0("chr", data_tmp$Chromosome[i])]][(data_tmp$Start_Position[i] - 1):(data_tmp$Start_Position[i] + 1)]
            }
            if(data_tmp$odds_ratio[i] == 1 & data_tmp$p_value[i] == 1){
              if(str_detect(data_tmp$unique_alt[i], "_trunc_NOS") | str_detect(data_tmp$unique_alt[i], "fs_NOS")){
                Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
                data_tmp$odds_ratio[data_tmp$unique_alt == data_tmp$unique_alt[i]] = exp(summary(Data_ORR)$coefficients["mutation",1])
                if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
                  fisher_tmp = matrix(c(data_tmp_fisher[i,1][[1]],
                                        data_tmp_fisher[i,2][[1]],
                                        data_tmp_fisher[i,3][[1]],
                                        data_tmp_fisher[i,4][[1]]),
                                      ncol = 2
                  )
                  data_tmp$p_value[data_tmp$unique_alt == data_tmp$unique_alt[i]] = fisher.test(fisher_tmp,)$p.value
                } else{
                  data_tmp$p_value[data_tmp$unique_alt == data_tmp$unique_alt[i]] = summary(Data_ORR)$coefficients["mutation",4]
                }
              }
              if(str_detect(data_tmp$unique_alt[i], "_trunc_CtoT")){
                Name_list = c(Name_list, data_tmp$unique_alt[i])
              }else {
                Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
                data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
                data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
                if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
                  fisher_tmp = matrix(c(data_tmp_fisher[i,1][[1]],
                                        data_tmp_fisher[i,2][[1]],
                                        data_tmp_fisher[i,3][[1]],
                                        data_tmp_fisher[i,4][[1]]),
                                      ncol = 2
                  )
                  data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
                }
              }
            }
            if(i %% 10 == 0){
              print(paste(i, ",", length(data_tmp$Hugo_Symbol)))
            }
          }
          triplet_post = 
            xscat(subseq(triplet_pre,1,1),
              DNAStringSet(data_tmp$Tumor_Seq_Allele2),
              subseq(triplet_pre,3,3))
          ID_revcom = subseq(triplet_pre,1,1) != DNAString("N") &
                      subseq(triplet_pre,2,2) %in% DNAStringSet(c("A", "G"))
          triplet_pre[ID_revcom] =reverseComplement(triplet_pre[ID_revcom])
          triplet_post[ID_revcom] =reverseComplement(triplet_post[ID_revcom])
          ID_CpG = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,3) == DNAString("CG") &
            subseq(triplet_post,2,3) == DNAString("TG")
          data_tmp$CpG_TpG[ID_CpG] = 1
          ID_CpC = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,3) == DNAString("CC") &
            subseq(triplet_post,2,3) == DNAString("TC")
          data_tmp$CpC_TpC[ID_CpC] = 1
          
          data_tmp$triplet_pre = as.character(triplet_pre)
          data_tmp$triplet_post = as.character(triplet_post)
  
          Name_list = unique(Name_list)
          Data_MAF_target_table_2_5 = data_tmp
          save(Data_MAF_target_table_2_5, file="source/Data_MAF_target_table_2_5.rda")
          save(Name_list, file="source/Name_list.rda")
        }
        for(i in 1:length(Name_list)){
          hotspot_CpG = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpG_TpG == 1)
          hotspot_CpC = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpC_TpC == 1)
          hotspot_nonCpG = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpG_TpG == 0& CpC_TpC == 0)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1L_position)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1L_position)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1L_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1_position)
          Data_MAF_target$No_found_in_F1L[Data_MAF_target$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1L_position)
          Data_MAF_target$No_found_in_F1L[Data_MAF_target$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1L_position)
          Data_MAF_target$No_found_in_F1L[Data_MAF_target$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1L_position)
          Data_MAF_target$No_found_in_F1[Data_MAF_target$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1_position)
          Data_MAF_target$No_found_in_F1[Data_MAF_target$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1_position)
          Data_MAF_target$No_found_in_F1[Data_MAF_target$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1_position)
          data_tmp$unique_alt[data_tmp$unique_position %in% hotspot_CpC$unique_position] = paste0(Name_list[i],"_CpC")
          data_tmp$unique_alt[data_tmp$unique_position %in% hotspot_CpG$unique_position] = paste0(Name_list[i],"_CpG")
          Data_MAF_target$unique_alt[Data_MAF_target$unique_position %in% hotspot_CpG$unique_position] = paste0(Name_list[i],"_CpG")
          Data_MAF_target$unique_alt[Data_MAF_target$unique_position %in% hotspot_CpC$unique_position] = paste0(Name_list[i],"_CpC")
          mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == Name_list[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_nonCpG$No_found_in_F1L_position) >= 1) &
              (sum(hotspot_nonCpG$No_found_in_F1L_position) >= Min_count |
               sum(hotspot_nonCpG$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == Name_list[i]] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_nonCpG$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_nonCpG$No_found_in_F1_position),
                                    sum(hotspot_nonCpG$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_nonCpG$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_MAF_target$non_germline_non_error[Data_MAF_target$unique_alt == Name_list[i]] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == Name_list[i]] = 1
            data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = 1
          }
          mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == paste0(Name_list[i],"_CpG")))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_CpG$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_CpG$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_CpG$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_CpG$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_CpG$No_found_in_F1_position),
                                    sum(hotspot_CpG$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_CpG$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_MAF_target$non_germline_non_error[Data_MAF_target$unique_alt == paste0(Name_list[i],"_CpG")] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = 1
            data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = 1
          }
          mut_ID = (Data_MAF_target %>% dplyr::filter(unique_alt == paste0(Name_list[i],"_CpC")))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_CpC$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_CpC$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_CpC$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_CpC$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_CpC$No_found_in_F1_position),
                                    sum(hotspot_CpC$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_CpC$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_MAF_target$non_germline_non_error[Data_MAF_target$unique_alt == paste0(Name_list[i],"_CpC")] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = 1
            data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = 1
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(Name_list)))
          }
        }
        Data_MAF_target_table_2 = data_tmp
        save(Data_MAF_target_table_2, file="source/Data_MAF_target_table_2.rda")
        save(Data_MAF_target, file="source/Data_MAF_target.rda")
      }
      data_tmp_save = data_tmp
      data_tmp = data_tmp %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::filter((No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1)) %>%
        dplyr::distinct(unique_alt, .keep_all = TRUE)
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^10] = 2^10
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-10] = 2^-10
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-30] = 10^-30
      data_tmp = data_tmp %>% dplyr::mutate(CtoT = case_when(str_sub(triplet_pre,1,1)!= "N" & str_sub(triplet_pre,2,2) == "C" & str_sub(triplet_post,2,2) == "T" ~ 1,TRUE ~ 0))
      data_tmp = data_tmp %>%
        dplyr::mutate(mut_pattern = case_when(
          CpG_TpG == 1 ~ "CpG>TpG",
          CpC_TpC == 1 ~ "CpC>TpC",
          CtoT == 1 ~ "C>T, NOS",
          SNV_check != 1 ~ "Indel",
          TRUE ~ "SNV, NOS"
        ))
      data_tmp$CH_genes = factor(data_tmp$CH_genes)
      data_tmp$mut_pattern = factor(data_tmp$mut_pattern)
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = ifelse(
                  (logfold > log2(threshold_OR) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                   logfold < -log2(threshold_OR_tumor) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                   logfold > log2(threshold_OR_OTHER) & CH_genes == 0 |
                     logfold < -log2(threshold_OR_tumor) & CH_genes == 0 |
                     logfold > log2(threshold_OR_TP53) & Hugo_Symbol == "TP53" |
                     logfold < -log2(threshold_OR_tumor) & Hugo_Symbol == "TP53") &
                    (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53"),
                  unique_alt, "" ),
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                   logfold > log2(threshold_OR_OTHER) & CH_genes == 0 |
                   logfold > log2(threshold_OR_TP53) & Hugo_Symbol == "TP53") &
                    (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                       logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                       logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53"), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                       # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                       #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                       #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                     ), "C",
                  # ifelse(
                  #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                  # ifelse(
                  #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
             aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = paste0(nrow(data_tmp), " hotspot variants found in F1 CDx or F1-Liquid CDx >= 10 times"),
             subtitle = "Variants found in F1 or F1L >= 10 times",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=c(log2(2^2), log2(threshold_OR)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4,-2,0,2,4,6,8,10),
                           labels = c("~-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
        #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
        #theme(legend.position = "none")
      plots[[8]] = g
      table_save_tmp = data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp, "source/Table_S5.csv")
      Summary_Table = data_tmp
      data_tmp_odds = data_tmp %>% dplyr::select(unique_alt, odds_ratio_adj, p_value_adj, aelabel, pnt_col, logfold, logpval, mut_pattern)
      data_tmp_save = data_tmp_save %>% dplyr::left_join(data_tmp_odds, by = "unique_alt")
      data_tmp = data_tmp_save
      Data_triplet = data_tmp %>% dplyr::select(unique_position, triplet_pre, triplet_post, CpG_TpG, CpC_TpC, odds_ratio, p_value, odds_ratio_adj, p_value_adj, pnt_col, logfold, logpval, mut_pattern, SNV_check)
      # g <- ggplot(data_tmp, aes(x = Variant_detection_rate_F1, y = Variant_detection_rate_F1L, color = pnt_col))
      # g <- g + geom_point() +
      #   ggtitle("Hotspot CH candidates and variant detection rate") +
      #   theme_bw() + 
      #   xlim(c(0,0.1)) + 
      #   ylim(c(0,0.1)) 
      # plots[[9]] = g
      
      Data_MAF_target_table_2 = data_tmp
      
      output$table_volcano_variant = downloadHandler(
        filename = function(){"volcano_plot_data.csv"}, 
        content = function(fname){
          write.csv(Data_MAF_target_table_2, fname)
        }
      )

      incProgress(1 / 8)
      setProgress(detail = "4203")
      

      Data_tmp = Data_MAF_target_table_2 %>% dplyr::filter(pnt_col %in% c("A"))
      CH_candidates = unique(Data_tmp$unique_alt)
      Data_tmp = Data_MAF_target %>% dplyr::filter(non_germline_non_error == 1) %>% dplyr::filter(unique_alt %in% CH_candidates & Panel == "F1Liquid CDx")
      
      gene_to_analyze = unique(names(sort(table(Data_tmp$Hugo_Symbol),
                                            decreasing = T)))
      Diseases = sort(unique(Data_tmp$Cancername))
      Summary_Gene_alteration = data.frame(matrix(0,
                                                  nrow=length(Diseases),
                                                  ncol=length(gene_to_analyze)))
      colnames(Summary_Gene_alteration) = gene_to_analyze
      rownames(Summary_Gene_alteration) = Diseases
      Data_tmp3 = Data_summary %>% dplyr::filter(Panel == "F1Liquid CDx")
      
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        for(j in gene_to_analyze){
          Summary_Gene_alteration[i,j] =
            length(unique((Data_tmp %>% dplyr::filter(
              Hugo_Symbol == j &
                Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
        }
      }
      Data_entropy = data.frame(gene_to_analyze)
      colnames(Data_entropy) = "Gene"
      Data_entropy$Entropy = 0
      for(i in 1:length(gene_to_analyze)){
        Data_entropy$Entropy[i] = shannon.entropy(Summary_Gene_alteration[,i], length(Summary_Gene_alteration,1))
      }

      Data_tmp2 = Summary_Gene_alteration
      Data_tmp2$Disease = rownames(Data_tmp2)
      Data_tmp2 = Data_tmp2 %>% gather(value = "freq", key = Gene, -Disease)
      Data_tmp2 <- transform(Data_tmp2, Disease= factor(Disease, levels = sort(unique(Disease), decreasing = TRUE)))
      Data_tmp2 <- Data_tmp2 %>%
        group_by(Gene) %>%
        mutate(Gene = factor(Gene, levels = names(sort(table(Gene), decreasing = TRUE))))
      g = ggplot(Data_tmp2, aes(Gene, as.factor(Disease))) +
        geom_tile(aes(fill = freq), color = "black",
                  #            lwd = 1,
                  linetype = 1) + 
        geom_text(aes(label = round(freq, 0))) +
        scale_fill_gradient(low = "white", high = "red", limit=c(0,100), name="Frequency (%)") +
        theme_classic() +
        labs(title = "Frequently detected genes with CH hotspot variants") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              axis.line.x = element_blank(),
              axis.line.y = element_blank())
      plots[[10]] = g
      g = ggplot(Data_entropy, aes(x=Gene, y=Entropy, fill="black")) +
        geom_point(stat = "identity") +
        coord_flip() +
        theme_classic() +
        theme(legend.position = "none")
      plots[[11]] = g
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates & Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      write_excel_csv(Data_tmp, "source/Table_S5_noise_excluded.csv")
      Data_tmp3 = Data_summary %>% dplyr::filter(Panel == "F1Liquid CDx")
      gene_to_analyze = unique(names(sort(table(Data_tmp$Hugo_Symbol),
                                          decreasing = T)))
      Diseases = sort(names(table(Data_tmp3$Cancername)[table(Data_tmp3$Cancername)>=50]))
      Data_tmp = Data_tmp %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "F1Liquid CDx")
      Summary_Gene_alteration = data.frame(matrix(0,
                                                  nrow=length(Diseases),
                                                  ncol=length(gene_to_analyze)))
      colnames(Summary_Gene_alteration) = gene_to_analyze
      rownames(Summary_Gene_alteration) = Diseases
      
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        for(j in gene_to_analyze){
          Summary_Gene_alteration[i,j] =
            length(unique((Data_tmp %>% dplyr::filter(
              Hugo_Symbol == j &
                Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
        }
      }
      Data_entropy = data.frame(gene_to_analyze)
      colnames(Data_entropy) = "Gene"
      Data_entropy$Entropy = 0
      for(i in 1:length(gene_to_analyze)){
        Data_entropy$Entropy[i] = shannon.entropy(Summary_Gene_alteration[,i], length(Summary_Gene_alteration,1))
      }
      
      Data_tmp2 = Summary_Gene_alteration
      Data_tmp2$Disease = rownames(Data_tmp2)
      Data_tmp2 = Data_tmp2 %>% gather(value = "freq", key = Gene, -Disease)
      Data_tmp2 <- transform(Data_tmp2, Disease= factor(Disease, levels = sort(unique(Disease), decreasing = TRUE)))
      Data_tmp2 <- Data_tmp2 %>%
        group_by(Gene) %>%
        mutate(Gene = factor(Gene, levels = names(sort(table(Gene), decreasing = TRUE))))
      g = ggplot(Data_tmp2, aes(Gene, as.factor(Disease))) +
        geom_tile(aes(fill = freq), color = "black",
                  #            lwd = 1,
                  linetype = 1) + 
        geom_text(aes(label = round(freq, 0))) +
        scale_fill_gradient(low = "white", high = "red", limit=c(0,100), name="Frequency (%)") +
        theme_classic() +
        labs(title = "Frequently mutated genes, CH-related hotspot variants") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              axis.line.x = element_blank(),
              axis.line.y = element_blank())
      plots[[12]] = g
      g = ggplot(Data_entropy, aes(x=Gene, y=Entropy, fill="black")) +
        geom_point(stat = "identity") +
        coord_flip() +
        theme_classic() +
        theme(legend.position = "none")
      plots[[13]] = g
      CH_gene_entropy_3 = Data_entropy[Data_entropy$Entropy > 2,"Gene"]
      CH_gene_entropy_3 = CH_gene_entropy_3[CH_gene_entropy_3 %in% CH_GENE_NatComms]
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx")
      write_excel_csv(Data_tmp, "source/Table_S5_low_entropy_gene_excluded.csv")
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      write_excel_csv(Data_tmp, "source/Table_S5_noise_and_low_entropy_gene_excluded.csv")
      Data_tmp = Data_tmp %>%
        dplyr::filter(!name %in% c(
          'EGFR_V441G', 'EGFR_V441D', 'EGFR_G465R', 'EGFR_G465E', 'EGFR_S464L', 'EGFR_G465V',
          'EGFR_S442R', 'EGFR_I491K', 'EGFR_I491R', 'EGFR_S492R'
        ))
      CH_candidates = Data_tmp$unique_alt
      CH_candidates_position = unique(Data_tmp$unique_position)
      gene_to_analyze = unique(names(sort(table(Data_tmp$Hugo_Symbol),
                                          decreasing = T)))
      Data_tmp3 = Data_summary %>% dplyr::filter(Panel == "F1Liquid CDx")
      Diseases = sort(names(table(Data_tmp3$Cancername)[table(Data_tmp3$Cancername)>=50]))
      Data_tmp = Data_tmp %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "F1Liquid CDx")
      Summary_Gene_alteration = data.frame(matrix(0,
                                                  nrow=length(Diseases),
                                                  ncol=length(gene_to_analyze)))
      colnames(Summary_Gene_alteration) = gene_to_analyze
      rownames(Summary_Gene_alteration) = Diseases
      
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        for(j in gene_to_analyze){
          Summary_Gene_alteration[i,j] =
            length(unique((Data_tmp %>% dplyr::filter(
              Hugo_Symbol == j &
                Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
        }
      }
      Data_entropy = data.frame(gene_to_analyze)
      colnames(Data_entropy) = "Gene"
      Data_entropy$Entropy = 0
      for(i in 1:length(gene_to_analyze)){
        Data_entropy$Entropy[i] = shannon.entropy(Summary_Gene_alteration[,i], length(Summary_Gene_alteration,1))
      }
      
      Data_tmp2 = Summary_Gene_alteration
      Data_tmp2$Disease = rownames(Data_tmp2)
      Data_tmp2 = Data_tmp2 %>% gather(value = "freq", key = Gene, -Disease)
      Data_tmp2 <- transform(Data_tmp2, Disease= factor(Disease, levels = sort(unique(Disease), decreasing = TRUE)))
      Data_tmp2 <- Data_tmp2 %>%
        group_by(Gene) %>%
        mutate(Gene = factor(Gene, levels = names(sort(table(Gene), decreasing = TRUE))))
      g = ggplot(Data_tmp2, aes(Gene, as.factor(Disease))) +
        geom_tile(aes(fill = freq), color = "black",
                  #            lwd = 1,
                  linetype = 1) + 
        geom_text(aes(label = round(freq, 0))) +
        scale_fill_gradient(low = "white", high = "red", limit=c(0,100), name="Frequency (%)") +
        theme_classic() +
        labs(title = paste0(nrow(Data_tmp), 
                            " LD-CH/LE-CH in F1-Liquid CDx, ",
                            nrow(Data_tmp3), " patients")) +
        theme(axis.text.x = element_text(angle = 45, hjust = 1),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              axis.line.x = element_blank(),
              axis.line.y = element_blank())
      plots[[14]] = g
      g = ggplot(Data_entropy, aes(x=Gene, y=Entropy, fill="black")) +
        geom_point(stat = "identity") +
        coord_flip() +
        theme_classic() +
        theme(legend.position = "none")
      plots[[15]] = g
      
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(CH_candidate_hotspot = case_when(
          unique_alt %in% CH_candidates & Hugo_Symbol %in% CH_gene_entropy_3 ~ "hotspot",
          TRUE ~ ""
        ))
      Data_MAF_target = Data_MAF_target %>% left_join(Data_triplet, by=c('unique_position'))
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(CH_hotspot_candidate = case_when(
          non_germline_non_error == 1 &
            unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 & 
          Panel == "F1Liquid CDx" ~ 1,
          TRUE ~ 0))
      Data_MAF_target_table = Data_MAF_target
      output$table_all_variants = downloadHandler(
        filename = function(){"all_variant_data.csv"}, 
        content = function(fname){
          write.csv(Data_MAF_target_table, fname)
        }
      )
      
      data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        #Panel == "F1Liquid CDx" &
                        Hugo_Symbol %in% CH_gene_entropy_3)
      
      CH_ID = unique(data_tmp$Tumor_Sample_Barcode)
      Data_summary = Data_summary %>%
        dplyr::mutate(CH_plus = case_when(
          Tumor_Sample_Barcode %in% CH_ID ~ 1,
          TRUE ~ 0
        ))
      
      data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        logfold > 8 & 
                        #Panel == "F1Liquid CDx" &
                        Hugo_Symbol %in% CH_gene_entropy_3)
      
      LOCH_ID = unique(data_tmp$Tumor_Sample_Barcode)
      Data_summary = Data_summary %>%
        dplyr::mutate(LOCH_plus = case_when(
          Tumor_Sample_Barcode %in% LOCH_ID ~ 1,
          TRUE ~ 0
        ))
      data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        #logfold > 8 &
                        Hugo_Symbol %in% CH_gene_entropy_3 & 
                        Panel == "F1Liquid CDx")
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "TP53", "CHEK2", "ATM", "RAD21", "MDM4", "PTPN11", "KRAS", "CBL", "SF3B1", "U2AF1", "EZH2", "KMT2D"))
      Data_tmp$CH_genes = 1
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "FoundationOne Liquid CDx",
                subtitle = "Epigenetic/DDR/MAPK/Splicing hotspot variants") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[16]] = g
      Data_tmp = data_tmp %>% dplyr::filter(!Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "TP53", "CHEK2", "ATM", "RAD21", "MDM4", "PTPN11", "KRAS", "CBL", "SF3B1", "U2AF1", "EZH2", "KMT2D"))
      Data_tmp$CH_genes = 1
      cor_A = Data_tmp %>%
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>%
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "FoundationOne Liquid CDx",
                subtitle = "Hotspot variants in other genes") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[17]] = g
      #plots[[17]] = gg_empty()
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% c("SF3B1", "U2AF1"))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "",
             subtitle = "Splicing factor hotspot variants (SF3B1, U2AF1)") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[133]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4"))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "",
             subtitle = "DNA repair gene  hotspot variants (TP53, CHEK2, ATM, RAD21, MDM4)") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[132]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D"))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "FoundationOne Liquid CDx",
                subtitle = "Epigenetic regulator hotspot variants (DNMT3A, TET2, ASXL1, IDH2, EZH2, KMT2D)") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[131]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "SF3B1")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "",
                subtitle = "CH hotspot variants (SF3B1) , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[18]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "DNMT3A")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "FoundationOne Liquid CDx",
                subtitle = "CH hotspot variants (DNMT3A) , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[19]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "TP53")
      cor_A = Data_tmp %>% 
          summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
          as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
          geom_point(alpha = .5) +
          ggtitle("CH hotspot variants (TP53) , FoundationOne Liquid") +
          ylim(c(0,1)) +
          xlim(c(0,100)) +
          theme_bw() +
          geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
          geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
          stat_poly_eq(formula = y ~ x,
                       aes(label = paste(stat(eq.label))),
                       parse = TRUE)
      plots[[20]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "ASXL1")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (ASXL1) , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[21]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "TET2")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (TET2) , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[22]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "MDM4")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (MDM4) , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[23]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "IDH2")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (IDH2) , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[24]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "CHEK2")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (CHEK2) , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
     plots[[25]] = g
     Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "ATM")
     cor_A = Data_tmp %>% 
       summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
       as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
     g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
       geom_point(alpha = .5) +
       ggtitle("CH hotspot variants (ATM) , FoundationOne Liquid") +
       ylim(c(0,1)) +
       xlim(c(0,100)) +
       theme_bw() +
       geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
       geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
       stat_poly_eq(formula = y ~ x,
                    aes(label = paste(stat(eq.label))),
                    parse = TRUE)
      plots[[26]] = g
      Data_tmp = data_tmp %>% dplyr::filter(logfold > 8)
      Data_tmp$CH_genes = as.factor(Data_tmp$CH_genes)
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF)) +
        geom_point(alpha = .5) +
        ggtitle(paste0(nrow(Data_tmp), " LD-CH, odds ratio > 256, F1-Liquid CDx")) +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 17, y = 0.74, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[27]] = g
      Data_tmp = data_tmp %>% dplyr::filter(logfold <= 8 & logfold > 2)
      Data_tmp$CH_genes = as.factor(Data_tmp$CH_genes)
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF)) +
        geom_point(alpha = .5) +
        ggtitle(paste0(nrow(Data_tmp), " LE-CH, 4< odds ratio <= 256, F1-Liquid CDx")) +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.74, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[28]] = g
      Data_tmp = data_tmp %>% dplyr::filter(logfold > 2) %>%
        dplyr::mutate(favor_only = case_when(
          logfold <= 8 ~ "Liquid-favor",
          TRUE ~ "Liquid-only"
        ))
      Data_tmp$favor_only = as.factor(Data_tmp$favor_only)
      summary_stats <- Data_tmp %>%
        group_by(favor_only) %>%
        summarise(
          mean_value = mean(VAF),
          sd_value = sd(VAF),
          .groups = "drop"
        )
      # 平均値 ± SD のラベル
      summary_stats <- summary_stats %>%
        mutate(label = paste0("Mean: ", round(mean_value, 3), "\nSD: ", round(sd_value, 3)))
      
      p_value <- wilcox.test(VAF ~ favor_only, data = Data_tmp)$p.value
      sig_label <- ifelse(p_value < 0.001, "***", "ns")  # *: 有意, ns: 有意差なし
      g <- ggplot(Data_tmp, aes(x = favor_only, y = VAF, fill = favor_only)) +
        geom_violin(alpha = 0.5) +                # バイオリンプロット
        geom_jitter(width = 0.2, size = 1, alpha = 0.05) +  # 実データ点
        geom_signif(comparisons = list(c("Liquid-favor", "Liquid-only")),  
                    annotations = sig_label, 
                    y_position = max(Data_tmp$VAF) * 1.1,  # P値の表示位置
                    tip_length = 0.02, vjust = 0.5) +
        geom_text(data = summary_stats, aes(x = favor_only, label = label), y = max(Data_tmp$VAF) * 1.07, 
                  color = "black", fontface = "bold") +  # 平均 ± SD を数値で表示
        scale_fill_manual(values = c("#00AFBB", "#E7B800")) +  # 色設定
        theme_minimal() +
        theme(legend.position = "none")  # 凡例を消す
      plots[[148]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "DNMT3A" & Smoking_years != 0 & !is.na(Smoking_years))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Smoking_years, VAF)$estimate,3), p = round(cor.test(Smoking_years, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Smoking_years, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (DNMT3A) and smoking years, only smokers, FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[33]] = g
      Data_tmp = data_tmp %>% dplyr::filter(CH_genes == 1 & Smoking_years != 0 & !is.na(Smoking_years))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Smoking_years, VAF)$estimate,3), p = round(cor.test(Smoking_years, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Smoking_years, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (CH genes) and smoking years, only smokers, FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[34]] = g
      # Data_tmp = data_tmp %>% dplyr::filter(CH_genes == 0 & Smoking_years != 0 & !is.na(Smoking_years))
      # cor_A = Data_tmp %>% 
      #   summarise(cor = round(cor.test(Smoking_years, VAF)$estimate,3), p = round(cor.test(Smoking_years, VAF)$p.value,3)) %>% 
      #   as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      # g <- ggplot(Data_tmp, aes(x = Smoking_years, y = VAF, color = CH_genes)) +
      #   geom_point(alpha = .5) +
      #   ggtitle("CH hotspot variants (non-CH genes) and smoking years, only smokers, FoundationOne Liquid") +
      #   ylim(c(0,1)) +
      #   xlim(c(0,100)) +
      #   theme_bw() +
      #   geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
      #   geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
      #   stat_poly_eq(formula = y ~ x,
      #                aes(label = paste(stat(eq.label))),
      #                parse = TRUE)
      # plots[[35]] = g
      # plots[[35]] = gg_empty()
      Data_tmp = data_tmp %>% dplyr::filter(Smoking_years != 0 & !is.na(Smoking_years))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Smoking_years, VAF)$estimate,3), p = round(cor.test(Smoking_years, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Smoking_years, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CH hotspot variants (all genes) and smoking years, only smokers, FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[36]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 & (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1)) %>%
        dplyr::filter(Panel == "F1Liquid CDx")
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "KRAS")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("KRAS hotspot variants , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[31]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "PIK3CA")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("PIK3CA hotspot variants , FoundationOne Liquid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[32]] = g
      data_tmp2 = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::mutate(
          Lines_and_CH = case_when(
            (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) & non_germline_non_error == 1 & CH_genes == 1 & Lines_pre_CGP_2_or_more == 1 ~ "CH hotspots (+)\n >1 lines",
            ((No_found_in_F1L < Min_count & No_found_in_F1 <= Min_count_F1) | non_germline_non_error != 1 | CH_genes == 0) & Lines_pre_CGP_2_or_more == 1 ~ "CH hotspots (-)\n >1 lines",
            (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) & non_germline_non_error == 1 & CH_genes == 1 & Lines_pre_CGP_2_or_more == 0 ~ "CH hotspots (+)\n 0~1 lines",
            ((No_found_in_F1L < Min_count & No_found_in_F1 <= Min_count_F1) | non_germline_non_error != 1 | CH_genes == 0) & Lines_pre_CGP_2_or_more == 0 ~ "CH hotspots (-)\n 0~1 lines"
          )
        )
      setProgress(detail = "4901")
      data_tmp = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::mutate(
          Lines_and_CH = case_when(
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(Lines_and_CH == "CH hotspots (+)\n >1 lines"))$Tumor_Sample_Barcode ~ "CH hotspots (+)\n >1 lines",
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(Lines_and_CH == "CH hotspots (-)\n >1 lines"))$Tumor_Sample_Barcode ~ "CH hotspots (-)\n >1 lines",
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(Lines_and_CH == "CH hotspots (+)\n 0~1 lines"))$Tumor_Sample_Barcode ~ "CH hotspots (+)\n 0~1 lines",
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(Lines_and_CH == "CH hotspots (-)\n 0~1 lines"))$Tumor_Sample_Barcode ~ "CH hotspots (-)\n 0~1 lines",
            Lines_pre_CGP_2_or_more == 1 ~ "CH hotspots (-)\n >1 lines",
            TRUE ~ "CH hotspots (-)\n 0~1 lines"
          )
        )
      
      sample_data = data.frame(
        sort(table(data_tmp$Lines_and_CH))
      )
      colnames(sample_data) = c("type","n")
      g = sample_data %>%
        mutate(per = n / sum(n)) %>%
        mutate(label = paste0(type, "\n", scales::percent(per, 0.1))) %>%
        ggplot(aes(x = 0, y = n, fill = factor(n))) +
        geom_col() +
        coord_polar("y") +
        geom_text(aes(label = label), position = position_stack(vjust = 0.5)) +
        theme_void() +
        theme(legend.position = "none") +
        ggtitle("LD-CH/LE-CH , F1-Liquid CDx")
      plots[[2]] = g
      data_tmp2 = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::mutate(
          oncogenic_varinats_and_CH = case_when(
            (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) & non_germline_non_error == 1 & CH_genes == 1 & oncogenic_muts_more_than_2 == 1 ~ "Pts with CH hotspots,\n >2 oncogenic variants",
            ((No_found_in_F1L < Min_count & No_found_in_F1 <= Min_count_F1) | non_germline_non_error != 1 | CH_genes == 0) & oncogenic_muts_more_than_2 == 1 ~ "Pts without CH hotspots,\n >2 oncogenic variants",
            (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) & non_germline_non_error == 1 & CH_genes == 1 & oncogenic_muts_more_than_2 == 0 ~ "Pts with CH hotspots,\n 0~1 oncogenic variants",
            ((No_found_in_F1L < Min_count & No_found_in_F1 <= Min_count_F1) | non_germline_non_error != 1 | CH_genes == 0) & oncogenic_muts_more_than_2 == 0 ~ "Pts without CH hotspots,\n 0~1 oncogenic variants",
          )
        )
      data_tmp = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::mutate(
          oncogenic_varinats_and_CH = case_when(
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(oncogenic_varinats_and_CH == "Pts with CH hotspots,\n >2 oncogenic variants"))$Tumor_Sample_Barcode ~ "Pts with CH hotspots,\n >2 oncogenic variants",
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(oncogenic_varinats_and_CH == "Pts without CH hotspots,\n >2 oncogenic variants"))$Tumor_Sample_Barcode ~ "Pts without CH hotspots,\n >2 oncogenic variants",
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(oncogenic_varinats_and_CH == "Pts with CH hotspots,\n 0~1 oncogenic variants"))$Tumor_Sample_Barcode ~ "Pts with CH hotspots,\n 0~1 oncogenic variants",
            Tumor_Sample_Barcode %in% (data_tmp2 %>% dplyr::filter(oncogenic_varinats_and_CH == "Pts without CH hotspots,\n 0~1 oncogenic variants"))$Tumor_Sample_Barcode ~ "Pts without CH hotspots,\n 0~1 oncogenic variants",
            TRUE ~ "Pts without CH hotspots,\n 0~1 oncogenic variants"
          )
        )
      sample_data = data.frame(
        sort(table(data_tmp$oncogenic_varinats_and_CH))
      )
      colnames(sample_data) = c("type","n")
      g = sample_data %>%
        mutate(per = n / sum(n)) %>%
        mutate(label = paste0(type, "\n", scales::percent(per, 0.1))) %>%
        ggplot(aes(x = 0, y = n, fill = factor(n))) +
        geom_col() +
        coord_polar("y") +
        geom_text(aes(label = label), position = position_stack(vjust = 0.5)) +
        theme_void() +
        theme(legend.position = "none") +
        ggtitle("LD-CH/LE-CH , F1-Liquid CDx")
      plots[[51]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" & name %in% c("APC_Q2322R", "STK11_F354L", "CHEK2_R519*", "CHEK2_R181H", "RNF43_P660fs*87", "EGFR_V742I"))
      g = data_tmp %>%
        ggplot(aes(x = name, y = VAF)) +
        geom_quasirandom(aes(color = name),
                         size = 2, 
                         alpha =.5) +
        theme_classic()+
        labs(x = "Variant",
             y = "Variant allele frequency",
             title = "Variant allele frequency") +
        theme(
          axis.title.x = element_blank(), 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank()
        )
      plots[[29]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" & unique_position %in% F1L_germline_error2) %>%
        dplyr::mutate(name = case_when(
          unique_position == "5_80654913_80654913_G_GGCCGCAGCG" ~ "MSH3_A61_A62insAAA",
          unique_position == "5_80654916_80654916_C_CGCAGCGCCCGCAGCGCCC" ~ "MSH3_A62_P63insPAAPAA",
          TRUE ~ name,
        ))
      g = data_tmp %>%
        ggplot(aes(x = name, y = VAF)) +
        geom_quasirandom(aes(color = name),
                         size = 2, 
                         alpha =.5) +
        theme_classic()+
        labs(x = "Variant",
             y = "Variant allele frequency",
             title = "Variant allele frequency") +
        theme(
          axis.title.x = element_blank(), 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank()
        )
      plots[[364]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" & name %in% c("MLH1_R9W", "CDKN2A_S12L", "CDK12_R99C", "CHEK2_R474C", "MYCN_R398W", "DNMT3A_I780T", "DNMT3A_R326C", "DNMT3A_F731del"))
      g = data_tmp %>%
        ggplot(aes(x = name, y = VAF)) +
        geom_quasirandom(aes(color = name),
                         size = 2, 
                         alpha =.5) +
        theme_classic()+
        labs(x = "Variant",
             y = "Variant allele frequency",
             title = "Variant allele frequency") +
        theme(
          axis.title.x = element_blank(), 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank()
        )
      g
      plots[[50]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" & name %in% c("KRAS_G12D", "KRAS_G12V*", "TERT_c.-79-45C>T", "TP53_R175H", "TP53_R273H", "PIK3CA_E545K"))
      g = data_tmp %>%
        ggplot(aes(x = name, y = VAF)) +
        geom_quasirandom(aes(color = name),
                         size = 2, 
                         alpha =.5) +
        theme_classic()+
        labs(x = "Variant",
             y = "Variant allele frequency",
             title = "Variant allele frequency") +
        theme(
          axis.title.x = element_blank(), 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank()
        )
      g
      plots[[30]] = g
      data_tmp = Data_MAF_target %>%
        dplyr::filter((non_germline_non_error == 1 |
                        # non_germline_non_error == 6) &
                         non_germline_non_error == 16) &
                        unique_alt %in% CH_candidates) %>%
        dplyr::filter(Panel == "FoundationOne CDx")
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "KRAS")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("KRAS hotspot variants found by FoundationOne solid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[37]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "DNMT3A")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("DNMT3A hotspot variants found by FoundationOne solid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[38]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "CHEK2")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("CHEK2 hotspot variants found by FoundationOne solid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[39]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% c("SF3B1", "U2AF1"))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "",
                subtitle = "Splicing factor hotspot variants (SF3B1, U2AF1)") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[136]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4"))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "",
                subtitle = "DNA repair gene  hotspot variants (TP53, CHEK2, ATM, RAD21, MDM4)") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[135]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "EZH2", "KMT2D"))
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        labs(title = "FoundationOne CDx (solid tumor)",
                subtitle = "Epigenetic regulator hotspot variants (DNMT3A, TET2, ASXL1, EZH2, KMT2D)") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[134]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "ATM")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("ATM hotspot variants found by FoundationOne solid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[39]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "TP53")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("TP53 hotspot variants found by FoundationOne solid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[40]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "ASXL1")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("ASXL1 hotspot variants found by FoundationOne solid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[41]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == "TET2")
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("TET2 hotspot variants found by FoundationOne solid") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[42]] = g
      Data_tmp = Data_MAF_target %>%
                 dplyr::filter(non_germline_non_error == 1 &
                                 Panel == "F1Liquid CDx" &
                                 unique_alt %in% CH_candidates &
                                 Hugo_Symbol %in% CH_gene_entropy_3) %>%
                 dplyr::arrange(SNV_check) %>%
                 dplyr::distinct(unique_alt, .keep_all = T)
      g <- ggplot(Data_tmp, aes(x = logfold, color = mut_pattern, fill = mut_pattern),alpha = 0.5) +
        geom_histogram() +
        facet_grid(mut_pattern ~ .) +
        #ylim(c(0,1)) +
        xlim(c(-3,10.5)) +
        labs(title = "Distribution of odds ratio for liquid/solid detection",
             subtitle = "CH hotspot candidate only",
             x="log2 odds ratio",
             y="Hotspot variant count") +
        geom_vline(xintercept=c(2,8), linetype = "dashed", col="blue") +
        theme_bw()
      plots[[52]] = g
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                      Panel == "F1Liquid CDx" &
                      Hugo_Symbol %in% CH_gene_entropy_3 &
                        (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1))# %>%
        # dplyr::arrange(SNV_check) %>%
        # dplyr::distinct(unique_alt, .keep_all = T)  
      g <- ggplot(Data_tmp, aes(x = logfold, color = mut_pattern, fill = mut_pattern),alpha = 0.5) +
        geom_histogram() +
        facet_grid(mut_pattern ~ .) +
        #ylim(c(0,1)) +
        xlim(c(-3,10.5)) +
        labs(title = "Distribution of odds ratio for liquid/solid detection",
             subtitle = paste0("All ", nrow(Data_tmp), " hotspot variants"),
             x="log2 odds ratio",
             y="Hotspot vasiant count") +
        geom_vline(xintercept=c(2,8), linetype = "dashed", col="blue") +
        theme_bw()
      plots[[44]] = g
      Data_tmp <- Data_tmp %>%
        mutate(logfold_group = cut(logfold, 
                                   breaks = c(-Inf, 2, 8, Inf), 
                                   labels = c("< 2", "2 ~ 8", "> 8"),
                                   right = FALSE))
      data_summary <- Data_tmp %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, tumor vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[76]] = g
      Gene_name = "DNMT3A"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, solid vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[86]] = g
      Gene_name = "TET2"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, solid vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[87]] = g
      Gene_name = "ASXL1"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, solid vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[88]] = g
      Gene_name = "ATM"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, solid vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[89]] = g
      Gene_name = "CHEK2"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, solid vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[90]] = g
      Gene_name = "TP53"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, solid vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[91]] = g
      Gene_name = "SF3B1"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT, solid vs liquid, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[92]] = g
      CH_candidates_base = unique(Data_MAF_target_table_2$unique_alt)
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot of hotspot variants for favor F1 CDx or F1-Liquid CDx",
             subtitle = paste0(nrow(Data_tmp), " hotspot variants found in F1 CDx or F1-Liquid CDx >= 10 times"),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[45]] = g
      Data_tmp = Data_MAF_target_all %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(#unique_alt %in% CH_candidates_base &
          #Hugo_Symbol %in% CH_gene_entropy_3 &
          !Hugo_Symbol %in% CH_GENE_NatComms &
          Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        #dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot of hotspot variants for favor F1 CDx or F1-Liquid CDx",
             subtitle = paste0(nrow(Data_tmp), " hotspot variants found in F1 or F1L >= 10 times, all genes"),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[368]] = g
      Data_tmp = Data_MAF_target_all %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(#unique_alt %in% CH_candidates_base &
          !Hugo_Symbol %in% CH_gene_entropy_3 &
          #!Hugo_Symbol %in% CH_GENE_NatComms &
            Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        #dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot of hotspot variants for favor F1 CDx or F1-Liquid CDx",
             subtitle = paste0(nrow(Data_tmp), " hotspot variants found in F1 or F1L >= 10 times, all genes"),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[369]] = g
      setProgress(detail = "5502")
      Data_tmp = Data_MAF_target_all %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(#unique_alt %in% CH_candidates_base &
          Hugo_Symbol %in% CH_gene_entropy_3 &
            #!Hugo_Symbol %in% CH_GENE_NatComms &
            Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        #dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot of hotspot variants for favor F1 CDx or F1-Liquid CDx",
             subtitle = paste0(nrow(Data_tmp), " hotspot variants found in F1 or F1L >= 10 times, all genes"),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[370]] = g
      data_tmp_uni_detail=Data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      table_save_tmp_A = Data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp_A, "source/Table_S5_selected.csv")
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx") %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        #dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      table_save_tmp_B = Data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp_B, "source/Table_S5_entropy.csv")
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        #Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx") %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      table_save_tmp_C = Data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp_C, "source/Table_S5_noise.csv")
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx") %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      table_save_tmp_D = Data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp_D, "source/Table_S5_entropy_noise.csv")
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        #Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx") %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        #dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      table_save_tmp_E = Data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp_E, "source/Table_S5_before_filter.csv")
      Gene_name = "DNMT3A"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 CDx or F1-Liquid CDx",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name, " LD-CH/LE-CH"),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[64]] = g
      Gene_name = "ATM"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L, after filtering",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[65]] = g
      setProgress(detail = "5678")
      Gene_name = "ASXL1"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L, after filtering",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[66]] = g
      Gene_name = "TET2"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L, after filtering",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[67]] = g
      Gene_name = "CHEK2"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L, after filtering",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[68]] = g
      Gene_name = "JAK2"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L, after filtering",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[69]] = g
      Gene_name = "IDH2"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      setProgress(detail = "5849")
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L, after filtering",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[70]] = g
      Gene_name = "KMT2D"
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol == Gene_name &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)  
      Data_tmp$CH_genes = factor(Data_tmp$CH_genes)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=CH_genes, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L, after filtering",
             subtitle = paste0("Variants found in F1 or F1L >= 10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[71]] = g
      setProgress(detail = "5915")
      
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates_base &
                      Hugo_Symbol == "DNMT3A" &
                      Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T)
      table_DNMT3A = read_csv("source/DNMT3A_INS.csv", show_col_types = FALSE)
      table_DNMT3A = table_DNMT3A %>%
        dplyr::select(Position, Destabilizing) %>%
        dplyr::mutate(Destabilizing = case_when(
          Destabilizing == "Yes" ~ "Instable",
          Destabilizing == "No" ~ "Stable"
        ))
      
      Data_tmp$Position = as.numeric(unlist(str_extract(Data_tmp$amino.acid.change, "\\d+")))
      Data_tmp = Data_tmp %>% left_join(table_DNMT3A, by = "Position")
      Data_tmp$Destabilizing[is.na(Data_tmp$Destabilizing)] = "Unknown"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "c")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "\\*")] = "Truncating"
      Data_tmp$Destabilizing = factor(Data_tmp$Destabilizing)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=Destabilizing, shape=Destabilizing, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot of DNMT3A LD-CH/LE-CH for favor F1 CDx or F1-Liquid CDx, grouped by stability",
             subtitle = "Variants found in F1 CDx or F1-Liquid CDx >= 10 times",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4, -2, 0,2,4,6,8,10),
                           labels = c("~-4", "-2", "0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = unique_alt), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[9]] = g
      setProgress(detail = "5970")
      
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx")
      Data_tmp$Position = as.numeric(unlist(str_extract(Data_tmp$amino.acid.change, "\\d+")))
      Data_tmp = Data_tmp %>% left_join(table_DNMT3A, by = "Position")
      Data_tmp$Destabilizing[is.na(Data_tmp$Destabilizing)] = "Unknown"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "c")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "\\*")] = "Truncating"
      Data_tmp$Destabilizing = factor(Data_tmp$Destabilizing)
      lolliplot_gene = input$lolliplot_gene
      if(is.null(lolliplot_gene)){
        lolliplot_gene = "DNMT3A"
      } else if(lolliplot_gene == ""){
        lolliplot_gene = "DNMT3A"
      }
      setProgress(detail = "5986")
      data_lolliplot = Data_tmp %>%
        dplyr::select(Chromosome,
                      Start_Position,
                      End_Position,
                      Reference_Allele,
                      Tumor_Seq_Allele2,
                      Tumor_Sample_Barcode,
                      Hugo_Symbol,
                      Variant_Classification,
                      amino.acid.change)
      if(lolliplot_gene == "DNMT3A"){
        data_lolliplot$Variant_Classification = Data_tmp$Destabilizing
      }
      data_lolliplot = data_lolliplot %>%
        dplyr::filter(!str_detect(amino.acid.change, "c.")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "_")) %>%
        dplyr::filter(Hugo_Symbol == lolliplot_gene) 
      if(length(data_lolliplot$Chromosome) == 0){
        plots[[3]] = gg_empty()
      } else {
        var<-unique(data_lolliplot[!duplicated(data_lolliplot),])#remove duplicates
        var[order(var$Hugo_Symbol,gsub("([A-Z]+)([0-9]+)","\\2",var$amino.acid.change)),]#order
        var.freq<-plyr::count(var,c("Hugo_Symbol","amino.acid.change","Variant_Classification"))
        var.aanum<-unlist(lapply(regmatches(var.freq$amino.acid.change,gregexpr(pattern="*(\\d+)",var.freq$amino.acid.change)),function(x) x[[1]]))
        var.plot.data<-cbind(var.freq,as.numeric(as.character(var.aanum)))#hard way to remove factor
        colnames(var.plot.data)[5]<-"var.aanum"
        var.plot.data = var.plot.data[var.plot.data$var.aanum >=1 & is.finite(var.plot.data$var.aanum) & !is.na(var.plot.data$var.aanum),]
        setProgress(detail = "6014")
        
        var.plot<-isolate(var.plot.data)
        var.plot<-var.plot[var.plot$Hugo_Symbol==lolliplot_gene,]
        var.plot<-var.plot[order(var.plot$var.aanum),]
        var.plot$amino.acid.change<-as.character(var.plot$amino.acid.change)
        
        incProgress(1 / 4)
        setProgress(detail = "6019")
        
        p<-ggplot(var.plot,aes(var.aanum,freq,color=Variant_Classification,label=amino.acid.change)) +
          geom_segment(aes(x=var.aanum,y=0,xend=var.aanum,yend=freq),
                       color=ifelse(var.plot$freq>=input$lolliplot_no,"orange","grey50"),
                       linewidth=ifelse(var.plot$freq>=input$lolliplot_no,1.2,0.6)) +
          geom_point(size=5) +
          geom_text_repel(nudge_y=max(var.plot$freq) * 0.05,
                          color=ifelse(var.plot$freq>=input$lolliplot_no,"orange","NA"),
                          size=ifelse(var.plot$freq>=input$lolliplot_no,4,2),
                          fontface="bold", max.overlaps=Inf) +
          theme_classic()
        setProgress(detail = "6031")
        proteins_acc<-read.table(file("source/UniProt.txt",
                                      encoding='UTF-8-BOM'),header=T)
        if(length(proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]) == 1){
          proteins_acc<-proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]
          proteins_acc_url<-gsub(" ","%2C",proteins_acc)
          baseurl<-"https://www.ebi.ac.uk/proteins/api/features?offset=0&size=100&accession="
          url<-paste0(baseurl,proteins_acc_url)
          incProgress(1 / 6)
          flag_lolliplot = 0
          if(file.exists(paste0("source/lolliplot/", lolliplot_gene, ".rda"))){
            load(paste0("source/lolliplot/", lolliplot_gene, ".rda"))
            flag_lolliplot = 1
          } else {
            prots_feat<-try(GET(url,accept_json()), silent = FALSE)
            if (class(prots_feat) != "try-error"){
              flag_lolliplot = 1
              save(prots_feat, file=paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))
            }
          }
          if (flag_lolliplot == 1){
            prots_feat_red<-(httr::content(prots_feat))
            incProgress(1 / 6)
            features_total_plot<-NULL
            for(i in 1:length(prots_feat_red)){ 
              features_temp<-drawProteins::extract_feat_acc(prots_feat_red[[i]])#the extract_feat_acc() function takes features into a data.frame
              features_temp$order<-i # this order is needed for plotting later
              features_total_plot<-rbind(features_total_plot,features_temp)
            }
            plot_start<-0#starts 0
            plot_end<-max(features_total_plot$end)
            if("CHAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="CHAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.1*max(var.plot$freq),ymax=-0.04*max(var.plot$freq)),
                          colour="grey",
                          fill="grey",
                          inherit.aes=F)
            }
            incProgress(1 / 6)
            if("DNA_BIND"%in%(unique(features_total_plot$type))) {
              features_total_plot[features_total_plot$type=="DNA_BIND","description"] <- features_total_plot[features_total_plot$type=="DNA_BIND","type"]
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DNA_BIND",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("DOMAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DOMAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("MOTIF"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="MOTIF",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("REPEAT"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="REPEAT",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            p <- p +
              scale_x_continuous(breaks=round(seq(plot_start,plot_end,by=50)),name="Amino acid number")
          }
        }
        p <- p +
          theme(plot.title=element_text(face="bold",size=(15),hjust=0),axis.text.x=element_text(angle=90,hjust=1)) +
          labs(y="Mutation frequency in our dataset",
               title=paste(nrow(data_lolliplot), " variants in ",lolliplot_gene,", C-CAT F1-Liquid CDx",sep="")) +
          scale_y_continuous(breaks=seq(0,max(var.plot$freq*1.05),max(floor(max(var.plot$freq)/100)*10,5)),name="Frequency")
        plots[[3]] = p
      }
      
      Data_maf = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        Panel == "F1Liquid CDx"&
                        unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3) %>%
        dplyr::arrange(Chromosome, Start_Position) %>%
        dplyr::left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      data_maf = Data_maf %>% dplyr::select(Hugo_Symbol)
      data_maf$Chromosome = paste0("chr", Data_maf$Chromosome)
      data_maf$Start_Position = as.integer(Data_maf$Start_Position)
      data_maf$End_Position = as.integer(Data_maf$Start_Position) + nchar(Data_maf$Reference_Allele)
      data_maf$Strand = "+"
      data_maf$Variant_Classification = "Missense_Mutation"
      data_maf$Variant_Type = ifelse(nchar(Data_maf$Reference_Allele) == 1 & nchar(Data_maf$Tumor_Seq_Allele2) == 1, "SNV",
                                     ifelse(nchar(Data_maf$Reference_Allele) > nchar(Data_maf$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf$Reference_Allele = Data_maf$Reference_Allele
      data_maf$Tumor_Seq_Allele1 = Data_maf$Reference_Allele
      data_maf$Tumor_Seq_Allele2 = Data_maf$Tumor_Seq_Allele2
      data_maf$Tumor_Sample_Barcode = Data_maf$Tumor_Sample_Barcode
      data_maf$favor = ifelse(Data_maf$logfold > 8, "A",
                              ifelse(Data_maf$logfold > 2, "B","C"))
      data_maf$Protein_Change = Data_maf$unique_alt
      data_maf$amino.acid.change = Data_maf$amino.acid.change
      setProgress(detail = "6132")
      
      all_samples = (Data_summary %>% dplyr::filter(Panel == "F1Liquid CDx"))$Tumor_Sample_Barcode
      no_mut_samples = all_samples[!all_samples %in% unique(data_maf$Tumor_Sample_Barcode)]
      dummy_data <- data.frame(
        Hugo_Symbol = "DUMMY",
        Chromosome = "chr1",
        Start_Position = 1,
        End_Position = 1,
        Strand = "+",
        Variant_Classification = "Missense_Mutation",
        Variant_Type = "dummy",
        Reference_Allele = "A",
        Tumor_Seq_Allele1 = "A",
        Tumor_Seq_Allele2 = "T",
        Tumor_Sample_Barcode = all_samples,
        favor = "dummy",
        Protein_Change = "dummy"
      )
      data_maf$Position = as.numeric(unlist(str_extract(data_maf$amino.acid.change, "\\d+")))
      data_maf = data_maf %>% left_join(table_DNMT3A, by = "Position")
      data_maf_rename = data_maf %>%
        dplyr::mutate(Hugo_Symbol = case_when(
          Hugo_Symbol == "DNMT3A" & amino.acid.change %in% c("R882C", "R882H", "R882P", "R882S") ~ "DNMT3A_R882",
          Hugo_Symbol == "DNMT3A" & amino.acid.change %in% c("Y735C") ~ "DNMT3A_Y735C",
          Hugo_Symbol == "DNMT3A" & Destabilizing == "Instable" ~ "DNMT3A_INS",
          Hugo_Symbol == "DNMT3A" & str_detect(amino.acid.change, "c") ~ "DNMT3A_splice",
          Hugo_Symbol == "DNMT3A" & str_detect(amino.acid.change, "\\*") ~ "DNMT3A_trunc",
          Hugo_Symbol == "DNMT3A" ~ "DNMT3A_NOS",
          TRUE ~ Hugo_Symbol
        ))
      Genes_maf_F1L = unique(data_maf_rename$Hugo_Symbol)[1:20]
      Genes_maf_F1L = Genes_maf_F1L[Genes_maf_F1L != "DUMMY"]
      maf_with_mut_samples_rename <- read.maf(maf = data_maf_rename)
      maf_without_mut_samples <- read.maf(maf = dummy_data)
      maf_with_all_samples_rename  <- merge_mafs(mafs = list(maf_with_mut_samples_rename, maf_without_mut_samples))
      maf_with_all_samples_rename <- subsetMaf(maf_with_all_samples_rename, genes = Genes_maf_F1L, dropLevels = FALSE)
      write_tsv(data_maf, "source/hotspot_maf_file.tsv")
      data_maf_CH_hotspot = data_maf %>% dplyr::filter(favor %in% c("A", "B"))
      maf_with_mut_samples <- read.maf(maf = data_maf_CH_hotspot)
      maf_with_all_samples  <- merge_mafs(mafs = list(maf_with_mut_samples, maf_without_mut_samples))
      maf_with_all_samples <- subsetMaf(maf_with_all_samples, genes = CH_gene_entropy_3, dropLevels = FALSE)
      output$figure_oncoplot1 = renderPlot({
        oncoplot(maf = maf_with_all_samples,
                 sortByAnnotation = TRUE,
                 #top = 10,
                 fontSize = 0.8,
                 drawColBar = FALSE,
                 drawRowBar = FALSE,
                 genesToIgnore = "DUMMY",
                 borderCol = NULL,
                 bgCol = "#EEEEEE",
                 drawBox = TRUE,
                 legendFontSize = 1.0)
        })
      output$figure_oncoplot2 = renderPlot({
        oncoplot(maf = maf_with_all_samples_rename,
                 top = 14,
                 sortByAnnotation = TRUE,
                 fontSize = 0.8,
                 drawColBar = FALSE,
                 drawRowBar = FALSE,
                 borderCol = NULL,
                 bgCol = "#EEEEEE",
                 drawBox = TRUE,
                 gene_mar = 15,
                 keepGeneOrder = FALSE,
                 GeneOrderSort = FALSE,
                 legendFontSize = 1.0)
      })
      output$figure_oncoplot3 = renderPlot({
        somaticInteractions(maf = maf_with_all_samples,
                            top = 10,
                            showCounts = FALSE,
                            pvalue = c(0.05, 0.01),
                            showSigSymbols = TRUE,
                            sigSymbolsSize = 2.0,
                            nShiftSymbols = 0,
                            pvSymbols = c(42, 1),
                            showSum = FALSE)
      })
      #par("mar"=c(1,1,1,1))
      # Summary_figure_oncoplot3 = somaticInteractions(maf = maf_with_all_samples,
      #                                                top = 10,
      #                                                showCounts = FALSE,
      #                                                pvalue = c(0.05, 0.01),
      #                                                showSigSymbols = TRUE,
      #                                                sigSymbolsSize = 2.0,
      #                                                nShiftSymbols = 0,
      #                                                pvSymbols = c(42, 1),
      #                                                showSum = FALSE)
      output$figure_oncoplot3_data = DT::renderDataTable(Summary_figure_oncoplot3,
                                                         server = FALSE,
                                                         filter = 'top', 
                                                         extensions = c('Buttons'), 
                                                         options = list(pageLength = 100, 
                                                                        scrollX = TRUE,
                                                                        scrollY = "1000px",
                                                                        scrollCollapse = TRUE,
                                                                        dom="Blfrtip",
                                                                        buttons = c('csv', 'copy')))
      output$figure_oncoplot4 = renderPlot({
        somaticInteractions(maf = maf_with_all_samples_rename,
                            top = length(Genes_maf_F1L),
                            pvalue = c(0.05, 0.01),
                            showSigSymbols = TRUE,
                            sigSymbolsSize = 2.0,
                            nShiftSymbols = 1,
                            pvSymbols = c(42, 1),
                            showSum = FALSE)
      })
      #par("mar"=c(1,1,1,1))
      # Summary_figure_oncoplot4 =
      #   somaticInteractions(maf = maf_with_all_samples_rename,
      #                     top = length(Genes_maf_F1L),
      #                     pvalue = c(0.05, 0.01),
      #                     showSigSymbols = TRUE,
      #                     sigSymbolsSize = 2.0,
      #                     nShiftSymbols = 1,
      #                     pvSymbols = c(42, 1),
      #                     showSum = FALSE)
      setProgress(detail = "6247")
      output$figure_oncoplot4_data = DT::renderDataTable(Summary_figure_oncoplot4,
                                                      server = FALSE,
                                                      filter = 'top', 
                                                      extensions = c('Buttons'), 
                                                      options = list(pageLength = 100, 
                                                                     scrollX = TRUE,
                                                                     scrollY = "1000px",
                                                                     scrollCollapse = TRUE,
                                                                     dom="Blfrtip",
                                                                     buttons = c('csv', 'copy')))
      data_maf = data_maf %>% 
        dplyr::arrange(Tumor_Sample_Barcode)# %>%
        #dplyr::distinct(Protein_Change,.keep_all = T)
      maf.dset <- as.data.frame(data_maf)
      maf.dset$Start_Position <- as.integer(maf.dset$Start_Position)
      maf.dset$End_Position   <- as.integer(maf.dset$End_Position)
      maf.dset <- filterSNV(dataSet = maf.dset,
                            seq_colNames = c("Reference_Allele",
                                             "Tumor_Seq_Allele1",
                                             "Tumor_Seq_Allele2"))
      hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      maf.dset <- attachContext(mutData = maf.dset,
                                chr_colName = "Chromosome",
                                start_colName = "Start_Position",
                                end_colName = "End_Position",
                                nucl_contextN = 3,
                                BSGenomeDb = hg38)
      setProgress(detail = "6283")
      maf.dset <- removeMismatchMut(mutData = maf.dset,
                                    refMut_colName = "Reference_Allele",
                                    context_colName = "context",
                                    refMut_format = "N")
      setProgress(detail = "6288")
      maf.dset <- attachMutType(mutData = maf.dset,
                                ref_colName = "Reference_Allele",
                                var_colName = "Tumor_Seq_Allele1",
                                var2_colName = "Tumor_Seq_Allele2",
                                context_colName = "context")
      setProgress(detail = "6294")
      maf.counts <- countMutTypes(mutTable = maf.dset,
                                  sample_colName = "favor",
                                  mutType_colName = "mutType")
      setProgress(detail = "6298")
      output$figure_inter_panel_a = renderPlot({
        grid.arrange(grobs=plots[1:25],
                     ncol = 5)
      })
      output$figure_inter_panel_b = renderPlot({
        grid.arrange(grobs=plots[26:50],
                     ncol = 5)
      })
      output$figure_inter_panel_c = renderPlot({
        grid.arrange(grobs=plots[51:75],
                     ncol = 5)
      })
      output$figure_inter_panel_d = renderPlot({
        grid.arrange(grobs=plots[76:100],
                     ncol = 5)
      })
      output$figure_inter_panel_e = renderPlot({
        grid.arrange(grobs=plots[101:125],
                     ncol = 5)
      })
      output$figure_inter_panel_f = renderPlot({
        grid.arrange(grobs=plots[126:150],
                     ncol = 5)
      })
      output$figure_inter_panel_g = renderPlot({
        grid.arrange(grobs=plots[c(3,43,124,125,126)],
                     ncol = 1)
      })
      output$figure_inter_panel_h = renderPlot({
        grid.arrange(grobs=plots[151:175],
                     ncol = 5)
      })
      output$figure_inter_panel_i = renderPlot({
        grid.arrange(grobs=plots[176:200],
                     ncol = 5)
      })
      output$figure_inter_panel_j = renderPlot({
        grid.arrange(grobs=plots[201:225],
                     ncol = 5)
      })
      output$figure_inter_panel_k = renderPlot({
        grid.arrange(grobs=plots[226:250],
                     ncol = 5)
      })
      output$figure_inter_panel_l = renderPlot({
        grid.arrange(grobs=plots[251:275],
                     ncol = 5)
      })
      output$figure_inter_panel_m = renderPlot({
        grid.arrange(grobs=plots[276:300],
                     ncol = 5)
      })
      output$figure_inter_panel_n = renderPlot({
        grid.arrange(grobs=plots[301:325],
                     ncol = 5)
      })
      output$figure_inter_panel_o = renderPlot({
        grid.arrange(grobs=plots[326:350],
                     ncol = 5)
      })
      output$figure_inter_panel_p = renderPlot({
        grid.arrange(grobs=plots[351:372],
                     ncol = 5)
      })
      output$figure_inter_panel_download_a <- downloadHandler(
        filename = function() {
          paste0("main_figure_a.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[1:25],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_b <- downloadHandler(
        filename = function() {
          paste0("main_figure_b.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[26:50],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_c <- downloadHandler(
        filename = function() {
          paste0("main_figure_c.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[51:75],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_d <- downloadHandler(
        filename = function() {
          paste0("main_figure_d.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[76:100],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_e <- downloadHandler(
        filename = function() {
          paste0("main_figure_e.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[101:125],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_f <- downloadHandler(
        filename = function() {
          paste0("main_figure_f.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[126:150],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_g <- downloadHandler(
        filename = function() {
          paste0("main_figure_g.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[c(3,43,124,125,126)],
                       ncol = 1)
          dev.off()
        }
      )
      output$figure_inter_panel_download_h <- downloadHandler(
        filename = function() {
          paste0("main_figure_h.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[151:175],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_i <- downloadHandler(
        filename = function() {
          paste0("main_figure_i.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[176:200],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_j <- downloadHandler(
        filename = function() {
          paste0("main_figure_j.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[201:225],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_k <- downloadHandler(
        filename = function() {
          paste0("main_figure_k.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[226:250],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_l <- downloadHandler(
        filename = function() {
          paste0("main_figure_l.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[251:275],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_m <- downloadHandler(
        filename = function() {
          paste0("main_figure_m.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[276:300],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_n <- downloadHandler(
        filename = function() {
          paste0("main_figure_n.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[301:325],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_o <- downloadHandler(
        filename = function() {
          paste0("main_figure_o.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[326:350],
                       ncol = 5)
          dev.off()
        }
      )
      output$figure_inter_panel_download_p <- downloadHandler(
        filename = function() {
          paste0("main_figure_p.pdf")
        },
        content = function(file) {
          pdf(file, width = 60, height = 60)
          grid.arrange(grobs=plots[351:372],
                       ncol = 5)
          dev.off()
        }
      )
      output$table_inter_panel1 = DT::renderDataTable(Summary_Table,
                                                    server = FALSE,
                                                    filter = 'top', 
                                                    extensions = c('Buttons'), 
                                                    options = list(pageLength = 100, 
                                                                   scrollX = TRUE,
                                                                   scrollY = "1000px",
                                                                   scrollCollapse = TRUE,
                                                                   dom="Blfrtip",
                                                                   buttons = c('csv', 'copy')))
      Data_tmp3 = Data_summary %>% dplyr::filter(Panel == "F1Liquid CDx")
      #Diseases = sort(names(table(Data_tmp3$Cancername)[table(Data_tmp3$Cancername)>=50]))
      setProgress(detail = "6551")
      data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        Hugo_Symbol %in% CH_gene_entropy_3 & 
                        Panel == "F1Liquid CDx" &
                        Smoking_history != "Unknown" &
                        Alcoholic_history != "Unknown" &
                        Double_cancer != "Unknown" &
                        Multiple_nodules != "Unknown"
                      ) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            unique_alt %in% CH_candidates &
              Hugo_Symbol %in% CH_gene_entropy_3 ~ 1,
            TRUE ~ 0)
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::mutate(Lines_pre_CGP = case_when(
          Lines_pre_CGP_2_or_more == 1 ~ ">=2",
          TRUE ~ "0~1"
        )) %>%
        dplyr::mutate(Years_from_1st_CTx = case_when(
          CTx_length >= 2 ~ ">=2",
          TRUE ~ "<2"
        )) %>%
        dplyr::mutate(VAF_10_or_more = case_when(
          VAF >= 0.1 ~ ">=10",
          TRUE ~ "<10"
        )) %>%
        dplyr::mutate(mutation_pattern = case_when(
          CpG_TpG == 1 ~ "CpG>TpG",
          CpC_TpC == 1 ~ "CpC>TpC",
          CtoT_mutation == "Yes" ~ "C>T, NOS",
          SNV_check == 1 ~ "SNV, NOS",
          TRUE ~ "Indel"
        )) %>%
        dplyr::select(Tumor_Sample_Barcode,
                      Age_decade, Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      #Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      Hugo_Symbol, amino.acid.change, VAF_10_or_more,
                      Years_from_1st_CTx, Lines_pre_CGP,
                      Hotspot_type,
                      mutation_pattern)
      setProgress(detail = "6605")
      data_tmp$Cancername = factor(data_tmp$Cancername)
      data_tmp$Cancername =  relevel(data_tmp$Cancername,
                                     ref=names(sort(table(data_tmp$Cancername),decreasing = T))[[1]]) 
      data_tmp$Hugo_Symbol = factor(data_tmp$Hugo_Symbol)
      data_tmp$Hugo_Symbol =  relevel(data_tmp$Hugo_Symbol,
                                     ref=names(sort(table(data_tmp$Hugo_Symbol),decreasing = T))[[1]]) 
      data_tmp_uni = data_tmp
      univ_tab <- data_tmp_uni %>%
        dplyr::select(-Tumor_Sample_Barcode, -amino.acid.change) %>%
        tbl_uvregression(
          method = glm,
          y = Hotspot_type,
          method.args = list(family = binomial),
          exponentiate = TRUE,
          vcov = ~ sandwich::vcovCL(., cluster = data_tmp_uni$Tumor_Sample_Barcode, data = data_tmp_uni),
          tidy_fun = broom.helpers::tidy_parameters
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      setProgress(detail = "6632")
      withProgress(message = "Logistic regression analysis", {
        dd <- datadist(data_tmp_uni)
        options(datadist = dd)
        my_formula <- as.formula(
          paste0("Hotspot_type ~ ", paste(colnames(data_tmp_uni)[!colnames(data_tmp_uni) %in% c("Tumor_Sample_Barcode", "amino.acid.change", "Hotspot_type")],collapse = " + "))
        )
        m1 <- lrm(my_formula, data = data_tmp_uni, x = TRUE, y = TRUE,penalty = Penal, tol=Toler)
        m1 <- robcov(m1, cluster = data_tmp_uni$Tumor_Sample_Barcode)
        bw_results <- fastbw(m1, rule = "p", sls = 0.05)
        final_formula <- as.formula(
          paste("Hotspot_type ~", paste(bw_results$names.kept, collapse = " + "))
        )
        final_mv_glm <- glm(final_formula, 
                            data = data_tmp_uni, 
                            family = binomial)
        vcov_cluster <- sandwich::vcovCL(final_mv_glm, 
                                         cluster = data_tmp_uni$Tumor_Sample_Barcode)
        final_mv_reg <- lrm(final_formula, 
                            data = data_tmp_uni, 
                            x = TRUE, y = TRUE,penalty = Penal, tol=Toler)
        final_mv_reg <- robcov(final_mv_reg, cluster = data_tmp_uni$Tumor_Sample_Barcode)
        
        significant_factors_nomo = bw_results$names.kept
      })
      setProgress(detail = "6659")
      output$table_inter_panel2 = render_gt({
        withProgress(message = "Rendering", {
          if(!is.null(significant_factors_nomo)){
            mv_tab = final_mv_glm |>
              tbl_regression(                         ## 多変量解析の表を生成
                exponentiate = TRUE,
                vcov = vcov_cluster,
                tidy_fun = broom.helpers::tidy_parameters,
                conf.level = 0.95,
                ci_method = "wald"
              ) |>
              add_global_p() |> # add global p-value
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
              modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels()
            tbl_merge(
              tbls = list(univ_tab, mv_tab),
              tab_spanner = c("**Univariate**", "**Multivariable**")) |>
              modify_caption("Factors, favor for CH hotspot") |> as_gt()
          } else {
            univ_tab |>
              modify_caption("Factors, favor for CH hotspot, no significant factor in multivariable analysis") |> as_gt()
          }
        })
      })
      
      setProgress(detail = "6689")
      if(!is.null(significant_factors_nomo)){
        mv_tab = final_mv_glm |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,
            vcov = vcov_cluster,
            tidy_fun = broom.helpers::tidy_parameters,
            conf.level = 0.95,
            ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        table_flextable <- tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors, favor for CH hotspot")
      } else {
        table_flextable <- univ_tab |>
          modify_caption("Factors, favor for CH hotspot, no significant factor in multivariable analysis")
      }
      as_hux_xlsx(table_flextable,file = "source/Table_odds_4.xlsx")
      setProgress(detail = "6715")
      
      logistic_data <- reactiveValues()
      if(!is.null(significant_factors_nomo)){
        log_nomogram <- nomogram(final_mv_reg, fun = plogis, lp=F, funlabel="Factors, favor for CH hotspot")
        options(datadist=NULL)
        # error in rms package
        # val <- try(rms::validate(m2, B=500), silent = FALSE)
        setProgress(detail = "6723")
        val <- try(rms::validate(final_mv_reg, B=500), silent = FALSE)
        if (class(val) == "try-error") {
          # val = rms::validate(m2, B=500)
          val = rms::validate(final_mv_reg, B=500)
        }
        # cal = rms::calibrate(m2, B=500)
        setProgress(detail = "6730")
        cal = rms::calibrate(final_mv_reg, B=500)
        # do not delete
        output$figure_inter_panel2 = renderPlot({
          par(mfcol = c(2, 1))
          plot(cal, lwd=2, lty=1,
               cex.lab=1.2, cex.axis=1, cex.main=1.2, cex.sub=0.6,
               xlim=c(0, 1), ylim= c(0, 1),
               xlab="Nomogram-Predicted Probability of LD-CH/LE-CH",
               ylab="Actual LD-CH/LE-CH rate",
               legend=FALSE)
  
          text(x=.4, y=.0, adj = c(0,0), paste0("LD-CH/LE-CH prediction\n",
                                                # "Model lokelihood ratio test P-value: ", format(m2$stats["P"][[1]], nsmall=3),'\n',
                                                "Model lokelihood ratio test P-value: ", format(final_mv_reg$stats["P"][[1]], nsmall=3),'\n',
                                                "500-time-bootstrapped-Brier score: ", round(val[9,5], digits=3),'\n',
                                                "500-time-bootstrapped-Nagelkerke R2 = ", round(val[2,5], digits=3),'\n',
                                                "500-time-bootstrapped-concordance index = ", round(((val[1,5] + 1) / 2), digits=3),"\n",
                                                "Intercept = ",round(val[3,5], digits=3),"\n",
                                                "Slope = ", round(val[4,5], digits=3)))
  
          abline(0, 1, lty=3, lwd=2,  col=c("#224444"))
          legend(x=.8, y=.2, legend=c("Apparent", "Bias-corrected", "Ideal"),
                 lty=c(3, 1, 2), bty="n")
          plot(log_nomogram)
        })
        logistic_data$log_nomogram = log_nomogram
        setProgress(detail = "6757")
        # predicted_probs <- predict(m2, type = "fitted")
        predicted_probs <- predict(final_mv_reg, type = "fitted")
        # 実際のラベルの取得
        # モデル m2 の y 成分から実際のラベルを取得
        # actual_labels <- m2$y
        actual_labels <- final_mv_reg$y
        # ROC 曲線の作成
        logistic_data$roc_obj <- roc(actual_labels, predicted_probs)
        logistic_data$auc_value <- auc(logistic_data$roc_obj)
      }
      setProgress(detail = "6768")
      
      output$figure_inter_panel2_2 = renderPlot({
        # ROC 曲線のプロット
        auc_val <- round(logistic_data$auc_value, 3)
        ggroc_obj = ggroc(logistic_data$roc_obj)
        ggroc_obj +
          geom_abline(intercept = 1, slope = 1, linetype = "dashed", color = "grey") +
          ggtitle("ROC Curve, Nomogram") +
          theme_minimal() +
          annotate("text", x = 0.75, y = 0.25, label = paste("AUC =", auc_val), size = 5)
      })
      output$text_inter_panel = renderText({
        if(!is.null(logistic_data$log_nomogram$Cancername$points)){
          text_point = paste0("Histology and points\n",
                 paste(paste(logistic_data$log_nomogram$Cancername$Cancername,
                             round(logistic_data$log_nomogram$Cancername$points,digits = 0),
                             sep="\t"), collapse = "\n"),
                 "\n")
        } else {
          text_point = ""
        }
        if(!is.null(logistic_data$log_nomogram$Hugo_Symbol$points)){
          text_point = paste0(text_point, "Gene name and points\n",
                              paste(paste(logistic_data$log_nomogram$Hugo_Symbol$Hugo_Symbol,
                                          round(logistic_data$log_nomogram$Hugo_Symbol$points,digits = 0),
                                          sep="\t"), collapse = "\n"))
        }
      })
 
      data_lgb = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        Hugo_Symbol %in% CH_gene_entropy_3 & 
                        Panel == "F1Liquid CDx" &
                        Smoking_history != "Unknown" &
                        Alcoholic_history != "Unknown" &
                        Double_cancer != "Unknown" &
                        Multiple_nodules != "Unknown"
        ) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            unique_alt %in% CH_candidates &
              Hugo_Symbol %in% CH_gene_entropy_3 ~ 1,
            TRUE ~ 0)
        ) %>%
        dplyr::mutate(mutation_pattern = case_when(
          CpG_TpG == 1 ~ "CpG>TpG",
          CpC_TpC == 1 ~ "CpC>TpC",
          CtoT_mutation == "Yes" ~ "C>T, NOS",
          SNV_check == 1 ~ "SNV, NOS",
          TRUE ~ "Indel"
        )) %>%
        dplyr::select(Tumor_Sample_Barcode, Age, Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      Hugo_Symbol, VAF,
                      CTx_length, Lines_pre_CGP,
                      Hotspot_type,
                      mutation_pattern)
      # --- 前処理 ---
      setProgress(detail = "6833")
      data_lgb$Cancername = factor(data_lgb$Cancername)
      data_lgb$Cancername = relevel(data_lgb$Cancername,
                                    ref = names(sort(table(data_lgb$Cancername), decreasing = TRUE))[[1]])
      data_lgb$Hugo_Symbol = factor(data_lgb$Hugo_Symbol)
      data_lgb$Hugo_Symbol = relevel(data_lgb$Hugo_Symbol,
                                     ref = names(sort(table(data_lgb$Hugo_Symbol), decreasing = TRUE))[[1]])
      names(data_lgb) <- gsub("[^A-Za-z0-9_]", "_", names(data_lgb))
      data_lgb$mutation_pattern <- gsub("[^A-Za-z0-9_]", "_", data_lgb$mutation_pattern)
      numeric_cols <- setdiff(names(data_lgb), 
                              c("Cancername", "Hugo_Symbol", "mutation_pattern", "Hotspot_type", "Tumor_Sample_Barcode"))
      data_processed = data_lgb

      # --- グループ付きCV設定 ---
      set.seed(123)
      folds_grouped <- rsample::group_vfold_cv(
        data = data_processed,
        group = Tumor_Sample_Barcode,
        v = 5
      )
      
      # --- LightGBM交差検証 ---
      preds_cv <- numeric(nrow(data_processed))
      importance_df <- NULL
      
      for (k in seq_along(folds_grouped$splits)) {
        split_obj <- folds_grouped$splits[[k]]
        
        train_data <- analysis(split_obj)
        valid_data <- assessment(split_obj)
        
        # 特徴量とラベル
        train_features <- data.matrix(train_data[, setdiff(names(train_data), c("Hotspot_type", "Tumor_Sample_Barcode")), with = FALSE])
        valid_features <- data.matrix(valid_data[, setdiff(names(valid_data), c("Hotspot_type", "Tumor_Sample_Barcode")), with = FALSE])
        train_label <- train_data$Hotspot_type
        
        # データセット作成
        dtrain_fold <- lgb.Dataset(data = train_features, label = train_label, params = list(min_data_in_bin = 1L))
        
        # モデル学習
        params <- list(
          objective = "binary",
          metric = c("auc", "binary_error"),
          num_leaves = 4L,
          min_data_in_leaf = 1L
        )
        bst <- lgb.train(
          params = params,
          data = dtrain_fold,
          nrounds = 1000,
          verbose = 0
        )
        
        # 重要度集計
        imp_tmp <- lightgbm::lgb.importance(bst)
        if (k == 1) {
          importance_df <- data.frame(Feature = imp_tmp$Feature, Importance = imp_tmp$Gain) %>%
            arrange(Feature)
        } else {
          imp_tmp <- imp_tmp %>% arrange(Feature)
          importance_df$Importance <- importance_df$Importance + imp_tmp$Gain
        }
        
        # 検証データで予測
        idx_valid <- which(data_lgb$Tumor_Sample_Barcode %in% valid_data$Tumor_Sample_Barcode)
        preds_cv[idx_valid] <- predict(bst, valid_features)
        
        print(k)
      }
      
      # --- ROCと可視化（元のまま） ---
      labels <- data_processed$Hotspot_type
      roc_obj <- roc(response = labels, predictor = preds_cv)
      ggroc_obj <- ggroc(roc_obj)
      auc_val <- round(auc(roc_obj), 3)
      g = ggroc_obj +
        geom_abline(intercept = 1, slope = 1, linetype = "dashed", color = "grey") +
        ggtitle("ROC Curve, lightGBM") +
        theme_minimal() +
        annotate("text", x = 0.75, y = 0.25, label = paste("AUC =", auc_val), size = 5)
      plots[[149]] = g
      setProgress(detail = "6914")
      
      importance_df$Importance <- importance_df$Importance / length(folds_grouped$splits)
      importance_df <- importance_df %>% arrange(desc(Importance))
      g = ggplot(importance_df, aes(x = reorder(Feature, Importance), y = Importance)) +
        geom_bar(stat = "identity", fill = "steelblue") +
        coord_flip() +
        xlab("Feature") +
        ylab("Importance (Gain)") +
        theme_minimal() +
        ggtitle("Feature Importance, lightGBM")
      plots[[150]] = g
      # Random forest model
      data <- data_lgb
      data$Hotspot_type = as.factor(data$Hotspot_type)
      levels(data$Hotspot_type) <- c("Class_0", "Class_1")
      target_variable <- "Hotspot_type"
      predictors <- setdiff(names(data), c(target_variable, "Tumor_Sample_Barcode"))
      # ---- group_vfold_cvを作成（Tumor_Sample_Barcode単位）----
      set.seed(123)
      folds <- group_vfold_cv(data, group = Tumor_Sample_Barcode, v = 5)
      
      # ---- 結果格納用 ----
      importance_df <- data.frame(Variable = predictors, Importance = 0)
      if (file.exists("source/results_rf_groupcv.rda")) {
        load(file = "source/results_rf_groupcv.rda")
      } else {
        cl <- makeCluster(5)
        registerDoParallel(cl)
        
        combine_func <- function(x, y) {
          list(predictions = c(x$predictions, y$predictions),
               labels = c(x$labels, y$labels),
               importance = c(x$importance, y$importance))
        }
        
        label_levels <- levels(data$Hotspot_type)
        init <- list(predictions = numeric(0),
                     labels = factor(levels = label_levels),
                     importance = numeric(0))
        
        # ---- 並列で各fold学習 ----
        results <- foreach(i = seq_len(length(folds$splits)), .combine = combine_func,
                           .init = init, .packages = c("randomForest", "dplyr", "rsample")) %dopar% {
                             
                             split <- folds$splits[[i]]
                             train_data <- analysis(split)
                             test_data <- assessment(split)
                             
                             # モデル学習
                             rf_model <- randomForest(
                               Hotspot_type ~ .,
                               data = train_data[, c(predictors, target_variable), drop = FALSE],
                               importance = TRUE,
                               ntree = 500,
                               mtry = 5
                             )
                             
                             # 検証
                             predictions <- predict(rf_model, newdata = test_data, type = "prob")[, 2]
                             labels <- test_data$Hotspot_type
                             importance <- importance(rf_model)[, 1]
                             
                             list(predictions = predictions, labels = labels, importance = importance)
                           }
        
        stopCluster(cl)
        save(results, file = "source/results_rf_groupcv.rda")
      }
      
      # ---- 結果まとめ ----
      imp_len <- length(results$importance) / 5
      importance_df_tmp <- vector("list", 5)
      for (i in 1:5) {
        importance_df_tmp[[i]] <- results$importance[(imp_len * (i - 1) + 1):(imp_len * i)]
      }
      
      importance_df$Importance <- Reduce(`+`, importance_df_tmp) / 5
      
      # ---- ROC ----
      predictions <- as.numeric(results$predictions)
      labels <- results$labels
      if (is.factor(labels)) labels <- as.numeric(labels) - 1
      
      roc_obj <- roc(labels, predictions)
      ggroc_obj <- ggroc(roc_obj)
      auc_val <- round(auc(roc_obj), 3)
      
      g1 <- ggroc_obj +
        geom_abline(intercept = 1, slope = 1, linetype = "dashed", color = "grey") +
        ggtitle("ROC Curve, random forest (group_vfold_cv)") +
        theme_minimal() +
        annotate("text", x = 0.75, y = 0.25, label = paste("AUC =", auc_val), size = 5)
      plots[[151]] <- g1
      setProgress(detail = "7008")
      
      # ---- Feature importance ----
      importance_df <- importance_df %>% arrange(desc(Importance))
      
      g2 <- ggplot(importance_df, aes(x = reorder(Variable, Importance), y = Importance)) +
        geom_bar(stat = "identity", fill = "steelblue") +
        coord_flip() +
        labs(title = "Feature Importance, random forest (group_vfold_cv)",
             x = "Feature", y = "Importance (MeanDecreaseAccuracy)") +
        theme_minimal()
      plots[[152]] <- g2
      # ---- データ準備 ----
      data <- data_lgb
      data$Hotspot_type <- as.factor(data$Hotspot_type)
      levels(data$Hotspot_type) <- c("Class_0", "Class_1")
      target_variable <- "Hotspot_type"
      predictors <- setdiff(names(data), c(target_variable, "Tumor_Sample_Barcode"))
      
      # ---- group_vfold_cvを作成 ----
      set.seed(123)
      folds <- group_vfold_cv(data, group = Tumor_Sample_Barcode, v = 5)
      
      # ---- 結果格納用 ----
      importance_df <- data.frame(Variable = predictors, Importance = 0)

      if (file.exists("source/results_logit_groupcv.rda")) {
        load(file = "source/results_logit_groupcv.rda")
      } else {
        cl <- makeCluster(5)
        registerDoParallel(cl)
        
        combine_func <- function(x, y) {
          list(predictions = c(x$predictions, y$predictions),
               labels = c(x$labels, y$labels),
               importance = cbind(x$importance, y$importance))
        }
        
        label_levels <- levels(data$Hotspot_type)
        init <- list(predictions = numeric(0),
                     labels = factor(levels = label_levels),
                     importance = matrix(0, nrow = length(predictors), ncol = 0,
                                         dimnames = list(predictors, NULL)))
        
        # ---- 各foldでロジスティック回帰 ----
        results <- foreach(i = seq_len(length(folds$splits)), .combine = combine_func,
                           .init = init, .packages = c("dplyr", "stats", "rsample")) %dopar% {
                             split <- folds$splits[[i]]
                             train_data <- analysis(split)
                             test_data <- assessment(split)
                             
                             # モデル学習
                             formula_str <- as.formula(paste("Hotspot_type ~", paste(predictors, collapse = " + ")))
                             glm_model <- glm(formula_str, data = train_data, family = binomial)
                             
                             # 検証データで予測（確率）
                             predictions <- predict(glm_model, newdata = test_data, type = "response")
                             labels <- test_data$Hotspot_type
                             
                             # 重要度（絶対係数値）
                             coef_vals <- abs(coef(glm_model)[-1])  # 切片を除く
                             names(coef_vals) <- predictors[predictors %in% names(coef_vals)]
                             importance_vec <- rep(0, length(predictors))
                             names(importance_vec) <- predictors
                             importance_vec[names(coef_vals)] <- coef_vals
                             
                             list(predictions = predictions, labels = labels, importance = importance_vec)
                           }
        
        stopCluster(cl)
        save(results, file = "source/results_logit_groupcv.rda")
      }
      setProgress(detail = "7050")
      
      # ---- 結果まとめ ----
      importance_mat <- results$importance
      if (is.matrix(importance_mat)) {
        importance_df$Importance <- rowMeans(importance_mat)
      } else {
        importance_df$Importance <- importance_mat
      }
      
      # ---- ROC ----
      predictions <- as.numeric(results$predictions)
      labels <- results$labels
      if (is.factor(labels)) labels <- as.numeric(labels) - 1
      
      roc_obj <- roc(labels, predictions)
      ggroc_obj <- ggroc(roc_obj)
      auc_val <- round(auc(roc_obj), 3)
      
      g1 <- ggroc_obj +
        geom_abline(intercept = 1, slope = 1, linetype = "dashed", color = "grey") +
        ggtitle("ROC Curve, Logistic Regression (group_vfold_cv)") +
        theme_minimal() +
        annotate("text", x = 0.75, y = 0.25, label = paste("AUC =", auc_val), size = 5)
      plots[[366]] <- g1
      setProgress(detail = "7105")
      
      # ---- Feature importance ----
      importance_df <- importance_df %>% arrange(desc(Importance))
      
      g2 <- ggplot(importance_df, aes(x = reorder(Variable, Importance), y = Importance)) +
        geom_bar(stat = "identity", fill = "steelblue") +
        coord_flip() +
        labs(title = "Feature Importance, Logistic Regression (group_vfold_cv)",
             x = "Feature", y = "|Coefficient|") +
        theme_minimal()
      plots[[367]] <- g2
      
      data_tmp_table = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        Panel == "F1Liquid CDx" &
                        Hugo_Symbol %in% CH_gene_entropy_3) %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            unique_alt %in% CH_candidates ~ "True hotspot variant in CH-genes",
            TRUE ~ "Filtered-out variants")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::mutate(Lines_pre_CGP_2_or_more = case_when(
          Lines_pre_CGP_2_or_more == 1 ~ ">=2",
          TRUE ~ "0~1"
        )) %>%
        dplyr::mutate(Years_from_1st_CTx = case_when(
          CTx_length >= 2 ~ ">=2",
          TRUE ~ "<2"
        )) %>%
        dplyr::mutate(VAF_10_or_more = case_when(
          VAF >= 0.1 ~ ">=10",
          TRUE ~ "<10"
        )) %>%
        dplyr::mutate(mutation_pattern = case_when(
          CpG_TpG == 1 ~ "CpG>TpG",
          CpC_TpC == 1 ~ "CpC>TpC",
          CtoT_mutation == "Yes" ~ "C>T, NOS",
          SNV_check == 1 ~ "SNV, NOS",
          TRUE ~ "Indel"
        )) %>%
        dplyr::select(Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      Hugo_Symbol, VAF_10_or_more,
                      CTx_length, Years_from_1st_CTx,
                      Lines_pre_CGP, Lines_pre_CGP_2_or_more,
                      Hotspot_type, TMB, Platinum,
                      mutation_pattern)
      output$table_inter_panel3 <- render_gt({
        data_tmp_table %>%
          tbl_summary(
            by = "Hotspot_type",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Hotspot variant, LD-CH/LE-CH or SEV** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_summary = data_tmp_table %>%
        tbl_summary(
          by = "Hotspot_type",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Hotspot variant, LD-CH/LE-CH or SEV** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_summary,file = "source/Table_odds_5.xlsx")

      
      ID_CH_patient = unique((Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx"))$Tumor_Sample_Barcode)
      data_tmp_table_patient = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            Tumor_Sample_Barcode %in% ID_CH_patient ~ "CH hotspot variant (+)",
            TRUE ~ "CH hotspot variant (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::mutate(Lines_pre_CGP_2_or_more = case_when(
          Lines_pre_CGP_2_or_more == 1 ~ ">=2",
          TRUE ~ "0~1"
        )) %>%
        dplyr::mutate(Years_from_1st_CTx = case_when(
          CTx_length >= 2 ~ ">=2",
          TRUE ~ "<2"
        )) %>%
        dplyr::select(Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CTx_length, Years_from_1st_CTx,
                      Lines_pre_CGP, Lines_pre_CGP_2_or_more,
                      Hotspot_type, TMB, Platinum)
      output$table_inter_panel4 <- render_gt({
        data_tmp_table_patient %>%
          tbl_summary(
            by = "Hotspot_type",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Patients with or without CH-related hotspot variants** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_summary = data_tmp_table_patient %>%
        tbl_summary(
          by = "Hotspot_type",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Patients with or without CH-related hotspot variants** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      
      as_hux_xlsx(table_summary,file = "source/Table_odds_6.xlsx")

      ID_CH_patient_detail = unique((Data_MAF_target %>%
                                dplyr::filter(non_germline_non_error == 1 &
                                                unique_alt %in% CH_candidates &
                                                logfold > 8 &
                                                Hugo_Symbol %in% CH_gene_entropy_3 &
                                                Panel == "F1Liquid CDx"))$Tumor_Sample_Barcode)
      data_tmp_table_patient_detail = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            Tumor_Sample_Barcode %in% ID_CH_patient_detail ~ "Liquid-only CH hotspot variant (+)",
            Tumor_Sample_Barcode %in% ID_CH_patient ~ "Tumor-observed CH hotspot variant (+)",
            TRUE ~ "CH hotspot variant (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::mutate(Lines_pre_CGP_2_or_more = case_when(
          Lines_pre_CGP_2_or_more == 1 ~ ">=2",
          TRUE ~ "0~1"
        )) %>%
        dplyr::mutate(Years_from_1st_CTx = case_when(
          CTx_length >= 2 ~ ">=2",
          TRUE ~ "<2"
        )) %>%
        dplyr::select(Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CTx_length, Years_from_1st_CTx,
                      Lines_pre_CGP, Lines_pre_CGP_2_or_more,
                      Hotspot_type, TMB, Platinum)
      output$table_inter_panel4_detail <- render_gt({
        data_tmp_table_patient_detail %>%
          tbl_summary(
            by = "Hotspot_type",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Patients with or without CH-related hotspot variants** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })      
      table_summary = data_tmp_table_patient_detail %>%
        tbl_summary(
          by = "Hotspot_type",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Patients with or without CH-related hotspot variants** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_summary,file = "source/Table_odds_7.xlsx")
      data_tmp_table_3 = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx") %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") ~ "DNMT3A/TET2/ASXL1/IDH2/EZH2/KMT2D",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") ~ "TP53/CHEK2/ATM/RAD21/MDM4",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") ~ "SF3B1/U2AF1",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") ~ "PTPN11/KRAS/CBL",
            TRUE ~ "Minor CH-gene hotspot")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::mutate(Lines_pre_CGP_2_or_more = case_when(
          Lines_pre_CGP_2_or_more == 1 ~ ">=2",
          TRUE ~ "0~1"
        )) %>%
        dplyr::mutate(Years_from_1st_CTx = case_when(
          CTx_length >= 2 ~ ">=2",
          TRUE ~ "<2"
        )) %>%
        dplyr::mutate(VAF_10_or_more = case_when(
          VAF >= 0.1 ~ ">=10",
          TRUE ~ "<10"
        )) %>%
        dplyr::mutate(mutation_pattern = case_when(
          CpG_TpG == 1 ~ "CpG>TpG",
          CpC_TpC == 1 ~ "CpC>TpC",
          CtoT_mutation == "Yes" ~ "C>T, NOS",
          SNV_check == 1 ~ "SNV, NOS",
          TRUE ~ "Indel"
        )) %>%
        dplyr::select(Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      Hugo_Symbol, VAF_10_or_more,
                      CTx_length, Years_from_1st_CTx,
                      Lines_pre_CGP, Lines_pre_CGP_2_or_more,
                      Hotspot_type, TMB, Platinum,
                      mutation_pattern)
      output$table_inter_panel5 <- render_gt({
        data_tmp_table_3 %>%
          tbl_summary(
            by = "Hotspot_type",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Hotspot variant, Epigenetic/DDR/Splicing or others** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_summary = data_tmp_table_3 %>%
        tbl_summary(
          by = "Hotspot_type",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Hotspot variant, Epigenetic/DDR/Splicing or others** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_summary,file = "source/Table_odds_8.xlsx")
      setProgress(detail = "7411")
      Data_summary_F1L= Data_summary %>% filter(Panel == "F1Liquid CDx")
      Data_summary_F1S = Data_summary %>% filter(Panel == "FoundationOne CDx")
      total_patients_L <- length(unique(Data_summary_F1L$Tumor_Sample_Barcode))
      total_patients_S <- length(unique(Data_summary_F1S$Tumor_Sample_Barcode))
      Data_MAF_DNMT3A_L = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      Data_MAF_DNMT3A_S = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "FoundationOne CDx" &
                        pnt_col %in% c("A"))
      Data_MAF_DNMT3A_L$color_group = ifelse(Data_MAF_DNMT3A_L$Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2"), "DNA_methyl",
                                             ifelse(Data_MAF_DNMT3A_L$Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D"), "Histone",
                                                ifelse(Data_MAF_DNMT3A_L$Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4"), "DNA_repair",
                                                    ifelse(Data_MAF_DNMT3A_L$Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL"), "MAPK",
                                                           ifelse(Data_MAF_DNMT3A_L$Hugo_Symbol %in% c("SF3B1", "U2AF1"), "Splicing","Other_gene")))))
      Data_MAF_DNMT3A_S$color_group = ifelse(Data_MAF_DNMT3A_S$Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2"), "DNA_methyl",
                                             ifelse(Data_MAF_DNMT3A_S$Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D"), "Histone",
                                                ifelse(Data_MAF_DNMT3A_S$Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4"), "DNA_repair",
                                                    ifelse(Data_MAF_DNMT3A_S$Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL"), "MAPK",
                                                           ifelse(Data_MAF_DNMT3A_S$Hugo_Symbol %in% c("SF3B1", "U2AF1"), "Splicing","Other_gene")))))
      # 2. LOCH+TOCH の患者数（全体）
      df_all <- Data_MAF_DNMT3A_L %>%
        dplyr::group_by(Hugo_Symbol) %>%
        dplyr::summarise(
          total_count = n_distinct(Tumor_Sample_Barcode),
          color_group = first(color_group)
        )
      # 3. TOCH だけを抽出（logfold ≤ 8 をTOCHと仮定、必要に応じて条件修正）
      df_toch <- Data_MAF_DNMT3A_L %>%
        dplyr::filter(logfold <= 8) %>%
        dplyr::group_by(Hugo_Symbol) %>%
        dplyr::summarise(
          toch_count = n_distinct(Tumor_Sample_Barcode)
        )
      # 4. 結合し、TOCH割合を計算
      mutation_ratio <- dplyr::left_join(df_all, df_toch, by = "Hugo_Symbol") %>%
        dplyr::mutate(
          toch_count = ifelse(is.na(toch_count), 0, toch_count),
          TOCH_ratio = toch_count / total_count * 100
        )
      # 5. プロット作成
      setProgress(detail = "7460")
      g <- ggplot(mutation_ratio, aes(x = reorder(Hugo_Symbol, -TOCH_ratio), y = TOCH_ratio, fill = color_group)) +
        geom_bar(stat = "identity") +
        geom_text(aes(label = paste0(round(TOCH_ratio, 1), "%")), vjust = -0.3, size = 3.5) +
        labs(
          title = "Proportion of TOCH variants among LOCH+TOCH",
          x = "Gene",
          y = "TOCH / (LOCH + TOCH) (%)"
        ) +
        theme_minimal() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        ylim(0, max(mutation_ratio$TOCH_ratio, na.rm = TRUE) * 1.1)
      # 6. 保存
      plots[[251]] <- g
      # 各遺伝子の出現症例数と割合（F1L）
      freq_L <- Data_MAF_DNMT3A_L %>%
        group_by(Hugo_Symbol) %>%
        summarise(count_L = n_distinct(Tumor_Sample_Barcode)) %>%
        mutate(freq_L = count_L / total_patients_L)
      # 各遺伝子の出現症例数と割合（F1S）
      freq_S <- Data_MAF_DNMT3A_S %>%
        group_by(Hugo_Symbol) %>%
        summarise(count_S = n_distinct(Tumor_Sample_Barcode)) %>%
        mutate(freq_S = count_S / total_patients_S)
      # F1L / F1S の比率を計算
      freq_compare <- inner_join(freq_L, freq_S, by = "Hugo_Symbol") %>%
        mutate(ratio = freq_L / freq_S)
      # 不要なNAやInfを除外
      freq_compare <- freq_compare %>%
        filter(is.finite(ratio), !is.na(ratio))
      freq_compare$color_group = ifelse(freq_compare$Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2"), "DNA_methyl",
                                        ifelse(freq_compare$Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D"), "Histone",
                                            ifelse(freq_compare$Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4"), "DNA_repair",
                                               ifelse(freq_compare$Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL"), "MAPK",
                                                      ifelse(freq_compare$Hugo_Symbol %in% c("SF3B1", "U2AF1"), "Splicing","Other_gene")))))
      # 棒グラフ作成
      g <- ggplot(freq_compare, aes(x = reorder(Hugo_Symbol, -ratio), y = ratio, fill = color_group)) +
        geom_bar(stat = "identity") +
        geom_text(aes(label = round(ratio, 2)), vjust = -0.3, size = 3.5) +
        labs(
          title = "Frequency Ratio (F1Liquid CDx / FoundationOne CDx)",
          x = "Gene",
          y = "F1L / F1S Ratio"
        ) +
        theme_minimal() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        ylim(0, max(freq_compare$ratio) * 1.1)
      # 結果表示または保存
      plots[[252]] <- g
      setProgress(detail = "7509")
      mutation_counts <- Data_MAF_DNMT3A_S %>%
        dplyr::group_by(Hugo_Symbol) %>%
        dplyr::summarise(
          patient_count = n_distinct(Tumor_Sample_Barcode),
          color_group = first(color_group)
        ) %>%
        dplyr::mutate(Percent = patient_count / total_patients_S * 100)
      mutation_counts$color_group = ifelse(mutation_counts$Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2"), "DNA_methyl",
                                        ifelse(mutation_counts$Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D"), "Histone",
                                               ifelse(mutation_counts$Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4"), "DNA_repair",
                                                      ifelse(mutation_counts$Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL"), "MAPK",
                                                             ifelse(mutation_counts$Hugo_Symbol %in% c("SF3B1", "U2AF1"), "Splicing","Other_gene")))))
      g <- ggplot(mutation_counts, aes(x = reorder(Hugo_Symbol, -Percent), y = Percent, fill = color_group)) +
        geom_bar(stat = "identity") +
        geom_text(aes(label = patient_count), vjust = -0.3, size = 3.5) +  # 上に症例数を表示
        labs(title = paste0("Number of patients with TI-CH variants, ",total_patients_S, " F1 CDx patients"),
             x = "Gene",
             y = "Variant rate (%)") +
        theme_minimal() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        ylim(0, max(mutation_counts$Percent) * 1.1)  # テキストスペース確保
      plots[[253]] = g
      Infiltrative_genes = (freq_compare %>% filter(ratio < 10))$Hugo_Symbol
      print(1)
      setProgress(detail = "7534")
      Target_drugs = c(
        #"Nivolumab,Ipilimumab",
        #"Ipilimumab,Nivolumab",
        "Nivolumab",
        #"Ipilimumab",
        #"Atezolizumab",
        #"Avelumab",
        "Pembrolizumab"
      )
      # Target_drugs = c(
      #   "Paclitaxel"
      # )
      Data_case_target = Data_case()
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor, final_observe) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      setProgress(detail = "7567")
      Data_ICI_pre = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      final_observe,
                      censor) %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in% c("緩和","その他") &
                        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% Target_drugs) %>%
        dplyr::select(-症例.EP前レジメン情報.実施目的.名称.) %>%
        dplyr::distinct()
      colnames(Data_ICI_pre) = c("Tumor_Sample_Barcode", "drug", "RECIST", "CTx_line", "entry_date", "CTx_start_date", "CTx_end_date", "ongoing", "final_observe", "censor")
      Data_ICI_pre = Data_ICI_pre %>%
        dplyr::filter(CTx_start_date != "") %>%
        dplyr::mutate(
          CTx_time_on_treatment = case_when(
            CTx_end_date != "" ~ as.integer(as.Date(CTx_end_date) - as.Date(CTx_start_date)),
            TRUE ~ as.integer(as.Date(entry_date) - as.Date(CTx_start_date))),
          CTx_censor  = case_when(
            !is.na(ongoing) ~ 0,
            TRUE ~ 1)
        ) %>%
        dplyr::select(-entry_date, -ongoing)
      Data_ICI_post = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      final_observe,
                      censor) %>%
        dplyr::filter(症例.EP後レジメン情報.薬剤名.YJ一般名.EN. %in% Target_drugs) %>%
        dplyr::distinct()
      colnames(Data_ICI_post) = c("Tumor_Sample_Barcode", "drug", "RECIST", "CTx_line", "CTx_start_date", "CTx_end_date", "final_observe", "censor")
      Data_ICI_post = Data_ICI_post %>%
        dplyr::filter(CTx_start_date != "" & CTx_end_date != "") %>%
        dplyr::mutate(CTx_time_on_treatment = as.integer(as.Date(CTx_end_date) - as.Date(CTx_start_date)))
      Data_ICI_post$CTx_censor = 1
      Data_ICI = rbind(Data_ICI_pre, Data_ICI_post)
      Data_ICI = Data_ICI %>%
        dplyr::filter(final_observe != "") %>%
        dplyr::mutate(Overall_survival = as.integer(as.Date(final_observe) - as.Date(CTx_start_date)))
      Data_ICI$Overall_survival[Data_ICI$Overall_survival < 0] <- NA
      Data_ICI$RECIST[is.na(Data_ICI$RECIST)] = "NE"
      Data_ICI$RECIST = factor(Data_ICI$RECIST, levels = c("CR","PR","SD","PD","NE"))
      Data_ICI = Data_ICI %>%
        dplyr::mutate(CTx_Effect = case_when(
          RECIST %in% c("CR","PR") ~ 1,
          RECIST %in% c("NE") ~ NA_integer_,
          TRUE ~ 0
        ))
      setProgress(detail = "7601")
      Data_ICI = Data_ICI %>%
        dplyr::mutate(CTx_line = case_when(
          CTx_line == "１次治療" ~ "1",
          CTx_line == "２次治療" ~ "2",
          CTx_line == "３次治療" ~ "3",
          CTx_line == "４次治療" ~ "4",
          CTx_line == "５次治療以降" ~ "5~",
          CTx_line == "不明" ~ "Unknown",
          TRUE ~ "Unknown",
        )) %>%
        dplyr::filter(CTx_line != "Unknown")
      Data_ICI = Data_ICI %>%
        dplyr::filter(CTx_start_date != "") %>%
        dplyr::arrange(CTx_start_date) %>%
        dplyr::filter(CTx_time_on_treatment>=0) %>%
        dplyr::distinct(Tumor_Sample_Barcode,.keep_all=T) %>%
        dplyr::select(Tumor_Sample_Barcode, CTx_time_on_treatment, CTx_censor, CTx_start_date, CTx_end_date, RECIST, CTx_Effect, CTx_line, Overall_survival)
      Data_summary_drug_effect = left_join(Data_summary, Data_ICI, by="Tumor_Sample_Barcode")
      # data.table に変換
      summary_dt <- as.data.table(Data_summary_drug_effect)
      case_dt <- as.data.table(Data_case())
      # 前後のレジメン情報を整形して結合
      pre_dt <- case_dt[, .(Tumor_Sample_Barcode = `C.CAT調査結果.基本項目.ハッシュID`,
                            CTx_start_date = `症例.EP前レジメン情報.投与開始日`,
                            drug = `症例.EP前レジメン情報.薬剤名.YJ一般名.EN.`)]
      post_dt <- case_dt[, .(Tumor_Sample_Barcode = `C.CAT調査結果.基本項目.ハッシュID`,
                             CTx_start_date = `症例.EP後レジメン情報.投与開始日`,
                             drug = `症例.EP後レジメン情報.薬剤名.YJ一般名.EN.`)]
      # 前後の情報を結合
      drug_dt <- rbind(pre_dt, post_dt)
      drug_dt <- drug_dt[!is.na(CTx_start_date) & drug != ""]
      # 同じ患者・投与日ごとに薬剤名をまとめる（重複除く）
      drug_agg <- drug_dt[, .(regimen = paste(unique(drug), collapse = "/")),
                          by = .(Tumor_Sample_Barcode, CTx_start_date)]
      # summary_dt に結合
      summary_dt <- merge(summary_dt, drug_agg,
                          by = c("Tumor_Sample_Barcode", "CTx_start_date"),
                          all.x = TRUE)
      # regimen 列がない行は空文字列に
      summary_dt[is.na(regimen), regimen := ""]
      summary_dt$regimen = str_replace(summary_dt$regimen, "Nivolumab,Ipilimumab", "Ipilimumab,Nivolumab")
      setProgress(detail = "7667")
      #summary_dt$Overall_survival = summary_dt$time_palliative_final
      # 結果を元の data.frame に戻すなら
      Data_summary_drug_effect <- as.data.frame(summary_dt)
      Data_summary_drug_effect = Data_summary_drug_effect %>% dplyr::filter(regimen %in% Target_drugs)
      Data_MAF_target_ICI = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel %in% c("F1Liquid CDx", "FoundationOne CDx")) %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") & logfold <= 8 ~ "Epigenetic_TOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold <= 8 ~ "DDR_TOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold <= 8 ~ "Splicing_TOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold <= 8 ~ "MAPK_TOCH",
            Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") & logfold > 8 ~ "Epigenetic_LOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold > 8 ~ "DDR_LOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold > 8 ~ "Splicing_LOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold > 8 ~ "MAPK_LOCH",
            logfold > 8 ~ "Minor CH_LOCH",
            TRUE ~ "Minor CH_TOCH"),
          Hotspot_type_2 = case_when(
            Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2") & logfold <= 8 ~ "DNA_methyl_TOCH",
            Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D") & logfold <= 8 ~ "Histone_TOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold <= 8 ~ "DDR_TOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold <= 8 ~ "Splicing_TOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold <= 8 ~ "MAPK_TOCH",
            Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2") & logfold > 8 ~ "DNA_methyl_LOCH",
            Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D") & logfold > 8 ~ "Histone_LOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold > 8 ~ "DDR_LOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold > 8 ~ "Splicing_LOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold > 8 ~ "MAPK_LOCH",
            logfold > 8 ~ "Minor CH_LOCH",
            TRUE ~ "Minor CH_TOCH"),
          VAF_class = case_when(
            VAF < 0.01 ~ "VAF<0.01",
            VAF < 0.02 ~ "VAF<0.02",
            VAF < 0.05 ~ "VAF<0.05",
            VAF < 0.10 ~ "VAF<0.10",
            VAF < 0.20 ~ "VAF<0.20",
            VAF < 0.40 ~ "VAF<0.40",
            TRUE ~ "VAF>=0.40"
          ),
          Infiltrative_genes = case_when(
            Hugo_Symbol %in% Infiltrative_genes ~ 1,
            TRUE ~ 0
          ),
          Important_spot = case_when(
            str_detect(unique_alt, "DNMT3A_") ~ "DNMT3A",
            str_detect(unique_alt, "CHEK2_") ~ "CHEK2",
            str_detect(unique_alt, "ATM_") ~ "ATM",
            str_detect(unique_alt, "ASXL1_") ~ "ASXL1",
            str_detect(unique_alt, "TET2_") ~ "TET2",
            str_detect(unique_alt, "TP53_") ~ "TP53",
            TRUE ~ "OTHER"
          )
        )
      setProgress(detail = "7718")
      add_hotspot_count <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Hotspot_type %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      add_hotspot_count2 <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Hotspot_type_2 %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      add_hotspot_count3 <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Important_spot %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      setProgress(detail = "7750")
      Data_summary_drug_effect <- Data_summary_drug_effect %>%
        add_hotspot_count(Data_MAF_target_ICI, unique(Data_MAF_target_ICI$Hotspot_type), "CH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("Epigenetic_TOCH", "DDR_TOCH", "MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH"), "TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("Epigenetic_LOCH", "DDR_LOCH", "MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH"), "LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "DDR_TOCH", "DDR_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "Epigenetic_TOCH", "Epigenetic_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH"), "Other_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "DDR_LOCH", "DDR_LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "Epigenetic_LOCH", "Epigenetic_LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH"), "Other_LOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "DNA_methyl_TOCH", "DNA_methyl_TOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "DNA_methyl_LOCH", "DNA_methyl_LOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "Histone_TOCH", "Histone_TOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "Histone_LOCH", "Histone_LOCH_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "DNMT3A", "DNMT3A_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "CHEK2", "CHEK2_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "ASXL1", "ASXL1_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "ATM", "ATM_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "TET2", "TET2_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "TP53", "TP53_count")
      mut_ID = (Data_MAF_target %>%
                  dplyr::filter(non_germline_non_error %in% c(0,1,6) &
                                !unique_alt %in% CH_candidates &
                                Hugo_Symbol %in% "TP53" &
                                Panel %in% c("F1Liquid CDx", "FoundationOne CDx")))$Tumor_Sample_Barcode
      # Data_ICI_ToT = Data_summary_drug_effect %>%
      #           dplyr::filter(Panel == "F1Liquid CDx") %>%
      #           #dplyr::filter(Panel == "FoundationOne CDx") %>%
      #           dplyr::mutate(CH_plus = case_when(
      #             Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(VAF_class %in%
      #                                                                                c("VAF>=0.40", "VAF<0.40", "VAF<0.20", "VAF<0.10", "VAF<0.05", "VAF<0.02", "VAF<0.01") &
      #                                                                                #c("VAF>=0.40", "VAF<0.40", "VAF<0.20", "VAF<0.10", "VAF<0.05", "VAF<0.02") &
      #                                                                                #c("VAF>=0.40", "VAF<0.40", "VAF<0.20") &
      #                                                                                #c("VAF<0.10", "VAF<0.05", "VAF<0.02", "VAF<0.01") &
      #                                                                                #c("VAF<0.01") &
      #                                                                                Hotspot_type %in%
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH", "DDR_TOCH", "DDR_LOCH", "MAPK_TOCH", "MAPK_LOCH", "Minor CH_TOCH", "Minor CH_LOCH", "Splicing_TOCH", "Splicing_LOCH")
      #                                                                                c("MAPK_TOCH", "MAPK_LOCH", "Minor CH_TOCH", "Minor CH_LOCH", "Splicing_TOCH", "Splicing_LOCH")
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH", "DDR_TOCH", "DDR_LOCH")
      #                                                                                #c("DDR_TOCH", "DDR_LOCH")
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH")
      #                                                                                #c("MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH")
      #                                                                                #c("DDR_LOCH")
      #                                                                                #c("Epigenetic_LOCH")
      #                                                                                #c("MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH")
      #                                                                                #c("DDR_TOCH")
      #                                                                                #c("Epigenetic_TOCH")
      #             ))$Tumor_Sample_Barcode ~ 1,
      #             TRUE ~ 0
      #           ))
      # Data_ICI_ToT = Data_summary_drug_effect %>%
      #   #dplyr::filter(Panel == "F1Liquid CDx") %>%
      #   dplyr::filter(Panel == "FoundationOne CDx") %>%
      #   dplyr::mutate(CH_plus = case_when(
      #     Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(Hotspot_type %in%
      #                                                                        # c("Epigenetic_TOCH", "DDR_TOCH", "Splicing_TOCH", "MAPK_TOCH", "Minor CH_TOCH",
      #                                                                        #   "Epigenetic_LOCH", "DDR_LOCH", "Splicing_LOCH", "MAPK_LOCH", "Minor CH_LOCH")
      #                                                                        "DDR_LOCH"
      #     ))$Tumor_Sample_Barcode ~ 1,
      #     TRUE ~ 0
      #   ))
      Data_summary_drug_effect = Data_summary_drug_effect %>%
        dplyr::mutate(
          TMB_class = case_when(
            TMB < 10 ~ "TMB<10",
            TMB < 14 ~ "10≤TMB<14",
            TMB < 20 ~ "14≤TMB<20",
            TRUE ~ "20<=TMB"
          )) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        ))
      Data_summary_drug_effect$TMB_class = factor(Data_summary_drug_effect$TMB_class,
                                                  levels = c("TMB<10", "10≤TMB<14", "14≤TMB<20", "20<=TMB"))
      Data_summary_drug_effect$Age_decade = factor(Data_summary_drug_effect$Age_decade,
                                                  levels = c("~49", "50~59", "60~69", "70~"))
      Data_summary_drug_effect$regimen = factor(Data_summary_drug_effect$regimen)
      Data_summary_drug_effect$regimen = relevel(Data_summary_drug_effect$regimen,
                                                 ref = names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])
      setProgress(detail = "7824")
      Data_summary_drug_effect = Data_summary_drug_effect %>%
        dplyr::mutate(Infiltraive_genes = case_when(
          Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(Infiltrative_genes == 1))$Tumor_Sample_Barcode ~ 1,
          TRUE ~ 0
        ))
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "SOFT_TISSUE", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
        # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
         filter(term != "(Intercept)") %>%
         mutate(term = case_when(
           grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
           grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
           grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
           grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
           grepl("^TMB_class", term) ~ gsub("TMB_class", "TMB: ", term),
           grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
           grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
           TRUE ~ term
         ))
      tidy_ORR <- tidy_ORR %>%
         mutate(term = factor(term, levels = tidy_ORR$term))
       # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      setProgress(detail = "7905")
      plots[[168]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
         geom_point(aes(color = p.value < 0.05), size = 3) +
         scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
         geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
         geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
         geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
         scale_y_log10() +
         coord_flip() +
         labs(
           title = "Odds Ratios for objective response, F1 CDx, any TI-CH, Nivo/Pembro",
           x = "",
           y = "Odds Ratio (95% CI)"
         ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[169]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-solid, any TICH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_class", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[170]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, any TICH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      #Data_ICI_ToT = Data_ICI_ToT %>% filter(TMB<10 & Age >49)# %>% 
        #dplyr::select(CTx_time_on_treatment, CTx_censor, CH_plus, CTx_line, Cancername, TMB_class, Age_decade, Sex, regimen)
      setProgress(detail = "7968")
      save(Data_ICI_ToT, file = "source/Data_ICI_ToT.rda")
      Data_ICI_ToT$Age_decade = factor(Data_ICI_ToT$Age_decade,
                                                   levels = c("70~", "60~69", "50~59", "~49"))
      Data_ICI_ToT$TMB_class = droplevels(Data_ICI_ToT$TMB_class)
      Data_ICI_ToT$Age_decade = droplevels(Data_ICI_ToT$Age_decade)
      Data_ICI_ToT$Age_decade = relevel(Data_ICI_ToT$Age_decade, ref = "70~")
      Data_ICI_ToT$CH_plus = factor(Data_ICI_ToT$CH_plus)
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + TMB_class + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      setProgress(detail = "7986") 
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        TMB_class = "<10",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.2", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "TMB_class"] = "TMB"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TI-CH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      setProgress(detail = "8078")
      ok <- is.na(df_plot$HR) | (
        df_plot$HR > 0 & df_plot$Lower > 0 & df_plot$Upper > 0 &
          df_plot$Lower <= df_plot$Upper
      )
      
      df_plot$HR[!ok] <- NA
      df_plot$Lower[!ok] <- NA
      df_plot$Upper[!ok] <- NA
      df_plot$HR_CI[!ok] <- ""
      df_plot$P_value_str[!ok] <- ""
      logf <- "source/app_debug.log"  # 書ける場所に変更可
      
      log_msg <- function(...) {
        msg <- paste0(format(Sys.time(), "%F %T"), " | ", paste(..., collapse=" "))
        cat(msg, "\n", file = logf, append = TRUE)
      }
      
      safe_step <- function(step, expr) {
        setProgress(detail = step)
        log_msg("STEP", step)
        tryCatch(expr,
                 error = function(e) {
                   log_msg("ERROR at", step, ":", conditionMessage(e))
                   showNotification(paste("ERROR at", step, ":", conditionMessage(e)),
                                    type = "error", duration = NULL)
                   stop(e)  # ここで止める
                 })
      }
      setProgress(detail = "8110") 
      safe_step("8078_before_forestplot", {
        forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1 CDx, time on immune checkpoint inhibitor trearmented",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" )
      })
      safe_step("8079_after_forestplot", {
        forest_plot <- forest_plot %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      })
      setProgress(detail = "8165") 
      safe_step("8166_grob_plot", {
        grob_plot <- grid.grabExpr(print(forest_plot))
      })
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[357]] <- ggdraw(ggplot_plot)
      setProgress(detail = "8093")
      
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      #subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        TMB_class = "<10",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        #df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            #df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      #df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TICH detection"
      df_all["CH_plus.2", "Subgroup"] = "TICH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "TMB_class"] = "TMB"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TICH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        #P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      setProgress(detail = "8192")
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1-solid, time on immune checkpoint inhibitor trearment",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black",
                               graphwidth = unit(6, "cm")) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)"#,
          # "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 4) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[358]] <- ggdraw(ggplot_plot)
      
      Data_ICI_ToT_save = Data_ICI_ToT
      Data_ICI_ToT = Data_ICI_ToT %>% dplyr::filter(TMB_class == "TMB<10")
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      setProgress(detail = "8257")
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.2", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TI-CH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1 CDx, TMB-low, time on immune checkpoint inhibitor trearmented",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" ) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      setProgress(detail = "8373")
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[371]] <- ggdraw(ggplot_plot)

      Data_ICI_ToT = Data_ICI_ToT_save %>% dplyr::filter(TMB_class != "TMB<10")
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.2", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      setProgress(detail = "8460")
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TI-CH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1 CDx, TMB-high, time on immune checkpoint inhibitor trearmented",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" ) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[372]] <- ggdraw(ggplot_plot)

      Data_ICI_ToT = Data_ICI_ToT_save
      data_tmp_table_TICH = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "Any TICH (+)",
            TRUE ~ "TICH (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_TICH <- render_gt({
        data_tmp_table_TICH %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and tumor-infiltrating CH, F1-solid** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      setProgress(detail = "8561")
      table_summary = data_tmp_table_TICH %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and tumor-infiltrating CH, F1-solid** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_summary,file = "source/Table_odds_9.xlsx")
      data_tmp_table_regression = data_tmp_table_TICH %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
          stats::step(direction = "backward", trace = FALSE)
          significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
          significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
          data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
          m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
          mv_tab = m2 |>
            tbl_regression(                         ## 多変量解析の表を生成
              exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
            ) |>
            add_global_p() |> # add global p-value
            add_q() |> # adjusts global p-values for multiple testing
            bold_p() |> # bold p-values under a given threshold (default 0.05)
            modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            bold_labels()
          tbl_TICH_ = tbl_merge(
            tbls = list(univ_tab, mv_tab),
            tab_spanner = c("**Univariate**", "**Multivariable**")) |>
            modify_caption("Factors for Nivo/Pembro objective response, F1-solid, any TICH")
          tbl_TICH = tbl_TICH_ |> as_gt()
      } else {
        tbl_TICH_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_TICH = tbl_TICH_ |> as_gt()
      }
      as_hux_xlsx(tbl_TICH_,file = "source/Table_odds_10.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
          diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                            data=Data_ICI_ToT, rho=0)
          diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                            data=Data_ICI_ToT, rho=1)
          diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                         data=Data_ICI_ToT)
          plots[[322]] = 
            surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-solid, any TICH (+)"), diff_0, diff_1, diff_2)
          diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
          # 2. HR・CI を整形
          tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
            filter(term != "(Intercept)") %>%
            mutate(term = case_when(
              grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
              grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
              grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
              grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
              grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
              grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
              grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
              TRUE ~ term
            )) %>%
            mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
          # データ結合・整形
          tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
            distinct(term, .keep_all = TRUE) %>%
            mutate(
              term = factor(term, levels = rev(sort(unique(term)))),
              label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
            )
          # プロット
          label_offset <- 0.4
          plots[[323]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
            geom_point(aes(color = p.value < 0.05), size = 3) +
            scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
            geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
            geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
            geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
            scale_y_log10() +
            coord_flip() +
            labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
                 subtitle = "F1-solid, any TICH, Nivo/Pembro") +
            theme_minimal() +
            theme(legend.position = "none")
          setProgress(detail = "8653")
          
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count >= 1 & DDR_LOCH_count == 0 ~ "1~",
          #TOCH_count >= 1 & LOCH_count == 0 ~ "1~",
          #LOCH_count >= 1 & TOCH_count == 0 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          TOCH_count >= 1 ~ "1~",
          #TOCH_count == 1 ~ "1",
          #LOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[254]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-liquid, any CH, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[255]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-Liquid CDx, LD-CH/LE-CH"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[256]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, any CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_TOCH = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "Any CH (+)",
            TRUE ~ "CH (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_TOCH <- render_gt({
        data_tmp_table_TOCH %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and any CH, F1-liquid** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_TOCH %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and any CH, F1-liquid** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_11.xlsx")

      data_tmp_table_regression = data_tmp_table_TOCH %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_TOCH_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1-liquid, any TICH")
        tbl_TOCH = tbl_TOCH_ |> as_gt()
      } else {
        tbl_TOCH_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_TOCH = tbl_TOCH_ |> as_gt()
      }
      as_hux_xlsx(tbl_TOCH_,file = "source/Table_odds_12.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[324]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-liquid, any CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[325]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, any CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT$Age_decade = factor(Data_ICI_ToT$Age_decade,
                                       levels = c("70~", "60~69", "50~59", "~49"))
      Data_ICI_ToT$TMB_class = droplevels(Data_ICI_ToT$TMB_class)
      Data_ICI_ToT$Age_decade = droplevels(Data_ICI_ToT$Age_decade)
      Data_ICI_ToT$Age_decade = relevel(Data_ICI_ToT$Age_decade, ref = "70~")
      Data_ICI_ToT$CH_plus = factor(Data_ICI_ToT$CH_plus)
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + TMB_class + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        TMB_class = "<10",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "LD-CH/LE-CH detection"
      df_all["CH_plus.2", "Subgroup"] = "LD-CH/LE-CH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "TMB_class"] = "TMB"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "LD-CH/LE-CH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1-Liquid CDx, time on immune checkpoint inhibitor trearment",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" ) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[360]] <- ggdraw(ggplot_plot)
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(TOCH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count >= 1 & DDR_LOCH_count == 0 ~ "1~",
          TOCH_count >= 2 ~ "2~",
          TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #LOCH_count >= 1 & TOCH_count == 0 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #LOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",,
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ),
        LOCH_plus = case_when(
            #DDR_TOCH_count >= 2 ~ "2~",
            #DDR_TOCH_count == 1 ~ "1",
            #DDR_TOCH_count >= 1 ~ "1~",
            #DDR_LOCH_count >= 1 ~ "1~",
            #Epigenetic_TOCH_count >= 1 ~ "1~",
            #Other_TOCH_count >= 1 ~ "1~",
            #DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
            #DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
            #DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
            #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
            #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
            #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
            #DDR_LOCH_count >= 1 ~ "1~",
            #DDR_TOCH_count >= 1 ~ "1~",
            #DDR_TOCH_count >= 1 & DDR_LOCH_count == 0 ~ "1~",
            LOCH_count >= 1 ~ "1~",
            #LOCH_count >= 1 & TOCH_count == 0 ~ "1~",
            #DNA_methyl_TOCH_count >= 1 ~ "1~",
            #Histone_TOCH_count >= 1 ~ "1~",
            #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
            #TOCH_count >= 2 ~ "2~",
            #TOCH_count == 1 ~ "1",
            #LOCH_count >= 1 ~ "1~",
            #CH_count >= 2 ~ "2~",
            #CH_count == 1 ~ "1",
            #CH_count >= 1 ~ "1~",
            #Infiltraive_genes == 1 ~ "1",
            #DNMT3A_R882_count >= 1 ~ "1~",
            #DNMT3A_Y735_count >= 1 ~ "1~",
            #DNMT3A_count >= 1 ~ "1~",
            #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
            #CHEK2_count >= 1 ~ "1~",
            #ATM_count >= 1 ~ "1~",
            #TP53_count >= 1 ~ "1~",
            #ASXL1_count >= 1 ~ "1~",
            #TET2_count >= 1 ~ "1~",
            TRUE ~ "0"
          ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ TOCH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("TOCH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        #term = c("TOCH count = 0", "LOCH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^TOCH_plus", term) ~ gsub("TOCH_plus", "TOCH count = ", term),
          # grepl("^LOCH_plus", term) ~ gsub("LOCH_plus", "LOCH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[281]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-liquid, any TOCH/LOCH, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      setProgress(detail = "9235")
      # survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus + LOCH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      # diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus + LOCH_plus,
      #                   data=Data_ICI_ToT, rho=0)
      # diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus + LOCH_plus,
      #                   data=Data_ICI_ToT, rho=1)
      # diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus + LOCH_plus,
      #                data=Data_ICI_ToT)
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus,
                     data=Data_ICI_ToT)
      plots[[282]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-liquid, TOCH (+)"), diff_0, diff_1, diff_2)
      # surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-liquid, TOCH/LOCH (+)"), diff_0, diff_1, diff_2)
      # diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus + LOCH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~ TOCH_plus + LOCH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^TOCH_plus", term) ~ gsub("TOCH_plus", "TOCH count = ", term),
          # grepl("^LOCH_plus", term) ~ gsub("LOCH_plus", "LOCH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[283]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, any TOCH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[257]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-solid, DDR TICH, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[258]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1 CDx, DDR TI-CH"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[259]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, DDR TICH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_DDR_TICH = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "DDR TICH (+)",
            TRUE ~ "DDR TICH (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_DDR_TICH <- render_gt({
        data_tmp_table_DDR_TICH %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and tumor-infiltrating DDR CH, F1-solid** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_DDR_TICH %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and tumor-infiltrating DDR CH, F1-solid** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_13.xlsx")

      data_tmp_table_regression = data_tmp_table_DDR_TICH %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_DDR_TICH_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1-solid, DDR TICH")
        tbl_DDR_TICH = tbl_DDR_TICH_ |> as_gt()
      } else {
        tbl_DDR_TICH = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_DDR_TICH = tbl_DDR_TICH_ |> as_gt()
      }
      as_hux_xlsx(tbl_DDR_TICH_,file = "source/Table_odds_14.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[326]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-solid, DDR CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[327]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, DDR CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT$Age_decade = factor(Data_ICI_ToT$Age_decade,
                                       levels = c("70~", "60~69", "50~59", "~49"))
      Data_ICI_ToT$TMB_class = droplevels(Data_ICI_ToT$TMB_class)
      Data_ICI_ToT$Age_decade = droplevels(Data_ICI_ToT$Age_decade)
      Data_ICI_ToT$Age_decade = relevel(Data_ICI_ToT$Age_decade, ref = "70~")
      Data_ICI_ToT$CH_plus = factor(Data_ICI_ToT$CH_plus)
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + TMB_class + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        TMB_class = "<10",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.2", "Subgroup"] = "TI-CH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "TMB_class"] = "TMB"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TI-CH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1 CDx, DDR TI-CH, time on immune checkpoint inhibitor trearment",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" ) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[359]] <- ggdraw(ggplot_plot)
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          DDR_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[260]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-liquid, DDR CH, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[261]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-Liquid, DDR LD-CH/LE-CH"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[262]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, DDR CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      setProgress(detail = "9865")
      data_tmp_table_DDR_TOCH = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "DDR CH (+)",
            TRUE ~ "DDR CH (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_DDR_TOCH <- render_gt({
        data_tmp_table_DDR_TOCH %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and DDR CH, F1-liquid** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_DDR_TOCH %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and DDR CH, F1-liquid** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_15.xlsx")

      data_tmp_table_regression = data_tmp_table_DDR_TOCH %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_DDR_TOCH_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1-liquid, DDR CH")
        tbl_DDR_TOCH = tbl_DDR_TOCH_ |> as_gt()
      } else {
        tbl_DDR_TOCH_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_DDR_TOCH = tbl_DDR_TOCH_ |> as_gt()
      }
      as_hux_xlsx(tbl_DDR_TOCH_,file = "source/Table_odds_16.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[328]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-liquid, DDR CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[329]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, DDR CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT$Age_decade = factor(Data_ICI_ToT$Age_decade,
                                       levels = c("70~", "60~69", "50~59", "~49"))
      Data_ICI_ToT$TMB_class = droplevels(Data_ICI_ToT$TMB_class)
      Data_ICI_ToT$Age_decade = droplevels(Data_ICI_ToT$Age_decade)
      Data_ICI_ToT$Age_decade = relevel(Data_ICI_ToT$Age_decade, ref = "70~")
      Data_ICI_ToT$CH_plus = factor(Data_ICI_ToT$CH_plus)
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + TMB_class + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        TMB_class = "<10",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "LD-CH/LE-CH detection"
      df_all["CH_plus.2", "Subgroup"] = "LD-CH/LE-CH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "TMB_class"] = "TMB"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "LD-CH/LE-CH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1-Liquid CDx, DDR LD-CH/LE-CH, time on immune checkpoint inhibitor trearment",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" ) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[361]] <- ggdraw(ggplot_plot)
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          Epigenetic_TOCH_count + Epigenetic_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[275]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-solid, Epigenetic TICH, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[276]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-solid, Epigenetic TI-CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[277]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, Epigenetic TICH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_Epi_TICH = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "DDR TICH (+)",
            TRUE ~ "DDR TICH (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_Epi_TICH <- render_gt({
        data_tmp_table_Epi_TICH %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and tumor-infiltrating Epigenetic CH, F1-solid** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_Epi_TICH %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and tumor-infiltrating Epigenetic CH, F1-solid** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_17.xlsx")

      data_tmp_table_regression = data_tmp_table_Epi_TICH %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_Epi_TICH_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1-solid, Epigenetic TICH")
        tbl_Epi_TICH = tbl_Epi_TICH_ |> as_gt()
      } else {
        tbl_Epi_TICH_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_Epi_TICH = tbl_Epi_TICH_ |> as_gt()
      }
      as_hux_xlsx(tbl_Epi_TICH_,file = "source/Table_odds_18.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[330]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-solid, Epigenetic CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      setProgress(detail = "10429")
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[331]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, epigenetic CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          Epigenetic_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[278]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-liquid, Epigenetic CH, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[279]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-liquid, Epigenetic CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[280]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, Epigenetic CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_Epi_TOCH = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "Epigenetic CH (+)",
            TRUE ~ "Epigenetic CH (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_Epi_TOCH <- render_gt({
        data_tmp_table_Epi_TOCH %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and Epigenetic CH, F1-liquid** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_Epi_TOCH %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and Epigenetic CH, F1-liquid** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_19.xlsx")

      data_tmp_table_regression = data_tmp_table_Epi_TOCH %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_Epi_TOCH_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1-liquid, Epigenetic CH")
        tbl_Epi_TOCH = tbl_Epi_TOCH_ |> as_gt()
      } else {
        tbl_Epi_TOCH_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_Epi_TOCH = tbl_Epi_TOCH_ |> as_gt()
      }
      as_hux_xlsx(tbl_Epi_TOCH_,file = "source/Table_odds_20.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[332]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-liquid, Epigenetic CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[333]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, epigenetic CH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[302]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-solid, ATM TICH variant, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[303]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-solid, ATM TICH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[304]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, ATM TICH variant, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[305]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-liquid, ATM CH variant, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      setProgress(detail = "10987")
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[306]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-liquid, ATM CH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[307]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, ATM CH variant, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[308]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-solid, TP53 TICH variant, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[309]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-solid, TP53 TICH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[310]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, TP53 TICH variant, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[311]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-liquid, TP53 CH variant, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[312]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-liquid, TP53 CH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[313]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, TP53 CH variant, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[314]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-solid, CHEK2 TICH variant, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[315]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-solid, CHEK2 TICH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[316]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, CHEK2 TICH variant, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[317]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-liquid, CHEK2 CH variant, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[318]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-liquid, CHEK2 CH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[319]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, CHEK2 CH variant, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      # print(1)
      # お蔵入り CHの重複変異を見たかったが感度不足
      # desired_columns <- c(
      #   "Gene", "Sample ID", "Cancer Type", "Protein Change", "Chromosome",
      #   "Start Pos", "End Pos", "Ref", "Var", "Allele Freq (T)",
      #   "Age", "Immunotherapy Treatment",
      #   "Time from Diagnosis", "Time to Blood Draw from Treatment",
      #   "Sex", "TMB (nonsynonymous)"
      # )
      # # 読み込み関数（必要な列だけ抽出）
      # read_selected_columns <- function(file) {
      #   df <- read_tsv(file, show_col_types = FALSE)
      #   # 使用可能な列だけ抽出
      #   df %>%
      #     select(any_of(desired_columns))
      # }
      # # ファイル名を作成
      # file_names <- paste0("source/NatGen/table-", 5:25, ".tsv")
      # # 一括読み込み
      # tables <- set_names(file_names) %>%
      #   map(read_selected_columns)
      # # 必要なら統合も可能
      # MSKCC_mutation_table <- bind_rows(tables, .id = "source_file")
      # MSKCC_clinical_table = read_tsv("source/NatGen/msk_ch_2020_clinical_data.tsv", show_col_types = FALSE)
      # MSKCC_clinical_table = MSKCC_clinical_table %>%
      #   dplyr::mutate(Age_decade = case_when(
      #     MSKCC_clinical_table$Age < 50 ~ "~49",
      #     MSKCC_clinical_table$Age < 60 ~ "50~59",
      #     MSKCC_clinical_table$Age < 70 ~ "60~69",
      #     TRUE ~ "70~",
      #   ))
      # MSKCC_clinical_table$TMB = MSKCC_clinical_table$`TMB (nonsynonymous)`
      # MSKCC_clinical_table = MSKCC_clinical_table %>%
      #   dplyr::mutate(TMB_class = case_when(
      #     MSKCC_clinical_table$TMB < 10 ~ "<10",
      #     MSKCC_clinical_table$TMB < 15 ~ "10~15",
      #     MSKCC_clinical_table$TMB < 20 ~ "15~20",
      #     TRUE ~ "20<=",
      #   ))
      # MSKCC_clinical_table = MSKCC_clinical_table %>%
      #   dplyr::filter(!is.na(`Cancer Type`))
      # MSKCC_clinical_table$Cancername = as.factor(MSKCC_clinical_table$`Cancer Type`)
      # MSKCC_clinical_table$Cancername = relevel(MSKCC_clinical_table$Cancername, ref = "Breast Cancer ")
      # generate_overlap_matrix <- function(mutation_data, gene_name1, gene_name2, mut1_filter, mut2_filter) {
      #   # 条件1に一致する症例ID
      #   ID_1 <- mutation_data %>%
      #     filter(Gene %in% gene_name1) %>%
      #     filter(!!mut1_filter) %>%
      #     pull(`Sample ID`) %>% unique()
      #     
      #   
      #   # 条件2に一致する症例ID
      #   ID_2 <- mutation_data %>%
      #     filter(Gene %in% gene_name2) %>%
      #     filter(!!mut2_filter) %>%
      #     pull(`Sample ID`) %>% unique()
      #   
      #   # 全体の対象数（その遺伝子の変異全体）
      #   total <- mutation_data %>%
      #     pull(`Sample ID`) %>% unique() %>% length()
      #   
      #   # 交差・差集合を計算してマトリクスに
      #   mat <- matrix(c(
      #     length(intersect(ID_1, ID_2)),
      #     length(setdiff(ID_1, ID_2)),
      #     length(setdiff(ID_2, ID_1)),
      #     total - length(unique(c(ID_1, ID_2)))
      #   ), nrow = 2, byrow = TRUE)
      #   
      #   rownames(mat) <- c(paste0("In ", deparse(substitute(mut1_filter))), paste0("Not in ", deparse(substitute(mut1_filter))))
      #   colnames(mat) <- c(paste0("In ", deparse(substitute(mut2_filter))), paste0("Not in ", deparse(substitute(mut2_filter))))
      #   
      #   return(mat)
      # }
      # library(rlang)
      # #mut1_cond <- expr(str_detect(`Protein Change`, "R") & !str_detect(`Protein Change`, "\\*|splice") )
      # mut1_cond <- expr(`Protein Change` == "Y735C")
      # mut2_cond <- expr(str_detect(`Protein Change`, "\\*|splice") & ((Ref == "C" & Var == "T") | (Ref == "G" & Var == "A")))
      # # 実行
      # generate_overlap_matrix(MSKCC_mutation_table, gene_name1 = "DNMT3A",  gene_name2 = c("DNMT3A", "CHEK2"), mut1_filter = mut1_cond, mut2_filter = mut2_cond)
      # print(1)
      print(1)
      setProgress(detail = "11660")
      desired_columns <- c(
        "Gene", "Sample ID", "Cancer Type", "Protein Change", "Chromosome",
        "Start Pos", "End Pos", "Ref", "Var", "Allele Freq (T)",
        "Age at sequencing", 
        "Months from start of Pembrolizumab...254", "PFS-I Status from start of Pembrolizumab",
        "Sex", "TMB (nonsynonymous)"
      )
      # 読み込み関数（必要な列だけ抽出）
      read_selected_columns <- function(file) {
        df <- read_tsv(file, show_col_types = FALSE)
        # 使用可能な列だけ抽出
        df %>%
          select(any_of(desired_columns))
      }
      # ファイル名を作成
      file_names <- paste0("source/GENIE/table-", 7:27, ".tsv")
      # 一括読み込み
      tables <- set_names(file_names) %>%
        map(read_selected_columns)
      # 必要なら統合も可能
      MSKCC_mutation_table <- bind_rows(tables, .id = "source_file")
      MSKCC_clinical_table = read_tsv("source/GENIE/combined_study_clinical_data.tsv", show_col_types = FALSE)
      MSKCC_clinical_table = MSKCC_clinical_table %>%
        dplyr::mutate(Age_decade = case_when(
          MSKCC_clinical_table$`Age at sequencing` < 50 ~ "~49",
          MSKCC_clinical_table$`Age at sequencing` < 60 ~ "50~59",
          MSKCC_clinical_table$`Age at sequencing` < 70 ~ "60~69",
          TRUE ~ "70~",
        ))
      MSKCC_clinical_table$TMB = MSKCC_clinical_table$`TMB (nonsynonymous)`
      MSKCC_clinical_table = MSKCC_clinical_table %>%
        dplyr::mutate(TMB_class = case_when(
          MSKCC_clinical_table$TMB < 10 ~ "<10",
          MSKCC_clinical_table$TMB < 15 ~ "10~15",
          MSKCC_clinical_table$TMB < 20 ~ "15~20",
          TRUE ~ "20<=",
        ))
      MSKCC_clinical_table = MSKCC_clinical_table %>%
        dplyr::filter(!is.na(`Cancer Type`)) %>%
        dplyr::filter(`Cancer Type` != "Colorectal Cancer") %>%
        dplyr::mutate(`Cancer Type Detailed` = case_when(
          `Cancer Type Detailed` %in% c("Lung Adenocarcinoma", "Lung Squamous Cell Carcinoma") ~ `Cancer Type Detailed`,
          TRUE ~ "Others"
        ))
      MSKCC_clinical_table$Cancername = as.factor(MSKCC_clinical_table$`Cancer Type Detailed`)
      MSKCC_clinical_table$Cancername = relevel(MSKCC_clinical_table$Cancername, ref = "Lung Adenocarcinoma")
      MSKCC_mutation_table$unique_alt = paste(MSKCC_mutation_table$Gene, MSKCC_mutation_table$`Protein Change`, sep="_")
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "TP53"))$`Sample ID`
      MSKCC_clinical_table$TP53 = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "ATM"))$`Sample ID`
      MSKCC_clinical_table$ATM = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "CHEK2"))$`Sample ID`
      MSKCC_clinical_table$CHEK2 = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4")))$`Sample ID`
      MSKCC_clinical_table$DDR = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D")))$`Sample ID`
      MSKCC_clinical_table$Epigenetic = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(!unique_alt %in% CH_candidates &
                                                         Gene == "TP53"))$`Sample ID`
      MSKCC_clinical_table$TP53_somatic = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, "Yes", "No")
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(unique_alt %in% CH_candidates |
                                                         (Gene == "CHEK2" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "RAD21" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "CBL" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "GNAS" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "MYCN" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "DNMT3A" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "TET2" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "ASXL1" & (str_detect(unique_alt, "trunc|fs|\\*|splice")))))$`Sample ID`
      MSKCC_clinical_table$TICH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = MSKCC_mutation_table$`Sample ID`
      MSKCC_clinical_table$CH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      MSKCC_clinical_table$CTx_time_on_treatment = MSKCC_clinical_table$`Months from start of Pembrolizumab...224`*365.25/12
      MSKCC_clinical_table$CTx_censor_tmp = MSKCC_clinical_table$`PFS-I Status from start of Pembrolizumab`
      MSKCC_clinical_table$CTx_censor = ifelse( MSKCC_clinical_table$CTx_censor_tmp == "1:DECEASED", 1, 0)
      Data_ICI_ToT = MSKCC_clinical_table %>%
        dplyr::mutate(CH_plus = case_when(
          TICH == 1 ~ "1~",
          TRUE ~ "0"
        ))# %>%
        # dplyr::filter(TMB<15)
      refs <- tibble::tibble(
        term = c("CH count = 0", "TP53: somatic mut (-)", "Cancer: Lung Adenocarcinoma", "Sex: Female", "TMB: <10", "Age: ~49"),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[355]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Overall survival, Pembro, GENIE BPC Lung, any TICH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + Cancername + TMB_class + Age_decade + Sex + TP53_somatic, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^TP53_somatic", term) ~ gsub("TP53_somatic", "TP53 somatic mut ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_class", "TMB: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[356]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "MSKCC, any TICH variant, PD-1i/PD-L1i") +
        theme_minimal() +
        theme(legend.position = "none")
      
      print(1)
      # 使用する列名（すべてのファイルで共通で使いたい列）
      desired_columns <- c(
        "Gene", "Sample ID", "Cancer Type", "Protein Change", "Chromosome",
        "Start Pos", "End Pos", "Ref", "Var", "Allele Freq (T)", "Allele Freq (N)",
        "Age at Which Sequencing was Reported (Days)", "Drug Type",
        "Overall Survival (Months)", "Overall Survival Status",
        "Sex", "TMB (nonsynonymous)"
      )
      # 読み込み関数（必要な列だけ抽出）
      read_selected_columns <- function(file) {
        df <- read_tsv(file, show_col_types = FALSE)
        # 使用可能な列だけ抽出
        df %>%
          select(any_of(desired_columns))
      }
      # ファイル名を作成
      file_names <- paste0("source/table-", 5:25, ".tsv")
      # 一括読み込み
      tables <- set_names(file_names) %>%
        map(read_selected_columns)
      # 必要なら統合も可能
      MSKCC_mutation_table <- bind_rows(tables, .id = "source_file")
      # # TICHのみ抽出
      # MSKCC_mutation_table = MSKCC_mutation_table %>%
      #   dplyr::filter(!is.na(`Allele Freq (N)`))
      MSKCC_clinical_table = read_tsv("source/tmb_mskcc_2018_clinical_data.tsv", show_col_types = FALSE)
      MSKCC_clinical_table = MSKCC_clinical_table %>%
        dplyr::mutate(Age_decade = case_when(
        MSKCC_clinical_table$`Age at Which Sequencing was Reported (Days)` < 50 ~ "~49",
        MSKCC_clinical_table$`Age at Which Sequencing was Reported (Days)` < 60 ~ "50~59",
        MSKCC_clinical_table$`Age at Which Sequencing was Reported (Days)` < 70 ~ "60~69",
        TRUE ~ "70~",
        ))
      MSKCC_clinical_table$TMB = MSKCC_clinical_table$`TMB (nonsynonymous)`
      MSKCC_clinical_table = MSKCC_clinical_table %>%
        dplyr::mutate(TMB_class = case_when(
          MSKCC_clinical_table$TMB < 10 ~ "<10",
          MSKCC_clinical_table$TMB < 15 ~ "10~15",
          MSKCC_clinical_table$TMB < 20 ~ "15~20",
          TRUE ~ "20<=",
        ))
      MSKCC_clinical_table = MSKCC_clinical_table %>%
        dplyr::filter(!is.na(`Cancer Type`))
      MSKCC_clinical_table = MSKCC_clinical_table %>%
        dplyr::filter(`Cancer Type` != "Skin Cancer, Non-Melanoma")
      MSKCC_clinical_table$Cancername = as.factor(MSKCC_clinical_table$`Cancer Type`)
      MSKCC_clinical_table$Cancername = relevel(MSKCC_clinical_table$Cancername, ref = "Non-Small Cell Lung Cancer")
      MSKCC_mutation_table$unique_alt = paste(MSKCC_mutation_table$Gene, MSKCC_mutation_table$`Protein Change`, sep="_")
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "TP53"))$`Sample ID`
      MSKCC_clinical_table$TP53 = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "TP53" & unique_alt %in% CH_candidates))$`Sample ID`
      MSKCC_clinical_table$TP53_TICH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "ATM"))$`Sample ID`
      MSKCC_clinical_table$ATM = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "ATM" & unique_alt %in% CH_candidates))$`Sample ID`
      MSKCC_clinical_table$ATM_TICH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "CHEK2"))$`Sample ID`
      MSKCC_clinical_table$CHEK2 = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene == "CHEK2" & (unique_alt %in% CH_candidates | str_detect(unique_alt, "trunc|fs|\\*|splice"))))$`Sample ID`
      MSKCC_clinical_table$CHEK2_TICH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4")))$`Sample ID`
      MSKCC_clinical_table$DDR = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & unique_alt %in% CH_candidates |
                                                         (Gene == "CHEK2" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "RAD21" & (str_detect(unique_alt, "trunc|fs|\\*|splice")))))$`Sample ID`
      MSKCC_clinical_table$DDR_TICH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D")))$`Sample ID`
      MSKCC_clinical_table$Epigenetic = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(Gene %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") & unique_alt %in% CH_candidates |
                                                         (Gene == "DNMT3A" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "TET2" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "ASXL1" & (str_detect(unique_alt, "trunc|fs|\\*|splice")))))$`Sample ID`
      MSKCC_clinical_table$Epigenetic_TICH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(unique_alt %in% CH_candidates |
                                                         (Gene == "CHEK2" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "RAD21" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "CBL" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "GNAS" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "MYCN" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "DNMT3A" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "TET2" & (str_detect(unique_alt, "trunc|fs|\\*|splice"))) |
                                                         (Gene == "ASXL1" & (str_detect(unique_alt, "trunc|fs|\\*|splice")))))$`Sample ID`
      MSKCC_clinical_table$TICH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = MSKCC_mutation_table$`Sample ID`
      MSKCC_clinical_table$CH = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, 1, 0)
      mut_ID = (MSKCC_mutation_table %>% dplyr::filter(!unique_alt %in% CH_candidates &
                                                         Gene == "TP53"))$`Sample ID`
      MSKCC_clinical_table$TP53_somatic = ifelse(MSKCC_clinical_table$`Sample ID` %in% mut_ID, "(+)", "(-)")
      MSKCC_clinical_table$CTx_time_on_treatment = MSKCC_clinical_table$`Overall Survival (Months)`*365.25/12
      MSKCC_clinical_table$CTx_censor_tmp = MSKCC_clinical_table$`Overall Survival Status`
      MSKCC_clinical_table$CTx_censor = ifelse( MSKCC_clinical_table$CTx_censor_tmp == "1:DECEASED", 1, 0)
      Data_ICI_ToT = MSKCC_clinical_table %>%
        dplyr::filter(`Drug Type` == "PD-1/PDL-1" & Cancername != "Skin Cancer, Non-Melanoma") %>%
        dplyr::mutate(CH_plus = case_when(
          Epigenetic_TICH == 1 ~ "1~",
          TRUE ~ "0"
        ))# %>%
        # dplyr::filter(TMB<15)
      Data_ICI_ToT$Cancername = as.factor(Data_ICI_ToT$`Cancer Type`)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "Non-Small Cell Lung Cancer")
      refs <- tibble::tibble(
        term = c("CH count = 0", "Cancer: Non-Small Cell Lung Cancer", "Sex: Female", "TMB: <10", "Age: ~49"),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[320]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival, PD-1i/PD-L1i, MSKCC, any TICH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + Cancername + TMB_class + Age_decade + Sex + TP53_somatic, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^TP53_somatic", term) ~ gsub("TP53_somatic", "TP53 somatic mut ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_class", "TMB: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[321]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "MSKCC, any TICH variant, PD-1i/PD-L1i") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = MSKCC_clinical_table %>%
        dplyr::filter(`Drug Type` == "PD-1/PDL-1" & Cancername != "Skin Cancer, Non-Melanoma") %>%
        dplyr::mutate(CH_plus = case_when(
          DDR_TICH == 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT$Cancername = as.factor(Data_ICI_ToT$`Cancer Type`)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "Non-Small Cell Lung Cancer")
      refs <- tibble::tibble(
        term = c("CH count = 0", "Cancer: Non-Small Cell Lung Cancer", "Sex: Female", "TMB: <10", "Age: ~49"),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[342]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival, PD-1i/PD-L1i, MSKCC, DDR TICH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_class", "TMB: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[343]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "MSKCC, DDR TICH variant, PD-1i/PD-L1i") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = MSKCC_clinical_table %>%
        dplyr::filter(`Drug Type` == "PD-1/PDL-1" & Cancername != "Skin Cancer, Non-Melanoma") %>%
        dplyr::mutate(CH_plus = case_when(
          Epigenetic_TICH == 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT$Cancername = as.factor(Data_ICI_ToT$`Cancer Type`)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "Non-Small Cell Lung Cancer")
      refs <- tibble::tibble(
        term = c("CH count = 0", "Cancer: Non-Small Cell Lung Cancer", "Sex: Female", "TMB: <10", "Age: ~49"),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[344]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival, PD-1i/PD-L1i, MSKCC, Epigenetic TICH variant (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      setProgress(detail = "12017")
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_class", "TMB: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[345]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "MSKCC, Epigenetic TICH variant, PD-1i/PD-L1i") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_summary_drug_effect <- as.data.frame(summary_dt)
      Data_summary_drug_effect = Data_summary_drug_effect %>% dplyr::filter(regimen %in% Target_drugs)
      Data_MAF_target_ICI = Data_MAF_target %>%
        dplyr::filter((non_germline_non_error <= 1 | non_germline_non_error == 6) &
                        !(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3) &
                        Panel %in% c("F1Liquid CDx", "FoundationOne CDx")) %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") & logfold <= 8 ~ "Epigenetic_TOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold <= 8 ~ "DDR_TOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold <= 8 ~ "Splicing_TOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold <= 8 ~ "MAPK_TOCH",
            Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") & logfold > 8 ~ "Epigenetic_LOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold > 8 ~ "DDR_LOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold > 8 ~ "Splicing_LOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold > 8 ~ "MAPK_LOCH",
            logfold > 8 ~ "Minor CH_LOCH",
            TRUE ~ "Minor CH_TOCH"),
          Hotspot_type_2 = case_when(
            Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2") & logfold <= 8 ~ "DNA_methyl_TOCH",
            Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D") & logfold <= 8 ~ "Histone_TOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold <= 8 ~ "DDR_TOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold <= 8 ~ "Splicing_TOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold <= 8 ~ "MAPK_TOCH",
            Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2") & logfold > 8 ~ "DNA_methyl_LOCH",
            Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D") & logfold > 8 ~ "Histone_LOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold > 8 ~ "DDR_LOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold > 8 ~ "Splicing_LOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold > 8 ~ "MAPK_LOCH",
            logfold > 8 ~ "Minor CH_LOCH",
            TRUE ~ "Minor CH_TOCH"),
          VAF_class = case_when(
            VAF < 0.01 ~ "VAF<0.01",
            VAF < 0.02 ~ "VAF<0.02",
            VAF < 0.05 ~ "VAF<0.05",
            VAF < 0.10 ~ "VAF<0.10",
            VAF < 0.20 ~ "VAF<0.20",
            VAF < 0.40 ~ "VAF<0.40",
            TRUE ~ "VAF>=0.40"
          ),
          Infiltrative_genes = case_when(
            Hugo_Symbol %in% Infiltrative_genes ~ 1,
            TRUE ~ 0
          ),
          Important_spot = case_when(
            str_detect(unique_alt, "DNMT3A_") ~ "DNMT3A",
            str_detect(unique_alt, "CHEK2_") ~ "CHEK2",
            str_detect(unique_alt, "ATM_") ~ "ATM",
            str_detect(unique_alt, "ASXL1_") ~ "ASXL1",
            str_detect(unique_alt, "TET2_") ~ "TET2",
            str_detect(unique_alt, "TP53_") ~ "TP53",
            TRUE ~ "OTHER"
          )
        )
      add_hotspot_count <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Hotspot_type %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      add_hotspot_count2 <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Hotspot_type_2 %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      add_hotspot_count3 <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Important_spot %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      Data_summary_drug_effect <- Data_summary_drug_effect %>%
        add_hotspot_count(Data_MAF_target_ICI, unique(Data_MAF_target_ICI$Hotspot_type), "CH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("Epigenetic_TOCH", "DDR_TOCH", "MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH"), "TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("Epigenetic_LOCH", "DDR_LOCH", "MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH"), "LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "DDR_TOCH", "DDR_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "Epigenetic_TOCH", "Epigenetic_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH"), "Other_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "DDR_LOCH", "DDR_LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "Epigenetic_LOCH", "Epigenetic_LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH"), "Other_LOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "DNA_methyl_TOCH", "DNA_methyl_TOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "DNA_methyl_LOCH", "DNA_methyl_LOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "Histone_TOCH", "Histone_TOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "Histone_LOCH", "Histone_LOCH_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "DNMT3A", "DNMT3A_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "CHEK2", "CHEK2_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "ASXL1", "ASXL1_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "ATM", "ATM_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "TET2", "TET2_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "TP53", "TP53_count")
      # Data_ICI_ToT = Data_summary_drug_effect %>%
      #           dplyr::filter(Panel == "F1Liquid CDx") %>%
      #           #dplyr::filter(Panel == "FoundationOne CDx") %>%
      #           dplyr::mutate(CH_plus = case_when(
      #             Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(VAF_class %in%
      #                                                                                c("VAF>=0.40", "VAF<0.40", "VAF<0.20", "VAF<0.10", "VAF<0.05", "VAF<0.02", "VAF<0.01") &
      #                                                                                #c("VAF>=0.40", "VAF<0.40", "VAF<0.20", "VAF<0.10", "VAF<0.05", "VAF<0.02") &
      #                                                                                #c("VAF>=0.40", "VAF<0.40", "VAF<0.20") &
      #                                                                                #c("VAF<0.10", "VAF<0.05", "VAF<0.02", "VAF<0.01") &
      #                                                                                #c("VAF<0.01") &
      #                                                                                Hotspot_type %in%
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH", "DDR_TOCH", "DDR_LOCH", "MAPK_TOCH", "MAPK_LOCH", "Minor CH_TOCH", "Minor CH_LOCH", "Splicing_TOCH", "Splicing_LOCH")
      #                                                                                c("MAPK_TOCH", "MAPK_LOCH", "Minor CH_TOCH", "Minor CH_LOCH", "Splicing_TOCH", "Splicing_LOCH")
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH", "DDR_TOCH", "DDR_LOCH")
      #                                                                                #c("DDR_TOCH", "DDR_LOCH")
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH")
      #                                                                                #c("MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH")
      #                                                                                #c("DDR_LOCH")
      #                                                                                #c("Epigenetic_LOCH")
      #                                                                                #c("MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH")
      #                                                                                #c("DDR_TOCH")
      #                                                                                #c("Epigenetic_TOCH")
      #             ))$Tumor_Sample_Barcode ~ 1,
      #             TRUE ~ 0
      #           ))
      # Data_ICI_ToT = Data_summary_drug_effect %>%
      #   #dplyr::filter(Panel == "F1Liquid CDx") %>%
      #   dplyr::filter(Panel == "FoundationOne CDx") %>%
      #   dplyr::mutate(CH_plus = case_when(
      #     Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(Hotspot_type %in%
      #                                                                        # c("Epigenetic_TOCH", "DDR_TOCH", "Splicing_TOCH", "MAPK_TOCH", "Minor CH_TOCH",
      #                                                                        #   "Epigenetic_LOCH", "DDR_LOCH", "Splicing_LOCH", "MAPK_LOCH", "Minor CH_LOCH")
      #                                                                        "DDR_LOCH"
      #     ))$Tumor_Sample_Barcode ~ 1,
      #     TRUE ~ 0
      #   ))
      Data_summary_drug_effect = Data_summary_drug_effect %>%
        dplyr::mutate(
          TMB_class = case_when(
            TMB < 10 ~ "TMB<10",
            TMB < 14 ~ "10≤TMB<14",
            TMB < 20 ~ "14≤TMB<20",
            TRUE ~ "20<=TMB"
          )) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        ))
      Data_summary_drug_effect$TMB_class = factor(Data_summary_drug_effect$TMB_class,
                                                  levels = c("TMB<10", "10≤TMB<14", "14≤TMB<20", "20<=TMB"))
      Data_summary_drug_effect$regimen = factor(Data_summary_drug_effect$regimen)
      Data_summary_drug_effect$regimen = relevel(Data_summary_drug_effect$regimen,
                                                 ref = names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])
      Data_summary_drug_effect = Data_summary_drug_effect %>%
        dplyr::mutate(Infiltraive_genes = case_when(
          Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(Infiltrative_genes == 1))$Tumor_Sample_Barcode ~ 1,
          TRUE ~ 0
        ))
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[263]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1 CDx, DDR SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[264]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1 CDx, DDR SEV"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[265]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1 CDx, DDR SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_DDR_SV = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "DDR SEV (+)",
            TRUE ~ "DDR SEV (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_DDR_SV <- render_gt({
        data_tmp_table_DDR_SV %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and DDR SEV, F1 CDx** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_DDR_SV %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and DDR SEV, F1 CDx** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_21.xlsx")

      data_tmp_table_regression = data_tmp_table_DDR_SV %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_DDR_SV_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1 CDx, DDR SEV")
        tbl_DDR_SV = tbl_DDR_SV_ |> as_gt()
      } else {
        tbl_DDR_SV_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_DDR_SV = tbl_DDR_SV_ |> as_gt()
      }
      as_hux_xlsx(tbl_DDR_SV_,file = "source/Table_odds_22.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[340]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1 CDx, DDR SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[341]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1 CDx, DDR SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT$Age_decade = factor(Data_ICI_ToT$Age_decade,
                                       levels = c("70~", "60~69", "50~59", "~49"))
      Data_ICI_ToT$TMB_class = droplevels(Data_ICI_ToT$TMB_class)
      Data_ICI_ToT$Age_decade = droplevels(Data_ICI_ToT$Age_decade)
      Data_ICI_ToT$Age_decade = relevel(Data_ICI_ToT$Age_decade, ref = "70~")
      Data_ICI_ToT$CH_plus = factor(Data_ICI_ToT$CH_plus)
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + TMB_class + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        TMB_class = "<10",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "DDR gene SEV"
      df_all["CH_plus.2", "Subgroup"] = "DDR gene SEV"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "TMB_class"] = "TMB"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "DDR gene SEV"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1 CDx, DDR SEV, time on immune checkpoint inhibitor trearment",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" ) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[362]] <- ggdraw(ggplot_plot)
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[290]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1 CDx, ATM SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      setProgress(detail = "12738")
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[291]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1 CDx, ATM SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[292]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1 CDx, ATM SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[293]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-Liquid, ATM SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[294]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-Liquid CDx, ATM SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[295]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-Liquid CDx, ATM SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[296]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1 CDx, TP53 SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[297]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1 CDx, TP53 SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      setProgress(detail = "13045")
      plots[[298]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1 CDx, TP53 SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[299]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-Liquid CDx, TP53 SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[300]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-Liquid, TP53 SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[301]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-Liquid CDx, TP53 SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[284]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1 CDx, CHEK2 SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[285]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1 CDx, CHEK2 SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[286]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1 CDx, CHEK2 SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[287]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-Liquid CDx, CHEK2 SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[288]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-Liquid CDx, CHEK2 SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[289]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-Liquid CDx, CHEK2 SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          DDR_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[266]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-Liquid CDx, DDR SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[267]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-Liquid CDx, DDR SEV"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[268]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-Liquid CDx, DDR SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_DDR_LV = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "DDR CH (+)",
            TRUE ~ "DDR CH (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_DDR_LV <- render_gt({
        data_tmp_table_DDR_LV %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and DDR SEV, F1-Liquid CDx** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_DDR_LV %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and DDR SEV, F1-Liquid CDx** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_23.xlsx")

      data_tmp_table_regression = data_tmp_table_DDR_LV %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_DDR_LV_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1-Liquid CDx, DDR SEV")
        tbl_DDR_LV = tbl_DDR_LV_ |> as_gt()
      } else {
        tbl_DDR_LV_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_DDR_LV = tbl_DDR_LV_ |> as_gt()
      }
      as_hux_xlsx(tbl_DDR_LV_,file = "source/Table_odds_24.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[338]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-Liquid CDx, DDR SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      setProgress(detail = "13749")
      plots[[339]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-Liquid CDx, DDR SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT$Age_decade = factor(Data_ICI_ToT$Age_decade,
                                       levels = c("70~", "60~69", "50~59", "~49"))
      Data_ICI_ToT$TMB_class = droplevels(Data_ICI_ToT$TMB_class)
      Data_ICI_ToT$Age_decade = droplevels(Data_ICI_ToT$Age_decade)
      Data_ICI_ToT$Age_decade = relevel(Data_ICI_ToT$Age_decade, ref = "70~")
      Data_ICI_ToT$CH_plus = factor(Data_ICI_ToT$CH_plus)
      surv_obj <- Surv(time = Data_ICI_ToT$CTx_time_on_treatment,
                       event = Data_ICI_ToT$CTx_censor)
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + Age_decade + Sex + TMB_class + regimen + Alcoholic_history + Smoking_history + Brain_met + Lymph_met + Lung_met + Bone_met + Liver_met")
      #form <- as.formula("surv_obj ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen")
      fit <- coxph(form, data = Data_ICI_ToT)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "regimen", "Age_decade", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "regimen", "Age_decade", "Sex", "TMB_class", "CTx_line")
      #subgroups <- c("CH_plus", "TMB_class", "regimen")
      #subgroups <- c("CH_plus", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "TMB_class", "Age_decade", "regimen")
      #subgroups <- c("CH_plus", "Cancername", "Sex", "CTx_line", "Age_decade", "regimen", "Brain_met", "Alcoholic_history", "Smoking_history", "Lymph_met", "Lung_met", "Bone_met", "Liver_met")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "LUNG",
        Sex = "Female",
        CTx_line = "1",
        TMB_class = "<10",
        Age_decade = "70~",
        regimen = "Nivolumab"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_ICI_ToT[[var]]))
        counts <- as.numeric(table(Data_ICI_ToT[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "DDR gene SEV"
      df_all["CH_plus.2", "Subgroup"] = "DDR gene SEV"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      df_all$Label[df_all$Label == "<10"] = "<10 (Reference)"
      df_all$Subgroup[df_all$Subgroup == "TMB_class"] = "TMB"
      df_all$Subgroup[df_all$Subgroup == "Cancername"] = "Cancer type"
      df_all$Subgroup[df_all$Subgroup == "regimen"] = "Regimen"
      df_all$Subgroup[df_all$Subgroup == "CTx_line"] = "CTx line"
      df_all$Subgroup[df_all$Subgroup == "Age_decade"] = "Age"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "DDR gene SEV"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      #labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               boxsize = .2,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model\nF1-Liquid CDx, DDR SEV, time on immune checkpoint inhibitor trearment",
                               is.summary = df_plot$Label %in% unique_subgroups,
                               lty.zero = 4,
                               lwd.zero = 2, 
                               col.zero = "black" ) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "darkblue", cex = .1),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = .1)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[363]] <- ggdraw(ggplot_plot)
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          Epigenetic_TOCH_count + Epigenetic_LOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[269]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1 CDx, Epigenetic SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[270]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1 CDx, Epigenetic SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[271]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1 CDx, Epigenetic SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_Epi_SV = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "Epigenetic SEV (+)",
            TRUE ~ "Epigenetic SEV (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_Epi_SV <- render_gt({
        data_tmp_table_Epi_SV %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and epigenetic SEV, F1 CDx** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_Epi_SV %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and epigenetic SEV, F1 CDx** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_25.xlsx")

      data_tmp_table_regression = data_tmp_table_Epi_SV %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      setProgress(detail = "14128")
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_Epi_SV_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1 CDx, Epigenetic SEV")
        tbl_Epi_SV = tbl_Epi_SV_ |> as_gt()
      } else {
        tbl_Epi_SV_ = univ_tab |>
          modify_caption("Factors, favor for Nivo/Pembro objective response, no significant factor in multivariable analysis")
        tbl_Epi_SV = tbl_Epi_SV_ |> as_gt()
      }
      as_hux_xlsx(tbl_Epi_SV_,file = "source/Table_odds_26.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[336]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1 CDx, Epigenetic SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[337]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1 CDx, epigenetic SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "PANCREAS", "PERITONEUM", "KIDNEY", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[272]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-Liquid CDx, Epigenetic SEV, Nivo/Pembro",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[273]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Nivo/Pembro, F1-Liquid CDx, Epigenetic SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[274]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-Liquid CDx, Epigenetic SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      data_tmp_table_Epi_LV = Data_ICI_ToT %>%
        dplyr::mutate(
          CH_status = case_when(
            CH_plus == "1~" ~ "Epigenetic SEV (+)",
            TRUE ~ "Epigenetic SEV (-)")
        ) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        )) %>%
        dplyr::filter(!is.na(CTx_start_date)) %>%
        dplyr::select(Panel,
                      Age, Age_decade, 
                      Sex, Cancername,
                      Smoking_history, Alcoholic_history,
                      Double_cancer, Multiple_nodules,
                      Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
                      CH_status, TMB, TMB_class, RECIST, CTx_Effect,
                      CTx_time_on_treatment, CTx_censor, CTx_line,
                      regimen)
      output$table_inter_panel_Epi_LV <- render_gt({
        data_tmp_table_Epi_LV %>%
          tbl_summary(
            by = "CH_status",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Nivo/Pembro effect and Epigenetic SEV, F1-Liquid CDx** (N = {N})") %>%
          add_n() %>%
          add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
          bold_labels() %>% as_gt()
      })
      table_flextable = data_tmp_table_Epi_LV %>%
        tbl_summary(
          by = "CH_status",
          statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                "{mean} ({sd})",
                                                "{median} ({p25}, {p75})", 
                                                "{min}, {max}"),
                           all_categorical() ~ "{n} / {N} ({p}%)"),
          type = list(all_continuous() ~ "continuous2",
                      all_dichotomous() ~ "categorical"),
          digits = all_continuous() ~ 2,
          missing = "ifany") %>%
        modify_caption("**Nivo/Pembro effect and Epigenetic SEV, F1-Liquid CDx** (N = {N})") %>%
        add_n() %>%
        add_p(test=list(all_continuous() ~ "t.test", all_categorical() ~ "chisq.test")) %>%
        bold_labels()
      as_hux_xlsx(table_flextable,file = "source/Table_odds_27.xlsx")

      data_tmp_table_regression = data_tmp_table_Epi_LV %>% dplyr::select(
        Age_decade, 
        Sex, Cancername,
        Smoking_history, Alcoholic_history,
        Double_cancer, Multiple_nodules,
        Lymph_met, Brain_met, Lung_met, Bone_met, Liver_met,
        CH_status, TMB_class, CTx_Effect,
        regimen
      )
      univ_tab <- data_tmp_table_regression %>%
        tbl_uvregression(
          method = glm,
          y = CTx_Effect,
          method.args = list(family = binomial),
          exponentiate = TRUE
        ) |>
        add_global_p() |> # add global p-value
        add_n(location = "level") |>
        add_nevent(location = "level") |> # add number of events of the outcome
        add_q() |> # adjusts global p-values for multiple testing
        bold_p() |> # bold p-values under a given threshold (default 0.05)
        modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
        bold_labels()
      m1 <- glm(CTx_Effect ~., data = data_tmp_table_regression, family = binomial(link = "logit"))
      final_mv_reg <- m1 %>%
        stats::step(direction = "backward", trace = FALSE)
      significant_factors = names(final_mv_reg$xlevels)
      if(!is.null(significant_factors)){
        significant_factors_CTx_Effect = c(significant_factors, "CTx_Effect")
        data_tmp_mv = data.table(data_tmp_table_regression)[,..significant_factors_CTx_Effect]
        m2 <- glm(CTx_Effect ~., data = data_tmp_mv, family = binomial(link = "logit"))
        mv_tab = m2 |>
          tbl_regression(                         ## 多変量解析の表を生成
            exponentiate = TRUE,tidy_fun = broom.helpers::tidy_parameters,conf.level = 0.95,ci_method = "wald"
          ) |>
          add_global_p() |> # add global p-value
          add_q() |> # adjusts global p-values for multiple testing
          bold_p() |> # bold p-values under a given threshold (default 0.05)
          modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
          bold_labels()
        tbl_Epi_LV_ = tbl_merge(
          tbls = list(univ_tab, mv_tab),
          tab_spanner = c("**Univariate**", "**Multivariable**")) |>
          modify_caption("Factors for Nivo/Pembro objective response, F1-Liquid CDx, Epigenetic SEV")
        tbl_Epi_LV = tbl_Epi_LV_ |> as_gt()
      } else {
        tbl_Epi_LV_ = univ_tab |>
          modify_caption("Factors, favor for Epigenetic SEV, no significant factor in multivariable analysis")
        tbl_Epi_LV = tbl_Epi_LV_ |> as_gt()
      }
      as_hux_xlsx(tbl_Epi_LV_,file = "source/Table_odds_28.xlsx")

      survfit_t <- survfit(Surv(Overall_survival, censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[334]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival from 1L-CTx initiation, Nivo/Pembro, F1-Liquid CDx, Epigenetic SEV (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[335]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for overall survival from ICI initiation", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-Liquid CDx, epigenetic SEV, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      Target_drugs = c(
        "nab-paclitaxel",
        "Paclitaxel",
        "Docetaxel"
      )
      Data_case_target = Data_case()
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor, final_observe) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      setProgress(detail = "14520")
      Data_ICI_pre = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      final_observe,
                      censor) %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in% c("緩和","その他") &
                        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% Target_drugs) %>%
        dplyr::select(-症例.EP前レジメン情報.実施目的.名称.) %>%
        dplyr::distinct()
      colnames(Data_ICI_pre) = c("Tumor_Sample_Barcode", "drug", "RECIST", "CTx_line", "entry_date", "CTx_start_date", "CTx_end_date", "ongoing", "final_observe", "censor")
      Data_ICI_pre = Data_ICI_pre %>%
        dplyr::filter(CTx_start_date != "") %>%
        dplyr::mutate(
          CTx_time_on_treatment = case_when(
            CTx_end_date != "" ~ as.integer(as.Date(CTx_end_date) - as.Date(CTx_start_date)),
            TRUE ~ as.integer(as.Date(entry_date) - as.Date(CTx_start_date))),
          CTx_censor  = case_when(
            !is.na(ongoing) ~ 0,
            TRUE ~ 1)
        ) %>%
        dplyr::select(-entry_date, -ongoing)
      Data_ICI_post = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      final_observe,
                      censor) %>%
        dplyr::filter(症例.EP後レジメン情報.薬剤名.YJ一般名.EN. %in% Target_drugs) %>%
        dplyr::distinct()
      colnames(Data_ICI_post) = c("Tumor_Sample_Barcode", "drug", "RECIST", "CTx_line", "CTx_start_date", "CTx_end_date", "final_observe", "censor")
      Data_ICI_post = Data_ICI_post %>%
        dplyr::filter(CTx_start_date != "" & CTx_end_date != "") %>%
        dplyr::mutate(CTx_time_on_treatment = as.integer(as.Date(CTx_end_date) - as.Date(CTx_start_date)))
      Data_ICI_post$CTx_censor = 1
      Data_ICI = rbind(Data_ICI_pre, Data_ICI_post)
      Data_ICI = Data_ICI %>%
        dplyr::filter(final_observe != "") %>%
        dplyr::mutate(Overall_survival = as.integer(as.Date(final_observe) - as.Date(CTx_start_date)))
      Data_ICI$Overall_survival[Data_ICI$Overall_survival < 0] <- NA
      Data_ICI$RECIST[is.na(Data_ICI$RECIST)] = "NE"
      Data_ICI$RECIST = factor(Data_ICI$RECIST, levels = c("CR","PR","SD","PD","NE"))
      Data_ICI = Data_ICI %>%
        dplyr::mutate(CTx_Effect = case_when(
          RECIST %in% c("CR","PR") ~ 1,
          RECIST %in% c("NE") ~ NA_integer_,
          TRUE ~ 0
        ))
      Data_ICI = Data_ICI %>%
        dplyr::mutate(CTx_line = case_when(
          CTx_line == "１次治療" ~ "1",
          CTx_line == "２次治療" ~ "2",
          CTx_line == "３次治療" ~ "3",
          CTx_line == "４次治療" ~ "4",
          CTx_line == "５次治療以降" ~ "5~",
          CTx_line == "不明" ~ "Unknown",
          TRUE ~ "Unknown",
        )) %>%
        dplyr::filter(CTx_line != "Unknown")
      Data_ICI = Data_ICI %>%
        dplyr::filter(CTx_start_date != "") %>%
        dplyr::arrange(CTx_start_date) %>%
        dplyr::filter(CTx_time_on_treatment>=0) %>%
        dplyr::distinct(Tumor_Sample_Barcode,.keep_all=T) %>%
        dplyr::select(Tumor_Sample_Barcode, CTx_time_on_treatment, CTx_censor, CTx_start_date, CTx_end_date, RECIST, CTx_Effect, CTx_line, Overall_survival)
      Data_summary_drug_effect = left_join(Data_summary, Data_ICI, by="Tumor_Sample_Barcode")
      # data.table に変換
      summary_dt <- as.data.table(Data_summary_drug_effect)
      case_dt <- as.data.table(Data_case())
      # 前後のレジメン情報を整形して結合
      pre_dt <- case_dt[, .(Tumor_Sample_Barcode = `C.CAT調査結果.基本項目.ハッシュID`,
                            CTx_start_date = `症例.EP前レジメン情報.投与開始日`,
                            drug = `症例.EP前レジメン情報.薬剤名.YJ一般名.EN.`)]
      post_dt <- case_dt[, .(Tumor_Sample_Barcode = `C.CAT調査結果.基本項目.ハッシュID`,
                             CTx_start_date = `症例.EP後レジメン情報.投与開始日`,
                             drug = `症例.EP後レジメン情報.薬剤名.YJ一般名.EN.`)]
      # 前後の情報を結合
      drug_dt <- rbind(pre_dt, post_dt)
      drug_dt <- drug_dt[!is.na(CTx_start_date) & drug != ""]
      # 同じ患者・投与日ごとに薬剤名をまとめる（重複除く）
      drug_agg <- drug_dt[, .(regimen = paste(unique(drug), collapse = "/")),
                          by = .(Tumor_Sample_Barcode, CTx_start_date)]
      # summary_dt に結合
      summary_dt <- merge(summary_dt, drug_agg,
                          by = c("Tumor_Sample_Barcode", "CTx_start_date"),
                          all.x = TRUE)
      # regimen 列がない行は空文字列に
      summary_dt[is.na(regimen), regimen := ""]
      summary_dt$regimen = str_replace(summary_dt$regimen, "Nivolumab,Ipilimumab", "Ipilimumab,Nivolumab")
      #summary_dt$Overall_survival = summary_dt$time_palliative_final
      # 結果を元の data.frame に戻すなら
      Data_summary_drug_effect <- as.data.frame(summary_dt)
      Data_summary_drug_effect = Data_summary_drug_effect %>% dplyr::filter(regimen %in% Target_drugs)
      Data_MAF_target_ICI = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1 &
                        unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel %in% c("F1Liquid CDx", "FoundationOne CDx")) %>%
        dplyr::mutate(
          Hotspot_type = case_when(
            Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") & logfold <= 8 ~ "Epigenetic_TOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold <= 8 ~ "DDR_TOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold <= 8 ~ "Splicing_TOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold <= 8 ~ "MAPK_TOCH",
            Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1", "IDH2", "EZH2", "KMT2D") & logfold > 8 ~ "Epigenetic_LOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold > 8 ~ "DDR_LOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold > 8 ~ "Splicing_LOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold > 8 ~ "MAPK_LOCH",
            logfold > 8 ~ "Minor CH_LOCH",
            TRUE ~ "Minor CH_TOCH"),
          Hotspot_type_2 = case_when(
            Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2") & logfold <= 8 ~ "DNA_methyl_TOCH",
            Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D") & logfold <= 8 ~ "Histone_TOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold <= 8 ~ "DDR_TOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold <= 8 ~ "Splicing_TOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold <= 8 ~ "MAPK_TOCH",
            Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2") & logfold > 8 ~ "DNA_methyl_LOCH",
            Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D") & logfold > 8 ~ "Histone_LOCH",
            Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") & logfold > 8 ~ "DDR_LOCH",
            Hugo_Symbol %in% c("SF3B1", "U2AF1") & logfold > 8 ~ "Splicing_LOCH",
            Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL") & logfold > 8 ~ "MAPK_LOCH",
            logfold > 8 ~ "Minor CH_LOCH",
            TRUE ~ "Minor CH_TOCH"),
          VAF_class = case_when(
            VAF < 0.01 ~ "VAF<0.01",
            VAF < 0.02 ~ "VAF<0.02",
            VAF < 0.05 ~ "VAF<0.05",
            VAF < 0.10 ~ "VAF<0.10",
            VAF < 0.20 ~ "VAF<0.20",
            VAF < 0.40 ~ "VAF<0.40",
            TRUE ~ "VAF>=0.40"
          ),
          Infiltrative_genes = case_when(
            Hugo_Symbol %in% Infiltrative_genes ~ 1,
            TRUE ~ 0
          ),
          Important_spot = case_when(
            str_detect(unique_alt, "DNMT3A_") ~ "DNMT3A",
            str_detect(unique_alt, "CHEK2_") ~ "CHEK2",
            str_detect(unique_alt, "ATM_") ~ "ATM",
            str_detect(unique_alt, "ASXL1_") ~ "ASXL1",
            str_detect(unique_alt, "TET2_") ~ "TET2",
            str_detect(unique_alt, "TP53_") ~ "TP53",
            TRUE ~ "OTHER"
          )
        )
      add_hotspot_count <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Hotspot_type %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      add_hotspot_count2 <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Hotspot_type_2 %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      add_hotspot_count3 <- function(df_summary, df_maf, hotspot_vec, new_colname) {
        barcode_counts <- df_maf %>%
          filter(Important_spot %in% hotspot_vec) %>%
          count(Tumor_Sample_Barcode, name = new_colname)
        df_summary %>%
          left_join(barcode_counts, by = "Tumor_Sample_Barcode") %>%
          tidyr::replace_na(setNames(list(0), new_colname))
      }
      Data_summary_drug_effect <- Data_summary_drug_effect %>%
        add_hotspot_count(Data_MAF_target_ICI, unique(Data_MAF_target_ICI$Hotspot_type), "CH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("Epigenetic_TOCH", "DDR_TOCH", "MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH"), "TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("Epigenetic_LOCH", "DDR_LOCH", "MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH"), "LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "DDR_TOCH", "DDR_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "Epigenetic_TOCH", "Epigenetic_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH"), "Other_TOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "DDR_LOCH", "DDR_LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, "Epigenetic_LOCH", "Epigenetic_LOCH_count") %>%
        add_hotspot_count(Data_MAF_target_ICI, c("MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH"), "Other_LOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "DNA_methyl_TOCH", "DNA_methyl_TOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "DNA_methyl_LOCH", "DNA_methyl_LOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "Histone_TOCH", "Histone_TOCH_count") %>%
        add_hotspot_count2(Data_MAF_target_ICI, "Histone_LOCH", "Histone_LOCH_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "DNMT3A", "DNMT3A_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "CHEK2", "CHEK2_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "ASXL1", "ASXL1_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "ATM", "ATM_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "TET2", "TET2_count") %>%
        add_hotspot_count3(Data_MAF_target_ICI, "TP53", "TP53_count")
      # Data_ICI_ToT = Data_summary_drug_effect %>%
      #           dplyr::filter(Panel == "F1Liquid CDx") %>%
      #           #dplyr::filter(Panel == "FoundationOne CDx") %>%
      #           dplyr::mutate(CH_plus = case_when(
      #             Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(VAF_class %in%
      #                                                                                c("VAF>=0.40", "VAF<0.40", "VAF<0.20", "VAF<0.10", "VAF<0.05", "VAF<0.02", "VAF<0.01") &
      #                                                                                #c("VAF>=0.40", "VAF<0.40", "VAF<0.20", "VAF<0.10", "VAF<0.05", "VAF<0.02") &
      #                                                                                #c("VAF>=0.40", "VAF<0.40", "VAF<0.20") &
      #                                                                                #c("VAF<0.10", "VAF<0.05", "VAF<0.02", "VAF<0.01") &
      #                                                                                #c("VAF<0.01") &
      #                                                                                Hotspot_type %in%
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH", "DDR_TOCH", "DDR_LOCH", "MAPK_TOCH", "MAPK_LOCH", "Minor CH_TOCH", "Minor CH_LOCH", "Splicing_TOCH", "Splicing_LOCH")
      #                                                                                c("MAPK_TOCH", "MAPK_LOCH", "Minor CH_TOCH", "Minor CH_LOCH", "Splicing_TOCH", "Splicing_LOCH")
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH", "DDR_TOCH", "DDR_LOCH")
      #                                                                                #c("DDR_TOCH", "DDR_LOCH")
      #                                                                                #c("Epigenetic_TOCH", "Epigenetic_LOCH")
      #                                                                                #c("MAPK_LOCH", "Minor CH_LOCH", "Splicing_LOCH")
      #                                                                                #c("DDR_LOCH")
      #                                                                                #c("Epigenetic_LOCH")
      #                                                                                #c("MAPK_TOCH", "Minor CH_TOCH", "Splicing_TOCH")
      #                                                                                #c("DDR_TOCH")
      #                                                                                #c("Epigenetic_TOCH")
      #             ))$Tumor_Sample_Barcode ~ 1,
      #             TRUE ~ 0
      #           ))
      # Data_ICI_ToT = Data_summary_drug_effect %>%
      #   #dplyr::filter(Panel == "F1Liquid CDx") %>%
      #   dplyr::filter(Panel == "FoundationOne CDx") %>%
      #   dplyr::mutate(CH_plus = case_when(
      #     Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(Hotspot_type %in%
      #                                                                        # c("Epigenetic_TOCH", "DDR_TOCH", "Splicing_TOCH", "MAPK_TOCH", "Minor CH_TOCH",
      #                                                                        #   "Epigenetic_LOCH", "DDR_LOCH", "Splicing_LOCH", "MAPK_LOCH", "Minor CH_LOCH")
      #                                                                        "DDR_LOCH"
      #     ))$Tumor_Sample_Barcode ~ 1,
      #     TRUE ~ 0
      #   ))
      Data_summary_drug_effect = Data_summary_drug_effect %>%
        dplyr::mutate(
          TMB_class = case_when(
            TMB < 10 ~ "TMB<10",
            TMB < 14 ~ "10≤TMB<14",
            TMB < 20 ~ "14≤TMB<20",
            TRUE ~ "20<=TMB"
          )) %>%
        dplyr::mutate(Age_decade = case_when(
          Age < 50 ~ "~49",
          Age < 60 ~ "50~59",
          Age < 70 ~ "60~69",
          TRUE ~ "70~"
        ))
      Data_summary_drug_effect$TMB_class = factor(Data_summary_drug_effect$TMB_class,
                                                  levels = c("TMB<10", "10≤TMB<14", "14≤TMB<20", "20<=TMB"))
      Data_summary_drug_effect$regimen = factor(Data_summary_drug_effect$regimen)
      Data_summary_drug_effect$regimen = relevel(Data_summary_drug_effect$regimen,
                                                 ref = names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])
      Data_summary_drug_effect = Data_summary_drug_effect %>%
        dplyr::mutate(Infiltraive_genes = case_when(
          Tumor_Sample_Barcode %in% (Data_MAF_target_ICI %>% dplyr::filter(Infiltrative_genes == 1))$Tumor_Sample_Barcode ~ 1,
          TRUE ~ 0
        ))
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        #dplyr::filter(Panel == "F1Liquid CDx") %>%
        dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          #TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "BRAIN", "KIDNEY", "CERVIX", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT, family=binomial)
      #Data_ORR = glm(CTx_Effect ~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT, family=binomial)
      # if(!is.na(exp(coef(Data_ORR))["CH_plus1"][[1]])){
      # 2. 結果を整形して OR・CI を計算
      refs <- tibble::tibble(
        term = c("CTx line = 1", "CH count = 0", "Cancer: LUNG", "Sex: Female", "TMB: <10", "Age: ~49", paste0("Drug: ", names(sort(table(Data_summary_drug_effect$regimen),decreasing = T))[[1]])),  # 基準カテゴリ
        estimate = 1, conf.low = 1, conf.high = 1, p.value = NA
      )
      tidy_ORR <- tidy(Data_ORR, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        ))
      tidy_ORR <- tidy_ORR %>%
        mutate(term = factor(term, levels = tidy_ORR$term))
      # 3. Forest plot
      # データ結合・整形
      tidy_ORR <- bind_rows(tidy_ORR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("OR = %.2f", estimate))
        )
      plots[[346]] = ggplot(tidy_ORR, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(
          title = "Odds Ratios for objective response, F1-solid, any TICH, Taxane",
          x = "",
          y = "Odds Ratio (95% CI)"
        ) +
        theme_minimal()
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[347]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Taxane, F1-solid, any TICH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[348]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, any TICH, Nivo/Pembro") +
        theme_minimal() +
        theme(legend.position = "none")
      survfit_t <- survfit(Surv(Overall_survival, censor)~ CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~ CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~ CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~ CH_plus,
                     data=Data_ICI_ToT)
      plots[[351]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival, Taxane, F1-solid, TICH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[352]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Overall survival", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-solid, any TICH, Taxane") +
        theme_minimal() +
        theme(legend.position = "none")
      Data_ICI_ToT = Data_summary_drug_effect %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::filter(Panel == "F1Liquid CDx") %>%
        #dplyr::filter(Panel == "FoundationOne CDx") %>%
        dplyr::mutate(CH_plus = case_when(
          #DDR_TOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count == 1 ~ "1",
          #DDR_TOCH_count >= 1 ~ "1~",
          #DDR_LOCH_count >= 1 ~ "1~",
          #Epigenetic_TOCH_count >= 1 ~ "1~",
          #Other_TOCH_count >= 1 ~ "1~",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 2 ~ "2~",
          # DDR_TOCH_count + Epigenetic_TOCH_count == 1 ~ "1",
          # DDR_TOCH_count + Epigenetic_TOCH_count >= 1 ~ "1~",
          #DDR_TOCH_count + DDR_LOCH_count >= 2 ~ "2~",
          #DDR_TOCH_count + DDR_LOCH_count == 1 ~ "1",
          #DDR_TOCH_count + DDR_LOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count >= 1 ~ "1~",
          #Histone_TOCH_count >= 1 ~ "1~",
          #DNA_methyl_TOCH_count + DDR_TOCH_count >= 1 ~ "1~",
          #TOCH_count >= 2 ~ "2~",
          #TOCH_count == 1 ~ "1",
          TOCH_count >= 1 ~ "1~",
          #CH_count >= 2 ~ "2~",
          #CH_count == 1 ~ "1",
          #CH_count >= 1 ~ "1~",
          #Infiltraive_genes == 1 ~ "1",
          #DNMT3A_R882_count >= 1 ~ "1~",
          #DNMT3A_Y735_count >= 1 ~ "1~",
          #DNMT3A_count >= 1 ~ "1~",
          #DNMT3A_count + DNMT3A_R882_count +DNMT3A_Y735_count  >= 1 ~ "1~",
          #CHEK2_count >= 1 ~ "1~",
          #ATM_count >= 1 ~ "1~",
          #TP53_count >= 1 ~ "1~",
          #ASXL1_count >= 1 ~ "1~",
          #TET2_count >= 1 ~ "1~",
          TRUE ~ "0"
        ))
      Data_ICI_ToT = Data_ICI_ToT %>%
        dplyr::mutate(Cancername = case_when(
          #Cancername %in% Diseases[!Diseases %in% c("THYMUS")] ~ Cancername,
          Cancername %in% Diseases[!Diseases %in% c("THYMUS", "OVARY", "BRAIN", "KIDNEY", "CERVIX", "LIVER", "THYROID", "BREAST")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_ICI_ToT$Cancername = factor(Data_ICI_ToT$Cancername)
      Data_ICI_ToT$Cancername = relevel(Data_ICI_ToT$Cancername, ref = "LUNG")
      survfit_t <- survfit(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus,
                     data=Data_ICI_ToT)
      plots[[349]] = 
        surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, Taxane, F1-liquid, any TICH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      #diff_2 = coxph(Surv(CTx_time_on_treatment, CTx_censor)~CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CTx_line", term) ~ gsub("CTx_line", "CTx line = ", term),
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[350]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Time on Treatment", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, any TICH, Taxane") +
        theme_minimal() +
        theme(legend.position = "none")
      survfit_t <- survfit(Surv(Overall_survival, censor)~ CH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      diff_0 = survdiff(Surv(Overall_survival, censor)~ CH_plus,
                        data=Data_ICI_ToT, rho=0)
      diff_1 = survdiff(Surv(Overall_survival, censor)~ CH_plus,
                        data=Data_ICI_ToT, rho=1)
      diff_2 = coxph(Surv(Overall_survival, censor)~ CH_plus,
                     data=Data_ICI_ToT)
      plots[[353]] = 
        surv_curv_ICI(survfit_t, Data_ICI_ToT, paste0("Overall survival, Taxane, F1-liquid, CH (+)"), diff_0, diff_1, diff_2)
      diff_2 = coxph(Surv(Overall_survival, censor)~ CH_plus + CTx_line + Cancername + TMB_class + Age_decade + Sex + regimen, data=Data_ICI_ToT)
      # 2. HR・CI を整形
      tidy_HR <- tidy(diff_2, conf.int = TRUE, exponentiate = TRUE) %>%
        filter(term != "(Intercept)") %>%
        mutate(term = case_when(
          grepl("^CH_plus", term) ~ gsub("CH_plus", "CH count = ", term),
          grepl("^Cancername", term) ~ gsub("Cancername", "Cancer: ", term),
          grepl("^Sex", term) ~ gsub("Sex", "Sex: ", term),
          grepl("^TMB_class", term) ~ gsub("TMB_classTMB", "TMB: ", term),
          grepl("^regimen", term) ~ gsub("regimen", "Drug: ", term),
          grepl("^Age_decade", term) ~ gsub("Age_decade", "Age: ", term),
          TRUE ~ term
        )) %>%
        mutate(term = factor(term, levels = rev(term)))  # 並び順の固定
      # データ結合・整形
      tidy_HR_all <- bind_rows(tidy_HR, refs) %>%
        distinct(term, .keep_all = TRUE) %>%
        mutate(
          term = factor(term, levels = rev(sort(unique(term)))),
          label = ifelse(is.na(p.value), "Reference", sprintf("HR = %.2f", estimate))
        )
      # プロット
      label_offset <- 0.4
      plots[[354]] <- ggplot(tidy_HR_all, aes(x = term, y = estimate)) +
        geom_point(aes(color = p.value < 0.05), size = 3) +
        scale_color_manual(values = c("FALSE" = "steelblue", "TRUE" = "firebrick")) +
        geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
        geom_text(aes(x = as.numeric(term) + label_offset, label = label), size = 3, hjust = 0) +
        geom_hline(yintercept = 1, linetype = "dashed", color = "gray40") +
        scale_y_log10() +
        coord_flip() +
        labs(title = "Hazard Ratios for Overall survival", x = "", y = "Hazard Ratio (95% CI, log scale)", color = "p < 0.05",
             subtitle = "F1-liquid, any CH, Taxane") +
        theme_minimal() +
        theme(legend.position = "none")
      output$table_inter_panel_TICH_2 = render_gt({tbl_TICH})
      output$table_inter_panel_TOCH_2 = render_gt({tbl_TOCH})
      output$table_inter_panel_DDR_TICH_2 = render_gt({tbl_DDR_TICH})
      output$table_inter_panel_DDR_TOCH_2 = render_gt({tbl_DDR_TOCH})
      output$table_inter_panel_Epi_TICH_2 = render_gt({tbl_Epi_TICH})
      output$table_inter_panel_Epi_TOCH_2 = render_gt({tbl_Epi_TOCH})
      output$table_inter_panel_DDR_SV_2 = render_gt({tbl_DDR_SV})
      output$table_inter_panel_DDR_LV_2 = render_gt({tbl_DDR_LV})
      output$table_inter_panel_Epi_SV_2 = render_gt({tbl_Epi_SV})
      output$table_inter_panel_Epi_LV_2 = render_gt({tbl_Epi_LV})
      # Data_ICI_ToT = Data_summary %>%
      #   dplyr::filter(Panel == "FoundationOne CDx")
      # survfit_t <- survfit(Surv(time_on_treatment, ICI_censor)~LOCH_plus, data=Data_ICI_ToT,type = "kaplan-meier", conf.type = "log-log")
      # diff_0 = survdiff(Surv(time_on_treatment, ICI_censor)~LOCH_plus,
      #                   data=Data_ICI_ToT, rho=0)
      # diff_1 = survdiff(Surv(time_on_treatment, ICI_censor)~LOCH_plus,
      #                   data=Data_ICI_ToT, rho=1)
      # diff_2 = coxph(Surv(time_on_treatment, ICI_censor)~LOCH_plus,
      #                data=Data_ICI_ToT)
      # plots[[171]] = surv_curv_drug(survfit_t, Data_ICI_ToT, paste0("Time on treatment, ICI, F1-solid, LOCH"), diff_0, diff_1, diff_2)
      print(1)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      plots[[172]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                              c("TOCH/LOCH (-)",
                                "TOCH/LOCH (+)"), diff_0)
      Data_survival_CGP = Data_survival_CGP %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$CH_plus = factor(Data_survival_CGP$CH_plus)
      surv_obj <- Surv(time = Data_survival_CGP$time_palliative_final,
                       event = Data_survival_CGP$censor)
      form <- as.formula("surv_obj ~ CH_plus + Age + Cancername + Sex + Platinum + Smoking_history + Alcoholic_history + Double_cancer + Multiple_nodules")
      fit <- coxph(form, data = Data_survival_CGP)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "Sex", "Platinum", "Smoking_history", "Alcoholic_history", "Double_cancer", "Multiple_nodules")
      setProgress(detail = "15171")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "BILIARY_TRACT",
        Sex = "Male",
        Platinum = "No",
        Smoking_history = "No",
        Alcoholic_history = "No",
        Double_cancer = "No",
        Multiple_nodules = "No"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_survival_CGP[[var]]))
        counts <- as.numeric(table(Data_survival_CGP[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      # ラベルの作成
      df_all["Age",] <- data.frame(
        Variable = "Age",
        Label = "1-Yr increase in age",
        N = length(Data_survival_CGP[["Age"]]),
        HR = df$HR[df$Variable == "Age"],
        Lower = df$Lower[df$Variable == "Age"],
        Upper = df$Upper[df$Variable == "Age"],
        P_value = df$P_value[df$Variable == "Age"],
        Star = FALSE,
        stringsAsFactors = FALSE
      )
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.2", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TOCH/LOCH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                                mean = df_plot$HR,
                                lower = df_plot$Lower,
                                upper = df_plot$Upper,
                                zero = 1,
                                xlog = TRUE,
                                col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                                xlab = "Hazard Ratio",
                                title = "Cox Proportional Hazards Model, F1-L after 1L CTx",
                                is.summary = df_plot$Label %in% unique_subgroups) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "black", cex = 3),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = 3)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[188]] <- ggdraw(ggplot_plot)
      
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~LOCH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~LOCH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~LOCH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~LOCH_plus,
                     data=Data_survival_CGP)
      plots[[173]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("LOCH (-)",
                                     "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "FoundationOne CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      plots[[174]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1 CDx after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("TI-CH (-)",
                                     "TI-CH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[175]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("LD-CH/LE-CH (-)",
                                     "LE-CH (+)",
                                     "LD-CH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 2) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "THYMUS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      plots[[207]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 2L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TOCH/LOCH (-)",
                                           "TOCH/LOCH (+)"), diff_0)
      Data_survival_CGP$CH_plus = factor(Data_survival_CGP$CH_plus)
      surv_obj <- Surv(time = Data_survival_CGP$time_palliative_final,
                       event = Data_survival_CGP$censor)
      form <- as.formula("surv_obj ~ CH_plus + Age + Cancername + Sex + Platinum + Smoking_history + Alcoholic_history + Double_cancer + Multiple_nodules")
      fit <- coxph(form, data = Data_survival_CGP)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "Sex", "Platinum", "Smoking_history", "Alcoholic_history", "Double_cancer", "Multiple_nodules")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "BILIARY_TRACT",
        Sex = "Male",
        Platinum = "No",
        Smoking_history = "No",
        Alcoholic_history = "No",
        Double_cancer = "No",
        Multiple_nodules = "No"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_survival_CGP[[var]]))
        counts <- as.numeric(table(Data_survival_CGP[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      # ラベルの作成
      df_all["Age",] <- data.frame(
        Variable = "Age",
        Label = "1-Yr increase in age",
        N = length(Data_survival_CGP[["Age"]]),
        HR = df$HR[df$Variable == "Age"],
        Lower = df$Lower[df$Variable == "Age"],
        Upper = df$Upper[df$Variable == "Age"],
        P_value = df$P_value[df$Variable == "Age"],
        Star = FALSE,
        stringsAsFactors = FALSE
      )
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.2", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TOCH/LOCH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model, F1-L after 2L CTx",
                               is.summary = df_plot$Label %in% unique_subgroups) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "black", cex = 3),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = 3)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[211]] <- ggdraw(ggplot_plot)
      
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 2) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "THYMUS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~LOCH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~LOCH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~LOCH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~LOCH_plus,
                     data=Data_survival_CGP)
      setProgress(detail = "15547")
      plots[[208]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 2L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("LOCH (-)",
                                           "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "FoundationOne CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 2) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "THYMUS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      plots[[209]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1-solid after 2L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TICH (-)",
                                           "TICH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 2) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "THYMUS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[210]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 2L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TOCH/LOCH (-)",
                                           "TOCH (+)",
                                           "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & Cancername == "LUNG"  & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[171]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, LUNG, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("TOCH/LOCH (-)",
                                     "TOCH (+)",
                                     "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & Cancername == "BILIARY_TRACT" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[200]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, BILIARY_TRACT, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("TOCH/LOCH (-)",
                                     "TOCH (+)",
                                     "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & Cancername == "BOWEL" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[201]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, BOWEL, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TOCH/LOCH (-)",
                                           "TOCH (+)",
                                           "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & Cancername == "BREAST" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[202]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, BREAST, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TOCH/LOCH (-)",
                                           "TOCH (+)",
                                           "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & Cancername == "STOMACH" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[203]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, STOMACH, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TOCH/LOCH (-)",
                                           "TOCH (+)",
                                           "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & Cancername == "PANCREAS" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[204]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, PANCREAS, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TOCH/LOCH (-)",
                                           "TOCH (+)",
                                           "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & Cancername == "PROSTATE" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      # survival_simple = survfit(Surv(time = time_palliative_enroll,
      #                                time2 = time_palliative_final, censor)~CH_plus,
      #                           data=Data_survival_CGP)
      # diff_0 = coxph(Surv(time = time_palliative_enroll,
      #                     time2 = time_palliative_final, censor)~CH_plus,
      #                data=Data_survival_CGP)
      survival_simple = survfit(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[205]] = surv_curv_CTx_naive(survival_simple, Data_survival_CGP, paste("Overall survival, PROSTATE, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                                   format(independence_check$PE,digits = 3)),
                                         c("TOCH/LOCH (-)",
                                           "TOCH (+)",
                                           "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_enroll,
                          time2 = time_palliative_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      setProgress(detail = "15850")
      plots[[180]] = surv_curv_CTx(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("TOCH/LOCH (-)",
                                     "TOCH/LOCH (+)"), diff_0)
      Data_survival_CGP = Data_survival_CGP %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$CH_plus = factor(Data_survival_CGP$CH_plus)
      surv_obj <- Surv(time = Data_survival_CGP$time_palliative_enroll,
                       time2 = Data_survival_CGP$time_palliative_final,
                       event = Data_survival_CGP$censor)
      form <- as.formula("surv_obj ~ CH_plus + Age + Cancername + Sex + Platinum + Smoking_history + Alcoholic_history + Double_cancer + Multiple_nodules")
      fit <- coxph(form, data = Data_survival_CGP)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "Sex", "Platinum", "Smoking_history", "Alcoholic_history", "Double_cancer", "Multiple_nodules")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "BILIARY_TRACT",
        Sex = "Female",
        Platinum = "No",
        Smoking_history = "No",
        Alcoholic_history = "No",
        Double_cancer = "No",
        Multiple_nodules = "No"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_survival_CGP[[var]]))
        counts <- as.numeric(table(Data_survival_CGP[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      # ラベルの作成
      df_all["Age",] <- data.frame(
        Variable = "Age",
        Label = "1-Yr increase in age",
        N = length(Data_survival_CGP[["Age"]]),
        HR = df$HR[df$Variable == "Age"],
        Lower = df$Lower[df$Variable == "Age"],
        Upper = df$Upper[df$Variable == "Age"],
        P_value = df$P_value[df$Variable == "Age"],
        Star = FALSE,
        stringsAsFactors = FALSE
      )
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.2", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TOCH/LOCH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                 mean = df_plot$HR,
                 lower = df_plot$Lower,
                 upper = df_plot$Upper,
                 zero = 1,
                 xlog = TRUE,
                 col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                 xlab = "Hazard Ratio",
                 title = "Cox Proportional Hazards Model, F1-L after 1L CTx, risk-set adjusted",
                 is.summary = df_plot$Label %in% unique_subgroups) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "black", cex = 3),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = 3)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[189]] <- ggdraw(ggplot_plot)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~LOCH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_enroll,
                          time2 = time_palliative_final, censor)~LOCH_plus,
                     data=Data_survival_CGP)
      plots[[181]] = surv_curv_CTx(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("TOCH/LOCH (-)",
                                     "TOCH/LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "FoundationOne CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_enroll,
                          time2 = time_palliative_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      plots[[182]] = surv_curv_CTx(survival_simple, Data_survival_CGP, paste("Overall survival, F1-solid after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("TICH (-)",
                                     "TICH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_palliative_final) & !is.na(time_enroll_final) & Lines_pre_CGP == 1) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      independence_check = with(Data_survival_CGP, cKendall(
        trun = time_palliative_enroll,
        obs = time_enroll_final,
        delta = censor,
        trans = "linear"))
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_palliative_enroll,
                          time2 = time_palliative_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[183]] = surv_curv_CTx(survival_simple, Data_survival_CGP, paste("Overall survival, F1-L after 1L CTx,", paste(survival_simple[[1]],collapse = "/"),"patients, cKendall tau=",
                                                                             format(independence_check$PE,digits = 3)),
                                   c("TOCH/LOCH (-)",
                                     "TOCH (+)",
                                     "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_enroll_final)) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      survival_simple = survfit(Surv(time = time_enroll_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_enroll_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      plots[[184]] = surv_curv_CGP(survival_simple, Data_survival_CGP, paste("Survival after CGP, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients"),
                                   c("TOCH/LOCH (-)",
                                     "TOCH/LOCH (+)"), diff_0)
      Data_survival_CGP$Lines_pre_CGP = as.character(Data_survival_CGP$Lines_pre_CGP)
      Data_survival_CGP = Data_survival_CGP %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS", "OVARY")] ~ Cancername,
          TRUE ~ "OTHER"
        ),
        Lines_pre_CGP = case_when(
          Lines_pre_CGP %in% c("0","1","2","3","4") ~ Lines_pre_CGP,
          TRUE ~ "5~"
        ))
      Data_survival_CGP$CH_plus = factor(Data_survival_CGP$CH_plus)
      surv_obj <- Surv(time = Data_survival_CGP$time_enroll_final,
                       event = Data_survival_CGP$censor)
      form <- as.formula("surv_obj ~ CH_plus + Age + Cancername + Sex + Lines_pre_CGP + Platinum + Smoking_history + Alcoholic_history + Double_cancer + Multiple_nodules")
      fit <- coxph(form, data = Data_survival_CGP)
      summary_fit <- summary(fit)
      # ハザード比と信頼区間の抽出
      hr <- summary_fit$coefficients[, "exp(coef)"]
      lower <- summary_fit$conf.int[, "lower .95"]
      upper <- summary_fit$conf.int[, "upper .95"]
      p_values <- summary_fit$coefficients[, "Pr(>|z|)"]
      # 変数名の取得
      variables <- rownames(summary_fit$coefficients)
      # データフレームの作成
      df <- data.frame(Variable = variables,
                       HR = hr,
                       Lower = lower,
                       Upper = upper,
                       P_value = p_values,
                       stringsAsFactors = FALSE)
      subgroups <- c("CH_plus", "Cancername", "Sex", "Lines_pre_CGP", "Platinum", "Smoking_history", "Alcoholic_history", "Double_cancer", "Multiple_nodules")
      reference_levels <- list(
        CH_plus = "0",
        Cancername = "BILIARY_TRACT",
        Sex = "Female",
        Lines_pre_CGP = "0",
        Platinum = "No",
        Smoking_history = "No",
        Alcoholic_history = "No",
        Double_cancer = "No",
        Multiple_nodules = "No"
      )
      df_list <- list()
      for (var in subgroups) {
        levels_var <- levels(factor(Data_survival_CGP[[var]]))
        counts <- as.numeric(table(Data_survival_CGP[[var]]))
        ref_level <- reference_levels[[var]]
        df_sub <- data.frame(
          Subgroup = var,
          Level = levels_var,
          Count = counts,
          stringsAsFactors = FALSE
        )
        df_sub$HR <- NA
        df_sub$Lower <- NA
        df_sub$Upper <- NA
        df_sub$P_value <- NA
        df_sub$Reference <- FALSE
        df_sub$Reference[df_sub$Level == ref_level] <- TRUE
        for (i in 1:nrow(df_sub)) {
          var_name <- paste0(var, df_sub$Level[i])
          if (var_name %in% df$Variable) {
            df_sub$HR[i] <- df$HR[df$Variable == var_name]
            df_sub$Lower[i] <- df$Lower[df$Variable == var_name]
            df_sub$Upper[i] <- df$Upper[df$Variable == var_name]
            df_sub$P_value[i] <- df$P_value[df$Variable == var_name]
          }
        }
        df_list[[var]] <- df_sub
      }
      df_all <- do.call(rbind, df_list)
      # ラベルの作成
      df_all["Age",] <- data.frame(
        Variable = "Age",
        Label = "1-Yr increase in age",
        N = length(Data_survival_CGP[["Age"]]),
        HR = df$HR[df$Variable == "Age"],
        Lower = df$Lower[df$Variable == "Age"],
        Upper = df$Upper[df$Variable == "Age"],
        P_value = df$P_value[df$Variable == "Age"],
        Star = FALSE,
        stringsAsFactors = FALSE
      )
      df_all$Label <- ifelse(df_all$Reference, paste0(df_all$Level, " (Reference)"), df_all$Level)
      df_all$HR_CI <- ifelse(is.na(df_all$HR), "", sprintf("%.2f (%.2f, %.2f)", df_all$HR, df_all$Lower, df_all$Upper))
      df_all$P_value_str <- ifelse(is.na(df_all$P_value), "", sprintf("%.3f", df_all$P_value))
      df_all["CH_plus.1", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.2", "Subgroup"] = "TOCH/LOCH detection"
      df_all["CH_plus.1", "Label"] = "No (Reference)"
      df_all["CH_plus.2", "Label"] = "Yes"
      # 見出し行の挿入
      unique_subgroups <- unique(df_all$Subgroup)
      unique_subgroups[unique_subgroups == "CH_plus"] = "TOCH/LOCH detection"
      header_rows <- data.frame(
        Subgroup = unique_subgroups,
        Label = unique_subgroups,
        Count = NA,
        HR_CI = NA,
        P_value_str = "",
        stringsAsFactors = FALSE
      )
      # 見出し行とデータの結合
      df_plot <- data.frame()
      for (subgroup in unique_subgroups) {
        header <- header_rows[header_rows$Subgroup == subgroup, ]
        data_sub <- df_all[df_all$Subgroup == subgroup, ]
        df_plot <- bind_rows(df_plot, header, data_sub,)
      }
      # ラベルテキストの作成
      labeltext <- cbind(df_plot$Label, df_plot$Count, df_plot$HR_CI, df_plot$P_value_str)
      # フォレストプロットの作成
      forest_plot = forestplot(labeltext = labeltext,
                               mean = df_plot$HR,
                               lower = df_plot$Lower,
                               upper = df_plot$Upper,
                               zero = 1,
                               xlog = TRUE,
                               col = fpColors(box = "royalblue", line = "darkblue", summary = "royalblue"),
                               xlab = "Hazard Ratio",
                               title = "Cox Proportional Hazards Model, F1-L, survival after CGP",
                               is.summary = df_plot$Label %in% unique_subgroups) %>%
        fp_set_style(
          box = gpar(fill = "black", col = "black", cex = 3),
          line = "black",
          summary = "black",
          align = "lccccc",
          hrz_lines = "black",
          txt_gp = fpTxtGp(ticks = gpar(cex = 1, box = gpar(cex = 3)))
        ) %>%
        fp_add_header(
          "Subgroup",
          "N",
          "HR (95%CI)",
          "P-value"
        ) %>%
        fp_decorate_graph(graph.pos = 5) %>%
        fp_set_zebra_style("#EFEFEF", ignore_subheaders = TRUE)
      grob_plot <- grid.grabExpr(print(forest_plot))
      # キャプチャした grob を ggplot2 の枠組みに組み込む
      ggplot_plot <- ggplot() +
        annotation_custom(grob_plot) +
        theme_void() # 必要に応じてテーマを調整
      # これで ggdraw で ggplot_plot を使用できるはずです
      plots[[206]] <- ggdraw(ggplot_plot)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_enroll_final)) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      survival_simple = survfit(Surv(time = time_enroll_final, censor)~LOCH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_enroll_final, censor)~LOCH_plus,
                     data=Data_survival_CGP)
      plots[[185]] = surv_curv_CGP(survival_simple, Data_survival_CGP, paste("Survival after CGP, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients"),
                                   c("LOCH (-)",
                                     "LOCH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "FoundationOne CDx" & !is.na(time_enroll_final)) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      survival_simple = survfit(Surv(time = time_enroll_final, censor)~CH_plus,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_enroll_final, censor)~CH_plus,
                     data=Data_survival_CGP)
      plots[[186]] = surv_curv_CGP(survival_simple, Data_survival_CGP, paste("Survival after CGP, F1-solid,", paste(survival_simple[[1]],collapse = "/"),"patients"),
                                   c("TICH (-)",
                                     "TICH (+)"), diff_0)
      Data_survival_CGP = Data_summary %>%
        dplyr::filter(Panel == "F1Liquid CDx" & !is.na(time_enroll_final)) %>%
        dplyr::mutate(TOCH_LOCH = case_when(
          LOCH_plus == 1 ~ "LOCH",
          CH_plus == 1 ~ "TOCH",
          TRUE ~ "No CH"
        )) %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases[!Diseases %in% c("EYE", "PNS")] ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_survival_CGP$TOCH_LOCH = factor(Data_survival_CGP$TOCH_LOCH, levels = c("No CH", "TOCH", "LOCH"))
      survival_simple = survfit(Surv(time = time_enroll_final, censor)~TOCH_LOCH,
                                data=Data_survival_CGP)
      diff_0 = coxph(Surv(time = time_enroll_final, censor)~TOCH_LOCH,
                     data=Data_survival_CGP)
      plots[[187]] = surv_curv_CGP(survival_simple, Data_survival_CGP, paste("Survival after CGP, F1-liquid,", paste(survival_simple[[1]],collapse = "/"),"patients"),
                                   c("TOCH/LOCH (-)",
                                     "TOCH (+)",
                                     "LOCH (+)"), diff_0)
      
      print(1)
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(Panel %in% c("F1Liquid CDx", "FoundationOne CDx"))
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::mutate(group = case_when(
          unique_alt %in% CH_candidates_base &
            Hugo_Symbol %in% CH_gene_entropy_3 & logfold > 8 ~ "LD-CH",
          unique_alt %in% CH_candidates_base &
            Hugo_Symbol %in% CH_gene_entropy_3 & logfold > 2 & Panel == "F1Liquid CDx" ~ "LE-CH",
          unique_alt %in% CH_candidates_base &
            Hugo_Symbol %in% CH_gene_entropy_3 & logfold > 2 & Panel == "FoundationOne CDx" ~ "TI-CH",
          Panel == "F1Liquid CDx" ~ "SEV, F1-Liquid",
          Panel == "FoundationOne CDx" ~ "SEV, F1",
          TRUE ~ "OTHER"
        ))
      Data_tmp$group = factor(Data_tmp$group,
                              levels = c("LD-CH", "LE-CH", "SEV, F1-Liquid","TI-CH", "SEV, F1"))
      my_comparisons <- list(
        c("LD-CH", "LE-CH"),
        c("LD-CH", "SEV, F1-Liquid"),
        c("LE-CH", "SEV, F1-Liquid"),
        c("TI-CH", "SEV, F1")
      )
      # 各群の平均と標準偏差を計算
      label_df <- Data_tmp %>%
        group_by(group) %>%
        summarise(
          mean_VAF = mean(VAF, na.rm = TRUE),
          sd_VAF = sd(VAF, na.rm = TRUE),
          y_pos = max(VAF, na.rm = TRUE) * 1.05  # 上に表示するためのy座標
        ) %>%
        mutate(label = sprintf("%.3f ± %.3f", mean_VAF, sd_VAF))
      # プロット
      plots[[176]] = ggplot(Data_tmp, aes(x = group, y = VAF, color = group)) +
        geom_boxplot(width = 0.1) +
        stat_compare_means(comparisons = my_comparisons, 
                           method = "t.test", 
                           label = "p.signif", 
                           tip.length = 0.01) +
        # stat_compare_means(method = "anova", 
        #                    label.y = max(Data_tmp$VAF, na.rm = TRUE) * 1.15) +  # 全体p値
        geom_text(data = label_df, aes(x = group, y = y_pos, label = label), 
                  inherit.aes = FALSE, vjust = -0.5, size = 3.5) +
        labs(title = "VAF by mutation type, mean ± SD", 
             x = "Mutation type", y = "Variant allele frequency") +
        theme_classic2() +
        theme(legend.position = "none") 
      target_genes <- c("DNMT3A", "TET2", "ASXL1", "CHEK2", "ATM")
      Data_tmp <- Data_tmp %>%
        mutate(gene_group = ifelse(Hugo_Symbol %in% target_genes, Hugo_Symbol, "Other"))
      Data_tmp$gene_group = factor(Data_tmp$gene_group,
                                   levels = c(target_genes, "Other"))
      # 平均 ± SD ラベルデータを計算
      label_df <- Data_tmp %>%
        group_by(group, gene_group) %>%
        summarise(
          mean_VAF = mean(VAF, na.rm = TRUE),
          sd_VAF = sd(VAF, na.rm = TRUE),
          y_pos = max(VAF, na.rm = TRUE) * 1.05,
          .groups = "drop"
        ) %>%
        mutate(label = sprintf("%.3f ± %.3f", mean_VAF, sd_VAF))
      # プロット
      plots[[177]] = ggplot(Data_tmp, aes(x = group, y = VAF, color = group)) +
        geom_boxplot(width = 0.1) +
        stat_compare_means(comparisons = my_comparisons,
                           method = "t.test",
                           label = "p.signif",
                           tip.length = 0.01) +
        geom_text(data = label_df,
                  aes(x = group, y = y_pos, label = label),
                  inherit.aes = FALSE,
                  vjust = -0.5, size = 3.2) +
        facet_wrap(~ gene_group) +
        labs(title = "VAF by mutation type and gene with significance and mean ± SD", x = "Mutation type", y = "VAF") +
        theme_minimal() +
        theme(legend.position = "none")


      # output$table_inter_panel4 = DT::renderDataTable(Data_MAF_target,
      #                                                 server = FALSE,
      #                                                 filter = 'top', 
      #                                                 extensions = c('Buttons'), 
      #                                                 options = list(pageLength = 100, 
      #                                                                scrollX = TRUE,
      #                                                                scrollY = "1000px",
      #                                                                scrollCollapse = TRUE,
      #                                                                dom="Blfrtip",
      #                                                                buttons = c('csv', 'copy')))
      Summary_GENIE = read_csv("source/GENIE_ALL.csv", show_col_types = FALSE)
      Summary_GENIE = Summary_GENIE %>%
        filter(Sample_type == "Blood_cancer")
      # Summary_GENIE = Summary_GENIE %>%
      #   filter(Age > 40 &
      #           Cancername %in% c(
      #             "B-Lymphoblastic Leukemia/Lymphoma",
      #             "T-Lymphoblastic Leukemia/Lymphoma",
      #             "Leukemia",
      #             "Myelodysplastic Syndromes",
      #             "Myelodysplastic/Myeloproliferative Neoplasms",
      #             "Myeloid Neoplasms with Germ Line Predisposition",
      #             "Myeloproliferative Neoplasms"))
      ID_Blood_cancer = unique(Summary_GENIE$Tumor_Sample_Barcode)
      if (!file.exists("source/GENIE_blood.tsv")) {
        library(rtracklayer)
        library(GenomicRanges)
        Data_GENIE_blood = read_tsv("source/GENIE_variants.txt", show_col_types = FALSE)
        Data_GENIE_blood = Data_GENIE_blood %>% dplyr::filter(Tumor_Sample_Barcode %in% ID_Blood_cancer)
        Data_GENIE_blood$Chromosome = paste0("chr", Data_GENIE_blood$Chromosome)
        Data_GENIE_blood$Chromosome[Data_GENIE_blood$Chromosome == "chrMT"]= "chrM"
        chain_file <- "source/hg19ToHg38.over.chain"
        if (!file.exists(chain_file)) {
          download.file("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz",
                        destfile = paste0(chain_file, ".gz"))
          system(paste0("gunzip -f ", chain_file, ".gz"))
        }
        gr <- GRanges(seqnames = Data_GENIE_blood$Chromosome,
                      ranges = IRanges(start = Data_GENIE_blood$Start_Position, end = Data_GENIE_blood$End_Position),
                      strand = "*")
        chain <- import.chain(chain_file)
        lifted_gr <- liftOver(gr, chain)
        tmp = lifted_gr@partitioning@end
        tmp0 = diff(tmp)
        tmp1 = tmp[tmp0 != 1]
        tmp2 = tmp1 + 1:length(tmp1)
        tmp3 = tmp[tmp0 == 0]
        tmp4 = tmp2[tmp1 %in% tmp3]
        Data_GENIE_blood = Data_GENIE_blood[-tmp4,]
        # chain <- import.chain(chain_file)
        # lifted_gr <- liftOver(gr, chain)
        # tmp = lifted_gr@partitioning@end
        # tmp0 = diff(tmp)
        # tmp1 = tmp[tmp0 != 1]
        # tmp2 = tmp1 - 1:length(tmp1) + 2
        # tmp3 = tmp1 + 2
        # lifted_gr <- unlist(lifted_gr)
        # lifted_gr = lifted_gr[-tmp3]
        # Data_GENIE_blood$Chromosome <- as.character(seqnames(lifted_gr))
        # Data_GENIE_blood$Start_Position <- start(lifted_gr)
        # Data_GENIE_blood$End_Position <- end(lifted_gr)
        # 座標は整数に（落ちにくくする：推奨）
        Data_GENIE_blood$Start_Position <- as.integer(Data_GENIE_blood$Start_Position)
        Data_GENIE_blood$End_Position   <- as.integer(Data_GENIE_blood$End_Position)
        gr <- GRanges(seqnames = Data_GENIE_blood$Chromosome,
                      ranges = IRanges(start = Data_GENIE_blood$Start_Position, end = Data_GENIE_blood$End_Position),
                      strand = "*")
        
        chain <- import.chain(chain_file)
        
        lifted_list <- liftOver(gr, chain)      # GRangesList
        nmap <- elementNROWS(lifted_list)       # 各入力が何件に変換されたか
        
        # 0件（liftOver失敗）を落とす
        keep <- which(nmap >= 1)
        Data_GENIE_blood <- Data_GENIE_blood[keep, , drop = FALSE]
        lifted_list <- lifted_list[keep]
        
        # 複数件は先頭1件だけ採用（= 1:1へ）
        lifted_1 <- endoapply(lifted_list, function(x) x[1])
        lifted_gr <- unlist(lifted_1, use.names = FALSE)  # GRanges（行数=Data_GENIE_blood）
        
        # 変換後座標を戻す
        Data_GENIE_blood$Chromosome <- as.character(seqnames(lifted_gr))
        Data_GENIE_blood$Start_Position <- start(lifted_gr)
        Data_GENIE_blood$End_Position <- end(lifted_gr)
        
        Data_GENIE_blood$HGVSp_Short[is.na(Data_GENIE_blood$HGVSp_Short)] = Data_GENIE_blood$HGVSc[is.na(Data_GENIE_blood$HGVSp_Short)]
        Data_GENIE_blood$HGVSp_Short[is.na(Data_GENIE_blood$HGVSp_Short)] = paste0(
          Data_GENIE_blood$Start_Position[is.na(Data_GENIE_blood$HGVSp_Short)],
          Data_GENIE_blood$Reference_Allele[is.na(Data_GENIE_blood$HGVSp_Short)],">",
          Data_GENIE_blood$Tumor_Seq_Allele2[is.na(Data_GENIE_blood$HGVSp_Short)]
        )
        Data_GENIE_blood$gnomAD_AF[is.na(Data_GENIE_blood$gnomAD_AF)] = 0
        Data_GENIE_blood = Data_GENIE_blood %>%
          dplyr::select(
            Hugo_Symbol,  Tumor_Sample_Barcode, HGVSp_Short, Variant_Classification, Chromosome, Start_Position, End_Position,
            Reference_Allele, Tumor_Seq_Allele2, t_alt_count, t_depth, gnomAD_AF) %>%
          dplyr::left_join(Summary_GENIE, by = "Tumor_Sample_Barcode")
        Data_GENIE_blood$VAF = Data_GENIE_blood$t_alt_count / Data_GENIE_blood$t_depth
        Data_GENIE_blood = Data_GENIE_blood[!is.na(Data_GENIE_blood$VAF),]
        Data_GENIE_blood = Data_GENIE_blood[Data_GENIE_blood$VAF != Inf,]
        Data_GENIE_blood = Data_GENIE_blood[Data_GENIE_blood$VAF != -Inf,]
        Data_GENIE_blood$HGVSp_Short = str_remove(Data_GENIE_blood$HGVSp_Short, "p.")
        Data_GENIE_blood = Data_GENIE_blood %>%
          dplyr::select(-t_alt_count, -t_depth)
        colname_data = colnames(Data_GENIE_blood)
        colname_data[colname_data == "HGVSp_Short"] = "amino.acid.change"
        colname_data[colname_data == "gnomAD_AF"] = "MinorAF"
        colnames(Data_GENIE_blood) = colname_data
        write_tsv(Data_GENIE_blood, "source/GENIE_blood.tsv")
      } else {
        Data_GENIE_blood = read_tsv("source/GENIE_blood.tsv", show_col_types = FALSE)
      }
      Data_GENIE_blood$Chromosome = str_remove(Data_GENIE_blood$Chromosome, "chr")
      Data_GENIE_blood$unique_position = paste0(Data_GENIE_blood$Chromosome, "_",
                                                Data_GENIE_blood$Start_Position, "_",
                                                Data_GENIE_blood$Reference_Allele, "_",
                                                Data_GENIE_blood$Tumor_Seq_Allele2
      )
      Data_GENIE_tmp = Data_MAF_target %>% 
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error <= 1) %>%
        dplyr::mutate(CH_hotspot_candidate = case_when(
          CH_hotspot_candidate == 1 & (max_VAF < 0.02 & !is.na(max_VAF)) ~ 2,
          CH_hotspot_candidate == 1 & !Hugo_Symbol %in% CH_gene_entropy_3 ~ 2,
          TRUE ~ CH_hotspot_candidate
        )) %>%
        dplyr::select(
          Hugo_Symbol, Tumor_Sample_Barcode, Cancername, amino.acid.change, Chromosome,
          Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, VAF, Age, unique_position,
          CH_hotspot_candidate, unique_alt, MinorAF
        )
      Data_GENIE_tmp$End_Position = as.integer(Data_GENIE_tmp$End_Position)
      Data_GENIE_tmp = Data_GENIE_tmp %>%
        dplyr::mutate(
          Start_Position = case_when(
            nchar(Reference_Allele) > 1 & nchar(Tumor_Seq_Allele2) == 1 ~ Start_Position + 1,
            TRUE ~ Start_Position
          ))
      # Data_GENIE_tmp = Data_GENIE_tmp %>%
      #   dplyr::mutate(
      #     End_Position = case_when(
      #       nchar(Reference_Allele) > 1 & nchar(Tumor_Seq_Allele2) == 1 ~ Start_Position + nchar(Reference_Allele) - 1,
      #       TRUE ~ End_Position
      #     ))
      setProgress(detail = "16473")
      Data_GENIE_tmp = Data_GENIE_tmp %>%
        dplyr::mutate(
          Tumor_Seq_Allele2 = case_when(
            nchar(Reference_Allele) > 1 & nchar(Tumor_Seq_Allele2) == 1 ~ "-",
            TRUE ~ Tumor_Seq_Allele2
          ))
      Data_GENIE_tmp = Data_GENIE_tmp %>%
        dplyr::mutate(
          Reference_Allele = case_when(
            nchar(Reference_Allele) > 1 & Tumor_Seq_Allele2 == "-" ~ str_sub(Reference_Allele, 2),
            TRUE ~ Reference_Allele
          ))
      Data_GENIE_tmp = Data_GENIE_tmp %>%
        dplyr::mutate(
          Reference_Allele = case_when(
            nchar(Reference_Allele) == 1 & nchar(Tumor_Seq_Allele2) > 1 ~ "-",
            TRUE ~ Reference_Allele
          ))
      Data_GENIE_tmp = Data_GENIE_tmp %>%
        dplyr::mutate(
          Tumor_Seq_Allele2 = case_when(
            Reference_Allele == "-" & nchar(Tumor_Seq_Allele2) > 1 ~ str_sub(Tumor_Seq_Allele2, 2),
            TRUE ~ Tumor_Seq_Allele2
          ))
      Data_GENIE_tmp$unique_position = paste0(Data_GENIE_tmp$Chromosome, "_",
                                              Data_GENIE_tmp$Start_Position, "_",
                                              Data_GENIE_tmp$Reference_Allele, "_",
                                              Data_GENIE_tmp$Tumor_Seq_Allele2
      )
      Data_GENIE_tmp$Sample_type = "CCAT_F1L"
      Data_GENIE_tmp$Variant_Classification = "SNV"
      Data_GENIE_blood$CH_hotspot_candidate = 0
      Data_GENIE_blood$CH_hotspot_candidate[Data_GENIE_blood$MinorAF >= 0.0001 & Data_GENIE_blood$VAF >= 0.25] = 2
      Data_GENIE_blood$CH_hotspot_candidate[!Data_GENIE_blood$Hugo_Symbol %in% CH_GENE_NatComms] = 2
      Data_GENIE_blood$unique_alt = ""
      
      # alt_replace = Data_GENIE_tmp %>% dplyr::select(unique_position, unique_alt) %>% dplyr::distinct(unique_position, .keep_all = T)
      # Data_GENIE_blood = Data_GENIE_blood %>%
      #   dplyr::left_join(alt_replace, by = "unique_position")
      # Data_GENIE_blood$unique_alt[is.na(Data_GENIE_blood$unique_alt)] = 
      #   paste0(Data_GENIE_blood$Hugo_Symbol[is.na(Data_GENIE_blood$unique_alt)], "_",
      #          Data_GENIE_blood$amino.acid.change[is.na(Data_GENIE_blood$unique_alt)])
      alt_replace = Data_GENIE_tmp %>%
        dplyr::select(unique_position, amino.acid.change) %>%
        dplyr::distinct(unique_position, .keep_all = T)
      colnames(alt_replace) = c("unique_position", "AAC")
      Data_GENIE_blood = Data_GENIE_blood %>%
        dplyr::left_join(alt_replace, by = "unique_position")
      Data_GENIE_blood$amino.acid.change[!is.na(Data_GENIE_blood$AAC)] = Data_GENIE_blood$AAC[!is.na(Data_GENIE_blood$AAC)]
      Data_GENIE_blood = Data_GENIE_blood %>% dplyr::select(-AAC)

      Data_GENIE_blood = Data_GENIE_blood %>%
        filter((!str_detect(amino.acid.change, "^\\*\\d+\\*") |
                  str_detect(amino.acid.change, "^\\*\\d+fs\\*")) &
                 !str_detect(amino.acid.change, "\\="))
      
      Data_GENIE = rbind(Data_GENIE_blood, Data_GENIE_tmp)
      Data_GENIE$name = paste0(Data_GENIE$Hugo_Symbol, "_", Data_GENIE$amino.acid.change)
      variants = sort(unique((Data_GENIE %>% filter(str_detect(name, "fs") & !str_detect(name, "_NOS")))$name))
      # 1文字の挿入・削除を検出する関数（末尾の数値は一致する必要あり）
      is_one_char_insert_or_delete <- function(shorter, longer) {
        # `fs*XX` の部分を抽出
        short_main <- str_extract(shorter, "^[^*]+")  # "ABCB11_K930Efs"
        short_suffix <- str_extract(shorter, "\\*\\d+$")  # "*79"
        long_main <- str_extract(longer, "^[^*]+")  # "ABCB11_K930Efs"
        long_suffix <- str_extract(longer, "\\*\\d+$")  # "*79"
        # 末尾の数値が一致しなければNG
        if (is.na(short_suffix) || is.na(long_suffix) || short_suffix != long_suffix) {
          return(FALSE)
        }
        # 長さチェック（1文字差でなければ除外）
        if (abs(nchar(short_main) - nchar(long_main)) != 1) {
          return(FALSE)
        }
        # 文字列を1文字ずつ比較
        short_chars <- strsplit(short_main, "")[[1]]
        long_chars  <- strsplit(long_main, "")[[1]]
        # どこかで1文字が余計に入っているか確認
        for (i in seq_along(short_chars)) {
          if (short_chars[i] != long_chars[i]) {
            return(identical(short_chars[i:length(short_chars)], long_chars[(i+1):length(long_chars)]))
          }
        }
        # 最後の1文字が追加されただけならOK
        return(TRUE)
      }
      pairs <- list()
      variant_lengths <- nchar(variants)
      i <- 1
      while (i < length(variants)) {
        if (nchar(variants[i]) < variants[i + 1]) {
          short <- variants[i]
          long <- variants[i + 1]
          if (is_one_char_insert_or_delete(short, long)) {
            pairs <- append(pairs, list(c(long, short)))  # 変更したペアを記録
          }
        } else if (nchar(variants[i]) > variants[i + 1]) {
          short <- variants[i + 1]
          long <- variants[i]
          if (is_one_char_insert_or_delete(short, long)) {
            pairs <- append(pairs, list(c(long, short)))  # 変更したペアを記録
          }
        }
        i <- i + 1
      }
      #unique_variants <- unique(apply(pairs, 1, function(x) paste(variants[x[1]], variants[x[2]], sep = " <-> ")))
      for (i in seq_len(length(pairs))) {
        Data_GENIE$name[Data_GENIE$name == pairs[[i]][2]] <- pairs[[i]][1]
        Data_GENIE$Start_Position[Data_GENIE$name == pairs[[i]][2]] <- Data_GENIE$Start_Position[Data_GENIE$name == pairs[[i]][1]][[1]]
        Data_GENIE$Reference_Allele[Data_GENIE$name == pairs[[i]][2]] <- Data_GENIE$Reference_Allele[Data_GENIE$name == pairs[[i]][1]][[1]]
        Data_GENIE$Tumor_Seq_Allele2[Data_GENIE$name == pairs[[i]][2]] <- Data_GENIE$Tumor_Seq_Allele2[Data_GENIE$name == pairs[[i]][1]][[1]]
        Data_GENIE$unique_position[Data_GENIE$name == pairs[[i]][2]] <- Data_GENIE$unique_position[Data_GENIE$name == pairs[[i]][1]][[1]]
        Data_GENIE$amino.acid.change[Data_GENIE$name == pairs[[i]][2]] <- Data_GENIE$amino.acid.change[Data_GENIE$name == pairs[[i]][1]][[1]]
      }
      Data_GENIE = Data_GENIE %>%
        dplyr::mutate(unique_alt = case_when(
          str_detect(amino.acid.change, "\\*") &
            str_sub(amino.acid.change, 1,1) != "*" &
            ((Reference_Allele) == "-" |
               (Tumor_Seq_Allele2) == "-") ~ paste0(Hugo_Symbol, "_fs_NOS"),
          str_detect(amino.acid.change, "\\*") &
            str_sub(amino.acid.change, 1,1) != "*" &
            Reference_Allele == "C" &
            Tumor_Seq_Allele2 == "T" ~ paste0(Hugo_Symbol, "_trunc_CtoT"),
          str_detect(amino.acid.change, "\\*") &
            str_sub(amino.acid.change, 1,1) != "*" ~ paste0(Hugo_Symbol, "_trunc_NOS"),
          TRUE ~ name
        ))
      Data_GENIE = Data_GENIE %>%
        dplyr::mutate(CtoT_mutation = case_when(
          Reference_Allele == "C" & Tumor_Seq_Allele2 == "T" |
            Reference_Allele == "G" & Tumor_Seq_Allele2 == "A" ~ "Yes",
          TRUE ~ "No"
        ))      
      table_GENIE_L = table((Data_GENIE %>% dplyr::filter(Sample_type == "Blood_cancer" & CH_hotspot_candidate <= 1))$unique_position)
      table_GENIE_S = table((Data_GENIE %>% dplyr::filter(Sample_type == "CCAT_F1L" & CH_hotspot_candidate <= 1))$unique_position)
      join_tmp_L = data.frame(
        unique_position = names(table_GENIE_L),
        No_found_in_F1L_position = as.numeric(table_GENIE_L)
      )
      Data_GENIE = Data_GENIE %>% dplyr::left_join(join_tmp_L, by = "unique_position")
      join_tmp_S = data.frame(
        unique_position = names(table_GENIE_S),
        No_found_in_F1_position = as.numeric(table_GENIE_S)
      )
      Data_GENIE = Data_GENIE %>% dplyr::left_join(join_tmp_S, by = "unique_position")
      Data_GENIE$No_found_in_F1L_position[is.na(Data_GENIE$No_found_in_F1L_position)] = 0
      Data_GENIE$No_found_in_F1_position[is.na(Data_GENIE$No_found_in_F1_position)] = 0
      Data_GENIE$VAF[is.na(Data_GENIE$VAF)] = 0

      Data_GENIE = Data_GENIE %>%
        dplyr::mutate(unique_alt = case_when(
          (No_found_in_F1L_position >= Min_count | No_found_in_F1_position >= Min_count) & !str_detect(name, "fs") & !str_detect(name, "trunc") ~ name,
          # (No_found_in_F1L_position >= Min_count | No_found_in_F1_position >= Min_count) ~ name,
          TRUE ~ unique_alt
        ))
      table_GENIE_L = table((Data_GENIE %>% dplyr::filter(Sample_type == "Blood_cancer" & CH_hotspot_candidate <= 1))$unique_alt)
      table_GENIE_S = table((Data_GENIE %>% dplyr::filter(Sample_type == "CCAT_F1L" & CH_hotspot_candidate <= 1))$unique_alt)
      join_tmp_L = data.frame(
        unique_alt = names(table_GENIE_L),
        No_found_in_F1L = as.numeric(table_GENIE_L)
      )
      Data_GENIE = Data_GENIE %>% dplyr::left_join(join_tmp_L, by = "unique_alt")
      join_tmp_S = data.frame(
        unique_alt = names(table_GENIE_S),
        No_found_in_F1 = as.numeric(table_GENIE_S)
      )
      Data_GENIE = Data_GENIE %>% dplyr::left_join(join_tmp_S, by = "unique_alt")
      Data_GENIE$No_found_in_F1L[is.na(Data_GENIE$No_found_in_F1L)] = 0
      Data_GENIE$No_found_in_F1[is.na(Data_GENIE$No_found_in_F1)] = 0
      Data_GENIE = Data_GENIE %>%
        dplyr::mutate(CH_hotspot_candidate = case_when(
          CH_hotspot_candidate == 0 &
            Sample_type == "Blood_cancer" &
            No_found_in_F1L >= Min_count ~ 1,
          TRUE ~ CH_hotspot_candidate
        ))
      Data_GENIE$F1L_patients = length((Summary_GENIE %>% dplyr::filter(Sample_type == "Blood_cancer"))$Age)
      Data_GENIE$F1_patients = length((Data_summary %>%
                                         dplyr::filter(Panel == "F1Liquid CDx") %>%
                                         dplyr::distinct(Tumor_Sample_Barcode))$Tumor_Sample_Barcode)
      Data_GENIE = Data_GENIE %>% dplyr::filter(Hugo_Symbol %in% sort(unique((Data_MAF %>% filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")))$Hugo_Symbol)))
      data_tmp = Data_GENIE %>%
        dplyr::filter(CH_hotspot_candidate == 1 &
                        Sample_type == "Blood_cancer")
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol %in% input$clonal_hematopoiesis_genes)
      Data_tmp$CH_genes = 1
      cor_A = Data_tmp %>% 
        summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
        as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
        geom_point(alpha = .5) +
        ggtitle("Hotspot variants in CH genes found in GENIE hematologic cohort") +
        ylim(c(0,1)) +
        xlim(c(0,100)) +
        theme_bw() +
        geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
        geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
        stat_poly_eq(formula = y ~ x,
                     aes(label = paste(stat(eq.label))),
                     parse = TRUE)
      plots[[48]] = g
      # Data_tmp = data_tmp %>% dplyr::filter(!Hugo_Symbol %in% input$clonal_hematopoiesis_genes)
      # Data_tmp$CH_genes = 0
      # cor_A = Data_tmp %>% 
      #   summarise(cor = round(cor.test(Age, VAF)$estimate,3), p = round(cor.test(Age, VAF)$p.value,3)) %>% 
      #   as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
      # g <- ggplot(Data_tmp, aes(x = Age, y = VAF, color = CH_genes)) +
      #   geom_point(alpha = .5) +
      #   ggtitle("Hotspot variants in non-CH genes found in GENIE hematologic cohort") +
      #   ylim(c(0,1)) +
      #   xlim(c(0,100)) +
      #   theme_bw() +
      #   geom_smooth(method = lm, se = FALSE, na.rm = T, linetype = "dashed", size = 0.5, colour = "red") +
      #   geom_text(data=cor_A, colour = "red", mapping = aes(x = 12, y = 0.94, label = paste("r =",cor, ", p =",p))) +
      #   stat_poly_eq(formula = y ~ x,
      #                aes(label = paste(stat(eq.label))),
      #                parse = TRUE)
      # plots[[49]] = g
      #plots[[49]] = gg_empty()
      data_tmp = Data_GENIE %>%
          dplyr::filter(CH_hotspot_candidate == 1) %>%
          dplyr::select(unique_position, unique_alt, name, Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, amino.acid.change, No_found_in_F1, No_found_in_F1L, No_found_in_F1_position, No_found_in_F1L_position, F1_patients, F1L_patients, unique_position) %>%
          dplyr::distinct(Hugo_Symbol, Chromosome, Start_Position, Reference_Allele, Tumor_Seq_Allele2, .keep_all = T) %>%
          dplyr::arrange(Chromosome, Start_Position)
      data_tmp$no_mut_F1 = data_tmp$F1_patients - data_tmp$No_found_in_F1
      data_tmp$no_mut_F1L = data_tmp$F1L_patients - data_tmp$No_found_in_F1L
      data_tmp_fisher = data_tmp %>% dplyr::select(No_found_in_F1L,No_found_in_F1,no_mut_F1L,no_mut_F1)
      data_tmp$p_value = 1
      data_tmp$odds_ratio = 1
      data_tmp$triplet_pre = ""
      data_tmp$triplet_post = ""
      data_tmp$CpG_TpG = 0
      data_tmp$CpC_TpC = 0
      data_tmp$SNV_check = ifelse(data_tmp$Reference_Allele == "-" |
                                  data_tmp$Tumor_Seq_Allele2 == "-" |
                                  nchar(data_tmp$Reference_Allele) > 1 |
                                  nchar(data_tmp$Tumor_Seq_Allele2)> 1, 2, 1)
      Name_list = NULL
      triplet_pre = DNAStringSet(rep("NNN",length(data_tmp$Hugo_Symbol)))
      triplet_post = DNAStringSet(rep("NNN",length(data_tmp$Hugo_Symbol)))
      Summary_GENIE$favor_liquid = 1
      Data_summary_tmp = Summary_GENIE %>%
        dplyr::select(Tumor_Sample_Barcode, Age, favor_liquid)
      Data_summary_tmp_F1L = (Data_summary %>% dplyr::filter(Panel == "F1Liquid CDx")) %>% dplyr::select(Tumor_Sample_Barcode, Age)
      Data_summary_tmp_F1L$favor_liquid = 0
      Data_summary_tmp = rbind(Data_summary_tmp, Data_summary_tmp_F1L)
      Summary_ALL = Data_summary_tmp
      ref_genome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      if(file.exists("source/Data_MAF_target_table_3.rda")){
        load(file="source/Data_MAF_target_table_3.rda")
        load(file="source/Data_GENIE.rda")
        load(file="source/Name_list_GENIE.rda")
        data_tmp = Data_MAF_target_table_3
      } else{
        if(file.exists("source/Data_MAF_target_table_3_5.rda")){
          load(file="source/Data_MAF_target_table_3_5.rda")
          load(file="source/Name_list_GENIE.rda")
          data_tmp = Data_MAF_target_table_3_5
        } else{
          Data_GENIE$Chromosome = str_replace_all(paste0("chr", str_replace_all(Data_GENIE$Chromosome, "chr", "")), "23", "X")
          data_tmp$Chromosome = str_replace_all(paste0("chr", str_replace_all(data_tmp$Chromosome, "chr", "")), "23", "X")
          for(i in 1:length(data_tmp$Hugo_Symbol)){
            mut_ID = (Data_GENIE %>% dplyr::filter(unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
            Data_summary_tmp = Data_summary_tmp %>%
              dplyr::mutate(mutation = case_when(
                Tumor_Sample_Barcode %in% mut_ID ~ 1,
                TRUE ~ 0))
            if(data_tmp$SNV_check[i] == 1){
              triplet_pre[[i]] = ref_genome[[data_tmp$Chromosome[i]]][(data_tmp$Start_Position[i] - 1):(data_tmp$Start_Position[i] + 1)]
            }
            if(data_tmp$odds_ratio[i] == 1 & data_tmp$p_value[i] == 1){
              if(str_detect(data_tmp$unique_alt[i], "_trunc_NOS") | str_detect(data_tmp$unique_alt[i], "fs_NOS")){
                Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
                data_tmp$odds_ratio[data_tmp$unique_alt == data_tmp$unique_alt[i]] = exp(summary(Data_ORR)$coefficients["mutation",1])
                if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
                  fisher_tmp = matrix(c(data_tmp_fisher[i,1][[1]],
                                        data_tmp_fisher[i,2][[1]],
                                        data_tmp_fisher[i,3][[1]],
                                        data_tmp_fisher[i,4][[1]]),
                                      ncol = 2
                  )
                  data_tmp$p_value[data_tmp$unique_alt == data_tmp$unique_alt[i]] = fisher.test(fisher_tmp,)$p.value
                } else{
                  data_tmp$p_value[data_tmp$unique_alt == data_tmp$unique_alt[i]] = summary(Data_ORR)$coefficients["mutation",4]
                }
              } 
              if(str_detect(data_tmp$unique_alt[i], "_trunc_CtoT")){
                Name_list = c(Name_list, data_tmp$unique_alt[i])
              }else {
                Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
                data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
                data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
                if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
                  fisher_tmp = matrix(c(data_tmp_fisher[i,1][[1]],
                                        data_tmp_fisher[i,2][[1]],
                                        data_tmp_fisher[i,3][[1]],
                                        data_tmp_fisher[i,4][[1]]),
                                      ncol = 2
                  )
                  data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
                }
              }
            }
            if(i %% 10 == 0){
              print(paste(i, ",", length(data_tmp$Hugo_Symbol)))
            }
          }
          triplet_post = 
            xscat(subseq(triplet_pre,1,1),
                  DNAStringSet(data_tmp$Tumor_Seq_Allele2),
                  subseq(triplet_pre,3,3))
          ID_revcom = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,2) %in% DNAStringSet(c("A", "G"))
          triplet_pre[ID_revcom] =reverseComplement(triplet_pre[ID_revcom])
          triplet_post[ID_revcom] =reverseComplement(triplet_post[ID_revcom])
          ID_CpG = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,3) == DNAString("CG") &
            subseq(triplet_post,2,3) == DNAString("TG")
          data_tmp$CpG_TpG[ID_CpG] = 1
          ID_CpC = subseq(triplet_pre,1,1) != DNAString("N") &
            subseq(triplet_pre,2,3) == DNAString("CC") &
            subseq(triplet_post,2,3) == DNAString("TC")
          data_tmp$CpC_TpC[ID_CpC] = 1
          
          data_tmp$triplet_pre = as.character(triplet_pre)
          data_tmp$triplet_post = as.character(triplet_post)
            
          Name_list = unique(Name_list)
          Data_MAF_target_table_3_5 = data_tmp
          save(Data_MAF_target_table_3_5, file="source/Data_MAF_target_table_3_5.rda")
          save(Name_list, file="source/Name_list_GENIE.rda")
        }
        for(i in 1:length(Name_list)){
          hotspot_CpG = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpG_TpG == 1)
          hotspot_CpC = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpC_TpC == 1)
          hotspot_nonCpG = data_tmp %>% dplyr::filter(unique_alt == Name_list[i] & CpG_TpG == 0 & CpC_TpC == 0)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1L_position)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1L_position)
          data_tmp$No_found_in_F1L[data_tmp$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1L_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1_position)
          data_tmp$No_found_in_F1[data_tmp$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1_position)
          Data_GENIE$No_found_in_F1L[Data_GENIE$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1L_position)
          Data_GENIE$No_found_in_F1L[Data_GENIE$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1L_position)
          Data_GENIE$No_found_in_F1L[Data_GENIE$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1L_position)
          Data_GENIE$No_found_in_F1[Data_GENIE$unique_position %in% hotspot_CpC$unique_position] = sum(hotspot_CpC$No_found_in_F1_position)
          Data_GENIE$No_found_in_F1[Data_GENIE$unique_position %in% hotspot_CpG$unique_position] = sum(hotspot_CpG$No_found_in_F1_position)
          Data_GENIE$No_found_in_F1[Data_GENIE$unique_position %in% hotspot_nonCpG$unique_position] = sum(hotspot_nonCpG$No_found_in_F1_position)
          data_tmp$unique_alt[data_tmp$unique_position %in% hotspot_CpG$unique_position] = paste0(Name_list[i],"_CpG")
          data_tmp$unique_alt[data_tmp$unique_position %in% hotspot_CpC$unique_position] = paste0(Name_list[i],"_CpC")
          Data_GENIE$unique_alt[Data_GENIE$unique_position %in% hotspot_CpG$unique_position] = paste0(Name_list[i],"_CpG")
          Data_GENIE$unique_alt[Data_GENIE$unique_position %in% hotspot_CpC$unique_position] = paste0(Name_list[i],"_CpC")
          mut_ID = (Data_GENIE %>% dplyr::filter(unique_alt == Name_list[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_nonCpG$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_nonCpG$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_nonCpG$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == Name_list[i]] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_nonCpG$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_nonCpG$No_found_in_F1_position),
                                    sum(hotspot_nonCpG$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_nonCpG$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_GENIE$CH_hotspot_candidate[Data_GENIE$unique_alt == Name_list[i]] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == Name_list[i]] = 1
            data_tmp$p_value[data_tmp$unique_alt == Name_list[i]] = 1
          }
          mut_ID = (Data_GENIE %>% dplyr::filter(unique_alt == paste0(Name_list[i],"_CpG")))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_CpG$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_CpG$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_CpG$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_CpG$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_CpG$No_found_in_F1_position),
                                    sum(hotspot_CpG$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_CpG$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_GENIE$CH_hotspot_candidate[Data_GENIE$unique_alt == paste0(Name_list[i],"_CpG")] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = 1
            data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpG")] = 1
          }
          mut_ID = (Data_GENIE %>% dplyr::filter(unique_alt == paste0(Name_list[i],"_CpC")))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          if((sum(hotspot_CpC$No_found_in_F1L_position) >= 1) &
             (sum(hotspot_CpC$No_found_in_F1L_position) >= Min_count |
              sum(hotspot_CpC$No_found_in_F1_position) >= Min_count_F1)){
            Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = exp(summary(Data_ORR)$coefficients["mutation",1])
            if(exp(summary(Data_ORR)$coefficients["mutation",1]) > 1024 | exp(summary(Data_ORR)$coefficients["mutation",1]) < 1/1024){
              fisher_tmp = matrix(c(sum(hotspot_CpC$No_found_in_F1_position),
                                    data_tmp$F1_patients[1] - sum(hotspot_CpC$No_found_in_F1_position),
                                    sum(hotspot_CpC$No_found_in_F1L_position),
                                    data_tmp$F1L_patients[1] - sum(hotspot_CpC$No_found_in_F1L_position)),
                                  ncol = 2
              )
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = fisher.test(fisher_tmp,)$p.value
            } else {
              data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = summary(Data_ORR)$coefficients["mutation",4]
            }
          } else {
            Data_GENIE$CH_hotspot_candidate[Data_GENIE$unique_alt == paste0(Name_list[i],"_CpC")] = 0
            data_tmp$odds_ratio[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = 1
            data_tmp$p_value[data_tmp$unique_alt == paste0(Name_list[i],"_CpC")] = 1
          }
          print(paste(i, ",", length(Name_list)))
        }
        Data_MAF_target_table_3 = data_tmp
        save(Data_MAF_target_table_3, file="source/Data_MAF_target_table_3.rda")
        save(Data_GENIE, file="source/Data_GENIE.rda")
        save(Name_list, file="source/Name_list_GENIE.rda")
      }
      data_tmp_save = data_tmp
      data_tmp = data_tmp %>%
        dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = TRUE)
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^10] = 2^10
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-10] = 2^-10
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-30] = 10^-30
      data_tmp = data_tmp %>% dplyr::mutate(CtoT = case_when(str_sub(triplet_pre,1,1)!= "N" & str_sub(triplet_pre,2,2) == "C" & str_sub(triplet_post,2,2) == "T" ~ 1,TRUE ~ 0))
      data_tmp = data_tmp %>%
        dplyr::mutate(mut_pattern = case_when(
          str_detect(unique_alt, "M1\\?") ~ "SNV, NOS",
          CpG_TpG == 1 ~ "CpG>TpG",
          CpC_TpC == 1 ~ "CpC>TpC",
          CtoT == 1 ~ "C>T, NOS",
          SNV_check != 1 ~ "Indel",
          TRUE ~ "SNV, NOS"
        ))
      data_tmp$mut_pattern = factor(data_tmp$mut_pattern)
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)) |
                    (logfold < -log2(threshold_OR)) &
                    (logpval > -log10(threshold_P))
                    ,
                  unique_alt, "" ),
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR) &
                     logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)
                    ), "C",
                    "B"))#))
        )
      g = ggplot(data_tmp %>% dplyr::filter(
        (No_found_in_F1L>=Min_count | No_found_in_F1 >= Min_count)),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = "Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      setProgress(detail = "16983")
      plots[[46]] = g
      table_save_tmp_F = data_tmp %>% dplyr::mutate(mutation_type = case_when(pnt_col == "A" ~ "Liquid dominant", pnt_col == "B" ~ "Liquid enriched", TRUE ~ "Somatic enriched"))
      write_excel_csv(table_save_tmp_F, "source/Table_S5_GENIE.csv")
      Gene_name = "DNMT3A"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        annotate("rect", xmin=-Inf, xmax=2, ymin=-Inf, ymax=Inf, fill="white", alpha=0.0) +
        annotate("rect", xmin=2, xmax=8, ymin=-Inf, ymax=Inf, fill="#00CED1", alpha=0.05) +  # 薄い青緑
        annotate("rect", xmin=8, xmax=Inf, ymin=-Inf, ymax=Inf, fill="#FF4500", alpha=0.08) + # 薄い赤オレンジ
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0(Gene_name, ", LD-CH/LE-CH vs GENIE hematologic cohort"),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[53]] = g
      Data_tmp = data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name)
      Data_tmp$Position = as.numeric(unlist(str_extract(str_remove(Data_tmp$unique_alt, "DNMT3A_"), "\\d+")))
      Data_tmp = Data_tmp %>% left_join(table_DNMT3A, by = "Position")
      Data_tmp$Destabilizing[is.na(Data_tmp$Destabilizing)] = "Unknown"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "c")] = "Splicing"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "\\*")] = "Truncating"
      Data_tmp$Destabilizing = factor(Data_tmp$Destabilizing)
      g = ggplot(Data_tmp,
                 aes(x=logfold, y=logpval, color=Destabilizing, shape=Destabilizing, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = "Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      plots[[72]] = g
      Gene_name = "ATM"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[54]] = g
      Gene_name = "ASXL1"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or GENIE hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[55]] = g
      Gene_name = "CHEK2"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[56]] = g
      Gene_name = "EGFR"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[57]] = g      
      Gene_name = "FLT3"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[58]] = g      
      Gene_name = "KDM6A"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[59]] = g      
      Gene_name = "EZH2"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[60]] = g      
      Gene_name = "TET2"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[61]] = g      
      Gene_name = "KMT2D"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[62]] = g      
      Gene_name = "IDH2"
      g = ggplot(data_tmp %>% dplyr::filter(Hugo_Symbol == Gene_name),
                 aes(x=logfold, y=logpval, color=mut_pattern, shape=mut_pattern, label = unique_alt)) +
        geom_point() +
        scale_shape_manual(values = c(3,1,4,6,8)) +
        scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor LD-CH/LE-CH or hematologic malignancy",
             subtitle = paste0("Variants found in C-CAT F1-liquid or GENIE hematologic cohort >=10 times, ", Gene_name),
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        # geom_vline(xintercept=c(log2(threshold_OR_TP53), log2(threshold_OR)), linetype = "dashed", col="blue") +
        #geom_vline(xintercept=c(-log2(2), log2(threshold_OR_TP53)), linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-10.5,12),
                           breaks = c(-10,-8,-6,-4,-2,0,2,4,6,8,10),
                           labels = c("~-10","-8","-6","-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,32),
                           breaks = c(0,10,20,30),
                           labels = c("0","10","20","30~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20)
      #theme(legend.position = "none")
      plots[[63]] = g
      data_tmp_join = data_tmp %>% dplyr::select(unique_alt, logfold, mut_pattern)
      Data_tmp = Data_GENIE %>%
        dplyr::filter(CH_hotspot_candidate == 1) %>%
        # dplyr::distinct(unique_alt, .keep_all = T)  %>%
        dplyr::filter((No_found_in_F1L>=Min_count | No_found_in_F1 >= Min_count)) %>%
        dplyr::left_join(data_tmp_join, by = "unique_alt")
      g <- ggplot(Data_tmp, aes(x = logfold, color = mut_pattern, fill = mut_pattern),alpha = 0.5) +
        geom_histogram() +
        facet_grid(mut_pattern ~ .) +
        #ylim(c(0,1)) +
        xlim(c(-10.5,10.5)) +
        labs(title = "Distribution of odds ratio for LD-CH/LE-CH or hematologic malignancy, C-CAT vs GENIE v17",
             subtitle = "LD-CH/LE-CH or hotspot variants found in GENIE hematologic cohort >=10 times",
             x="log2 odds ratio",
             y="Hotspot variant count") +
        geom_vline(xintercept=c(-2,2), linetype = "dashed", col="blue") +
        theme_bw()
      plots[[47]] = g
      Data_tmp <- Data_tmp %>%
        mutate(logfold_group = cut(logfold, 
                                   breaks = c(-Inf, -8, -2, 2, 8, Inf), 
                                   labels = c("< -8", "-8 ~ -2", "-2 ~ 2", "2 ~ 8", "> 8"),
                                   right = FALSE))
      data_summary <- Data_tmp %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[77]] = g
      Gene_name = "DNMT3A"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[78]] = g
      Gene_name = "TET2"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[79]] = g
      Gene_name = "ASXL1"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[80]] = g
      Gene_name = "ATM"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[81]] = g
      Gene_name = "CHEK2"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[82]] = g
      Gene_name = "TP53"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[83]] = g
      Gene_name = "U2AF1"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      plots[[84]] = g
      Gene_name = "SF3B1"
      data_summary <- Data_tmp %>%
        filter(Hugo_Symbol == Gene_name) %>%
        group_by(logfold_group, mut_pattern) %>%
        summarise(count = n(), .groups = "drop") %>%
        group_by(logfold_group) %>%
        mutate(prop = count / sum(count) * 100)  # 割合（%）を計算
      g = ggplot(data_summary, aes(x = logfold_group, y = prop, fill = mut_pattern)) +
        geom_bar(stat = "identity", position = "stack") +  # 積み上げ棒グラフ
        labs(title = "C-CAT vs GENIE, Mutational Pattern Proportion by Log2 Odds Ratio",
             subtitle = Gene_name,
             x = "Log2 Odds Ratio Group",
             y = "Proportion (%)",
             fill = "Mutation Pattern") +
        theme_minimal()
      setProgress(detail = "17484")
      plots[[85]] = g
      # Data_MAF_target_table_GENIE = Data_GENIE %>%
      #   dplyr::filter(CH_hotspot_candidate == 1) %>%
      #   dplyr::arrange(Chromosome, Start_Position) %>%
      #   dplyr::distinct(unique_alt, .keep_all = T) %>%
      #   dplyr::left_join(data_tmp_join, by = "unique_alt")
      # output$table_all_variants_GENIE = downloadHandler(
      #   filename = function(){"all_hotspot_data_GENIE.csv"}, 
      #   content = function(fname){
      #     write.csv(Data_MAF_target_table_GENIE, fname)
      #   }
      # )
      # Data_maf = Data_GENIE %>%
      #   dplyr::filter(CH_hotspot_candidate == 1 &
      #                   Sample_type == "Blood_cancer") %>%
      #   dplyr::arrange(Chromosome, Start_Position) %>%
      #   dplyr::left_join(data_tmp_join, by = "unique_alt")
      # data_maf = Data_maf %>% dplyr::select(Hugo_Symbol)
      # data_maf$Chromosome = Data_maf$Chromosome
      # data_maf$Start_Position = as.integer(Data_maf$Start_Position)
      # data_maf$End_Position = as.integer(Data_maf$Start_Position) + nchar(Data_maf$Reference_Allele)
      # data_maf$Strand = "+"
      # data_maf$Variant_Classification = "Missense_Mutation"
      # data_maf$Variant_Type = Data_maf$Variant_type
      # data_maf$Reference_Allele = Data_maf$Reference_Allele
      # data_maf$Tumor_Seq_Allele1 = Data_maf$Reference_Allele
      # data_maf$Tumor_Seq_Allele2 = Data_maf$Tumor_Seq_Allele2
      # data_maf$Tumor_Sample_Barcode = Data_maf$Tumor_Sample_Barcode
      # data_maf$favor = ifelse(Data_maf$logfold > 8, "A",
      #                         ifelse(Data_maf$logfold > 2, "B","C"))
      # data_maf$Protein_Change = Data_maf$unique_alt
      # all_samples = (Summary_GENIE %>% dplyr::filter(Sample_type == "Blood_cancer"))$Tumor_Sample_Barcode
      # no_mut_samples = all_samples[!all_samples %in% unique(data_maf$Tumor_Sample_Barcode)]
      # dummy_data <- data.frame(
      #   Hugo_Symbol = "DUMMY",
      #   Chromosome = "chr1",
      #   Start_Position = 1,
      #   End_Position = 1,
      #   Strand = "+",
      #   Variant_Classification = "Missense_Mutation",
      #   Variant_Type = "dummy",
      #   Reference_Allele = "A",
      #   Tumor_Seq_Allele1 = "A",
      #   Tumor_Seq_Allele2 = "T",
      #   Tumor_Sample_Barcode = all_samples,
      #   favor = "dummy",
      #   Protein_Change = "dummy"
      # )
      # Liquid_only_DNMT3A_alt = (Data_tmp %>%
      #                             dplyr::filter(logfold > 8 & Hugo_Symbol == "DNMT3A"))$unique_alt
      # Liquid_favor_DNMT3A_alt = (Data_tmp %>%
      #                              dplyr::filter(logfold > 2 & logfold <= 8 & Hugo_Symbol == "DNMT3A"))$unique_alt
      # Liquid_only_CHEK2_alt = (Data_tmp %>%
      #                            dplyr::filter(logfold > 8 & Hugo_Symbol == "CHEK2"))$unique_alt
      # Liquid_favor_CHEK2_alt = (Data_tmp %>%
      #                             dplyr::filter(logfold > 2 & logfold <= 8 & Hugo_Symbol == "CHEK2"))$unique_alt
      # Liquid_only_ATM_alt = (Data_tmp %>%
      #                          dplyr::filter(logfold > 8 & Hugo_Symbol == "ATM"))$unique_alt
      # Liquid_favor_ATM_alt = (Data_tmp %>%
      #                           dplyr::filter(logfold > 2 & logfold <= 8 & Hugo_Symbol == "ATM"))$unique_alt
      # Liquid_only_TET2_alt = (Data_tmp %>%
      #                           dplyr::filter(logfold > 8 & Hugo_Symbol == "TET2"))$unique_alt
      # Liquid_favor_TET2_alt = (Data_tmp %>%
      #                            dplyr::filter(logfold > 2 & logfold <= 8 & Hugo_Symbol == "TET2"))$unique_alt
      # Liquid_only_ASXL1_alt = (Data_tmp %>%
      #                            dplyr::filter(logfold > 8 & Hugo_Symbol == "ASXL1"))$unique_alt
      # Liquid_favor_ASXL1_alt = (Data_tmp %>%
      #                             dplyr::filter(logfold > 2 & logfold <= 8 & Hugo_Symbol == "ASXL1"))$unique_alt
      # data_maf_rename = data_maf %>%
      #   dplyr::mutate(Hugo_Symbol = case_when(
      #     Protein_Change %in% Liquid_only_DNMT3A_alt ~ "DNMT3A_onlyL",
      #     Protein_Change %in% Liquid_favor_DNMT3A_alt ~ "DNMT3A_favorL",
      #     Protein_Change %in% Liquid_only_CHEK2_alt ~ "CHEK2_onlyL",
      #     Protein_Change %in% Liquid_favor_CHEK2_alt ~ "CHEK2_favorL",
      #     Protein_Change %in% Liquid_only_ATM_alt ~ "ATM_onlyL",
      #     Protein_Change %in% Liquid_favor_ATM_alt ~ "ATM_favorL",
      #     Protein_Change %in% Liquid_only_TET2_alt ~ "TET2_onlyL",
      #     Protein_Change %in% Liquid_favor_TET2_alt ~ "TET2_favorL",
      #     Protein_Change %in% Liquid_only_ASXL1_alt ~ "ASXL1_onlyL",
      #     Protein_Change %in% Liquid_favor_ASXL1_alt ~ "ASXL1_favorL",
      #     TRUE ~ "DUMMY"
      #   ))
      # Genes_maf = unique(data_maf_rename$Hugo_Symbol)
      # Genes_maf = Genes_maf[Genes_maf != "DUMMY"]
      # maf_with_mut_samples_rename <- read.maf(maf = data_maf_rename)
      # maf_without_mut_samples <- read.maf(maf = dummy_data)
      # maf_with_all_samples_rename_GENIE  <- merge_mafs(mafs = list(maf_with_mut_samples_rename, maf_without_mut_samples))
      # maf_with_all_samples_rename_GENIE <- subsetMaf(maf_with_all_samples_rename_GENIE, genes = Genes_maf, dropLevels = FALSE)
      # #write_tsv(data_maf, "source/hotspot_maf_file_GENIE.tsv")
      # output$figure_oncoplot5 = renderPlot({
      #   oncoplot(maf = maf_with_all_samples_rename_GENIE,
      #            sortByAnnotation = TRUE,
      #            genes = Genes_maf,
      #            fontSize = 0.8,
      #            drawColBar = FALSE,
      #            drawRowBar = FALSE,
      #            genesToIgnore = c("DUMMY"),
      #            borderCol = NULL,
      #            bgCol = "#EEEEEE",
      #            drawBox = TRUE,
      #            gene_mar = 15,
      #            keepGeneOrder = FALSE,
      #            GeneOrderSort = FALSE,
      #            legendFontSize = 1.0)
      # })
      # output$figure_oncoplot6 = renderPlot({
      #   somaticInteractions(maf = maf_with_all_samples_rename_GENIE,
      #                               top = length(Genes_maf),
      #                               pvalue = c(0.05, 0.01),
      #                               showSigSymbols = TRUE,
      #                               sigSymbolsSize = 2.0,
      #                               nShiftSymbols = 1,
      #                               pvSymbols = c(42, 1),
      #                               showSum = FALSE)
      # })
      # par("mar"=c(1,1,1,1))
      # Summary_figure_oncoplot6 =
      #   somaticInteractions(maf = maf_with_all_samples_rename_GENIE,
      #                               top = length(Genes_maf),
      #                               pvalue = c(0.05, 0.01),
      #                               showSigSymbols = TRUE,
      #                               sigSymbolsSize = 2.0,
      #                               nShiftSymbols = 1,
      #                               pvSymbols = c(42, 1),
      #                               showSum = FALSE)
      # output$figure_oncoplot6_data = DT::renderDataTable(Summary_figure_oncoplot6,
      #                                                    server = FALSE,
      #                                                    filter = 'top', 
      #                                                    extensions = c('Buttons'), 
      #                                                    options = list(pageLength = 100, 
      #                                                                   scrollX = TRUE,
      #                                                                   scrollY = "1000px",
      #                                                                   scrollCollapse = TRUE,
      #                                                                   dom="Blfrtip",
      #                                                                   buttons = c('csv', 'copy')))
      # Data_GENIE = Data_GENIE %>%
      #   dplyr::filter(non_germline_non_error == 1 & No_found_in_F1L >= Min_count &
      #                   Sample_type == "Blood_cancer") %>%
      #   dplyr::arrange(Chromosome, Start_Position) %>%
      #   dplyr::left_join(F1L_germline_error, by="Hugo_Symbol") %>%
      #   dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
      #   dplyr::left_join(data_tmp_join, by = "unique_alt")
      # CH_gene_compare = CH_gene_entropy_3
      # CH_gene_compare = c("DNMT3A", "TET2", "ASXL1", "CHEK2", "SF3B1", "JAK2",
      #                     "IDH2", "CBL", "U2AF1", "ATM", "MDM4", "MPL",
      #                     "KMT2D", "MYD88")
      # CH_gene_compare = input$clonal_hematopoiesis_genes
      # C_CAT_alt_only_liquid = unique((Data_MAF_target %>%
      #                                   filter(pnt_col == "A" &
      #                                          non_germline_non_error == 1 &
      #                                          Panel == "F1Liquid CDx" &
      #                                          Hugo_Symbol %in% CH_gene_compare &
      #                                          unique_alt %in% CH_candidates &
      #                                          logfold > 8))$unique_alt)
      # C_CAT_alt_favor_liquid = unique((Data_MAF_target %>%
      #                                   filter(pnt_col == "A" &
      #                                            non_germline_non_error == 1 &
      #                                            Panel == "F1Liquid CDx" &
      #                                            Hugo_Symbol %in% CH_gene_compare &
      #                                            unique_alt %in% CH_candidates &
      #                                            logfold <= 8))$unique_alt)
      # GENIE_alt_only_liquid = unique((Data_GENIE %>%
      #                   dplyr::filter(non_germline_non_error == 1 & 
      #                                 Hugo_Symbol %in% CH_gene_compare &
      #                                 logfold > 8 &
      #                                 No_found_in_F1L >= Min_count &
      #                                 Sample_type == "Blood_cancer"))$unique_alt)
      # GENIE_alt_favor_liquid = unique((Data_GENIE %>%
      #                                   dplyr::filter(non_germline_non_error == 1 & 
      #                                                   Hugo_Symbol %in% CH_gene_compare &
      #                                                   logfold > 2 &
      #                                                   logfold <= 8 &
      #                                                   No_found_in_F1L >= Min_count &
      #                                                   Sample_type == "Blood_cancer"))$unique_alt)
      # C_CAT_alt_extra_liquid = unique((Data_MAF_target %>%
      #                                    filter(pnt_col != "A" &
      #                                             non_germline_non_error == 1 &
      #                                             Panel == "F1Liquid CDx" &
      #                                             Hugo_Symbol %in% CH_gene_compare &
      #                                             logfold <= 2))$unique_alt)
      # GENIE_alt_extra_liquid = unique((Data_GENIE %>%
      #                                    dplyr::filter(non_germline_non_error == 1 & 
      #                                                    Hugo_Symbol %in% CH_gene_compare &
      #                                                    logfold <= 2 &
      #                                                    No_found_in_F1L >= Min_count &
      #                                                    Sample_type == "Blood_cancer"))$unique_alt)
      # A1_B1 <- intersect(C_CAT_alt_only_liquid, GENIE_alt_only_liquid)
      # A1_B2 <- intersect(C_CAT_alt_only_liquid, GENIE_alt_favor_liquid)
      # A1_B3 <- intersect(C_CAT_alt_only_liquid, GENIE_alt_extra_liquid)
      # A1_only <- setdiff(C_CAT_alt_only_liquid, union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid))
      # A2_B1 <- intersect(C_CAT_alt_favor_liquid, GENIE_alt_only_liquid)
      # A2_B2 <- intersect(C_CAT_alt_favor_liquid, GENIE_alt_favor_liquid)
      # A2_B3 <- intersect(C_CAT_alt_favor_liquid, GENIE_alt_extra_liquid)
      # A2_only <- setdiff(C_CAT_alt_favor_liquid, union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid))
      # A3_B1 <- intersect(C_CAT_alt_extra_liquid, GENIE_alt_only_liquid)
      # A3_B2 <- intersect(C_CAT_alt_extra_liquid, GENIE_alt_favor_liquid)
      # A3_B3 <- intersect(C_CAT_alt_extra_liquid, GENIE_alt_extra_liquid)
      # A3_only <- setdiff(C_CAT_alt_extra_liquid, union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid))
      # B1_only <- setdiff(GENIE_alt_only_liquid, union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid))
      # B2_only <- setdiff(GENIE_alt_favor_liquid, union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid))
      # B3_only <- setdiff(GENIE_alt_extra_liquid, union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid))
      # all_genes <- unique(c(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid, C_CAT_alt_extra_liquid, GENIE_alt_only_liquid, GENIE_alt_favor_liquid, GENIE_alt_extra_liquid))
      # neither <- setdiff(all_genes, union(union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid), union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid)))
      # table <- matrix(
      #   c(length(A1_B1), length(A1_B2), length(A1_B3), length(A1_only),
      #     length(A2_B1), length(A2_B2), length(A2_B3), length(A2_only),
      #     length(A3_B1), length(A3_B2), length(A3_B3), length(A3_only),
      #     length(B1_only), length(B2_only), length(B3_only), length(neither)
      #   ),
      #   nrow = 4,
      #   byrow = TRUE
      # )
      # table_df <- data.frame(
      #   Col = rep(c("Only_liquid", "Favor_liquid", "Neurtal", "Not_in_C_CAT"), times = 4),
      #   Row = rep(c("Only_blood", "Favor_blood", "Neurtal", "Not_in_GENIE"), each = 4),
      #   Value = as.vector(table)
      # )
      # table_df$Col = factor(table_df$Col, levels = c("Only_liquid", "Favor_liquid", "Neurtal", "Not_in_C_CAT"))
      # table_df$Row = factor(table_df$Row, levels = c("Only_blood", "Favor_blood", "Neurtal", "Not_in_GENIE"))
      # g = ggplot(table_df, aes(x = Col, y = Row, fill = Value)) +
      #   geom_tile(color = "black") +
      #   geom_text(aes(label = Value), color = "black", size = 6) +
      #   scale_fill_gradient(low = "white", high = "red") +
      #   labs(title = "Comparison of C-CAT and GENIE",
      #        subtitle = paste0("Genes: ", paste(sort(CH_gene_compare), collapse = "/")),
      #        x = "C-CAT",
      #        y = "GENIE",
      #        fill = "Variant counts") +
      #   theme_minimal() +
      #   theme(
      #     axis.text.x = element_text(angle = 45, hjust = 1),
      #     panel.grid = element_blank()  # グリッド線を消す
      #   )
      # plots[[50]] = g
      # subgroups <- list(
      #   "Only_liquid_Only_blood" = intersect(C_CAT_alt_only_liquid, GENIE_alt_only_liquid),
      #   "Only_liquid_Favor_blood" = intersect(C_CAT_alt_only_liquid, GENIE_alt_favor_liquid),
      #   "Only_liquid_Neutral" = intersect(C_CAT_alt_only_liquid, GENIE_alt_extra_liquid),
      #   "Only_liquid_Not_in_GENIE" = setdiff(C_CAT_alt_only_liquid, union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid)),
      #   
      #   "Favor_liquid_Only_blood" = intersect(C_CAT_alt_favor_liquid, GENIE_alt_only_liquid),
      #   "Favor_liquid_Favor_blood" = intersect(C_CAT_alt_favor_liquid, GENIE_alt_favor_liquid),
      #   "Favor_liquid_Neutral" = intersect(C_CAT_alt_favor_liquid, GENIE_alt_extra_liquid),
      #   "Favor_liquid_Not_in_GENIE" = setdiff(C_CAT_alt_favor_liquid, union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid)),
      #   
      #   "Neutral_Only_blood" = intersect(C_CAT_alt_extra_liquid, GENIE_alt_only_liquid),
      #   "Neutral_Favor_blood" = intersect(C_CAT_alt_extra_liquid, GENIE_alt_favor_liquid),
      #   "Neutral_Neutral" = intersect(C_CAT_alt_extra_liquid, GENIE_alt_extra_liquid),
      #   "Neutral_Not_in_GENIE" = setdiff(C_CAT_alt_extra_liquid, union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid)),
      #   
      #   "Not_in_C_CAT_Only_blood" = setdiff(GENIE_alt_only_liquid, union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid)),
      #   "Not_in_C_CAT_Favor_blood" = setdiff(GENIE_alt_favor_liquid, union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid)),
      #   "Not_in_C_CAT_Neutral" = setdiff(GENIE_alt_extra_liquid, union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid)),
      #   "Not_in_C_CAT_Not_in_GENIE" = setdiff(all_genes, union(union(union(C_CAT_alt_only_liquid, C_CAT_alt_favor_liquid), C_CAT_alt_extra_liquid), union(union(GENIE_alt_only_liquid, GENIE_alt_favor_liquid), GENIE_alt_extra_liquid)))
      # )
      # gene_subgroup_df <- bind_rows(lapply(names(subgroups), function(name) {
      #   genes <- subgroups[[name]]
      #   if (length(genes) == 0) {
      #     return(data.frame(Gene = character(), Subgroup = character()))
      #   } else {
      #     return(data.frame(Gene = genes, Subgroup = name))
      #   }
      # }))
      # output$comparison_CCAT_GENIE = DT::renderDataTable(gene_subgroup_df,
      #                                                    server = FALSE,
      #                                                    filter = 'top', 
      #                                                    extensions = c('Buttons'), 
      #                                                    options = list(pageLength = 100, 
      #                                                                   scrollX = TRUE,
      #                                                                   scrollY = "1000px",
      #                                                                   scrollCollapse = TRUE,
      #                                                                   dom="Blfrtip",
      #                                                                   buttons = c('csv', 'copy')))
      Data_tmp = Data_GENIE %>%
        dplyr::filter(Sample_type == "Blood_cancer")
      Data_tmp$Position = as.numeric(unlist(str_extract(Data_tmp$amino.acid.change, "\\d+")))
      Data_tmp = Data_tmp %>% left_join(table_DNMT3A, by = "Position")
      Data_tmp$Destabilizing[is.na(Data_tmp$Destabilizing)] = "Unknown"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "c")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "splice")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "\\*")] = "Truncating"
      Data_tmp$Destabilizing = factor(Data_tmp$Destabilizing)
      lolliplot_gene = input$lolliplot_gene
      if(is.null(lolliplot_gene)){
        lolliplot_gene = "DNMT3A"
      } else if(lolliplot_gene == ""){
        lolliplot_gene = "DNMT3A"
      }
      data_lolliplot = Data_tmp %>%
        dplyr::select(Chromosome,
                      Start_Position,
                      End_Position,
                      Reference_Allele,
                      Tumor_Seq_Allele2,
                      Tumor_Sample_Barcode,
                      Hugo_Symbol,
                      amino.acid.change)
      if(lolliplot_gene == "DNMT3A"){
        data_lolliplot$Variant_Classification = Data_tmp$Destabilizing
      } else{
        data_lolliplot$Variant_Classification = "Missense_Mutation"
      }
      data_lolliplot = data_lolliplot %>%
        dplyr::filter(!str_detect(amino.acid.change, "c.")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "_")) %>%
        dplyr::filter(Hugo_Symbol == lolliplot_gene) 
      if(length(data_lolliplot$Chromosome) == 0){
        plots[[43]] = gg_empty()
      } else {
        var<-unique(data_lolliplot[!duplicated(data_lolliplot),])#remove duplicates
        var[order(var$Hugo_Symbol,gsub("([A-Z]+)([0-9]+)","\\2",var$amino.acid.change)),]#order
        var.freq<-plyr::count(var,c("Hugo_Symbol","amino.acid.change","Variant_Classification"))
        var.aanum<-unlist(lapply(regmatches(var.freq$amino.acid.change,gregexpr(pattern="*(\\d+)",var.freq$amino.acid.change)),function(x) x[[1]]))
        var.plot.data<-cbind(var.freq,as.numeric(as.character(var.aanum)))#hard way to remove factor
        colnames(var.plot.data)[5]<-"var.aanum"
        var.plot.data = var.plot.data[var.plot.data$var.aanum >=1 & is.finite(var.plot.data$var.aanum) & !is.na(var.plot.data$var.aanum),]
        
        var.plot<-isolate(var.plot.data)
        var.plot<-var.plot[var.plot$Hugo_Symbol==lolliplot_gene,]
        var.plot<-var.plot[order(var.plot$var.aanum),]
        var.plot$amino.acid.change<-as.character(var.plot$amino.acid.change)
        
        incProgress(1 / 4)
        
        p<-ggplot(var.plot,aes(var.aanum,freq,color=Variant_Classification,label=amino.acid.change)) +
          geom_segment(aes(x=var.aanum,y=0,xend=var.aanum,yend=freq),
                       color=ifelse(var.plot$freq>=input$lolliplot_no,"orange","grey50"),
                       linewidth=ifelse(var.plot$freq>=input$lolliplot_no,1.2,0.6)) +
          geom_point(size=5) +
          geom_text_repel(nudge_y=max(var.plot$freq) * 0.05,
                          color=ifelse(var.plot$freq>=input$lolliplot_no,"orange","NA"),
                          size=ifelse(var.plot$freq>=input$lolliplot_no,4,2),
                          fontface="bold", max.overlaps=Inf) +
          theme_classic()
        proteins_acc<-read.table(file("source/UniProt.txt",
                                      encoding='UTF-8-BOM'),header=T)
        if(length(proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]) == 1){
          proteins_acc<-proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]
          proteins_acc_url<-gsub(" ","%2C",proteins_acc)
          baseurl<-"https://www.ebi.ac.uk/proteins/api/features?offset=0&size=100&accession="
          url<-paste0(baseurl,proteins_acc_url)
          incProgress(1 / 6)
          flag_lolliplot = 0
          if(file.exists(paste0("source/lolliplot/", lolliplot_gene, ".rda"))){
            load(paste0("source/lolliplot/", lolliplot_gene, ".rda"))
            flag_lolliplot = 1
          } else {
            prots_feat<-try(GET(url,accept_json()), silent = FALSE)
            if (class(prots_feat) != "try-error"){
              flag_lolliplot = 1
              save(prots_feat, file=paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))
            }
          }
          if (flag_lolliplot == 1){
            prots_feat_red<-(httr::content(prots_feat))
            incProgress(1 / 6)
            features_total_plot<-NULL
            for(i in 1:length(prots_feat_red)){ 
              features_temp<-drawProteins::extract_feat_acc(prots_feat_red[[i]])#the extract_feat_acc() function takes features into a data.frame
              features_temp$order<-i # this order is needed for plotting later
              features_total_plot<-rbind(features_total_plot,features_temp)
            }
            plot_start<-0#starts 0
            plot_end<-max(features_total_plot$end)
            if("CHAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="CHAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.1*max(var.plot$freq),ymax=-0.04*max(var.plot$freq)),
                          colour="grey",
                          fill="grey",
                          inherit.aes=F)
            }
            incProgress(1 / 6)
            if("DNA_BIND"%in%(unique(features_total_plot$type))) {
              features_total_plot[features_total_plot$type=="DNA_BIND","description"] <- features_total_plot[features_total_plot$type=="DNA_BIND","type"]
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DNA_BIND",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("DOMAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DOMAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("MOTIF"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="MOTIF",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("REPEAT"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="REPEAT",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            p <- p +
              scale_x_continuous(breaks=round(seq(plot_start,plot_end,by=50)),name="Amino acid number")
          }
        }
        p <- p +
          theme(plot.title=element_text(face="bold",size=(15),hjust=0),axis.text.x=element_text(angle=90,hjust=1)) +
          labs(y="Mutation frequency in AACR project GENIE dataset",
               title=paste(nrow(data_lolliplot), " variants in ",lolliplot_gene,", GENIE hematologic cohort",sep="")) +
          scale_y_continuous(breaks=seq(0,max(var.plot$freq*1.05),max(floor(max(var.plot$freq)/100)*10,5)),name="Frequency")
        plots[[43]] = p
      }

      
      DNMT3A_GENIE = sort(table((Data_GENIE %>%
                    dplyr::filter(Sample_type == "Blood_cancer" &
                                    CH_hotspot_candidate <= 1 &
                                    #non_germline_non_error == 1 &
                                    # No_found_in_F1L >= Min_count &
                                    Hugo_Symbol == "DNMT3A"))$amino.acid.change))
      DNMT3A_CCAT = sort(table((Data_MAF_target %>%
                                   dplyr::filter(Panel == "F1Liquid CDx" &
                                                   # CH_hotspot_candidate == 1 &
                                                   non_germline_non_error <= 1 &
                                                   # unique_alt %in% CH_candidates &
                                                   # Hugo_Symbol %in% CH_gene_entropy_3 &
                                                   # No_found_in_F1L >= Min_count &
                                                   Hugo_Symbol == "DNMT3A"))$amino.acid.change))
      DNMT3A_GENIE = data.frame(names(DNMT3A_GENIE), as.integer(DNMT3A_GENIE))
      DNMT3A_CCAT = data.frame(names(DNMT3A_CCAT), as.integer(DNMT3A_CCAT))
      colnames(DNMT3A_GENIE) = c("amino.acid.change", "count_GENIE")
      colnames(DNMT3A_CCAT) = c("amino.acid.change", "count_CCAT")
      DNMT3A_count_blood = dplyr::full_join(DNMT3A_GENIE, DNMT3A_CCAT, by = "amino.acid.change")
      DNMT3A_count_blood$count_GENIE[is.na(DNMT3A_count_blood$count_GENIE)] = 0
      DNMT3A_count_blood$count_CCAT[is.na(DNMT3A_count_blood$count_CCAT)] = 0
      DNMT3A_count_blood = DNMT3A_count_blood[DNMT3A_count_blood$count_CCAT>=Min_count | DNMT3A_count_blood$count_GENIE>=Min_count,]
      DNMT3A_count_blood = DNMT3A_count_blood %>%
        dplyr::filter(!str_detect(amino.acid.change, "\\*|\\>|splice"))
      DNMT3A_count_blood$GENIE_negative = length((Summary_GENIE %>% dplyr::filter(Sample_type == "Blood_cancer"))$Age) - DNMT3A_count_blood$count_GENIE
      DNMT3A_count_blood$CCAT_negative = length((Data_summary %>% filter(Panel == "F1Liquid CDx"))$Panel) - DNMT3A_count_blood$count_CCAT
      DNMT3A_count_blood$ratio = fun_zero(DNMT3A_count_blood$count_CCAT,DNMT3A_count_blood$count_GENIE) * fun_zero(DNMT3A_count_blood$GENIE_negative,DNMT3A_count_blood$CCAT_negative)
      # DNMT3A_favor_Liquid_to_Blood = (DNMT3A_count_blood %>% filter(ratio > 4 | ratio == 0))$amino.acid.change
      DNMT3A_favor_Liquid_to_Blood = (DNMT3A_count_blood %>% filter((ratio >= 4 | ratio == 0) & count_CCAT >= 30))$amino.acid.change
      output$table_DNMT3A_odds_blood = DT::renderDataTable(DNMT3A_count_blood,
                                                     server = FALSE,
                                                     filter = 'top', 
                                                     extensions = c('Buttons'), 
                                                     options = list(pageLength = 100, 
                                                                    scrollX = TRUE,
                                                                    scrollY = "1000px",
                                                                    scrollCollapse = TRUE,
                                                                    dom="Blfrtip",
                                                                    buttons = c('csv', 'excel', 'copy')))
      Data_GENIE$Position = as.numeric(unlist(str_extract(Data_GENIE$amino.acid.change, "\\d+")))
      Data_GENIE = Data_GENIE %>% left_join(table_DNMT3A, by = "Position")
      Data_GENIE$Destabilizing[is.na(Data_GENIE$Destabilizing)] = "Unknown"
      Data_GENIE$Destabilizing[str_detect(Data_GENIE$name, "c")] = "Splice"
      Data_GENIE$Destabilizing[str_detect(Data_GENIE$name, "\\*")] = "Truncating"
      Data_GENIE$Destabilizing = factor(Data_GENIE$Destabilizing)
      DNMT3A_high = unique((Data_GENIE %>%
                              dplyr::filter(Sample_type == "Blood_cancer" &
                                              CH_hotspot_candidate == 1 &
                                              # non_germline_non_error == 1 &
                                              # No_found_in_F1L >= Min_count &
                                              Hugo_Symbol == "DNMT3A" &
                                              # CpC_TpC == 1 &
                                              # logfold > 8))$unique_alt)
                                              # Destabilizing == "Instable"))$unique_alt)
                                              amino.acid.change %in% DNMT3A_favor_Liquid_to_Blood))$unique_alt)

                                              # str_detect(unique_alt, "Y735C|R635W|P904L")))$unique_alt)
                                              # str_detect(unique_alt, "Y735C")))$unique_alt)
      DNMT3A_882 = unique((Data_GENIE %>%
                             dplyr::filter(Sample_type == "Blood_cancer" &
                                             CH_hotspot_candidate == 1 &
                                             # non_germline_non_error == 1 &
                                             # No_found_in_F1L >= Min_count &
                                             Hugo_Symbol == "DNMT3A" &
                                             # str_detect(unique_alt, "I780T|V657M|R326C|I705T|F755S|M801V|P700L|Y908C|Y660C|V649M|L639F|G332R|G796D|G706R|S337L|R729G|R635W|S638C|P799S|P904L")))$unique_alt)
                                             # str_detect(unique_alt, "I780T|V657M|R635W|R736H|R736C|P904L")))$unique_alt)
                                             # str_detect(unique_alt, "882")))$unique_alt)
                                             str_detect(unique_alt, "R882")))$unique_alt)
      ID_DNMT3A_high_and_882 = intersect(
        unique((Data_GENIE %>%
                  dplyr::filter(Sample_type == "Blood_cancer" &
                                  CH_hotspot_candidate == 1 &
                                  unique_alt %in% DNMT3A_high))$Tumor_Sample_Barcode),
        unique((Data_GENIE %>%
                  dplyr::filter(Sample_type == "Blood_cancer" &
                                  CH_hotspot_candidate == 1 &
                                  unique_alt %in% DNMT3A_882))$Tumor_Sample_Barcode))
      ID_DNMT3A_high = setdiff(
        unique((Data_GENIE %>%
                  dplyr::filter(Sample_type == "Blood_cancer" &
                                  unique_alt %in% DNMT3A_high))$Tumor_Sample_Barcode),
        ID_DNMT3A_high_and_882)
      ID_DNMT3A_882_blood = setdiff(
        unique((Data_GENIE %>%
                  dplyr::filter(Sample_type == "Blood_cancer" &
                                  CH_hotspot_candidate == 1 &
                                  unique_alt %in% DNMT3A_882))$Tumor_Sample_Barcode),
        ID_DNMT3A_high_and_882)
      ID_DNMT3A_other_blood = setdiff(
        unique((Data_GENIE %>%
                  dplyr::filter(Sample_type == "Blood_cancer" &
                                  CH_hotspot_candidate == 1 &
                                  str_detect(Hugo_Symbol, "DNMT3A")))$Tumor_Sample_Barcode),
        c(ID_DNMT3A_high_and_882, ID_DNMT3A_high, ID_DNMT3A_882_blood))
      ID_DNMT3A_none_blood = setdiff(
        ID_Blood_cancer,
        c(ID_DNMT3A_high_and_882, ID_DNMT3A_high, ID_DNMT3A_882_blood, ID_DNMT3A_other_blood))
      Data_GENIE = Data_GENIE %>%
        dplyr::mutate(DNMT3A_status = case_when(
          Tumor_Sample_Barcode %in% ID_DNMT3A_high_and_882 ~ "INS and 882",
          Tumor_Sample_Barcode %in% ID_DNMT3A_high ~ "INS",
          Tumor_Sample_Barcode %in% ID_DNMT3A_882_blood ~ "882",
          Tumor_Sample_Barcode %in% ID_DNMT3A_other_blood ~ "other mut",
          Tumor_Sample_Barcode %in% ID_DNMT3A_none_blood ~ "none",
          TRUE ~ "solid sample"
        ))
      Data_GENIE$DNMT3A_status_all = "all"
      df_plot <- Data_GENIE %>%
        dplyr::filter(DNMT3A_status != "solid sample") %>%
        group_by(DNMT3A_status, Cancername) %>%
        summarise(count = n(), .groups = "drop") %>%
        mutate(percentage = count / sum(count) * 100)  # 割合を計算
      
      # ggplotで可視化
      g = ggplot(df_plot, aes(x = DNMT3A_status, y = percentage, fill = Cancername)) +
        geom_bar(stat = "identity", position = "fill") +
        scale_y_continuous(labels = scales::percent_format(scale = 100)) +
        labs(title = "Cancer frequency by DNMT3A status",
             x = "Cancer type",
             y = "Percentage (%)") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +  # x軸ラベルを回転
        theme_bw() +
        scale_fill_viridis_d(option = "turbo")  # カラーパレット適用
      plots[[75]] = g
      Data_maf_DNMT3A_GENIE = Data_GENIE %>%
        #dplyr::filter(max_VAF >= 0.02 & logfold > 2) %>%
        dplyr::filter(!unique_alt %in% c(DNMT3A_high, DNMT3A_882) &
                        #Hugo_Symbol != "DNMT3A" &
                        #Hugo_Symbol %in% CH_gene_compare &
                        Sample_type == "Blood_cancer") %>%
        dplyr::arrange(Chromosome, Start_Position)
      data_maf_DNMT3A_GENIE = data.frame(Data_maf_DNMT3A_GENIE %>% dplyr::select(Hugo_Symbol))
      data_maf_DNMT3A_GENIE$Chromosome = Data_maf_DNMT3A_GENIE$Chromosome
      data_maf_DNMT3A_GENIE$Start_Position = as.integer(Data_maf_DNMT3A_GENIE$Start_Position)
      data_maf_DNMT3A_GENIE$End_Position = as.integer(Data_maf_DNMT3A_GENIE$Start_Position) + nchar(Data_maf_DNMT3A_GENIE$Reference_Allele) - 1
      data_maf_DNMT3A_GENIE$Strand = "+"
      data_maf_DNMT3A_GENIE$Variant_Classification = "Missense_Mutation"
      data_maf_DNMT3A_GENIE$Variant_Type = ifelse(nchar(Data_maf_DNMT3A_GENIE$Reference_Allele) == 1 & nchar(Data_maf_DNMT3A_GENIE$Tumor_Seq_Allele2) == 1, "SNV",
                                            ifelse(nchar(Data_maf_DNMT3A_GENIE$Reference_Allele) > nchar(Data_maf_DNMT3A_GENIE$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf_DNMT3A_GENIE$Reference_Allele = Data_maf_DNMT3A_GENIE$Reference_Allele
      data_maf_DNMT3A_GENIE$Tumor_Seq_Allele1 = Data_maf_DNMT3A_GENIE$Reference_Allele
      data_maf_DNMT3A_GENIE$Tumor_Seq_Allele2 = Data_maf_DNMT3A_GENIE$Tumor_Seq_Allele2
      data_maf_DNMT3A_GENIE$Tumor_Sample_Barcode = Data_maf_DNMT3A_GENIE$Tumor_Sample_Barcode
      data_maf_DNMT3A_GENIE$DNMT3A_status = Data_maf_DNMT3A_GENIE$DNMT3A_status
      data_maf_DNMT3A_GENIE$Protein_Change = Data_maf_DNMT3A_GENIE$unique_alt
      maf.dset_GENIE <- data_maf_DNMT3A_GENIE %>%
        dplyr::filter(Reference_Allele != "-" & Tumor_Seq_Allele2 != "-" &
                        nchar(Reference_Allele) == 1 &
                        nchar(Tumor_Seq_Allele2) == 1)
      maf.dset_GENIE = as.data.frame(maf.dset_GENIE)
      maf.dset_GENIE <- filterSNV(dataSet = maf.dset_GENIE,
                            seq_colNames = c("Reference_Allele",
                                             "Tumor_Seq_Allele2"))
      hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      maf.dset_GENIE <- attachContext(mutData = maf.dset_GENIE,
                                chr_colName = "Chromosome",
                                start_colName = "Start_Position",
                                end_colName = "End_Position",
                                nucl_contextN = 3,
                                BSGenomeDb = hg38)
      maf.dset_GENIE <- removeMismatchMut(mutData = maf.dset_GENIE,
                                    refMut_colName = "Reference_Allele",
                                    context_colName = "context",
                                    refMut_format = "N")
      maf.dset_GENIE <- attachMutType(mutData = maf.dset_GENIE,
                                ref_colName = "Reference_Allele",
                                var_colName = "Tumor_Seq_Allele1",
                                var2_colName = "Tumor_Seq_Allele2",
                                context_colName = "context")
      maf.counts_DNMT3A_GENIE <- countMutTypes(mutTable = maf.dset_GENIE,
                                         sample_colName = "DNMT3A_status",
                                         mutType_colName = "mutType")
      # mouCancer.assess <- prelimProcessAssess(input = maf.counts,
      #                                         approach = "counts")
      # maf.params_DNMT3A_GENIE <- setMutClusterParams(num_processesToExtract = 3,
      #                                          approach = "counts",
      #                                          num_totIterations = 5,
      #                                          num_parallelCores = 6, debug = FALSE,
      #                                          algorithm = "alexa")
      # maf.counts.analysis_DNMT3A_GENIE <- decipherMutationalProcesses(input = maf.counts_DNMT3A_GENIE,
      #                                                           params = maf.params_DNMT3A_GENIE)
      # ocd.signs_DNMT3A_GENIE <- maf.counts.analysis_DNMT3A_GENIE$Results$signatures
      # ocd.expos_DNMT3A_GENIE <- maf.counts.analysis_DNMT3A_GENIE$Results$exposures
      # msig1_DNMT3A_GENIE <- matchSignatures(mutSign = ocd.signs_DNMT3A_GENIE, reference = cosmixSigs,
      #                                 threshold = 0.45, plot = TRUE)
      # hm1_DNMT3A_GENIE <- msig1_DNMT3A_GENIE$plot + ggtitle("Match to COSMIC signs.")
      # x_DNMT3A_GENIE <- maf.counts_DNMT3A_GENIE
      # xx_DNMT3A_GENIE <- x_DNMT3A_GENIE#as.mutation.counts(x)
      # blca.expo1_DNMT3A_GENIE <- resolveMutSignatures(mutCountData = x_DNMT3A_GENIE,
      #                                           signFreqData = cosmixSigs)
      # blca.exp.1x_DNMT3A_GENIE <- blca.expo1_DNMT3A_GENIE$Results$count.result
      # bp1_DNMT3A_GENIE <- msigPlot(blca.exp.1x_DNMT3A_GENIE, main = "All variants, DMNT3A mutation status | COSMIC sigs.")
      # cosmx <-cosmixSigs[c(1,3,6,26,29,30)]
      # msig1 <- matchSignatures(mutSign = ocd.signs, reference = cosmx, 
      #                          threshold = 0.45, plot = TRUE) 
      # hm2 <- msig1$plot + ggtitle("Match to COSMIC signs.")
      # x <- maf.counts
      # xx <- x#as.mutation.counts(x)
      # blca.expo1 <- resolveMutSignatures(mutCountData = x, 
      #                                    signFreqData = cosmx)
      # blca.exp.1x <- blca.expo1$Results$count.result
      # bp2 <- msigPlot(blca.exp.1x, main = "Hotspot variants, favor | COSMIC sigs.")
      # fig_msig = msigPlot(ocd.expos, main = "Hotspot variants, favor liquid/tumor/NOS") + 
      #   scale_fill_manual(values = c("#1f78b4", "#cab2d6", "#ff7f00", "#a6cee3"))
      output$figure_signature_DNMT3A_GENIE = renderPlot({
        # par(mfrow = c(5, 1))
        # msigPlot(maf.counts_DNMT3A_GENIE, sample = "INS and 882", main = "R882 & Instable variant compound", ylim = c(0, 0.25))
        # msigPlot(maf.counts_DNMT3A_GENIE, sample = "INS", main = "Instable variant", ylim = c(0, 0.25))
        # msigPlot(maf.counts_DNMT3A_GENIE, sample = "882", main = "R882 mutation", ylim = c(0, 0.25))
        # msigPlot(maf.counts_DNMT3A_GENIE, sample = "other mut", main = "Other DNMT3A mutation", ylim = c(0, 0.25))
        # msigPlot(maf.counts_DNMT3A_GENIE, sample = "none", main = "No DNMT3A mutation", ylim = c(0, 0.25))
        setProgress(detail = "18106")
        par(mfrow = c(5, 1))
        msigPlot(maf.counts_DNMT3A_GENIE,
                 sample = "INS and 882",
                 main = paste0("GENIE patients with DNMT3A CH-favor mut and R882 mut, ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_high_and_882),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_high_and_882)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_high_and_882)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A_GENIE,
                 sample = "INS",
                 main = paste0("GENIE patients with DNMT3A CH-favor mut, ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_high),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_high)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "INS" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_high)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A_GENIE,
                 sample = "882",
                 main = paste0("GENIE patients with DNMT3A R882 mut, ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_882_blood),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_882_blood)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_882_blood)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A_GENIE,
                 sample = "other mut",
                 main = paste0("GENIE patients with other DNMT3A hotspot mutations, ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_other_blood),
                               " pts (mean ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_other_blood)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_other_blood)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A_GENIE,
                 sample = "none",
                 main = paste0("GENIE patients with no DNMT3A hotspot mutations, ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_none_blood),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_none_blood)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_GENIE %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_none_blood)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        # prelimProcessAssess(input = maf.counts,
        #                     approach = "counts")
        # msigPlot(ocd.signs_DNMT3A_GENIE, signature = 1, ylim = c(0, 0.15))
        # msigPlot(ocd.signs_DNMT3A_GENIE, signature = 2, ylim = c(0, 0.15))
        # msigPlot(ocd.signs_DNMT3A_GENIE, signature = 3, ylim = c(0, 0.15))
        # msigPlot(ocd.signs_DNMT3A_GENIE, signature = 4, ylim = c(0, 0.15))
      })
      # output$figure_signature2_DNMT3A_GENIE = renderPlot({
      #   grid.arrange(hm1_DNMT3A_GENIE, bp1_DNMT3A_GENIE, ncol = 1)
      # })
      output$table_summary_GENIE = DT::renderDataTable(data_maf_DNMT3A_GENIE,
                                                   server = FALSE,
                                                   filter = 'top', 
                                                   extensions = c('Buttons'), 
                                                   options = list(pageLength = 100, 
                                                                  scrollX = TRUE,
                                                                  scrollY = "1000px",
                                                                  scrollCollapse = TRUE,
                                                                  dom="Blfrtip",
                                                                  buttons = c('csv', 'excel', 'copy')))
      
      data_maf_DNMT3A_GENIE$DNMT3A_status = Data_maf_DNMT3A_GENIE$DNMT3A_status_all
      data_maf_DNMT3A_GENIE_add = data_maf_DNMT3A_GENIE[1,]
      data_maf_DNMT3A_GENIE_add$DNMT3A_status = "DUMMY"
      data_maf_DNMT3A_GENIE_dummy = rbind(data_maf_DNMT3A_GENIE, data_maf_DNMT3A_GENIE_add)
      maf.dset_GENIE_all <- data_maf_DNMT3A_GENIE_dummy %>%
        dplyr::filter(Reference_Allele != "-" & Tumor_Seq_Allele2 != "-" &
                        nchar(Reference_Allele) == 1 &
                        nchar(Tumor_Seq_Allele2) == 1)
      maf.dset_GENIE_all = as.data.frame(maf.dset_GENIE_all)
      maf.dset_GENIE_all <- filterSNV(dataSet = maf.dset_GENIE_all,
                                  seq_colNames = c("Reference_Allele",
                                                   "Tumor_Seq_Allele2"))
      hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      maf.dset_GENIE_all <- attachContext(mutData = maf.dset_GENIE_all,
                                      chr_colName = "Chromosome",
                                      start_colName = "Start_Position",
                                      end_colName = "End_Position",
                                      nucl_contextN = 3,
                                      BSGenomeDb = hg38)
      maf.dset_GENIE_all <- removeMismatchMut(mutData = maf.dset_GENIE_all,
                                          refMut_colName = "Reference_Allele",
                                          context_colName = "context",
                                          refMut_format = "N")
      maf.dset_GENIE_all <- attachMutType(mutData = maf.dset_GENIE_all,
                                      ref_colName = "Reference_Allele",
                                      var_colName = "Tumor_Seq_Allele1",
                                      var2_colName = "Tumor_Seq_Allele2",
                                      context_colName = "context")
      maf.counts_DNMT3A_GENIE_all <- countMutTypes(mutTable = maf.dset_GENIE_all,
                                               sample_colName = "DNMT3A_status",
                                               mutType_colName = "mutType")
      # mouCancer.assess <- prelimProcessAssess(input = maf.counts,
      #                                         approach = "counts")
      
      # A[C>T]C rich: Y735C, R635W, P904L
      CCAT_liquid_only_mut = unique((Data_MAF_target %>%
                                       left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                                       dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                                       dplyr::filter(Panel == "F1Liquid CDx" &
                                                       non_germline_non_error == 1 &
                                                       unique_alt %in% CH_candidates &
                                                       Hugo_Symbol %in% CH_gene_entropy_3 &
                                                       #CpC_TpC == 1 &
                                                       #Hugo_Symbol == "DNMT3A" &
                                                       logfold > 8))$unique_alt)
      CCAT_liquid_only_DNMT3A = unique((Data_MAF_target %>%
                                          left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                                          dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                                          dplyr::filter(Panel == "F1Liquid CDx" &
                                                          non_germline_non_error == 1 &
                                                          unique_alt %in% CH_candidates &
                                                          Hugo_Symbol %in% CH_gene_entropy_3 &
                                                          #CpC_TpC == 1 &
                                                          Hugo_Symbol == "DNMT3A" &
                                                          logfold > 8))$unique_alt)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(with_liquid_mut = case_when(
          unique_alt %in% CCAT_liquid_only_mut ~ "Yes",
          TRUE ~ "No"
        ))
      ID_tmp = unique((Data_MAF_target %>% filter(with_liquid_mut == "Yes"))$Tumor_Sample_Barcode)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(with_liquid_mut = case_when(
          Tumor_Sample_Barcode %in% ID_tmp ~ "Yes",
          TRUE ~ "No"
        ))
      Data_with_liquid_only = Data_MAF_target %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        filter(with_liquid_mut == "Yes" & Panel == "F1Liquid CDx")
      Data_without_liquid_only = Data_MAF_target %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        filter(with_liquid_mut == "No" & Panel == "F1Liquid CDx")
      Data_with_liquid_only_DNMT3A = sort(table((Data_with_liquid_only %>%
                                                   filter(Hugo_Symbol == "DNMT3A" & !name %in% CCAT_liquid_only_mut))$name))
      Data_without_liquid_only_DNMT3A = sort(table((Data_without_liquid_only %>%
                                                      filter(Hugo_Symbol == "DNMT3A"))$name))
      DNMT3A_with_liquid_only = data.frame(names(Data_with_liquid_only_DNMT3A), as.integer(Data_with_liquid_only_DNMT3A))
      DNMT3A_without_liquid_only = data.frame(names(Data_without_liquid_only_DNMT3A), as.integer(Data_without_liquid_only_DNMT3A))
      colnames(DNMT3A_with_liquid_only) = c("name", "count_with_liquid_only")
      colnames(DNMT3A_without_liquid_only) = c("name", "count_without_liquid_only")
      DNMT3A_count_with_liquid = dplyr::full_join(DNMT3A_with_liquid_only, DNMT3A_without_liquid_only, by = "name")
      DNMT3A_count_with_liquid$count_with_liquid_only[is.na(DNMT3A_count_with_liquid$count_with_liquid_only)] = 0
      DNMT3A_count_with_liquid$count_without_liquid_only[is.na(DNMT3A_count_with_liquid$count_without_liquid_only)] = 0
      DNMT3A_count_with_liquid = DNMT3A_count_with_liquid[DNMT3A_count_with_liquid$count_with_liquid_only>=5,]
      DNMT3A_count_with_liquid = DNMT3A_count_with_liquid %>%
        dplyr::filter(!str_detect(name, "\\*|\\>"))
      DNMT3A_count_with_liquid$with_liquid_negative = length(unique(Data_with_liquid_only$Tumor_Sample_Barcode)) - DNMT3A_count_with_liquid$count_with_liquid_only
      DNMT3A_count_with_liquid$without_liquid_negative = length(unique(Data_without_liquid_only$Tumor_Sample_Barcode)) - DNMT3A_count_with_liquid$count_without_liquid_only
      DNMT3A_count_with_liquid$ratio = fun_zero(DNMT3A_count_with_liquid$count_with_liquid_only,DNMT3A_count_with_liquid$count_without_liquid_only) * fun_zero(DNMT3A_count_with_liquid$without_liquid_negative,DNMT3A_count_with_liquid$with_liquid_negative)
      DNMT3A_with_liquid_only_mut = (DNMT3A_count_with_liquid %>% filter(ratio > 1.5))$name
      output$table_DNMT3A_odds = DT::renderDataTable(DNMT3A_count_with_liquid,
                                                     server = FALSE,
                                                     filter = 'top', 
                                                     extensions = c('Buttons'), 
                                                     options = list(pageLength = 100, 
                                                                    scrollX = TRUE,
                                                                    scrollY = "1000px",
                                                                    scrollCollapse = TRUE,
                                                                    dom="Blfrtip",
                                                                    buttons = c('csv', 'excel', 'copy')))
      #DNMT3A_with_liquid_only_mut = c('Y735C', 'R635W', 'P904L')
      DNMT3A_735 = unique((Data_MAF_target %>%
                             left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                             dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                             dplyr::filter(Panel == "F1Liquid CDx" &
                                             non_germline_non_error == 1 &
                                             # unique_alt %in% CH_candidates &
                                             Hugo_Symbol == "DNMT3A" &
                                             # CpC_TpC == 1 &
                                             # logfold > 8))$unique_alt)
                                             #name %in% DNMT3A_with_liquid_only_mut))$unique_alt)
                                             name %in% "DNMT3A_Y735C"))$unique_alt)
                                             # str_detect(unique_alt, "Y735C|R635W|P904L")))$unique_alt)
                                             #amino.acid.change %in% DNMT3A_favor_Liquid_to_Blood))$unique_alt)
      # str_detect(unique_alt, "Y735C")))$unique_alt)
      DNMT3A_882 = unique((Data_MAF_target %>%
                             left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                             dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                             dplyr::filter(Panel == "F1Liquid CDx" &
                                             non_germline_non_error == 1 &
                                             unique_alt %in% CH_candidates &
                                             #amino.acid.change %in% DNMT3A_favor_Liquid_to_Blood &
                                             Hugo_Symbol == "DNMT3A" &
                                             # str_detect(unique_alt, "I780T|V657M|R326C|I705T|F755S|M801V|P700L|Y908C|Y660C|V649M|L639F|G332R|G796D|G706R|S337L|R729G|R635W|S638C|P799S|P904L")))$unique_alt)
                                             # str_detect(unique_alt, "I780T|V657M|R635W|R736H|R736C|P904L")))$unique_alt)
                                             # str_detect(unique_alt, "882")))$unique_alt)
                                             str_detect(unique_alt, "R882H|R882C|R882P|R882S")))$unique_alt)
      DNMT3A_count = table((Data_MAF_target %>%
                              left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                              dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                              dplyr::filter(Panel == "F1Liquid CDx" &
                                              non_germline_non_error == 1 &
                                              unique_alt %in% CH_candidates &
                                              Hugo_Symbol == "DNMT3A"))$Tumor_Sample_Barcode)
      # DNMT3A_count_single = names(DNMT3A_count[DNMT3A_count==1])
      # DNMT3A_count_multi = names(DNMT3A_count[DNMT3A_count>1])
      ID_DNMT3A_735_and_882 = intersect(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_735))$Tumor_Sample_Barcode),
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_882))$Tumor_Sample_Barcode))
      ID_DNMT3A_735 = setdiff(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_735))$Tumor_Sample_Barcode),
        ID_DNMT3A_735_and_882)
      ID_DNMT3A_882 = setdiff(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_882))$Tumor_Sample_Barcode),
        ID_DNMT3A_735_and_882)
      ID_DNMT3A_other = setdiff(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% CH_candidates &
                                  str_detect(Hugo_Symbol, "DNMT3A")))$Tumor_Sample_Barcode),
        c(ID_DNMT3A_735_and_882, ID_DNMT3A_735, ID_DNMT3A_882))
      ID_DNMT3A_none = setdiff(
        unique((Data_summary %>%
                  dplyr::filter(Panel == "F1Liquid CDx"))$Tumor_Sample_Barcode),
        c(ID_DNMT3A_735_and_882, ID_DNMT3A_735, ID_DNMT3A_882, ID_DNMT3A_other))
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(DNMT3A_status = case_when(
          Tumor_Sample_Barcode %in% ID_DNMT3A_735_and_882 ~ "CH_favor and 882",
          Tumor_Sample_Barcode %in% ID_DNMT3A_735 ~ "CH_favor",
          Tumor_Sample_Barcode %in% ID_DNMT3A_882 ~ "882",
          Tumor_Sample_Barcode %in% ID_DNMT3A_other ~ "other mut",
          Tumor_Sample_Barcode %in% ID_DNMT3A_none ~ "none",
          TRUE ~ "solid sample or error"
        ))
      df_plot <- Data_MAF_target %>%
        dplyr::filter(DNMT3A_status != "solid sample or error") %>%
        group_by(DNMT3A_status, Cancername) %>%
        summarise(count = n(), .groups = "drop") %>%
        mutate(percentage = count / sum(count) * 100)  # 割合を計算
      
      # ggplotで可視化
      g = ggplot(df_plot, aes(x = DNMT3A_status, y = percentage, fill = Cancername)) +
        geom_bar(stat = "identity", position = "fill") +
        scale_y_continuous(labels = scales::percent_format(scale = 100)) +
        labs(title = "Cancer frequency by DNMT3A status",
             x = "Cancer type",
             y = "Percentage (%)") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +  # x軸ラベルを回転
        theme_bw() +
        scale_fill_viridis_d(option = "turbo")  # カラーパレット適用
      plots[[74]] = g
      # Data_MAF_target = Data_MAF_target %>%
      #   dplyr::mutate(DNMT3A_status = case_when(
      #     Tumor_Sample_Barcode %in% DNMT3A_count_single ~ "DNMT3A single mutation",
      #     Tumor_Sample_Barcode %in% DNMT3A_count_multi ~ "DNMT3A multi mutations",
      #     TRUE ~ "No DNMT3A mutations"
      #   ))
      # Data_MAF_target = Data_MAF_target %>%
      #   dplyr::mutate(DNMT3A_status = case_when(
      #     with_liquid_mut == "Yes" ~ "with liquid only",
      #     TRUE ~ "without liquid only"
      #   ))
      Data_maf_DNMT3A = Data_MAF_target %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::filter(#Hugo_Symbol == "CHEK2" &
          unique_alt %in% CH_candidates &
            #!unique_alt %in% c("DNMT3A_R882H", "DNMT3A_R882C", DNMT3A_with_liquid_only_mut) &
            #!unique_alt %in% CCAT_liquid_only_mut &
            !unique_alt %in% c(DNMT3A_735, DNMT3A_882) &
            #!Hugo_Symbol == "DNMT3A" &
            Hugo_Symbol %in% CH_gene_entropy_3 &
            Panel == "F1Liquid CDx") %>%
        dplyr::arrange(Chromosome, Start_Position)
      data_maf_DNMT3A = Data_maf_DNMT3A %>% dplyr::select(Hugo_Symbol)
      data_maf_DNMT3A$Chromosome = paste0("chr", Data_maf_DNMT3A$Chromosome)
      data_maf_DNMT3A$Start_Position = as.integer(Data_maf_DNMT3A$Start_Position)
      data_maf_DNMT3A$End_Position = as.integer(Data_maf_DNMT3A$Start_Position) + nchar(Data_maf_DNMT3A$Reference_Allele) - 1
      data_maf_DNMT3A$Strand = "+"
      data_maf_DNMT3A$Variant_Classification = "Missense_Mutation"
      data_maf_DNMT3A$Variant_Type = ifelse(nchar(Data_maf_DNMT3A$Reference_Allele) == 1 & nchar(Data_maf_DNMT3A$Tumor_Seq_Allele2) == 1, "SNV",
                                            ifelse(nchar(Data_maf_DNMT3A$Reference_Allele) > nchar(Data_maf_DNMT3A$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf_DNMT3A$Reference_Allele = Data_maf_DNMT3A$Reference_Allele
      data_maf_DNMT3A$Tumor_Seq_Allele1 = Data_maf_DNMT3A$Reference_Allele
      data_maf_DNMT3A$Tumor_Seq_Allele2 = Data_maf_DNMT3A$Tumor_Seq_Allele2
      data_maf_DNMT3A$Tumor_Sample_Barcode = Data_maf_DNMT3A$Tumor_Sample_Barcode
      data_maf_DNMT3A$DNMT3A_status = Data_maf_DNMT3A$DNMT3A_status
      data_maf_DNMT3A$Protein_Change = Data_maf_DNMT3A$unique_alt
      maf.dset <- as.data.frame(data_maf_DNMT3A)
      maf.dset <- filterSNV(dataSet = maf.dset,
                            seq_colNames = c("Reference_Allele",
                                             "Tumor_Seq_Allele1",
                                             "Tumor_Seq_Allele2"))
      hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      maf.dset <- attachContext(mutData = maf.dset,
                                chr_colName = "Chromosome",
                                start_colName = "Start_Position",
                                end_colName = "End_Position",
                                nucl_contextN = 3,
                                BSGenomeDb = hg38)
      maf.dset <- removeMismatchMut(mutData = maf.dset,
                                    refMut_colName = "Reference_Allele",
                                    context_colName = "context",
                                    refMut_format = "N")
      maf.dset <- attachMutType(mutData = maf.dset,
                                ref_colName = "Reference_Allele",
                                var_colName = "Tumor_Seq_Allele1",
                                var2_colName = "Tumor_Seq_Allele2",
                                context_colName = "context")
      maf.counts_DNMT3A <- countMutTypes(mutTable = maf.dset,
                                         sample_colName = "DNMT3A_status",
                                         mutType_colName = "mutType")
      # mouCancer.assess <- prelimProcessAssess(input = maf.counts,
      #                                         approach = "counts")
      # maf.params_DNMT3A <- setMutClusterParams(num_processesToExtract = 3,
      #                                          approach = "counts",
      #                                          num_totIterations = 5,
      #                                          num_parallelCores = 6, debug = FALSE,
      #                                          algorithm = "alexa")
      # maf.counts.analysis_DNMT3A <- decipherMutationalProcesses(input = maf.counts_DNMT3A,
      #                                                           params = maf.params_DNMT3A)
      # ocd.signs_DNMT3A <- maf.counts.analysis_DNMT3A$Results$signatures
      # ocd.expos_DNMT3A <- maf.counts.analysis_DNMT3A$Results$exposures
      # if(file.exists("source/cosmixSigs.rda")){
      #   load(file="source/cosmixSigs.rda")
      # } else{
      #   cosmixSigs <- getCosmicSignatures()
      #   save(cosmixSigs, file="source/cosmixSigs.rda")
      # }
      # msig1_DNMT3A <- matchSignatures(mutSign = ocd.signs_DNMT3A, reference = cosmixSigs,
      #                                 threshold = 0.45, plot = TRUE)
      # hm1_DNMT3A <- msig1_DNMT3A$plot + ggtitle("Match to COSMIC signs.")
      # x_DNMT3A <- maf.counts_DNMT3A
      # xx_DNMT3A <- x_DNMT3A#as.mutation.counts(x)
      # blca.expo1_DNMT3A <- resolveMutSignatures(mutCountData = x_DNMT3A,
      #                                           signFreqData = cosmixSigs)
      # blca.exp.1x_DNMT3A <- blca.expo1_DNMT3A$Results$count.result
      # bp1_DNMT3A <- msigPlot(blca.exp.1x_DNMT3A, main = "All variants, DMNT3A mutation status | COSMIC sigs.")
      # cosmx <-cosmixSigs[c(1,3,6,26,29,30)]
      # msig1 <- matchSignatures(mutSign = ocd.signs, reference = cosmx, 
      #                          threshold = 0.45, plot = TRUE) 
      # hm2 <- msig1$plot + ggtitle("Match to COSMIC signs.")
      # x <- maf.counts
      # xx <- x#as.mutation.counts(x)
      # blca.expo1 <- resolveMutSignatures(mutCountData = x, 
      #                                    signFreqData = cosmx)
      # blca.exp.1x <- blca.expo1$Results$count.result
      # bp2 <- msigPlot(blca.exp.1x, main = "Hotspot variants, favor | COSMIC sigs.")
      # fig_msig = msigPlot(ocd.expos, main = "Hotspot variants, favor liquid/tumor/NOS") + 
      #   scale_fill_manual(values = c("#1f78b4", "#cab2d6", "#ff7f00", "#a6cee3"))
      output$table_summary_GENIE = DT::renderDataTable(data_maf_DNMT3A,
                                                       server = FALSE,
                                                       filter = 'top', 
                                                       extensions = c('Buttons'), 
                                                       options = list(pageLength = 100, 
                                                                      scrollX = TRUE,
                                                                      scrollY = "1000px",
                                                                      scrollCollapse = TRUE,
                                                                      dom="Blfrtip",
                                                                      buttons = c('csv', 'excel', 'copy')))
      
      output$figure_signature_DNMT3A = renderPlot({
        # par(mfrow = c(3, 1))
        # msigPlot(maf.counts_DNMT3A, sample = "DNMT3A single mutation", main = "DNMT3A single mutation", ylim = c(0, 0.20))
        # msigPlot(maf.counts_DNMT3A, sample = "DNMT3A multi mutations", main = "DNMT3A multi mutations", ylim = c(0, 0.20))
        # msigPlot(maf.counts_DNMT3A, sample = "No DNMT3A mutations", main = "No DNMT3A mutations", ylim = c(0, 0.20))
        par(mfrow = c(5, 1))
        msigPlot(maf.counts_DNMT3A,
                 sample = "CH_favor and 882",
                 main = paste0("C-CAT patients with DNMT3A CH-favor mut and R882 mut, ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_735_and_882),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735_and_882)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735_and_882)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A,
                 sample = "CH_favor",
                 main = paste0("C-CAT patients with DNMT3A CH-favor mut, ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_735),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "CH_favor" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A,
                 sample = "882",
                 main = paste0("C-CAT patients with DNMT3A R882 mut, ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_882),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_882)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_882)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A,
                 sample = "other mut",
                 main = paste0("C-CAT patients with other DNMT3A hotspot mutations, ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_other),
                               " pts (mean ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_other)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_other)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        setProgress(detail = "18607")
        msigPlot(maf.counts_DNMT3A,
                 sample = "none",
                 main = paste0("C-CAT patients with no DNMT3A hotspot mutations, ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-CH-favor/882-DNMT3A and ",
                               length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_none),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_none)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_none)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        # par(mfrow = c(2, 1))
        # msigPlot(maf.counts_DNMT3A, sample = "with liquid only", main = "with liquid only", ylim = c(0, 0.20))
        # msigPlot(maf.counts_DNMT3A, sample = "without liquid only", main = "without liquid only", ylim = c(0, 0.20))
        # prelimProcessAssess(input = maf.counts,
        #                     approach = "counts")
        # msigPlot(ocd.signs, signature = 1, ylim = c(0, 0.15))
        # msigPlot(ocd.signs, signature = 2, ylim = c(0, 0.15))
        # msigPlot(ocd.signs, signature = 3, ylim = c(0, 0.15))
        # msigPlot(ocd.signs, signature = 4, ylim = c(0, 0.15))
      })
      # output$figure_signature2_DNMT3A = renderPlot({
      #   grid.arrange(hm1_DNMT3A, bp1_DNMT3A, ncol = 1)
      # })
      
      DNMT3A_735_2 = unique((Data_MAF_target %>%
                             left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                             dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                             dplyr::filter(Panel == "F1Liquid CDx" &
                                             non_germline_non_error == 1 &
                                             unique_alt %in% CH_candidates &
                                             Hugo_Symbol == "DNMT3A" &
                                             # CpC_TpC == 1 &
                                             # logfold > 8))$unique_alt)
                                             # unique_alt %in% DNMT3A_with_liquid_only_mut))$unique_alt)
                                             str_detect(unique_alt, "Y735C")))$unique_alt)
                                             #amino.acid.change %in% DNMT3A_favor_Liquid_to_Blood))$unique_alt)
      # str_detect(unique_alt, "Y735C")))$unique_alt)
      DNMT3A_882_2 = unique((Data_MAF_target %>%
                             left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                             dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                             dplyr::filter(Panel == "F1Liquid CDx" &
                                             non_germline_non_error == 1 &
                                             unique_alt %in% CH_candidates &
                                             #amino.acid.change %in% DNMT3A_favor_Liquid_to_Blood &
                                             Hugo_Symbol == "DNMT3A" &
                                             # str_detect(unique_alt, "I780T|V657M|R326C|I705T|F755S|M801V|P700L|Y908C|Y660C|V649M|L639F|G332R|G796D|G706R|S337L|R729G|R635W|S638C|P799S|P904L")))$unique_alt)
                                             # str_detect(unique_alt, "I780T|V657M|R635W|R736H|R736C|P904L")))$unique_alt)
                                             # str_detect(unique_alt, "882")))$unique_alt)
                                             str_detect(unique_alt, "R882H|R882C|R882P|R882S|R882G")))$unique_alt)
      DNMT3A_count = table((Data_MAF_target %>%
                              left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                              dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                              dplyr::filter(Panel == "F1Liquid CDx" &
                                              non_germline_non_error == 1 &
                                              unique_alt %in% CH_candidates &
                                              Hugo_Symbol == "DNMT3A"))$Tumor_Sample_Barcode)
      # DNMT3A_count_single = names(DNMT3A_count[DNMT3A_count==1])
      # DNMT3A_count_multi = names(DNMT3A_count[DNMT3A_count>1])
      ID_DNMT3A_735_and_882_2 = intersect(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_735_2))$Tumor_Sample_Barcode),
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_882_2))$Tumor_Sample_Barcode))
      ID_DNMT3A_735_2 = setdiff(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_735_2))$Tumor_Sample_Barcode),
        ID_DNMT3A_735_and_882)
      ID_DNMT3A_882_2 = setdiff(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_882_2))$Tumor_Sample_Barcode),
        ID_DNMT3A_735_and_882)
      ID_DNMT3A_other_2 = setdiff(
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% CH_candidates &
                                  str_detect(Hugo_Symbol, "DNMT3A")))$Tumor_Sample_Barcode),
        c(ID_DNMT3A_735_and_882_2, ID_DNMT3A_735_2, ID_DNMT3A_882_2))
      ID_DNMT3A_none_2 = setdiff(
        unique((Data_summary %>%
                  dplyr::filter(Panel == "F1Liquid CDx"))$Tumor_Sample_Barcode),
        c(ID_DNMT3A_735_and_882_2, ID_DNMT3A_735_2, ID_DNMT3A_882_2, ID_DNMT3A_other_2))
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(DNMT3A_status = case_when(
          Tumor_Sample_Barcode %in% ID_DNMT3A_735_and_882_2 ~ "735 and 882",
          Tumor_Sample_Barcode %in% ID_DNMT3A_735_2 ~ "735",
          Tumor_Sample_Barcode %in% ID_DNMT3A_882_2 ~ "882",
          Tumor_Sample_Barcode %in% ID_DNMT3A_other_2 ~ "other mut",
          Tumor_Sample_Barcode %in% ID_DNMT3A_none_2 ~ "none",
          TRUE ~ "solid sample or error"
        ))
      # Cancernameの割合を計算
      df_plot <- Data_MAF_target %>%
        dplyr::filter(DNMT3A_status != "solid sample or error") %>%
        group_by(DNMT3A_status, Cancername) %>%
        summarise(count = n(), .groups = "drop") %>%
        mutate(percentage = count / sum(count) * 100)  # 割合を計算
      
      # ggplotで可視化
      g = ggplot(df_plot, aes(x = DNMT3A_status, y = percentage, fill = Cancername)) +
        geom_bar(stat = "identity", position = "fill") +
        scale_y_continuous(labels = scales::percent_format(scale = 100)) +
        labs(title = "Cancer frequency by DNMT3A status",
             x = "Cancer type",
             y = "Percentage (%)") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +  # x軸ラベルを回転
        theme_bw() +
        scale_fill_viridis_d(option = "turbo")  # カラーパレット適用
      plots[[73]] = g
      # Data_MAF_target = Data_MAF_target %>%
      #   dplyr::mutate(DNMT3A_status = case_when(
      #     Tumor_Sample_Barcode %in% DNMT3A_count_single ~ "DNMT3A single mutation",
      #     Tumor_Sample_Barcode %in% DNMT3A_count_multi ~ "DNMT3A multi mutations",
      #     TRUE ~ "No DNMT3A mutations"
      #   ))
      # Data_MAF_target = Data_MAF_target %>%
      #   dplyr::mutate(DNMT3A_status = case_when(
      #     with_liquid_mut == "Yes" ~ "with liquid only",
      #     TRUE ~ "without liquid only"
      #   ))
      Data_maf_DNMT3A_2 = Data_MAF_target %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::filter(#Hugo_Symbol == "CHEK2" &
          unique_alt %in% CH_candidates &
            #!unique_alt %in% c("DNMT3A_R882H", "DNMT3A_R882C", DNMT3A_with_liquid_only_mut) &
            #!unique_alt %in% CCAT_liquid_only_mut &
            !unique_alt %in% c(DNMT3A_735, DNMT3A_882) &
            #!Hugo_Symbol == "DNMT3A" &
            Hugo_Symbol %in% CH_gene_entropy_3 &
            Panel == "F1Liquid CDx") %>%
        dplyr::arrange(Chromosome, Start_Position)
      data_maf_DNMT3A_2 = Data_maf_DNMT3A_2 %>% dplyr::select(Hugo_Symbol)
      data_maf_DNMT3A_2$Chromosome = paste0("chr", Data_maf_DNMT3A_2$Chromosome)
      data_maf_DNMT3A_2$Start_Position = as.integer(Data_maf_DNMT3A_2$Start_Position)
      data_maf_DNMT3A_2$End_Position = as.integer(Data_maf_DNMT3A_2$Start_Position) + nchar(Data_maf_DNMT3A_2$Reference_Allele) - 1
      data_maf_DNMT3A_2$Strand = "+"
      data_maf_DNMT3A_2$Variant_Classification = "Missense_Mutation"
      data_maf_DNMT3A_2$Variant_Type = ifelse(nchar(Data_maf_DNMT3A_2$Reference_Allele) == 1 & nchar(Data_maf_DNMT3A_2$Tumor_Seq_Allele2) == 1, "SNV",
                                            ifelse(nchar(Data_maf_DNMT3A_2$Reference_Allele) > nchar(Data_maf_DNMT3A_2$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf_DNMT3A_2$Reference_Allele = Data_maf_DNMT3A_2$Reference_Allele
      data_maf_DNMT3A_2$Tumor_Seq_Allele1 = Data_maf_DNMT3A_2$Reference_Allele
      data_maf_DNMT3A_2$Tumor_Seq_Allele2 = Data_maf_DNMT3A_2$Tumor_Seq_Allele2
      data_maf_DNMT3A_2$Tumor_Sample_Barcode = Data_maf_DNMT3A_2$Tumor_Sample_Barcode
      data_maf_DNMT3A_2$DNMT3A_status = Data_maf_DNMT3A_2$DNMT3A_status
      data_maf_DNMT3A_2$Protein_Change = Data_maf_DNMT3A_2$unique_alt
      maf.dset_2 <- as.data.frame(data_maf_DNMT3A_2)
      maf.dset_2 <- filterSNV(dataSet = maf.dset_2,
                            seq_colNames = c("Reference_Allele",
                                             "Tumor_Seq_Allele1",
                                             "Tumor_Seq_Allele2"))
      hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      maf.dset_2 <- attachContext(mutData = maf.dset_2,
                                chr_colName = "Chromosome",
                                start_colName = "Start_Position",
                                end_colName = "End_Position",
                                nucl_contextN = 3,
                                BSGenomeDb = hg38)
      maf.dset_2 <- removeMismatchMut(mutData = maf.dset_2,
                                    refMut_colName = "Reference_Allele",
                                    context_colName = "context",
                                    refMut_format = "N")
      maf.dset_2 <- attachMutType(mutData = maf.dset_2,
                                ref_colName = "Reference_Allele",
                                var_colName = "Tumor_Seq_Allele1",
                                var2_colName = "Tumor_Seq_Allele2",
                                context_colName = "context")
      maf.counts_DNMT3A_2 <- countMutTypes(mutTable = maf.dset_2,
                                         sample_colName = "DNMT3A_status",
                                         mutType_colName = "mutType")
      # mouCancer.assess <- prelimProcessAssess(input = maf.counts,
      #                                         approach = "counts")
      # maf.params_DNMT3A_2 <- setMutClusterParams(num_processesToExtract = 3,
      #                                          approach = "counts",
      #                                          num_totIterations = 5,
      #                                          num_parallelCores = 6, debug = FALSE,
      #                                          algorithm = "alexa")
      # maf.counts.analysis_DNMT3A_2 <- decipherMutationalProcesses(input = maf.counts_DNMT3A_2,
      #                                                           params = maf.params_DNMT3A_2)
      # ocd.signs_DNMT3A_2 <- maf.counts.analysis_DNMT3A_2$Results$signatures
      # ocd.expos_DNMT3A_2 <- maf.counts.analysis_DNMT3A_2$Results$exposures
      # msig1_DNMT3A_2 <- matchSignatures(mutSign = ocd.signs_DNMT3A_2, reference = cosmixSigs,
      #                                 threshold = 0.45, plot = TRUE)
      # hm1_DNMT3A_2 <- msig1_DNMT3A_2$plot + ggtitle("Match to COSMIC signs.")
      # x_DNMT3A_2 <- maf.counts_DNMT3A_2
      # xx_DNMT3A_2 <- x_DNMT3A_2#as.mutation.counts(x)
      # blca.expo1_DNMT3A_2 <- resolveMutSignatures(mutCountData = x_DNMT3A_2,
      #                                           signFreqData = cosmixSigs)
      # blca.exp.1x_DNMT3A_2 <- blca.expo1_DNMT3A_2$Results$count.result
      # bp1_DNMT3A_2 <- msigPlot(blca.exp.1x_DNMT3A_2, main = "All variants, DMNT3A mutation status | COSMIC sigs.")
      # cosmx <-cosmixSigs[c(1,3,6,26,29,30)]
      # msig1 <- matchSignatures(mutSign = ocd.signs, reference = cosmx, 
      #                          threshold = 0.45, plot = TRUE) 
      # hm2 <- msig1$plot + ggtitle("Match to COSMIC signs.")
      # x <- maf.counts
      # xx <- x#as.mutation.counts(x)
      # blca.expo1 <- resolveMutSignatures(mutCountData = x, 
      #                                    signFreqData = cosmx)
      # blca.exp.1x <- blca.expo1$Results$count.result
      # bp2 <- msigPlot(blca.exp.1x, main = "Hotspot variants, favor | COSMIC sigs.")
      # fig_msig = msigPlot(ocd.expos, main = "Hotspot variants, favor liquid/tumor/NOS") + 
      #   scale_fill_manual(values = c("#1f78b4", "#cab2d6", "#ff7f00", "#a6cee3"))
      output$table_summary_GENIE_2 = DT::renderDataTable(data_maf_DNMT3A_2,
                                                       server = FALSE,
                                                       filter = 'top', 
                                                       extensions = c('Buttons'), 
                                                       options = list(pageLength = 100, 
                                                                      scrollX = TRUE,
                                                                      scrollY = "1000px",
                                                                      scrollCollapse = TRUE,
                                                                      dom="Blfrtip",
                                                                      buttons = c('csv', 'excel', 'copy')))
      
      output$figure_signature_DNMT3A_2 = renderPlot({
        # par(mfrow = c(3, 1))
        # msigPlot(maf.counts_DNMT3A, sample = "DNMT3A single mutation", main = "DNMT3A single mutation", ylim = c(0, 0.20))
        # msigPlot(maf.counts_DNMT3A, sample = "DNMT3A multi mutations", main = "DNMT3A multi mutations", ylim = c(0, 0.20))
        # msigPlot(maf.counts_DNMT3A, sample = "No DNMT3A mutations", main = "No DNMT3A mutations", ylim = c(0, 0.20))
        par(mfrow = c(5, 1))
        msigPlot(maf.counts_DNMT3A_2,
                 sample = "735 and 882",
                 main = paste0("C-CAT patients with DNMT3A Y735C mut and R882 mut, ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-Y735C/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_735_and_882_2),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735_and_882_2)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735_and_882_2)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.30))
        msigPlot(maf.counts_DNMT3A_2,
                 sample = "735",
                 main = paste0("C-CAT patients with DNMT3A Y735C mut, ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-Y735C/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_735_2),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735_2)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_735_2)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A_2,
                 sample = "882",
                 main = paste0("C-CAT patients with DNMT3A R882 mut, ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-Y735C/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_882_2),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_882_2)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_882_2)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A_2,
                 sample = "other mut",
                 main = paste0("C-CAT patients with other DNMT3A hotspot mutations, ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-Y735C/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_other_2),
                               " pts (mean ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_other_2)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_other_2)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        msigPlot(maf.counts_DNMT3A_2,
                 sample = "none",
                 main = paste0("C-CAT patients with no DNMT3A hotspot mutations, ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                               " non-Y735C/882-DNMT3A and ",
                               length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                               " other hotspot muts in ",
                               length(ID_DNMT3A_none_2),
                               " pts (mean, ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_none_2)),
                                      digits=2),
                               " and ",
                               format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                               length(ID_DNMT3A_none_2)),
                                      digits=2),
                               " muts/pt)"),
                 ylim = c(0, 0.20))
        # par(mfrow = c(2, 1))
        # msigPlot(maf.counts_DNMT3A, sample = "with liquid only", main = "with liquid only", ylim = c(0, 0.20))
        # msigPlot(maf.counts_DNMT3A, sample = "without liquid only", main = "without liquid only", ylim = c(0, 0.20))
        # prelimProcessAssess(input = maf.counts,
        #                     approach = "counts")
        # msigPlot(ocd.signs, signature = 1, ylim = c(0, 0.15))
        # msigPlot(ocd.signs, signature = 2, ylim = c(0, 0.15))
        # msigPlot(ocd.signs, signature = 3, ylim = c(0, 0.15))
        # msigPlot(ocd.signs, signature = 4, ylim = c(0, 0.15))
      })
      # output$figure_signature2_DNMT3A_2 = renderPlot({
      #   grid.arrange(hm1_DNMT3A_2, bp1_DNMT3A_2, ncol = 1)
      # })
      ID_DNMT3A_735_3 = 
        unique((Data_MAF_target %>%
                  left_join(F1L_germline_error, by="Hugo_Symbol") %>%
                  dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
                  dplyr::filter(Panel == "F1Liquid CDx" &
                                  unique_alt %in% DNMT3A_735_2))$Tumor_Sample_Barcode)
      ID_DNMT3A_none_3 = setdiff(
        unique((Data_summary %>%
                  dplyr::filter(Panel == "F1Liquid CDx"))$Tumor_Sample_Barcode),
        ID_DNMT3A_735_3)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::mutate(DNMT3A_status = case_when(
          Tumor_Sample_Barcode %in% ID_DNMT3A_735_3 ~ "735",
          Tumor_Sample_Barcode %in% ID_DNMT3A_none_3 ~ "other",
          TRUE ~ "solid sample or error"
        ))
      Data_maf_DNMT3A_3 = Data_MAF_target %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF)) %>%
        dplyr::filter(#Hugo_Symbol == "CHEK2" &
          unique_alt %in% CH_candidates &
            #!unique_alt %in% c("DNMT3A_R882H", "DNMT3A_R882C", DNMT3A_with_liquid_only_mut) &
            #!unique_alt %in% CCAT_liquid_only_mut &
            !unique_alt %in% c(DNMT3A_735) &
            #!Hugo_Symbol == "DNMT3A" &
            Hugo_Symbol %in% CH_gene_entropy_3 &
            Panel == "F1Liquid CDx") %>%
        dplyr::arrange(Chromosome, Start_Position)
      data_maf_DNMT3A_3 = Data_maf_DNMT3A_3 %>% dplyr::select(Hugo_Symbol)
      data_maf_DNMT3A_3$Chromosome = paste0("chr", Data_maf_DNMT3A_3$Chromosome)
      data_maf_DNMT3A_3$Start_Position = as.integer(Data_maf_DNMT3A_3$Start_Position)
      data_maf_DNMT3A_3$End_Position = as.integer(Data_maf_DNMT3A_3$Start_Position) + nchar(Data_maf_DNMT3A_3$Reference_Allele) - 1
      data_maf_DNMT3A_3$Strand = "+"
      data_maf_DNMT3A_3$Variant_Classification = "Missense_Mutation"
      data_maf_DNMT3A_3$Variant_Type = ifelse(nchar(Data_maf_DNMT3A_3$Reference_Allele) == 1 & nchar(Data_maf_DNMT3A_3$Tumor_Seq_Allele2) == 1, "SNV",
                                              ifelse(nchar(Data_maf_DNMT3A_3$Reference_Allele) > nchar(Data_maf_DNMT3A_3$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf_DNMT3A_3$Reference_Allele = Data_maf_DNMT3A_3$Reference_Allele
      data_maf_DNMT3A_3$Tumor_Seq_Allele1 = Data_maf_DNMT3A_3$Reference_Allele
      data_maf_DNMT3A_3$Tumor_Seq_Allele2 = Data_maf_DNMT3A_3$Tumor_Seq_Allele2
      data_maf_DNMT3A_3$Tumor_Sample_Barcode = Data_maf_DNMT3A_3$Tumor_Sample_Barcode
      data_maf_DNMT3A_3$DNMT3A_status = Data_maf_DNMT3A_3$DNMT3A_status
      data_maf_DNMT3A_3$Protein_Change = Data_maf_DNMT3A_3$unique_alt
      maf.dset_3 <- as.data.frame(data_maf_DNMT3A_3)
      maf.dset_3 <- filterSNV(dataSet = maf.dset_3,
                              seq_colNames = c("Reference_Allele",
                                               "Tumor_Seq_Allele1",
                                               "Tumor_Seq_Allele2"))
      hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      maf.dset_3 <- attachContext(mutData = maf.dset_3,
                                  chr_colName = "Chromosome",
                                  start_colName = "Start_Position",
                                  end_colName = "End_Position",
                                  nucl_contextN = 3,
                                  BSGenomeDb = hg38)
      maf.dset_3 <- removeMismatchMut(mutData = maf.dset_3,
                                      refMut_colName = "Reference_Allele",
                                      context_colName = "context",
                                      refMut_format = "N")
      maf.dset_3 <- attachMutType(mutData = maf.dset_3,
                                  ref_colName = "Reference_Allele",
                                  var_colName = "Tumor_Seq_Allele1",
                                  var2_colName = "Tumor_Seq_Allele2",
                                  context_colName = "context")
      maf.counts_DNMT3A_3 <- countMutTypes(mutTable = maf.dset_3,
                                           sample_colName = "DNMT3A_status",
                                           mutType_colName = "mutType")
      # mouCancer.assess <- prelimProcessAssess(input = maf.counts,
      #                                         approach = "counts")
      # maf.params_DNMT3A_3 <- setMutClusterParams(num_processesToExtract = 3,
      #                                            approach = "counts",
      #                                            num_totIterations = 5,
      #                                            num_parallelCores = 6, debug = FALSE,
      #                                            algorithm = "alexa")
      # maf.counts.analysis_DNMT3A_3 <- decipherMutationalProcesses(input = maf.counts_DNMT3A_3,
      #                                                             params = maf.params_DNMT3A_3)
      # ocd.signs_DNMT3A_3 <- maf.counts.analysis_DNMT3A_3$Results$signatures
      # ocd.expos_DNMT3A_3 <- maf.counts.analysis_DNMT3A_3$Results$exposures
      # msig1_DNMT3A_3 <- matchSignatures(mutSign = ocd.signs_DNMT3A_3, reference = cosmixSigs,
      #                                   threshold = 0.45, plot = TRUE)
      setProgress(detail = "19030")
      Data_MAF_DNMT3A = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$logfold > 8, "LOCH", "TOCH")
      # Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$mut_pattern %in% c("CpC>TpC"), "LOCH", "TOCH")
      # Data_MAF_DNMT3A$color_group = ifelse(!Data_MAF_DNMT3A$mut_pattern %in% c("CpG>TpG"), "LOCH", "TOCH")
      # Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$Hugo_Symbol %in% c("SF3B1", "U2AF1"), "LOCH", "TOCH")
      Data_MAF_DNMT3A_join = Data_MAF_DNMT3A %>%
        add_count(Tumor_Sample_Barcode, name = "CH_count") %>%
        dplyr::select(Tumor_Sample_Barcode, CH_count) %>%
        dplyr::distinct()
      Data_summary_F1L = Data_summary %>% filter(Panel == "F1Liquid CDx")
      Data_summary_F1L = left_join(Data_summary_F1L, Data_MAF_DNMT3A_join, by="Tumor_Sample_Barcode")
      Data_summary_F1L$CH_count[is.na(Data_summary_F1L$CH_count)] = 0
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("LOCH", "TOCH"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      # 5. 年齢ごと & `color_group` ごとに統計情報を計算（全症例基準）
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        ) %>% distinct()
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0, total_patients = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      # 8. LOCH率（LOCH / (LOCH + TOCH)）の計算
      # LOCH割合を計算
      loh_rate_by_age <- final_data %>%
        select(Age, color_group, mean_mutation) %>%
        pivot_wider(names_from = color_group, values_from = mean_mutation) %>%
        mutate(
          LOCH_ratio = if_else((LOCH + TOCH) > 0, LOCH / (LOCH + TOCH), NA_real_)
        )
      # loessでスムージング（NAを除いて）
      loess_fit <- loess(LOCH_ratio ~ Age, data = loh_rate_by_age, span = 0.8)
      loh_rate_by_age$LOCH_ratio_smooth <- predict(loess_fit, newdata = loh_rate_by_age$Age)
      # 9. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Count per Patient by Age",
             x = "Age",
             y = "Average LD-CH/LE-CH Count per Patient") +
        geom_line(data = loh_rate_by_age,
                  aes(x = Age, y = LOCH_ratio_smooth * 3),
                  color = "black", linetype = "dashed", size = 1.2) +
        scale_y_continuous(
          name = "Average LD-CH/LE-CH Count per Patient",
          sec.axis = sec_axis(~ . / 3,
                              name = "Proportion of LOCH (LOCH / (LOCH + TOCH))")
        ) +
        theme_minimal()
      plots[[213]] = g
      Data_MAF_DNMT3A = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$logfold > 8, "LOCH", "TOCH")
      # Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$mut_pattern %in% c("CpC>TpC"), "LOCH", "TOCH")
      # Data_MAF_DNMT3A$color_group = ifelse(!Data_MAF_DNMT3A$mut_pattern %in% c("CpG>TpG"), "LOCH", "TOCH")
      # Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$Hugo_Symbol %in% c("SF3B1", "U2AF1"), "LOCH", "TOCH")
      Data_MAF_DNMT3A_join = Data_MAF_DNMT3A %>%
        add_count(Tumor_Sample_Barcode, name = "CH_count") %>%
        dplyr::select(Tumor_Sample_Barcode, CH_count) %>%
        dplyr::distinct()
      Data_summary_F1L = Data_summary %>% filter(Panel == "F1Liquid CDx")
      Data_summary_F1L = left_join(Data_summary_F1L, Data_MAF_DNMT3A_join, by="Tumor_Sample_Barcode")
      Data_summary_F1L$CH_count[is.na(Data_summary_F1L$CH_count)] = 0
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      mutation_per_case$mutation_count = 1
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("LOCH", "TOCH", "Either"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      loch_toch_sum <- all_mutations %>%
        filter(color_group %in% c("LOCH", "TOCH")) %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(loch_toch_count = sum(mutation_count), .groups = "drop")
      # Either行に結合して処理
      all_mutations <- all_mutations %>%
        left_join(loch_toch_sum, by = "Tumor_Sample_Barcode") %>%
        mutate(mutation_count = if_else(
          color_group == "Either",
          if_else(loch_toch_count == 0, 0, 1),
          mutation_count
        )) %>%
        select(-loch_toch_count)  # 不要になった列を削除
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      # 5. 年齢ごと & `color_group` ごとに統計情報を計算（全症例基準）
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        ) %>% distinct()
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0, total_patients = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      # 8. LOCH率（LOCH / (LOCH + TOCH)）の計算
      # LOCH割合を計算
      loh_rate_by_age <- final_data %>%
        select(Age, color_group, mean_mutation) %>%
        pivot_wider(names_from = color_group, values_from = mean_mutation) %>%
        mutate(
          LOCH_ratio = if_else(Either > 0, LOCH / Either, NA_real_)
        )
      # loessでスムージング（NAを除いて）
      loess_fit <- loess(LOCH_ratio ~ Age, data = loh_rate_by_age, span = 0.8)
      loh_rate_by_age$LOCH_ratio_smooth <- predict(loess_fit, newdata = loh_rate_by_age$Age)
      # 9. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Count per Patient by Age",
             x = "Age",
             y = "Proportion of Patients with CH Variant") +
        geom_line(data = loh_rate_by_age,
                  aes(x = Age, y = LOCH_ratio_smooth * 2.5),
                  color = "black", linetype = "dashed", size = 1.2) +
        scale_y_continuous(
          name = "Proportion of Patients with CH Variant",
          sec.axis = sec_axis(~ . / 2.5,
                              name = "Proportion of LOCH (LOCH / (LOCH or TOCH))")
        ) +
        theme_minimal()
      plots[[216]] = g
      Data_MAF_DNMT3A = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4") &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$logfold > 8, "LOCH", "TOCH")
      Data_MAF_DNMT3A_join = Data_MAF_DNMT3A %>%
        add_count(Tumor_Sample_Barcode, name = "CH_count") %>%
        dplyr::select(Tumor_Sample_Barcode, CH_count) %>%
        dplyr::distinct()
      Data_summary_F1L = Data_summary %>% filter(Panel == "F1Liquid CDx")
      Data_summary_F1L = left_join(Data_summary_F1L, Data_MAF_DNMT3A_join, by="Tumor_Sample_Barcode")
      Data_summary_F1L$CH_count[is.na(Data_summary_F1L$CH_count)] = 0
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("LOCH", "TOCH"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      # 5. 年齢ごと & `color_group` ごとに統計情報を計算（全症例基準）
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        )
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0, total_patients = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      # 8. LOCH率（LOCH / (LOCH + TOCH)）の計算
      # LOCH割合を計算
      loh_rate_by_age <- final_data %>%
        select(Age, color_group, mean_mutation) %>%
        pivot_wider(names_from = color_group, values_from = mean_mutation) %>%
        mutate(
          LOCH_ratio = if_else((LOCH + TOCH) > 0, LOCH / (LOCH + TOCH), NA_real_)
        )
      # loessでスムージング（NAを除いて）
      loess_fit <- loess(LOCH_ratio ~ Age, data = loh_rate_by_age, span = 0.8)
      loh_rate_by_age$LOCH_ratio_smooth <- predict(loess_fit, newdata = loh_rate_by_age$Age)
      # 9. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Count per Patient by Age",
             x = "Age",
             y = "Average LD-CH/LE-CH Count per Patient") +
        geom_line(data = loh_rate_by_age,
                  aes(x = Age, y = LOCH_ratio_smooth * 1),
                  color = "black", linetype = "dashed", size = 1.2) +
        scale_y_continuous(
          name = "Average LD-CH/LE-CH DNA_repair gene variant Count per Patient",
          sec.axis = sec_axis(~ . / 1,
                              name = "Proportion of LOCH (LOCH / (LOCH + TOCH))")
        ) +
        theme_minimal()
      plots[[214]] = g
      
      Data_MAF_DNMT3A = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Hugo_Symbol %in% c("DNMT3A", "TET2", "ASXL1") &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$logfold > 8, "LOCH", "TOCH")
      Data_MAF_DNMT3A_join = Data_MAF_DNMT3A %>%
        add_count(Tumor_Sample_Barcode, name = "CH_count") %>%
        dplyr::select(Tumor_Sample_Barcode, CH_count) %>%
        dplyr::distinct()
      Data_summary_F1L = Data_summary %>% filter(Panel == "F1Liquid CDx")
      Data_summary_F1L = left_join(Data_summary_F1L, Data_MAF_DNMT3A_join, by="Tumor_Sample_Barcode")
      Data_summary_F1L$CH_count[is.na(Data_summary_F1L$CH_count)] = 0
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("LOCH", "TOCH"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      # 5. 年齢ごと & `color_group` ごとに統計情報を計算（全症例基準）
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        )
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0, total_patients = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      # 8. LOCH率（LOCH / (LOCH + TOCH)）の計算
      # LOCH割合を計算
      loh_rate_by_age <- final_data %>%
        select(Age, color_group, mean_mutation) %>%
        pivot_wider(names_from = color_group, values_from = mean_mutation) %>%
        mutate(
          LOCH_ratio = if_else((LOCH + TOCH) > 0, LOCH / (LOCH + TOCH), NA_real_)
        )
      # loessでスムージング（NAを除いて）
      loess_fit <- loess(LOCH_ratio ~ Age, data = loh_rate_by_age, span = 0.8)
      loh_rate_by_age$LOCH_ratio_smooth <- predict(loess_fit, newdata = loh_rate_by_age$Age)
      # 9. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Count per Patient by Age",
             x = "Age",
             y = "Average LD-CH/LE-CH Count per Patient") +
        geom_line(data = loh_rate_by_age,
                  aes(x = Age, y = LOCH_ratio_smooth * 2),
                  color = "black", linetype = "dashed", size = 1.2) +
        scale_y_continuous(
          name = "Average LD-CH/LE-CH Epigenetic gene variant Count per Patient",
          sec.axis = sec_axis(~ . / 2,
                              name = "Proportion of LOCH (LOCH / (LOCH + TOCH))")
        ) +
        theme_minimal()
      plots[[215]] = g

      Data_MAF_DNMT3A = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
            Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      Data_MAF_DNMT3A$color_group = ifelse(Data_MAF_DNMT3A$Hugo_Symbol %in% c("DNMT3A", "TET2", "IDH2"), "DNA_methyl",
                                    ifelse(Data_MAF_DNMT3A$Hugo_Symbol %in% c("ASXL1", "EZH2", "KMT2D"), "Histone",
                                    ifelse(Data_MAF_DNMT3A$Hugo_Symbol %in% c("TP53", "CHEK2", "ATM", "RAD21", "MDM4"), "DNA_repair",
                                    ifelse(Data_MAF_DNMT3A$Hugo_Symbol %in% c("PTPN11", "KRAS", "CBL"), "MAPK",
                                    ifelse(Data_MAF_DNMT3A$Hugo_Symbol %in% c("SF3B1", "U2AF1"), "Splicing","Other_gene")))))
      Data_MAF_DNMT3A_join = Data_MAF_DNMT3A %>%
        add_count(Tumor_Sample_Barcode, name = "CH_count") %>%
        dplyr::select(Tumor_Sample_Barcode, CH_count) %>%
        dplyr::distinct()
      Data_summary_F1L = Data_summary %>% filter(Panel == "F1Liquid CDx")
      Data_summary_F1L = left_join(Data_summary_F1L, Data_MAF_DNMT3A_join, by="Tumor_Sample_Barcode")
      Data_summary_F1L$CH_count[is.na(Data_summary_F1L$CH_count)] = 0
      patient_mut_count <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age, CH_count)
      age_avg_mutation <- Data_summary_F1L %>%
        filter(!is.na(CH_count)) %>%
        group_by(Age) %>%
        summarise(avg_mutation = mean(CH_count), .groups = "drop",
                  lower = ifelse(n() > 1, avg_mutation - 1.96 * sd(CH_count, na.rm = TRUE) / sqrt(n()), avg_mutation),
                  upper = ifelse(n() > 1, avg_mutation + 1.96 * sd(CH_count, na.rm = TRUE) / sqrt(n()), avg_mutation)) %>%
        arrange(Age)
      g = ggplot(age_avg_mutation, aes(x = Age, y = avg_mutation)) +
        geom_point(alpha = 0.3, color = "grey") + # 生データを薄くプロット
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, color = "blue") +  # 信頼区間の線
        geom_smooth(method = "loess", span = 0.8, color = "red", size = 1.2) + # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Count per Patient by Age",
             x = "Age",
             y = "Average LD-CH/LE-CH Count per Patient") +
        theme_minimal()
      plots[[93]] = g
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("DNA_methyl", "Histone", "DNA_repair", "Splicing", "MAPK", "Other_gene"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      
      # 5. 年齢ごと & `color_group` ごとに統計情報を計算（全症例基準）
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        )
      
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0, total_patients = 0))
      
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Count per Patient by Age",
             x = "Age",
             y = "Average LD-CH/LE-CH Count per Patient") +
        theme_minimal()
      plots[[97]] = g
      # 1. 全症例リスト（年齢情報含む）
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("DNA_methyl", "Histone", "DNA_repair", "Splicing", "MAPK", "Other_gene"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      all_mutations$mutation_count = fun_zero(all_mutations$mutation_count, all_mutations$mutation_count)
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      # 5. 年齢ごと & `color_group` ごとに変異を持つ症例の割合を計算
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        )
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = paste0("LD-CH/LE-CH presence ratio, ", length(unique(Data_summary_F1L$Tumor_Sample_Barcode)), " patients"),
             x = "Age",
             y = "LD-CH/LE-CH presence rate") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())
      plots[[98]] = g
      # 1. 全症例リスト（年齢情報含む）
      all_cases <- Data_MAF_DNMT3A %>%
        select(Tumor_Sample_Barcode, Age)
      all_cases_color_group <- Data_MAF_DNMT3A %>%
        select(Tumor_Sample_Barcode, Age, color_group)
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = VAF, .groups = "drop")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("DNA_methyl", "Histone", "DNA_repair", "Splicing", "MAPK", "Other_gene"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases_color_group %>%
        group_by(Age, color_group) %>%
        summarise(total_patients = n(), .groups = "drop")
      # 5. 年齢ごと & `color_group` ごとに変異を持つ症例の割合を計算
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = c("Age", "color_group")) %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = total_patients,  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        )
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = paste0("LD-CH/LE-CH variant allele frequency, ", length(unique(Data_summary_F1L$Tumor_Sample_Barcode)), "patients"),
             x = "Age",
             y = "Mean variant allele frequency") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())      
      plots[[99]] = g
      patient_mut_status <- patient_mut_count
      patient_mut_status$CH_count = fun_zero(patient_mut_status$CH_count,patient_mut_status$CH_count)
      age_mutation_rate <- patient_mut_status %>%
        group_by(Age) %>%
        summarise(mutation_rate = mean(CH_count), .groups = "drop",
                  lower = ifelse(n() > 1, mutation_rate - 1.96 * sd(CH_count, na.rm = TRUE) / sqrt(n()), mutation_rate),
                  upper = ifelse(n() > 1, mutation_rate + 1.96 * sd(CH_count, na.rm = TRUE) / sqrt(n()), mutation_rate)) %>%
        arrange(Age)
      g = ggplot(age_mutation_rate, aes(x = Age, y = mutation_rate)) +
        geom_point(alpha = 0.3, color = "grey") + # 生データを薄くプロット
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, color = "blue") +  # 信頼区間の線
        geom_smooth(method = "loess", span = 0.8, color = "red", size = 1.2) + # loess平滑化
        labs(title = paste0("LD-CH/LE-CH occurrence rate, ", length(unique(Data_summary_F1L$Tumor_Sample_Barcode)), " patients"),
             x = "Age",
             y = "Proportion of patients with LD-CH/LE-CH") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())
      plots[[94]] = g
      age_vaf_summary <- Data_MAF_DNMT3A %>%
        group_by(Age) %>%
        summarise(mean_VAF = mean(VAF, na.rm = TRUE), .groups = "drop",
                  lower = ifelse(n() > 1, mean_VAF - 1.96 * sd(VAF, na.rm = TRUE) / sqrt(n()), mean_VAF),
                  upper = ifelse(n() > 1, mean_VAF + 1.96 * sd(VAF, na.rm = TRUE) / sqrt(n()), mean_VAF)) %>%
        arrange(Age)
      g = ggplot(age_vaf_summary, aes(x = Age, y = mean_VAF)) +
        geom_point(alpha = 0.3, color = "grey") + # 生データを薄くプロット
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, color = "blue") +  # 信頼区間の線
        geom_smooth(method = "loess", span = 0.8, color = "red", size = 1.2) +  # なめらかな曲線
        labs(title = "C-CAT, variant allele frequency by age",
             subtitle = "Average VAF across different ages",
             x = "Age",
             y = "Average VAF") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())
      plots[[95]] = g
      total_patients <- length(unique(Data_summary_F1L$Tumor_Sample_Barcode))
      mutation_counts <- Data_MAF_DNMT3A %>%
        dplyr::group_by(Hugo_Symbol) %>%
        dplyr::summarise(
          patient_count = n_distinct(Tumor_Sample_Barcode),
          color_group = first(color_group)
        ) %>%
        dplyr::mutate(Percent = patient_count / total_patients * 100)
      g <- ggplot(mutation_counts, aes(x = reorder(Hugo_Symbol, -Percent), y = Percent, fill = color_group)) +
        geom_bar(stat = "identity") +
        geom_text(aes(label = patient_count), vjust = -0.3, size = 3.5) +  # 上に症例数を表示
        labs(title = paste0("Number of patients with LD-CH/LE-CH variants, ",total_patients, " F1-Liquid CDx patients"),
             x = "Gene",
             y = "Variant rate (%)") +
        theme_minimal() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        ylim(0, max(mutation_counts$Percent) * 1.1)  # テキストスペース確保
      plots[[96]] = g
      total_patients <- length(unique(Data_summary_F1L$Tumor_Sample_Barcode))
      mutation_counts <- Data_MAF_DNMT3A %>%
        dplyr::filter(logfold > 8) %>%
        dplyr::group_by(Hugo_Symbol) %>%
        dplyr::summarise(
          patient_count = n_distinct(Tumor_Sample_Barcode),
          color_group = first(color_group)
        ) %>%
        dplyr::mutate(Percent = patient_count / total_patients * 100)
      g <- ggplot(mutation_counts, aes(x = reorder(Hugo_Symbol, -Percent), y = Percent, fill = color_group)) +
        geom_bar(stat = "identity") +
        geom_text(aes(label = patient_count), vjust = -0.3, size = 3.5) +  # 上に症例数を表示
        labs(title = paste0("Number of patients with LD-CH variants, ",total_patients, " F1-Liquid CDx patients"),
             x = "Gene",
             y = "Variant rate (%)") +
        theme_minimal() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        ylim(0, max(mutation_counts$Percent) * 1.1)  # テキストスペース確保
      plots[[35]] = g
      # 1. 頻度の高い遺伝子トップ10を抽出
      top10_genes <- Data_MAF_DNMT3A %>%
        count(Hugo_Symbol, sort = TRUE) %>%
        top_n(10, n) %>%
        pull(Hugo_Symbol)
      gene_freq <- names(sort(table(Data_MAF_DNMT3A$Hugo_Symbol),decreasing = T)[10:1])
      # 変異情報を症例ごとにまとめる（変異がある場合 1, ない場合 0）
      mutation_status <- Data_MAF_DNMT3A %>%
        filter(Hugo_Symbol %in% top10_genes) %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol) %>%
        distinct() %>%
        mutate(Mutation = 1) %>%
        pivot_wider(names_from = Hugo_Symbol, values_from = Mutation, values_fill = list(Mutation = 0))
      # 症例データと結合（Smoking, Alcohol, Treatment情報を含む）
      Data_taxane_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "taxel")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Taxane = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_taxane_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_Topo2_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Etoposide|ubicin")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Topo2 = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_Topo2_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_Topo1_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "otecan")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Topo1 = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_Topo1_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_nucleic_acid_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "tabine|racil")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Nucleic_acid = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_nucleic_acid_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_alkyl_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
      dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Ifosfamide|Cyclophosphamide")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Alkylate = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_alkyl_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_ICI_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Nivolumab|Ipilimumab|Pembrolizumab|Atezolizumab|Tremelimumab|Avelumab|Durvalumab")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$ICI = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_ICI_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_summary_F1L$Age_10Year_increase = Data_summary_F1L$Age / 10
      all_data <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age_10Year_increase, Sex, Smoking_history, Alcoholic_history, Lines_pre_CGP_2_or_more, Platinum,
               Taxane, Topo1, Topo2, Nucleic_acid, Alkylate, ICI) %>%
        left_join(mutation_status, by = "Tumor_Sample_Barcode") %>%
        mutate(across(all_of(top10_genes), ~replace_na(.x, 0)))  # 変異のない遺伝子を0にする
      all_data = all_data %>%
        dplyr::mutate(
          Lines_pre_CGP_2_or_more = case_when(
            Lines_pre_CGP_2_or_more == 1 ~ "Yes",
            TRUE ~"No"
          )
        )
      filtered_data <- all_data %>%
        filter(
          Smoking_history != "Unknown", 
          Alcoholic_history != "Unknown", 
          Lines_pre_CGP_2_or_more != "Unknown", 
          Platinum != "Unknown"
        )
      filtered_data$`Smoking history` = filtered_data$Smoking_history
        filtered_data$`Treatment >=2 lines` = filtered_data$Lines_pre_CGP_2_or_more
        filtered_data$`Topo-II inhibitor` = filtered_data$Topo2
        filtered_data$`Age, 10-Year increase` = filtered_data$Age_10Year_increase
        filtered_data$`Immune checkpoint inhibitor` = filtered_data$ICI
        count_yes <- filtered_data %>%
        summarise(
          Age_yes = sum(!is.na(Age_10Year_increase)),
          Smoking_yes = sum(Smoking_history == "Yes"),
          Alcohol_yes = sum(Alcoholic_history == "Yes"),
          Lines_yes = sum(Lines_pre_CGP_2_or_more == "Yes"),
          Platinum_yes = sum(Platinum == "Yes"),
          Taxane_yes = sum(Taxane == "Yes"),
          Topo1_yes = sum(Topo1 == "Yes"),
          Topo2_yes = sum(Topo2 == "Yes"),
          Nucleic_acid_yes = sum(Nucleic_acid == "Yes"),
          ICI_yes = sum(ICI == "Yes"),
          Alkylate_yes = sum(Alkylate == "Yes")
                  )
      # 各因子のYes症例数をまとめたデータフレームを作成
      count_yes_long <- tibble(
        # term = c("Smoking_history", "Alcoholic_history", "Lines_pre_CGP_2_or_more", "Platinum",
        #          "Taxane", "Topo1", "Topo2", "Nucleic_acid", "Alkylate"),
        # Yes_count = c(count_yes$Smoking_yes, count_yes$Alcohol_yes, count_yes$Lines_yes, count_yes$Platinum_yes,
        #               count_yes$Taxane_yes, count_yes$Topo1_yes, count_yes$Topo2_yes, count_yes$Nucleic_acid_yes, count_yes$Alkylate_yes)
        term = c("Smoking history", "Age, 10-Year increase", "Treatment >=2 lines", "Platinum",
                 "Taxane", "Topo-II inhibitor", "Alkylate", "Immune checkpoint inhibitor"),
        Yes_count = c(count_yes$Smoking_yes, count_yes$Age_yes, count_yes$Lines_yes, count_yes$Platinum_yes,
                      count_yes$Taxane_yes, count_yes$Topo2_yes, count_yes$Alkylate_yes, count_yes$ICI_yes)
      )
      n_subjects <- nrow(filtered_data)
      # 因子リスト
      factors <- c("Age, 10-Year increase", "Sex", "Smoking history", "Alcoholic_history", "Treatment >=2 lines","Platinum",
                   "Taxane", "Topo1", "Topo-II inhibitor", "Nucleic_acid", "Alkylate", "Immune checkpoint inhibitor")
      logistic_regression <- function(gene, data) {
        # 遺伝子の変異情報を2値に変換
        data$mutation <- ifelse(data[[gene]] == 1, 1, 0)  # 変異があれば1、なければ0
        # ロジスティック回帰モデル
        model <- glm(mutation ~ `Smoking history` + Alcoholic_history + `Treatment >=2 lines` + Platinum + `Age, 10-Year increase` + Sex +
                       Taxane + Topo1 + `Topo-II inhibitor` + Nucleic_acid + Alkylate +`Immune checkpoint inhibitor`, 
                     data = data, 
                     family = binomial)
        # 結果の抽出
        tidy_model <- tidy(model)
        # オッズ比の計算（係数の指数）
        tidy_model <- tidy_model %>%
          mutate(
            OR = exp(estimate),
            CI_lower = exp(estimate - 1.96 * std.error),  # 信頼区間の下限
            CI_upper = exp(estimate + 1.96 * std.error),  # 信頼区間の上限
            p_value = p.value
          ) %>%
          filter(term != "(Intercept)")  # 切片項を除外
        tidy_model$Gene <- gene
        return(tidy_model)
      }
      # 各遺伝子に対してロジスティック回帰を実行
      logistic_results <- lapply(top10_genes, logistic_regression, data = filtered_data)
      # 結果をデータフレームにまとめる
      logistic_results_df <- bind_rows(logistic_results)
      # Q値を計算 (Benjamini-Hochberg法)
      logistic_results_df <- logistic_results_df %>%
        mutate(Q_value = p.adjust(p_value, method = "none"))
      # 有意水準を追加
      logistic_results_df <- logistic_results_df %>%
        mutate(
          Significance = case_when(
            Q_value < 0.001 ~ "***",
            Q_value < 0.01  ~ "**",
            Q_value < 0.05  ~ "*",
            TRUE            ~ ""
          )
        ) %>%
      dplyr::filter(!term %in% c("SexMale", "Topo1Yes", "Nucleic_acidYes", "Alcoholic_historyYes"))
      logistic_results_df <- logistic_results_df %>%
        left_join(Data_MAF_DNMT3A %>% select(Hugo_Symbol, color_group) %>% distinct(), 
                  by = c("Gene" = "Hugo_Symbol"))
      logistic_results_df$Gene <- factor(logistic_results_df$Gene, levels = gene_freq)
      logistic_results_df$term = str_remove(logistic_results_df$term, "Yes")
      logistic_results_df$term = str_remove_all(logistic_results_df$term, '`')
      # 図の作成
      g = ggplot(data = logistic_results_df, aes(x = OR, y = Gene, color = color_group)) +
        geom_point(size = 2) +  # 推定値を点で表示
        geom_text(aes(label = Significance), hjust = -0.4, size = 6, color = "black") +  # アスタリスク追加
        geom_errorbarh(aes(xmin = CI_lower, xmax = CI_upper), height = 0.2, size = 1.2) +  # 水平エラーバー
        facet_wrap(~ term, scales = "free_y", ncol = 4) +  # Factorごとに4つの図を縦に並べる
        scale_x_continuous(
          trans = 'log',  # 対数スケール
          breaks = c(0.2, 0.5, 1, 2, 5),  # 指定された目盛り
          labels = c("0.2", "0.5", "1", "2.0", "5.0")  # 数値表示に変更
        ) +
        labs(
          title = paste0("C-CAT, odds ratios for LD-CH/LE-CH variants, ", n_subjects, " patients with clinical information"),
          subtitle = "Q-values, FDR; logistic regression adjusted for shown 8 factors, sex, History of alcohol, nucleic acid drugs, and Topo-I inhibitors.",
          x = "Odds ratio (log scale)",
          y = "Gene"
        ) +
        theme_minimal() +
        theme(
          axis.text.y = element_text(size = 12),  # Y軸のラベルサイズ調整
          strip.text = element_text(size = 12),  # Factorのラベルサイズ調整
          axis.title = element_text(size = 14),  # 軸タイトルのサイズ調整
          axis.text = element_text(size = 12),  # 軸ラベルのサイズ調整
          panel.spacing = unit(1, "lines"),  # 図と図の間のスペース調整
          panel.grid.major = element_blank(),  # 主グリッド線を非表示
          panel.grid.minor = element_blank(),  # 副グリッド線を非表示
          axis.line = element_line(color = "black", size = 1)  # 軸線を表示
        ) +
        geom_vline(xintercept = 1, linetype = "dashed", color = "black") +
        # 'Yes'症例数を図に追加
        geom_text(data = count_yes_long, aes(x = 4.5, y = 11, label = paste("\nYes:", Yes_count)), 
                  size = 3.5, color = "black", inherit.aes = FALSE)
      # x = 1の位置に縦線を追加
      plots[[100]] = g
      
      # 1. 頻度の高い遺伝子トップ10を抽出
      top10_genes <- Data_MAF_DNMT3A %>%
        dplyr::filter(logfold > 8) %>%
        count(Hugo_Symbol, sort = TRUE) %>%
        top_n(4, n) %>%
        pull(Hugo_Symbol)
      gene_freq <- names(sort(table(Data_MAF_DNMT3A$Hugo_Symbol),decreasing = T)[10:1])
      # 変異情報を症例ごとにまとめる（変異がある場合 1, ない場合 0）
      mutation_status <- Data_MAF_DNMT3A %>%
        filter(Hugo_Symbol %in% top10_genes) %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol) %>%
        distinct() %>%
        mutate(Mutation = 1) %>%
        pivot_wider(names_from = Hugo_Symbol, values_from = Mutation, values_fill = list(Mutation = 0))
      # 症例データと結合（Smoking, Alcohol, Treatment情報を含む）
      Data_taxane_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "taxel")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Taxane = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_taxane_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_Topo2_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Etoposide|ubicin")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Topo2 = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_Topo2_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_Topo1_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "otecan")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Topo1 = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_Topo1_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_nucleic_acid_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "tabine|racil")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Nucleic_acid = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_nucleic_acid_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_alkyl_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Ifosfamide|Cyclophosphamide")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Alkylate = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_alkyl_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_ICI_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Nivolumab|Ipilimumab|Pembrolizumab|Atezolizumab|Tremelimumab|Avelumab|Durvalumab")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$ICI = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_ICI_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_summary_F1L$Age_10Year_increase = Data_summary_F1L$Age / 10
      all_data <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age_10Year_increase, Sex, Smoking_history, Alcoholic_history, Lines_pre_CGP_2_or_more, Platinum,
               Taxane, Topo1, Topo2, Nucleic_acid, Alkylate, ICI) %>%
        left_join(mutation_status, by = "Tumor_Sample_Barcode") %>%
        mutate(across(all_of(top10_genes), ~replace_na(.x, 0)))  # 変異のない遺伝子を0にする
      all_data = all_data %>%
        dplyr::mutate(
          Lines_pre_CGP_2_or_more = case_when(
            Lines_pre_CGP_2_or_more == 1 ~ "Yes",
            TRUE ~"No"
          )
        )
      filtered_data <- all_data %>%
        filter(
          Smoking_history != "Unknown", 
          Alcoholic_history != "Unknown", 
          Lines_pre_CGP_2_or_more != "Unknown", 
          Platinum != "Unknown"
        )
      count_yes <- filtered_data %>%
        summarise(
          Age_yes = sum(!is.na(Age_10Year_increase)),
          Smoking_yes = sum(Smoking_history == "Yes"),
          Alcohol_yes = sum(Alcoholic_history == "Yes"),
          Lines_yes = sum(Lines_pre_CGP_2_or_more == "Yes"),
          Platinum_yes = sum(Platinum == "Yes"),
          Taxane_yes = sum(Taxane == "Yes"),
          Topo1_yes = sum(Topo1 == "Yes"),
          Topo2_yes = sum(Topo2 == "Yes"),
          Nucleic_acid_yes = sum(Nucleic_acid == "Yes"),
          ICI_yes = sum(ICI == "Yes"),
          Alkylate_yes = sum(Alkylate == "Yes")
        )
      # 各因子のYes症例数をまとめたデータフレームを作成
      count_yes_long <- tibble(
        # term = c("Smoking_history", "Alcoholic_history", "Lines_pre_CGP_2_or_more", "Platinum",
        #          "Taxane", "Topo1", "Topo2", "Nucleic_acid", "Alkylate"),
        # Yes_count = c(count_yes$Smoking_yes, count_yes$Alcohol_yes, count_yes$Lines_yes, count_yes$Platinum_yes,
        #               count_yes$Taxane_yes, count_yes$Topo1_yes, count_yes$Topo2_yes, count_yes$Nucleic_acid_yes, count_yes$Alkylate_yes)
        term = c("Smoking_history", "Age_10Year_increase", "Lines_pre_CGP_2_or_more", "Platinum",
                 "Taxane", "Topo2", "Alkylate", "ICI"),
        Yes_count = c(count_yes$Smoking_yes, count_yes$Age_yes, count_yes$Lines_yes, count_yes$Platinum_yes,
                      count_yes$Taxane_yes, count_yes$Topo2_yes, count_yes$Alkylate_yes, count_yes$ICI_yes)
      )
      n_subjects <- nrow(filtered_data)
      # 因子リスト
      factors <- c("Age_10Year_increase", "Sex", "Smoking_history", "Alcoholic_history", "Lines_pre_CGP_2_or_more","Platinum",
                   "Taxane", "Topo1", "Topo2", "Nucleic_acid", "Alkylate", "ICI")
      logistic_regression <- function(gene, data) {
        # 遺伝子の変異情報を2値に変換
        data$mutation <- ifelse(data[[gene]] == 1, 1, 0)  # 変異があれば1、なければ0
        # ロジスティック回帰モデル
        model <- glm(mutation ~ Smoking_history + Alcoholic_history + Lines_pre_CGP_2_or_more + Platinum + Age_10Year_increase + Sex +
                       Taxane + Topo1 + Topo2 + Nucleic_acid + Alkylate + ICI, 
                     data = data, 
                     family = binomial)
        # 結果の抽出
        tidy_model <- tidy(model)
        # オッズ比の計算（係数の指数）
        tidy_model <- tidy_model %>%
          mutate(
            OR = exp(estimate),
            CI_lower = exp(estimate - 1.96 * std.error),  # 信頼区間の下限
            CI_upper = exp(estimate + 1.96 * std.error),  # 信頼区間の上限
            p_value = p.value
          ) %>%
          filter(term != "(Intercept)")  # 切片項を除外
        tidy_model$Gene <- gene
        return(tidy_model)
      }
      # 各遺伝子に対してロジスティック回帰を実行
      logistic_results <- lapply(top10_genes, logistic_regression, data = filtered_data)
      # 結果をデータフレームにまとめる
      logistic_results_df <- bind_rows(logistic_results)
      # Q値を計算 (Benjamini-Hochberg法)
      logistic_results_df <- logistic_results_df %>%
        mutate(Q_value = p.adjust(p_value, method = "none"))
      # 有意水準を追加
      logistic_results_df <- logistic_results_df %>%
        mutate(
          Significance = case_when(
            Q_value < 0.001 ~ "***",
            Q_value < 0.01  ~ "**",
            Q_value < 0.05  ~ "*",
            TRUE            ~ ""
          )
        ) %>%
        dplyr::filter(!term %in% c("SexMale", "Topo1Yes", "Nucleic_acidYes", "Alcoholic_historyYes"))
      logistic_results_df <- logistic_results_df %>%
        left_join(Data_MAF_DNMT3A %>% select(Hugo_Symbol, color_group) %>% distinct(), 
                  by = c("Gene" = "Hugo_Symbol"))
      logistic_results_df$Gene <- factor(logistic_results_df$Gene, levels = gene_freq)
      logistic_results_df$term = str_remove(logistic_results_df$term, "Yes")
      # 図の作成
      g = ggplot(data = logistic_results_df, aes(x = OR, y = Gene, color = color_group)) +
        geom_point(size = 2) +  # 推定値を点で表示
        geom_text(aes(label = Significance), hjust = -0.4, size = 6, color = "black") +  # アスタリスク追加
        geom_errorbarh(aes(xmin = CI_lower, xmax = CI_upper), height = 0.2) +  # 水平エラーバー
        facet_wrap(~ term, scales = "free_y", ncol = 4) +  # Factorごとに4つの図を縦に並べる
        scale_x_continuous(
          trans = 'log',  # 対数スケール
          breaks = c(0.2, 0.5, 1, 2, 5),  # 指定された目盛り
          labels = c("0.2", "0.5", "1", "2.0", "5.0")  # 数値表示に変更
        ) +
        labs(
          title = paste0("C-CAT, Odds Ratios for LOCH variants, ", n_subjects, " patients"),
          subtitle = "Q-values, FDR; logistic regression adjusted for shown 4 factors, age, and sex.",
          x = "Odds Ratio (Log Scale)",
          y = "Gene"
        ) +
        theme_minimal() +
        theme(
          axis.text.y = element_text(size = 12),  # Y軸のラベルサイズ調整
          strip.text = element_text(size = 12),  # Factorのラベルサイズ調整
          axis.title = element_text(size = 14),  # 軸タイトルのサイズ調整
          axis.text = element_text(size = 12),  # 軸ラベルのサイズ調整
          panel.spacing = unit(1, "lines"),  # 図と図の間のスペース調整
          panel.grid.major = element_blank(),  # 主グリッド線を非表示
          panel.grid.minor = element_blank(),  # 副グリッド線を非表示
          axis.line = element_line(color = "black", size = 1)  # 軸線を表示
        ) +
        geom_vline(xintercept = 1, linetype = "dashed", color = "black") +
        # 'Yes'症例数を図に追加
        geom_text(data = count_yes_long, aes(x = 2, y = length(top10_genes)+1, label = paste("\nYes:", Yes_count)), 
                  size = 3, color = "black", inherit.aes = FALSE)
      # x = 1の位置に縦線を追加
      plots[[49]] = g
      
      # 1. 頻度の高い遺伝子トップ10を抽出
      top10_genes <- Data_MAF_DNMT3A %>%
        filter(Hugo_Symbol == "DNMT3A" & non_germline_non_error == 1) %>%
        count(amino.acid.change, sort = TRUE) %>%
        top_n(10, n) %>%
        pull(amino.acid.change)
      gene_freq <- (Data_MAF_DNMT3A %>%
                      filter(Hugo_Symbol == "DNMT3A" & non_germline_non_error == 1) %>%
                      count(amino.acid.change, sort = TRUE) %>%
                      top_n(10, n))$amino.acid.change[10:1]
      # 変異情報を症例ごとにまとめる（変異がある場合 1, ない場合 0）
      mutation_status <- Data_MAF_DNMT3A %>%
        filter(amino.acid.change %in% top10_genes) %>%
        select(Tumor_Sample_Barcode, amino.acid.change) %>%
        distinct() %>%
        mutate(Mutation = 1) %>%
        pivot_wider(names_from = amino.acid.change, values_from = Mutation, values_fill = list(Mutation = 0))
      colname_tmp = colnames(mutation_status)
      colname_tmp[colname_tmp == "R771*"] = "R771Ter"
      colnames(mutation_status) = colname_tmp
      top10_genes[top10_genes == "R771*"] = "R771Ter"
      gene_freq[gene_freq == "R771*"] = "R771Ter"
      setProgress(detail = "20045")
      # 症例データと結合（Smoking, Alcohol, Treatment情報を含む）
      Data_taxane_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "taxel")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Taxane = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_taxane_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_Topo2_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Etoposide|ubicin")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Topo2 = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_Topo2_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_Topo1_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "otecan")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Topo1 = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_Topo1_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_nucleic_acid_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "tabine|racil")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Nucleic_acid = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_nucleic_acid_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_alkyl_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Ifosfamide|Cyclophosphamide")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$Alkylate = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_alkyl_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_ICI_pre = Data_case() %>%
        dplyr::filter(症例.検体情報.パネル.名称. == "F1Liquid CDx") %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
        dplyr::filter(str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Nivolumab|Ipilimumab|Pembrolizumab|Atezolizumab|Tremelimumab|Avelumab|Durvalumab")) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID)
      Data_summary_F1L$ICI = ifelse(Data_summary_F1L$Tumor_Sample_Barcode %in% Data_ICI_pre$C.CAT調査結果.基本項目.ハッシュID, "Yes", "No")
      Data_summary_F1L$Age_10Year_increase = Data_summary_F1L$Age / 10
      all_data <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age_10Year_increase, Sex, Smoking_history, Alcoholic_history, Lines_pre_CGP_2_or_more, Platinum,
               Taxane, Topo1, Topo2, Nucleic_acid, Alkylate, ICI) %>%
        left_join(mutation_status, by = "Tumor_Sample_Barcode") %>%
        mutate(across(all_of(top10_genes), ~replace_na(.x, 0)))  # 変異のない遺伝子を0にする
      all_data = all_data %>%
        dplyr::mutate(
          Lines_pre_CGP_2_or_more = case_when(
            Lines_pre_CGP_2_or_more == 1 ~ "Yes",
            TRUE ~"No"
          )
        )
      filtered_data <- all_data %>%
        filter(
          Smoking_history != "Unknown", 
          Alcoholic_history != "Unknown", 
          Lines_pre_CGP_2_or_more != "Unknown", 
          Platinum != "Unknown"
        )
      count_yes <- filtered_data %>%
        summarise(
          Age_yes = sum(!is.na(Age_10Year_increase)),
          Smoking_yes = sum(Smoking_history == "Yes"),
          Alcohol_yes = sum(Alcoholic_history == "Yes"),
          Lines_yes = sum(Lines_pre_CGP_2_or_more == "Yes"),
          Platinum_yes = sum(Platinum == "Yes"),
          Taxane_yes = sum(Taxane == "Yes"),
          Topo1_yes = sum(Topo1 == "Yes"),
          Topo2_yes = sum(Topo2 == "Yes"),
          Nucleic_acid_yes = sum(Nucleic_acid == "Yes"),
          ICI_yes = sum(ICI == "Yes"),
          Alkylate_yes = sum(Alkylate == "Yes")
        )
      # 各因子のYes症例数をまとめたデータフレームを作成
      count_yes_long <- tibble(
        # term = c("Smoking_history", "Alcoholic_history", "Lines_pre_CGP_2_or_more", "Platinum",
        #          "Taxane", "Topo1", "Topo2", "Nucleic_acid", "Alkylate"),
        # Yes_count = c(count_yes$Smoking_yes, count_yes$Alcohol_yes, count_yes$Lines_yes, count_yes$Platinum_yes,
        #               count_yes$Taxane_yes, count_yes$Topo1_yes, count_yes$Topo2_yes, count_yes$Nucleic_acid_yes, count_yes$Alkylate_yes)
        term = c("Smoking_history", "Age_10Year_increase", "Lines_pre_CGP_2_or_more", "Platinum",
                 "Taxane", "Topo2", "Alkylate", "ICI"),
        Yes_count = c(count_yes$Smoking_yes, count_yes$Age_yes, count_yes$Lines_yes, count_yes$Platinum_yes,
                      count_yes$Taxane_yes, count_yes$Topo2_yes, count_yes$Alkylate_yes, count_yes$ICI_yes)
      )
      n_subjects <- nrow(filtered_data)
      # 因子リスト
      factors <- c("Age_10Year_increase", "Sex", "Smoking_history", "Alcoholic_history", "Lines_pre_CGP_2_or_more","Platinum",
                   "Taxane", "Topo1", "Topo2", "Nucleic_acid", "Alkylate", "ICI")
      logistic_regression <- function(gene, data) {
        # 遺伝子の変異情報を2値に変換
        data$mutation <- ifelse(data[[gene]] == 1, 1, 0)  # 変異があれば1、なければ0
        # ロジスティック回帰モデル
        model <- glm(mutation ~ Smoking_history + Alcoholic_history + Lines_pre_CGP_2_or_more + Platinum + Age_10Year_increase + Sex +
                       Taxane + Topo1 + Topo2 + Nucleic_acid + Alkylate + ICI, 
                     data = data, 
                     family = binomial)
        # 結果の抽出
        tidy_model <- tidy(model)
        # オッズ比の計算（係数の指数）
        tidy_model <- tidy_model %>%
          mutate(
            OR = exp(estimate),
            CI_lower = exp(estimate - 1.96 * std.error),  # 信頼区間の下限
            CI_upper = exp(estimate + 1.96 * std.error),  # 信頼区間の上限
            p_value = p.value
          ) %>%
          filter(term != "(Intercept)")  # 切片項を除外
        tidy_model$Gene <- gene
        return(tidy_model)
      }
      # 各遺伝子に対してロジスティック回帰を実行
      logistic_results <- lapply(top10_genes, logistic_regression, data = filtered_data)
      # 結果をデータフレームにまとめる
      logistic_results_df <- bind_rows(logistic_results)
      # Q値を計算 (Benjamini-Hochberg法)
      logistic_results_df <- logistic_results_df %>%
        mutate(Q_value = p.adjust(p_value, method = "none"))
      # 有意水準を追加
      logistic_results_df <- logistic_results_df %>%
        mutate(
          Significance = case_when(
            Q_value < 0.001 ~ "***",
            Q_value < 0.01  ~ "**",
            Q_value < 0.05  ~ "*",
            TRUE            ~ ""
          )
        ) %>%
        dplyr::filter(!term %in% c("SexMale", "Topo1Yes", "Nucleic_acidYes", "Alcoholic_historyYes"))
      logistic_results_df <- logistic_results_df %>%
        left_join(Data_MAF_DNMT3A %>% select(Hugo_Symbol, color_group) %>% distinct(), 
                  by = c("Gene" = "Hugo_Symbol"))
      logistic_results_df$Gene <- factor(logistic_results_df$Gene, levels = gene_freq)
      logistic_results_df$term = str_remove(logistic_results_df$term, "Yes")
      # 図の作成
      g = ggplot(data = logistic_results_df, aes(x = OR, y = Gene)) +
        geom_point(size = 2) +  # 推定値を点で表示
        geom_text(aes(label = Significance), hjust = -0.4, size = 6, color = "black") +  # アスタリスク追加
        geom_errorbarh(aes(xmin = CI_lower, xmax = CI_upper), height = 0.2) +  # 水平エラーバー
        facet_wrap(~ term, scales = "free_y", ncol = 2) +  # Factorごとに4つの図を縦に並べる
        scale_x_continuous(
          trans = 'log',  # 対数スケール
          breaks = c(0.2, 0.5, 1, 1.5, 2),  # 指定された目盛り
          labels = c("0.2", "0.5", "1", "1.5", "2.0")  # 数値表示に変更
        ) +
        labs(
          title = paste0("C-CAT, Odds Ratios for DNMT3A Mutations, ", n_subjects, " patients"),
          subtitle = "Benjamini-Hochberg adjusted Q-values, FDR; logistic regression adjusted for shown 4 factors, age, and sex.",
          x = "Odds Ratio (Log Scale)",
          y = "Gene"
        ) +
        theme_minimal() +
        theme(
          axis.text.y = element_text(size = 12),  # Y軸のラベルサイズ調整
          strip.text = element_text(size = 12),  # Factorのラベルサイズ調整
          axis.title = element_text(size = 14),  # 軸タイトルのサイズ調整
          axis.text = element_text(size = 12),  # 軸ラベルのサイズ調整
          panel.spacing = unit(1, "lines"),  # 図と図の間のスペース調整
          panel.grid.major = element_blank(),  # 主グリッド線を非表示
          panel.grid.minor = element_blank(),  # 副グリッド線を非表示
          axis.line = element_line(color = "black", size = 1)  # 軸線を表示
        ) +
        geom_vline(xintercept = 1, linetype = "dashed", color = "black") +
        # 'Yes'症例数を図に追加
        geom_text(data = count_yes_long, aes(x = 2, y = nrow(count_yes_long)+0.5, label = paste("\nYes:", Yes_count)), 
                  size = 3, color = "black", inherit.aes = FALSE)
      # x = 1の位置に縦線を追加
      plots[[101]] = g
      
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          TRUE ~ "Others"
        )) %>% 
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      gene_freq = c("Y735C", "R882H", "R882C", "R635W", "I780T", "Others")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("Y735C", "R882H", "R882C", "R635W", "I780T", "Others"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      Data_DNMT3A_count = 
        Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          TRUE ~ "Others"
        ))
      table_count = data.frame(gene_freq)
      colnames(table_count) = "color_group"
      table_count$count = 0
      for(i in 1:length(table_count$count)){
        table_count$count[i] = length(unique((Data_DNMT3A_count %>% filter(color_group == gene_freq[i]))$unique_position))
      }
      
      # 5. 年齢ごと & `color_group` ごとに統計情報を計算（全症例基準）
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        left_join(table_count, by = "color_group") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients / count,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count / count), 0),
          .groups = "drop"
        )
      
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0, total_patients = 0))
      
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      final_data$color_group <- factor(final_data$color_group, levels = gene_freq)      
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Count per Patient by Age",
             x = "Age",
             y = "Average LD-CH/LE-CH Count per Patient") +
        theme_minimal()
      plots[[102]] = g
      # 1. 全症例リスト（年齢情報含む）
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # all_cases$Age = round(all_cases$Age/5)*5
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          TRUE ~ "Others"
        )) %>% 
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      gene_freq = c("Y735C", "R882H", "R882C", "R635W", "I780T", "Others")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("Y735C", "R882H", "R882C", "R635W", "I780T", "Others"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      all_mutations$mutation_count = fun_zero(all_mutations$mutation_count, all_mutations$mutation_count)
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age) %>%
        summarise(total_patients = n(), .groups = "drop")
      Data_DNMT3A_count = 
        Data_MAF_DNMT3A %>%
          dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
          dplyr::mutate(color_group = case_when(
            name == "DNMT3A_Y735C" ~ "Y735C",
            str_detect(name, "R882H") ~ "R882H",
            str_detect(name, "R882C") ~ "R882C",
            str_detect(name, "R635W") ~ "R635W",
            str_detect(name, "I780T") ~ "I780T",
            TRUE ~ "Others"
          ))
      table_count = data.frame(gene_freq)
      colnames(table_count) = "color_group"
      table_count$count = 0
      for(i in 1:length(table_count$count)){
        table_count$count[i] = length(unique((Data_DNMT3A_count %>% filter(color_group == gene_freq[i]))$unique_position))
      }
      # 5. 年齢ごと & `color_group` ごとに変異を持つ症例の割合を計算
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = "Age") %>% 
        left_join(table_count, by = "color_group") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients / count,  # その年齢での color_group の変異総数
          sample_size = n(),  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count / count), 0),
          .groups = "drop"
        )
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # age_range <- tibble(Age = 0:20*5)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      final_data$color_group <- factor(final_data$color_group, levels = gene_freq)
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = paste0("C-CAT, TOCH/LOCH presence ratio by age, ", length(unique(Data_summary_F1L$Tumor_Sample_Barcode)), "patients"),
             x = "Age",
             y = "TOCH/LOCH presence ratio") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())
      plots[[103]] = g
      # 1. 全症例リスト（年齢情報含む）
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      all_cases_color_group <- Data_MAF_DNMT3A %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          str_detect(name, "DNMT3A_") ~ "Others",
          TRUE ~ "None"
        )) %>%
        select(Tumor_Sample_Barcode, Age, color_group) %>%
        filter(color_group != "None")
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          str_detect(name, "DNMT3A_") ~ "Others",
          TRUE ~ "None"
        )) %>% 
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = VAF, .groups = "drop")
      gene_freq = c("Y735C", "R882H", "R882C", "R635W", "I780T", "Others")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("Y735C", "R882H", "R882C", "R635W", "I780T", "Others"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases_color_group %>%
        group_by(Age, color_group) %>%
        summarise(total_patients = n(), .groups = "drop")
      Data_DNMT3A_count = 
        Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          TRUE ~ "Others"
        ))
      table_count = data.frame(gene_freq)
      colnames(table_count) = "color_group"
      table_count$count = 0
      for(i in 1:length(table_count$count)){
        table_count$count[i] = length(unique((Data_DNMT3A_count %>% filter(color_group == gene_freq[i]))$unique_position))
      }
      
      # 5. 年齢ごと & `color_group` ごとに変異を持つ症例の割合を計算
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = c("Age", "color_group")) %>% 
        left_join(table_count, by = "color_group") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients / count,  # その年齢での color_group の変異総数
          sample_size = total_patients,  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count / count), 0),
          .groups = "drop"
        )
      
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0))
      
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      final_data$color_group <- factor(final_data$color_group, levels = gene_freq)
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 0.8, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Variant Allele Frequency by Age",
             x = "Age",
             y = "Mean variant allele frequency") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())      
      plots[[104]] = g
      
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
            Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      mutation_data_DNMT3A <- mutation_data %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "DNMT3A_R882") ~ "R882",
          TRUE ~ "Others"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode &
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R882"))$Tumor_Sample_Barcode ~ 2,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R882"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 2
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!name %in% c("DNMT3A_Y735C", "DNMT3A_R882H", "DNMT3A_R882C", "DNMT3A_R882S", "DNMT3A_R882P"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-8] = 10^-8
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor DNMT3A R882 or Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[105]] = g
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
          Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      mutation_data_DNMT3A <- mutation_data %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "DNMT3A_R635W") ~ "R635W",
          TRUE ~ "Others"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode &
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R635W"))$Tumor_Sample_Barcode ~ 2,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R635W"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 2
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!name %in% c("DNMT3A_Y735C", "DNMT3A_R635W"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-8] = 10^-8
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor DNMT3A R635W or Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[106]] = g
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
          Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      mutation_data_DNMT3A <- mutation_data %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "DNMT3A_I780T") ~ "I780T",
          TRUE ~ "Others"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode &
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "I780T"))$Tumor_Sample_Barcode ~ 2,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "I780T"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 2
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!name %in% c("DNMT3A_Y735C", "DNMT3A_I780T"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-8] = 10^-8
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor DNMT3A I780T or Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[107]] = g
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
          Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      data_maf_ACC = mutation_data %>% dplyr::select(Hugo_Symbol)
      data_maf_ACC$Chromosome = paste0("chr", mutation_data$Chromosome)
      data_maf_ACC$Start_Position = as.integer(mutation_data$Start_Position)
      data_maf_ACC$End_Position = as.integer(mutation_data$Start_Position) + nchar(mutation_data$Reference_Allele) - 1
      data_maf_ACC$Strand = "+"
      data_maf_ACC$Variant_Classification = "Missense_Mutation"
      data_maf_ACC$Variant_Type = ifelse(nchar(mutation_data$Reference_Allele) == 1 & nchar(mutation_data$Tumor_Seq_Allele2) == 1, "SNV",
                                              ifelse(nchar(mutation_data$Reference_Allele) > nchar(mutation_data$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf_ACC$Reference_Allele = mutation_data$Reference_Allele
      data_maf_ACC$Tumor_Seq_Allele1 = mutation_data$Reference_Allele
      data_maf_ACC$Tumor_Seq_Allele2 = mutation_data$Tumor_Seq_Allele2
      data_maf_ACC$Tumor_Sample_Barcode = mutation_data$Tumor_Sample_Barcode
      data_maf_ACC$name = mutation_data$name
      data_maf_ACC$Cancername = mutation_data$Cancername
      data_maf_ACC$Age = mutation_data$Age
      data_maf_ACC$pnt_col = mutation_data$pnt_col
      hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      data_maf_ACC = as.data.frame(data_maf_ACC)
      maf_ACC <- filterSNV(dataSet = data_maf_ACC,
                              seq_colNames = c("Reference_Allele",
                                               "Tumor_Seq_Allele1",
                                               "Tumor_Seq_Allele2"))
      maf_ACC <- attachContext(mutData = maf_ACC,
                                      chr_colName = "Chromosome",
                                      start_colName = "Start_Position",
                                      end_colName = "End_Position",
                                      nucl_contextN = 3,
                                      BSGenomeDb = hg38)
      maf_ACC <- removeMismatchMut(mutData = maf_ACC,
                                          refMut_colName = "Reference_Allele",
                                          context_colName = "context",
                                          refMut_format = "N")
      mutation_data_signature <- attachMutType(mutData = maf_ACC,
                                      ref_colName = "Reference_Allele",
                                      var_colName = "Tumor_Seq_Allele1",
                                      var2_colName = "Tumor_Seq_Allele2",
                                      context_colName = "context")
      mutation_data_signature_IDH2 = mutation_data_signature
      mutation_data_DNMT3A <- mutation_data_signature %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "DNMT3A_I780T") ~ "I780T",
          TRUE ~ "Others"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data_signature %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode &
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "I780T"))$Tumor_Sample_Barcode ~ 2,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "I780T"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 2
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!name %in% c("DNMT3A_Y735C", "DNMT3A_I780T"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$mutType))
      colnames(data_tmp) = "mutType"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$mutType)){
        mut_ID = (mutation_data %>% dplyr::filter(mutType == data_tmp$mutType[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-8] = 10^-8
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = mutType,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor DNMT3A I780T or Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor variants",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[108]] = g
      mutation_data_DNMT3A <- mutation_data_signature %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "DNMT3A_R635W") ~ "R635W",
          TRUE ~ "Others"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data_signature %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode &
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R635W"))$Tumor_Sample_Barcode ~ 2,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R635W"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 2
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!name %in% c("DNMT3A_Y735C", "DNMT3A_R635W"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$mutType))
      colnames(data_tmp) = "mutType"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$mutType)){
        mut_ID = (mutation_data %>% dplyr::filter(mutType == data_tmp$mutType[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-8] = 10^-8
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = mutType,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor DNMT3A R635W or Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor variants",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[109]] = g
      mutation_data_DNMT3A <- mutation_data_signature %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "DNMT3A_R882") ~ "R882",
          TRUE ~ "Others"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data_signature %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode &
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R882"))$Tumor_Sample_Barcode ~ 2,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R882"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 2
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!name %in% c("DNMT3A_Y735C", "DNMT3A_R882H", "DNMT3A_R882C", "DNMT3A_R882S", "DNMT3A_R882P", "DNMT3A_R882G"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$mutType))
      colnames(data_tmp) = "mutType"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$mutType)){
        mut_ID = (mutation_data %>% dplyr::filter(mutType == data_tmp$mutType[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-8] = 10^-8
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = mutType,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor DNMT3A R882 or Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor variants",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[110]] = g
      mutation_data_DNMT3A <- mutation_data_signature %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          TRUE ~ "R882"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data_signature %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "R882"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 0
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!str_detect(name, "DNMT3A_Y735C"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$mutType))
      colnames(data_tmp) = "mutType"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$mutType)){
        mut_ID = (mutation_data %>% dplyr::filter(mutType == data_tmp$mutType[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-8] = 10^-8
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = mutType,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor with or without DNMT3A Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor variants",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      setProgress(detail = "21223")
      plots[[111]] = g
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
          Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      mutation_data_DNMT3A <- mutation_data %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          TRUE ~ "I780T"
        )) %>%
        dplyr::filter(color_group != "Others")
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "I780T"))$Tumor_Sample_Barcode ~ 0,
          TRUE ~ 0
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!str_detect(name, "DNMT3A_Y735C"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-2] = 10^-2
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor with or without DNMT3A Y735C",
             subtitle = "Only liquid-favor variants, right: Y735C-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,2),
                           breaks = c(0,1,2),
                           labels = c("0","1","2~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[112]] = g
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
          Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      mutation_data_DNMT3A <- mutation_data %>%
        dplyr::mutate(color_group = case_when(
          str_detect(name, "\\+1G>A") ~ "Y735C",
          TRUE ~ "I780T"
        ))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          TRUE ~ 0
        )) %>%
        filter(DNMT3A_735_882 != 2) %>%
        filter(!str_detect(name, "\\+1G>A"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$unique_alt))
      colnames(data_tmp) = "unique_alt"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      if(file.exists("source/splice_site.rda")){
        load("source/splice_site.rda")
      } else{
        for(i in 1:length(data_tmp$odds_ratio)){
          mut_ID = (mutation_data %>% dplyr::filter(unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
          Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$unique_alt)))
          }
        }
        save(data_tmp, file="source/splice_site.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-6] = 10^-6
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = unique_alt,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        guides(color = guide_legend(override.aes = list(shape = NA))) +
        labs(title = "Volcano plot of LD-CH/LE-CH for favor splice site +1G>A co-variants",
             subtitle = "right: +1G>A-favor variant",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,6),
                           breaks = c(0,2,4,6),
                           labels = c("0","2","4","6~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      plots[[113]] = g
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      if(file.exists("source/splice_site_gene.rda")){
        load("source/splice_site_gene.rda")
      } else{
        for(i in 1:length(data_tmp$odds_ratio)){
          mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
          Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$Hugo_Symbol)))
          }
        }
        save(data_tmp, file="source/splice_site_gene.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor splice site +1G>A variants",
             subtitle = "Only liquid-favor variants, right: +1G>A-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2),
                           labels = c("~-4","-2","0","2~")) +
        scale_y_continuous(limits = c(0,2),
                           breaks = c(0,1,2),
                           labels = c("0","1","2~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[114]] = g
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1 | non_germline_non_error == 6) %>%
        dplyr::filter(#unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "FoundationOne CDx")
      gene_to_analyze = unique(names(sort(table(Data_tmp$Hugo_Symbol),
                                          decreasing = T)))
      Data_tmp3 = Data_summary %>% dplyr::filter(Panel == "FoundationOne CDx")
      #Diseases = sort(names(table(Data_tmp3$Cancername)[table(Data_tmp3$Cancername)>=50]))
      Data_tmp = Data_tmp %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "FoundationOne CDx")
      Summary_Gene_alteration = data.frame(matrix(0,
                                                  nrow=length(Diseases),
                                                  ncol=length(gene_to_analyze)))
      colnames(Summary_Gene_alteration) = gene_to_analyze
      rownames(Summary_Gene_alteration) = Diseases
      
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        for(j in gene_to_analyze){
          Summary_Gene_alteration[i,j] =
            length(unique((Data_tmp %>% dplyr::filter(
              Hugo_Symbol == j &
                Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
        }
      }
      Data_tmp2 = Summary_Gene_alteration
      Data_tmp2$Disease = rownames(Data_tmp2)
      Data_tmp2 = Data_tmp2 %>% gather(value = "freq", key = Gene, -Disease)
      Data_tmp2 <- transform(Data_tmp2, Disease= factor(Disease, levels = sort(unique(Disease), decreasing = TRUE)))
      g = ggplot(Data_tmp2, aes(as.factor(Gene), as.factor(Disease))) +
        geom_tile(aes(fill = freq), color = "black",
                  #            lwd = 1,
                  linetype = 1) + 
        geom_text(aes(label = round(freq, 0))) +
        scale_fill_gradient(low = "white", high = "red", limit=c(0,100), name="Frequency (%)") +
        theme_classic() +
        labs(title = "Non-germline variant in CH-related genes, solid tumor samples",
             subtitle = "Including non-hotspot variants") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              axis.line.x = element_blank(),
              axis.line.y = element_blank())
      plots[[115]] = g
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "FoundationOne CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "FoundationOne CDx")
      Summary_Gene_alteration = data.frame(matrix(0,
                                                  nrow=length(Diseases),
                                                  ncol=length(gene_to_analyze)))
      colnames(Summary_Gene_alteration) = gene_to_analyze
      rownames(Summary_Gene_alteration) = Diseases
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        for(j in gene_to_analyze){
          Summary_Gene_alteration[i,j] =
            length(unique((Data_tmp %>% dplyr::filter(
              Hugo_Symbol == j &
                Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
        }
      }
      Data_tmp2 = Summary_Gene_alteration
      Data_tmp2$Disease = rownames(Data_tmp2)
      Data_tmp2 = Data_tmp2 %>% gather(value = "freq", key = Gene, -Disease)
      Data_tmp2 <- transform(Data_tmp2, Disease= factor(Disease, levels = sort(unique(Disease), decreasing = TRUE)))
      g = ggplot(Data_tmp2, aes(as.factor(Gene), as.factor(Disease))) +
        geom_tile(aes(fill = freq), color = "black",
                  #            lwd = 1,
                  linetype = 1) + 
        geom_text(aes(label = round(freq, 0))) +
        scale_fill_gradient(low = "white", high = "red", limit=c(0,100), name="Frequency (%)") +
        theme_classic() +
        labs(title = paste0(nrow(Data_tmp), " TI-CH in F1 CDx, solid tumor ", nrow(Data_tmp3), " samples")) +
        theme(axis.text.x = element_text(angle = 45, hjust = 1),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              axis.line.x = element_blank(),
              axis.line.y = element_blank())
      plots[[178]] = g
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "FoundationOne CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "FoundationOne CDx")
      Summary_Gene_alteration = data.frame(rep(0, length(Diseases)))
      colnames(Summary_Gene_alteration) = "TICH"
      Summary_Gene_alteration$TICH_No = 0
      Summary_Gene_alteration$TOCH = 0
      Summary_Gene_alteration$TOCH_No = 0
      Summary_Gene_alteration$LOCH = 0
      Summary_Gene_alteration$LOCH_No = 0
      rownames(Summary_Gene_alteration) = Diseases
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        Summary_Gene_alteration[i,"TICH_No"] = patient_no
        Summary_Gene_alteration[i,"TICH"] =
            length(unique((Data_tmp %>% dplyr::filter(
                Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
      }
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx")
      ID_tmp = unique((Data_tmp %>%
        dplyr::filter(logfold > 8))$Tumor_Sample_Barcode)
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(logfold > 2 &
                      !Tumor_Sample_Barcode %in% ID_tmp)
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "F1Liquid CDx")
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        Summary_Gene_alteration[i,"TOCH_No"] = patient_no
        Summary_Gene_alteration[i,"TOCH"] =
          length(unique((Data_tmp %>% dplyr::filter(
            Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
      }
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        logfold > 8 &
                        Panel == "F1Liquid CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "F1Liquid CDx")
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        Summary_Gene_alteration[i,"LOCH_No"] = patient_no
        Summary_Gene_alteration[i,"LOCH"] =
          length(unique((Data_tmp %>% dplyr::filter(
            Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
      }
      df_plot <- Summary_Gene_alteration %>%
        mutate(Disease = rownames(.)) %>%
        select(Disease, TICH, TOCH, LOCH, TICH_No, TOCH_No, LOCH_No) %>%
        mutate(
          label = sprintf("%s (Solid: %d, Liquid: %d)",
                          Disease, TICH_No, TOCH_No)
        ) %>%
        pivot_longer(cols = c(TICH, TOCH, LOCH), names_to = "Type", values_to = "Frequency")
      
      # y軸ラベルに症例数を含めた並び順の制御
      df_plot$label <- factor(df_plot$label, levels = unique(df_plot$label[order(df_plot$Disease, decreasing = TRUE)]))
      df_plot$Type <- factor(df_plot$Type, levels = c("TOCH", "LOCH", "TICH"))
      df_upper <- df_plot %>% filter(Type == "TICH")
      df_lower <- df_plot %>% filter(Type != "TICH")
      g <- ggplot() +
        geom_bar(data = df_lower, aes(x = Frequency, y = label, fill = Type),
                 stat = "identity", position = "stack", width = 0.3) +
        geom_bar(data = df_upper, aes(x = Frequency, y = label, fill = Type),
                 stat = "identity", position = position_nudge(y = 0.4), width = 0.3) +
        scale_fill_manual(values = c("TICH" = "#1b9e77", "TOCH" = "#d95f02", "LOCH" = "#7570b3")) +
        labs(x = "Frequency (%)", y = NULL, title = "TICH/TOCH/LOCH Frequency by Cancer Type") +
        theme_classic() +
        theme(legend.position = "top")

      plots[[179]] = g
      df_age <- Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        logfold > 2 &
                        Panel == "F1Liquid CDx") %>%
        group_by(Cancername) %>%
        summarise(mean_age = mean(Age, na.rm = TRUE), .groups = "drop")
      df_VAF <- Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        logfold > 2 &
                        #logfold <= 8 &
                        Panel == "F1Liquid CDx") %>%
        group_by(Cancername) %>%
        summarise(mean_TOCH_VAF = mean(VAF, na.rm = TRUE), .groups = "drop")
      # LOCH / TOCH 比を計算
      df_ratio <- Summary_Gene_alteration %>%
        mutate(Cancername = rownames(.),
               LOCH_TOCH_ratio = LOCH / TOCH)
      # 結合してプロット用データフレーム作成
      df_plot2 <- df_ratio %>%
        inner_join(df_age, by = "Cancername") %>%
        filter(!is.infinite(LOCH_TOCH_ratio), !is.na(LOCH_TOCH_ratio), !is.na(mean_age))
      df_plot3 <- df_ratio %>%
        inner_join(df_VAF, by = "Cancername") %>%
        filter(!is.infinite(LOCH_TOCH_ratio), !is.na(LOCH_TOCH_ratio), !is.na(mean_TOCH_VAF))
      # 散布図 + 相関係数 + 回帰線
      plots[[190]] = ggplot(df_plot2, aes(x = mean_age, y = LOCH_TOCH_ratio, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "Mean Age", y = "LOCH / TOCH Ratio",
             title = "Correlation between Age and LOCH/TOCH Ratio") +
        theme_minimal()
      plots[[191]] = ggplot(df_plot2, aes(x = TOCH, y = LOCH_TOCH_ratio, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "TOCH", y = "LOCH / TOCH Ratio",
             title = "Correlation between TOCH and LOCH/TOCH Ratio") +
        theme_minimal()
      plots[[192]] = ggplot(df_plot3, aes(x = mean_TOCH_VAF, y = LOCH_TOCH_ratio, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "Mean VAF of TOCH", y = "LOCH / TOCH Ratio",
             title = "Correlation between Mean VAF of CH and LOCH/TOCH Ratio") +
        theme_minimal()
      plots[[193]] = ggplot(df_plot2, aes(x = mean_age, y = LOCH, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "Mean Age", y = "LOCH Rate",
             title = "Correlation between Age and LOCH Rate") +
        theme_minimal()
      plots[[194]] = ggplot(df_plot2, aes(x = TOCH, y = LOCH, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "TOCH", y = "LOCH Rate",
             title = "Correlation between TOCH and LOCH Rate") +
        theme_minimal()
      plots[[195]] = ggplot(df_plot3, aes(x = mean_TOCH_VAF, y = LOCH, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "Mean VAF of TOCH", y = "LOCH Rate",
             title = "Correlation between Mean VAF of CH and LOCH Rate") +
        theme_minimal()
      plots[[196]] = ggplot(df_plot2, aes(x = mean_age, y = TOCH, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "Mean Age", y = "TOCH Rate",
             title = "Correlation between Age and TOCH Rate") +
        theme_minimal()
      plots[[197]] = ggplot(df_plot3, aes(x = mean_TOCH_VAF, y = TOCH, label = Cancername)) +
        geom_point(size = 3, color = "#7570b3") +
        geom_text(vjust = -0.7, size = 3) +
        geom_smooth(method = "lm", se = TRUE, color = "gray40", linetype = "dashed") +
        stat_cor(method = "pearson", label.x.npc = "left", label.y.npc = "top") +
        labs(x = "Mean VAF of TOCH", y = "TOCH Rate",
             title = "Correlation between Mean VAF of TOCH Rate") +
        theme_minimal()
      
      # ===== 1. LOCH優先で症例分類 =====
      LOCH_samples <- Data_MAF_target %>%
        filter(non_germline_non_error == 1,
               unique_alt %in% CH_candidates,
               Hugo_Symbol %in% CH_gene_entropy_3,
               logfold > 8,
               Panel == "F1Liquid CDx") %>%
        distinct(Tumor_Sample_Barcode) %>%
        mutate(group = "LOCH")
      TOCH_samples <- Data_MAF_target %>%
        filter(non_germline_non_error == 1,
               unique_alt %in% CH_candidates,
               Hugo_Symbol %in% CH_gene_entropy_3,
               logfold > 2, logfold <= 8,
               Panel == "F1Liquid CDx") %>%
        distinct(Tumor_Sample_Barcode) %>%
        filter(!(Tumor_Sample_Barcode %in% LOCH_samples$Tumor_Sample_Barcode)) %>%
        mutate(group = "TOCH")
      sample_groups <- bind_rows(LOCH_samples, TOCH_samples)
      # ===== 2. 各症例の全変異数 =====
      mutation_counts <- Data_MAF_target %>%
        filter(non_germline_non_error == 1,
               unique_alt %in% CH_candidates,
               Hugo_Symbol %in% CH_gene_entropy_3,
               logfold > 2,
               Panel == "F1Liquid CDx") %>%
        filter(Tumor_Sample_Barcode %in% sample_groups$Tumor_Sample_Barcode) %>%
        count(Tumor_Sample_Barcode, name = "mutation_count") %>%
        left_join(sample_groups, by = "Tumor_Sample_Barcode")
      # ===== 3. 統計量（n, mean, sd, median）をラベル化 =====
      summary_stats <- mutation_counts %>%
        group_by(group) %>%
        summarise(
          mean_val = mean(mutation_count),
          sd_val = sd(mutation_count),
          median_val = median(mutation_count),
          n = n(),
          .groups = "drop"
        ) %>%
        mutate(
          label = sprintf("n=%d, median=%d\nmean=%.1f, sd=%.1f", n, median_val, mean_val, sd_val)
        )
      # ── ① group2 を作成：LOCH のみ抽出して "LOCH"、全症例を "ALL" とラベル付けして結合 ──
      loch_only <- mutation_counts %>%
        filter(group == "LOCH") %>%
        mutate(group2 = "LOCH")
      all_samples <- mutation_counts %>%
        mutate(group2 = "ALL")
      mutation_counts_all <- bind_rows(loch_only, all_samples)
      # ── ② 新グループで統計量を再計算 ──
      summary_stats2 <- mutation_counts_all %>%
        group_by(group2) %>%
        summarise(
          n        = n(),
          mean_val = mean(mutation_count),
          sd_val   = sd(mutation_count),
          median_val = median(mutation_count),
          .groups = "drop"
        ) %>%
        mutate(
          label = sprintf("n=%d, median=%d\nmean=%.1f, sd=%.1f",
                          n, median_val, mean_val, sd_val)
        )
      # ── ③ LOCH vs ALL でヴァイオリン＋ボックスプロット ──
      plots[[198]] <- ggplot(mutation_counts_all, aes(x = group2, y = mutation_count, fill = group2)) +
        geom_violin(trim = FALSE, alpha = 0.6) +
        geom_boxplot(width = 0.1, outlier.shape = NA, color = "black") +
        stat_summary(fun = mean, geom = "point", shape = 20, size = 2, color = "red") +
        stat_compare_means(
          method      = "wilcox.test",
          label       = "p.signif",
          comparisons = list(c("LOCH", "ALL")),
          label.y     = max(mutation_counts$mutation_count) * 1.2
        ) +
        geom_text(
          data = summary_stats2,
          aes(x = group2, y = max(mutation_counts$mutation_count) * 1.15, label = label),
          inherit.aes = FALSE,
          size = 3.5
        ) +
        labs(
          title = "Mutation count per patient: LOCH vs LOCH/TOCH (ALL)",
          x     = "Patients with LOCH or with LOCH or TOCH",
          y     = "Number of mutations per patient"
        ) +
        theme_minimal() +
        theme(legend.position = "none")
      # # ===== 4. 描画 =====
      # plots[[198]] = ggplot(mutation_counts, aes(x = group, y = mutation_count, fill = group)) +
      #   geom_violin(trim = FALSE, alpha = 0.6) +
      #   geom_boxplot(width = 0.1, outlier.shape = NA, color = "black") +
      #   stat_summary(fun = mean, geom = "point", shape = 20, size = 2, color = "red") +
      #   stat_compare_means(
      #     method = "wilcox.test",
      #     label = "p.signif",
      #     comparisons = list(c("LOCH", "TOCH")),
      #     label.y = (max(mutation_counts$mutation_count)) * 1.2
      #   ) +
      #   geom_text(
      #     data = summary_stats,
      #     #aes(x = group, y = (max(mutation_counts$mutation_count)) * 1.15, label = label),
      #     aes(x = group, y = 13.5, label = label),
      #     inherit.aes = FALSE,
      #     size = 3.5
      #   ) +
      #   #scale_y_log10() +
      #   labs(
      #     title = "Mutation count per patient: LOCH vs TOCH",
      #     x = "Group", y = "Number of mutations per patient"
      #   ) +
      #   theme_minimal() +
      #   theme(legend.position = "none")  
      # ===== 1. LOCH優先で症例分類 =====
      # ===== 1. サンプルリスト作成 =====
      # LOCH 全例
      LOCH_samples <- Data_MAF_target %>%
        filter(non_germline_non_error == 1,
               unique_alt %in% CH_candidates,
               Hugo_Symbol %in% CH_gene_entropy_3,
               logfold > 8,
               Panel == "F1Liquid CDx") %>%
        distinct(Tumor_Sample_Barcode) %>%
        mutate(group = "LOCH")
      
      # ALL = LOCH + TOCH 条件をゆるめて抽出
      ALL_samples <- Data_MAF_target %>%
        filter(non_germline_non_error == 1,
               unique_alt %in% CH_candidates,
               Hugo_Symbol %in% CH_gene_entropy_3,
               logfold > 2,         # ここに TOCH を含む
               Panel == "F1Liquid CDx") %>%
        distinct(Tumor_Sample_Barcode) %>%
        mutate(group = "ANY")
      
      # CH(-) 群：LOCH も TOCH も検出されない症例
      CH_negative_samples <- Data_MAF_target %>%
        filter(Panel == "F1Liquid CDx") %>%
        distinct(Tumor_Sample_Barcode) %>%
        filter(!(Tumor_Sample_Barcode %in% ALL_samples$Tumor_Sample_Barcode)) %>%
        mutate(group = "CH(-)")
      
      # bind_rows でまとめ（LOCHはLOCHとALLで二重に、TOCHはALLのみ、CH(-)は一度だけ）
      mutation_groups <- bind_rows(
        CH_negative_samples,
        LOCH_samples,
        ALL_samples
      )
      
      # ===== 2. TMB データと結合 =====
      TMB_data <- Data_MAF_target %>%
        filter(Panel == "F1Liquid CDx") %>%
        distinct(Tumor_Sample_Barcode, TMB) %>%
        inner_join(mutation_groups, by = "Tumor_Sample_Barcode")
      
      # ===== 3. 統計量ラベル作成 =====
      summary_stats <- TMB_data %>%
        group_by(group) %>%
        summarise(
          n          = n(),
          mean_val   = mean(TMB, na.rm = TRUE),
          sd_val     = sd(TMB, na.rm = TRUE),
          median_val = median(TMB, na.rm = TRUE),
          .groups = "drop"
        ) %>%
        mutate(
          label = sprintf("n=%d, median=%.1f\nmean=%.1f, sd=%.1f",
                          n, median_val, mean_val, sd_val)
        )
      
      # ===== 4. 描画 =====
      # グループ順を CH(-), LOCH, ANY に固定
      TMB_data <- TMB_data %>%
        mutate(group = factor(group, levels = c("CH(-)", "LOCH", "ANY")))
      setProgress(detail = "22001")
      
      # y 軸テキスト位置
      label_y <- max(TMB_data$TMB, na.rm = TRUE) * 1.2
      
      plots[[212]] <- ggplot(TMB_data, aes(x = group, y = TMB, fill = group)) +
        geom_violin(trim = FALSE, alpha = 0.6) +
        geom_boxplot(width = 0.1, outlier.shape = NA, color = "black") +
        stat_summary(fun = mean, geom = "point", shape = 20, size = 2, color = "red") +
        # ３つの組み合わせでWilcoxon検定
        stat_compare_means(
          method      = "wilcox.test",
          comparisons = list(
            c("CH(-)", "LOCH"),
            c("CH(-)", "ANY"),
            c("LOCH", "ANY")
          ),
          label       = "p.signif",
          label.y     = c(log10(label_y * 1.05), log10(label_y * 1.1), log10(label_y * 1.05))
        ) +
        geom_text(
          data = summary_stats,
          aes(x = group, y = label_y, label = label),
          inherit.aes = FALSE,
          size = 3.5
        ) +
        scale_y_log10() +
        labs(
          title = "TMB per patient: CH(-) vs LOCH(+) vs CH(+) (LOCH or TOCH)",
          x     = "CH variant type",
          y     = "TMB (log10 scale)"
        ) +
        theme_minimal() +
        theme(legend.position = "none")
      # ===== 2. 各症例の全変異数 =====
      # 1. 対象症例（DNMT3A の LOCH/TOCH）を先に定義
      dnmt3a_samples <- sample_groups %>%
        filter(Tumor_Sample_Barcode %in% Data_MAF_target$Tumor_Sample_Barcode)
      # 2. DNMT3A の変異カウントを取得
      dnmt3a_mut_counts <- Data_MAF_target %>%
        filter(non_germline_non_error == 1,
               unique_alt %in% CH_candidates,
               Hugo_Symbol == "DNMT3A",
               logfold > 2,
               Panel == "F1Liquid CDx") %>%
        count(Tumor_Sample_Barcode, name = "mutation_count")
      # 3. 全症例に結合し、変異がないものは0とする
      mutation_counts <- dnmt3a_samples %>%
        left_join(dnmt3a_mut_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(mutation_count = ifelse(is.na(mutation_count), 0, mutation_count))
      # ===== 3. 統計量（n, mean, sd, median）をラベル化 =====
      summary_stats <- mutation_counts %>%
        group_by(group) %>%
        summarise(
          mean_val = mean(mutation_count),
          sd_val = sd(mutation_count),
          median_val = median(mutation_count),
          n = n(),
          .groups = "drop"
        ) %>%
        mutate(
          label = sprintf("n=%d, median=%d\nmean=%.1f, sd=%.1f", n, median_val, mean_val, sd_val)
        )
      # ===== 4. 描画 =====
      plots[[199]] = ggplot(mutation_counts, aes(x = group, y = mutation_count, fill = group)) +
        geom_violin(trim = FALSE, alpha = 0.6) +
        geom_boxplot(width = 0.1, outlier.shape = NA, color = "black") +
        stat_summary(fun = mean, geom = "point", shape = 20, size = 2, color = "red") +
        stat_compare_means(
          method = "wilcox.test",
          label = "p.signif",
          comparisons = list(c("LOCH", "TOCH")),
          label.y = (max(mutation_counts$mutation_count)) *1.2
        ) +
        geom_text(
          data = summary_stats,
          aes(x = group, y = (max(mutation_counts$mutation_count)) * 1.16, label = label),
          inherit.aes = FALSE,
          size = 3.5
        ) +
        # scale_y_log10() +
        labs(
          title = "DNMT3A mutation count per patient: LOCH vs TOCH",
          x = "Group", y = "Number of DNMT3A mutations per patient"
        ) +
        theme_minimal() +
        theme(legend.position = "none")  
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" &
                        non_germline_non_error == 1 &
                        Hugo_Symbol %in% CH_GENE_NatComms &
                        No_found_in_F1L >= Min_count) %>%
        dplyr::select(unique_position, unique_alt, name, Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, amino.acid.change, Evidence_level, CH_genes, No_found_in_F1, No_found_in_F1L, No_found_in_F1_position, No_found_in_F1L_position, F1_patients, F1L_patients, Variant_detection_rate_F1, Variant_detection_rate_F1L, unique_position) %>%
        dplyr::distinct(Hugo_Symbol) %>%
        dplyr::arrange(Hugo_Symbol)
      data_tmp = data.frame(sort(unique(data_tmp$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary$favor_liquid = ifelse(Data_summary$Panel == "FoundationOne CDx", 0, 1)
      Data_summary_tmp = Data_summary %>%
        dplyr::filter(Panel %in% c("FoundationOne CDx", "F1Liquid CDx")) %>%
        dplyr::select(Tumor_Sample_Barcode, Age, Cancername, favor_liquid)
      if(file.exists("source/volcano_CCAT_gene.rda")){
        load("source/volcano_CCAT_gene.rda")
      } else{
        for(i in 1:length(data_tmp$Hugo_Symbol)){
          mut_ID = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
          Data_ORR = speedglm(favor_liquid ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$Hugo_Symbol)))
          }
        }
        save(data_tmp, file = "source/volcano_CCAT_gene.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^5] = 2^5
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-2] = 2^-2
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-50] = 10^-50
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor F1 or F1L", subtitle = "Genes with hotspot variants found in F1L",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-2.5,5.5),
                           breaks = c(-2,0,2,4),
                           labels = c("~-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,52),
                           breaks = c(0,10,20,30,40,50),
                           labels = c("0","10","20","30","40","50~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[116]] = g
      data_tmp = Data_GENIE %>%
        dplyr::filter(CH_hotspot_candidate == 1)
      data_tmp = data.frame(sort(unique(data_tmp$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_ALL
      if(file.exists("source/volcano_GENIE_gene.rda")){
        load("source/volcano_GENIE_gene.rda")
      } else{
        for(i in 1:length(data_tmp$Hugo_Symbol)){
          mut_ID = (Data_GENIE %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
          Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$Hugo_Symbol)))
          }
        }
        save(data_tmp, file = "source/volcano_GENIE_gene.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^5] = 2^5
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-5] = 2^-5
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-50] = 10^-50
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "Volcano plot for favor CH or blood tumor",
             subtitle = "Genes with hotspot variants, right: blood tumor favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-5.5,5.5),
                           breaks = c(-5,0,5),
                           labels = c("~-5","0","5~")) +
        scale_y_continuous(limits = c(0,52),
                           breaks = c(0,10,20,30,40,50),
                           labels = c("0","10","20","30","40","50~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[117]] = g      
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(#unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
            Panel == "F1Liquid CDx")
      gene_to_analyze = unique(names(sort(table(Data_tmp$Hugo_Symbol),
                                          decreasing = T)))
      Data_tmp3 = Data_summary %>% dplyr::filter(Panel == "F1Liquid CDx")
      #Diseases = sort(names(table(Data_tmp3$Cancername)[table(Data_tmp3$Cancername)>=50]))
      Data_tmp = Data_tmp %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        ))
      Data_tmp3 = Data_summary %>%
        dplyr::mutate(Cancername = case_when(
          Cancername %in% Diseases ~ Cancername,
          TRUE ~ "OTHER"
        )) %>% dplyr::filter(Panel == "F1Liquid CDx")
      Summary_Gene_alteration = data.frame(matrix(0,
                                                  nrow=length(Diseases),
                                                  ncol=length(gene_to_analyze)))
      colnames(Summary_Gene_alteration) = gene_to_analyze
      rownames(Summary_Gene_alteration) = Diseases
      
      for(i in Diseases){
        Data_tmp2 = Data_tmp3 %>%
          dplyr::filter(Cancername == i)
        Data_tmp2 = unique(Data_tmp2$Tumor_Sample_Barcode)
        patient_no = length(Data_tmp2)
        for(j in gene_to_analyze){
          Summary_Gene_alteration[i,j] =
            length(unique((Data_tmp %>% dplyr::filter(
              Hugo_Symbol == j &
                Tumor_Sample_Barcode %in% Data_tmp2))$Tumor_Sample_Barcode)) / patient_no * 100
        }
      }
      Data_tmp2 = Summary_Gene_alteration
      Data_tmp2$Disease = rownames(Data_tmp2)
      Data_tmp2 = Data_tmp2 %>% gather(value = "freq", key = Gene, -Disease)
      Data_tmp2 <- transform(Data_tmp2, Disease= factor(Disease, levels = sort(unique(Disease), decreasing = TRUE)))
      Data_tmp2 <- Data_tmp2 %>%
        group_by(Gene) %>%
        mutate(Gene = factor(Gene, levels = names(sort(table(Gene), decreasing = TRUE))))
      g = ggplot(Data_tmp2, aes(Gene, as.factor(Disease))) +
        geom_tile(aes(fill = freq), color = "black",
                  #            lwd = 1,
                  linetype = 1) + 
        geom_text(aes(label = round(freq, 0))) +
        scale_fill_gradient(low = "white", high = "red", limit=c(0,100), name="Frequency (%)") +
        theme_classic() +
        labs(title = "Non-germline variant in CH-related genes, F1 liquid",
             subtitle = "Including non-hotspot variants") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              axis.line.x = element_blank(),
              axis.line.y = element_blank())
      plots[[118]] = g
      Data_tmp = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
            Panel == "F1Liquid CDx")
      Data_tmp$color_group = ifelse(Data_tmp$logfold > 8, "LOCH", "TOCH")
      barcode_counts <- as.data.frame(table(Data_tmp$Tumor_Sample_Barcode))
      colnames(barcode_counts) <- c("Patient", "Freq")
      total_patients <- length(unique(Data_summary_F1L$Tumor_Sample_Barcode))  # 全症例数
      mutated_patients <- nrow(barcode_counts)  # 変異がある症例
      N_zero <- total_patients - mutated_patients  # 変異ゼロの症例数
      # カウントゼロのデータを追加
      barcode_counts <- rbind(barcode_counts, data.frame(Patient = rep("0", N_zero), Freq = rep(0, N_zero)))
      # 図を作成
      # カウントを再集計してから割合を計算
      plot_data <- barcode_counts %>%
        dplyr::group_by(Freq) %>%
        dplyr::summarise(Patients = n()) %>%
        dplyr::mutate(Percent = Patients / total_patients * 100)
      # 描画
      g <- ggplot(plot_data, aes(x = factor(Freq), y = Percent)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        geom_text(aes(label = Patients), vjust = -0.3, size = 3.5) +  # 上部に症例数
        labs(x = "LD-CH/LE-CH count per patient", 
             y = "Patients (%)", 
             title = paste0("LD-CH/LE-CH detected by F1-Liquid CDx ", nrow(barcode_counts), " patients")) +
        theme_classic() +
        ylim(0, max(plot_data$Percent) * 1.1)  # テキストスペースのために少し余裕
      plots[[119]] = g
      # 1. DNMT3AのTOCH変異数を各患者ごとにカウント
      dnmt3a_counts <- Data_tmp %>%
        #filter(color_group == "TOCH") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "CH variant count per patient",
          y = "Total CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "Total CH variants by CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "CH count in any genes per patient",
          y = "Mean LOCH variant count",
          title = "Mean LOCH vairants by CH variant count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "CH variant count per patient",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by Total CH count (Any genes)"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[216]] = g1
      plots[[217]] = g2
      plots[[226]] = g3
      # 1. DNMT3AのTOCH変異数を各患者ごとにカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "DNMT3A") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "DNMT3A CH variant count per patient",
          y = "Total CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "Total CH variants by DNMT3A CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.3)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "DNMT3A CH count per patient",
          y = "Mean LOCH variant count",
          title = "Mean LOCH vairants by DNMT3A CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "CH variant count per patient",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by Total CH count (DNMT3A)"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[218]] = g1
      plots[[219]] = g2
      plots[[227]] = g3
      # 1. DNMT3AのTOCH変異数を各患者ごとにカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "CHEK2") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "CHEK2 CH variant count per patient",
          y = "Total CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "Total CH variants by CHEK2 CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "CHEK2 CH count per patient",
          y = "Mean LOCH variant count",
          title = "Mean LOCH vairants by CHEK2 CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "CH variant count per patient",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by Total CH count (CHEK2)"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[220]] = g1
      plots[[221]] = g2
      plots[[228]] = g3
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "TET2") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "TET2 CH variant count per patient",
          y = "Total CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "Total CH variants by TET2 CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "TET2 CH count per patient",
          y = "Mean LOCH variant count",
          title = "Mean LOCH vairants by TET2 CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "CH variant count per patient",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by Total CH count (TET2)"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[222]] = g1
      plots[[223]] = g2
      plots[[229]] = g3
      setProgress(detail = "22811")
      # 1. 各患者ごとに DNMT3A の変異数をカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "ASXL1") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "ASXL1 CH variant count per patient",
          y = "Total CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "Total CH variants by ASXL1 CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "ASXL1 CH count per patient",
          y = "Mean LOCH variant count",
          title = "Mean LOCH vairants by ASXL1 CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "CH variant count per patient",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by Total CH count (ASXL1)"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[224]] = g1
      plots[[225]] = g2
      plots[[230]] = g3
      # 1. 各患者ごとに DNMT3A の変異数をカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "ATM") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "ATM CH variant count per patient",
          y = "Total CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "Total CH variants by ATM CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "ATM CH count per patient",
          y = "Mean LOCH variant count",
          title = "Mean LOCH vairants by ATM CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "CH variant count per patient",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by Total CH count (ATM)"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[231]] = g1
      plots[[232]] = g2
      plots[[233]] = g3      
      # 1. DNMT3AのTOCH変異数を各患者ごとにカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "DNMT3A") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "DNMT3A") %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "DNMT3A CH variant count per patient",
          y = "DNMT3A CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "DNMT3A CH variants by DNMT3A CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.3)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "DNMT3A CH count per patient",
          y = "Mean DNMT3A LOCH variant count",
          title = "Mean DNMT3A LOCH vairants by DNMT3A CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "DNMT3A CH variant count per patient",
          y = "Proportion of DNMT3A LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by DNMT3A CH count"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[234]] = g1
      plots[[235]] = g2
      plots[[236]] = g3
      # 1. DNMT3AのTOCH変異数を各患者ごとにカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "CHEK2") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "CHEK2") %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "CHEK2 CH variant count per patient",
          y = "CHEK2 CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "CHEK2 CH variants by CHEK2 CH variant count (with LOCH mean)"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "CHEK2 CH count per patient",
          y = "Mean CHEK2 LOCH variant count",
          title = "Mean CHEK2 LOCH vairants by CHEK2 CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "CHEK2 CH variant count per patient",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by CHEK2 CH count"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      setProgress(detail = "23239")
      plots[[237]] = g1
      plots[[238]] = g2
      plots[[239]] = g3
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "TET2") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "TET2") %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "TET2 CH variant count per patient",
          y = "TET2 CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "TET2 CH variants by TET2 CH variant count"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "TET2 CH count per patient",
          y = "Mean TET2 LOCH variant count",
          title = "Mean TET2 LOCH vairants by TET2 CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "TET2 CH variant count per patient",
          y = "Proportion of TET2 LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by TET2 CH count"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[240]] = g1
      plots[[241]] = g2
      plots[[242]] = g3
      # 1. 各患者ごとに DNMT3A の変異数をカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "ASXL1") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "ASXL1") %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "ASXL1 CH variant count per patient",
          y = "ASXL1 CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "ASXL1 CH variants by ASXL1 CH variant count"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "ASXL1 CH count per patient",
          y = "Mean ASXL1 LOCH variant count",
          title = "Mean ASXL1 LOCH vairants by ASXL1 CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "ASXL1 CH variant count per patient",
          y = "Proportion of ASXL1 LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by ASXL1 CH count"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[243]] = g1
      plots[[244]] = g2
      plots[[245]] = g3
      # 1. 各患者ごとに DNMT3A の変異数をカウント
      dnmt3a_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "ATM") %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        filter(Hugo_Symbol == "ATM") %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "ATM CH variant count per patient",
          y = "ATM CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "ATM CH variants by ATM CH variant count"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.2)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "ATM CH count per patient",
          y = "Mean ATM LOCH variant count",
          title = "Mean ATM LOCH vairants by ATM CH count"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "ATM CH variant count per patient",
          y = "Proportion of ATM LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by ATM CH count"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[246]] = g1
      plots[[247]] = g2
      plots[[248]] = g3      
      # 1. DNMT3AのTOCH変異数を各患者ごとにカウント
      dnmt3a_counts <- Data_tmp %>%
        # filter(color_group == "TOCH" & Hugo_Symbol == "DNMT3A") %>%
        filter(color_group == "TOCH" & str_detect(unique_alt, "DNMT3A_Y735")) %>%
        group_by(Tumor_Sample_Barcode) %>%
        summarise(DNMT3A_count = n(), .groups = "drop")
      # 2. LOCH / TOCH の変異数を患者ごとにカウント
      group_counts <- Data_tmp %>%
        #filter(Hugo_Symbol == "DNMT3A") %>%
        # filter(color_group == "TOCH" & Hugo_Symbol == "DNMT3A") %>%
        #filter(str_detect(unique_alt, ">|fs|trun")) %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(Mut_count = n(), .groups = "drop")
      # 3. 患者 × color_group の全組み合わせを作成（全症例考慮）
      all_patients <- unique(Data_summary_F1L$Tumor_Sample_Barcode)
      color_levels <- c("LOCH", "TOCH")
      full_combos <- expand.grid(Tumor_Sample_Barcode = all_patients,
                                 color_group = color_levels,
                                 stringsAsFactors = FALSE)
      # 4. group_countsと結合し、存在しない組み合わせは0補完
      group_counts_full <- full_combos %>%
        left_join(group_counts, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        mutate(Mut_count = replace_na(Mut_count, 0))
      # 5. DNMT3A変異数を結合（存在しない患者には0補完）
      plot_df <- group_counts_full %>%
        left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
        mutate(DNMT3A_count = replace_na(DNMT3A_count, 0))
      # 6a. 症例数も集計（DNMT3A_count ごと）
      patient_counts <- plot_df %>%
        group_by(DNMT3A_count) %>%
        summarise(PatientN = n_distinct(Tumor_Sample_Barcode), .groups = "drop")
      # 6b. LOCH だけ抽出して、DNMT3A_count ごとの平均を計算
      loch_means <- plot_df %>%
        filter(color_group == "LOCH") %>%
        group_by(DNMT3A_count) %>%
        summarise(Mean_LOCH = mean(Mut_count), .groups = "drop")
      # 6c. 棒グラフの x 軸に使う summary に平均LOCHを結合
      plot_summary_with_mean <- plot_df %>%
        group_by(DNMT3A_count, color_group) %>%
        summarise(TotalMutations = sum(Mut_count), .groups = "drop") %>%
        left_join(loch_means, by = "DNMT3A_count") %>%
        left_join(patient_counts, by = "DNMT3A_count")
      # 7. 積み上げ棒グラフ作成（LOCH 平均を上に表示）
      g1 <- ggplot(plot_summary_with_mean, aes(x = factor(DNMT3A_count), y = TotalMutations, fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        geom_text(aes(label = TotalMutations),
                  position = position_stack(vjust = 0.5),
                  size = 4.5, color = "black") +
        # LOCHの平均を上に表示（LOCHのバーの上にだけ表示）
        geom_text(data = loch_means,
                  aes(x = factor(DNMT3A_count), y = Inf,
                      label = sprintf("Mean LOCH: %.2f", Mean_LOCH)),
                  vjust = -0.5, inherit.aes = FALSE, size = 4, fontface = "italic") +
        labs(
          x = "DNMT3A Y735 variant",
          y = "CH variants (LOCH + TOCH)",
          fill = "CH type",
          title = "CH variant proportion by DNMT3A Y735 variant"
        ) +
        theme_minimal() +
        ylim(0, max(plot_summary_with_mean$TotalMutations) * 1.3)  # 余白調整
      # 比率プロット用データを作成（すでに plot_summary_with_mean がある場合）
      plot_prop <- plot_summary_with_mean %>%
        group_by(DNMT3A_count) %>%
        mutate(Proportion = TotalMutations / sum(TotalMutations)) %>%
        ungroup()
      # 割合グラフの作成
      g2 <- ggplot(loch_means, aes(x = factor(DNMT3A_count), y = Mean_LOCH)) +
        geom_bar(stat = "identity", fill = "steelblue", color = "black") +
        # バーの上に Mean_LOCH の値を表示
        geom_text(aes(label = sprintf("%.2f", Mean_LOCH)),
                  vjust = -0.3, size = 4, color = "black") +
        labs(
          x = "DNMT3A Y735 variant",
          y = "Mean LOCH variant count",
          title = "Mean LOCH vairants by DNMT3A Y735 variant"
        ) +
        theme_minimal() +
        ylim(0, max(loch_means$Mean_LOCH) * 1.1) +
        theme(
          axis.text.x = element_text(size = 12),
          axis.title    = element_text(size = 14),
          plot.title    = element_text(size = 16, face = "bold")
        )
      g3 <- ggplot(plot_prop, 
                   aes(x = factor(DNMT3A_count), 
                       y = Proportion, 
                       fill = color_group)) +
        geom_bar(stat = "identity", color = "black") +
        # 各セグメントに割合ラベル
        geom_text(aes(label = sprintf("%.1f%%", Proportion * 100)),
                  position = position_stack(vjust = 0.5),
                  size = 3) +
        scale_y_continuous(labels = percent_format(accuracy = 1)) +
        labs(
          x = "DNMT3A Y735 variant",
          y = "Proportion of LOCH/TOCH variants",
          fill = "CH type",
          title = "Proportion of LOCH vs TOCH by DNMT3A Y735 variant"
        ) +
        theme_minimal() +
        theme(
          axis.text.x  = element_text(size = 12),
          axis.title   = element_text(size = 14),
          plot.title   = element_text(size = 16, face = "bold"),
          legend.title = element_text(size = 13),
          legend.text  = element_text(size = 11)
        )
      plots[[249]] = g1
      plots[[250]] = g2
      plots[[147]] = g3
      
      # # 1. DNMT3AのTOCH変異を抽出し、mut_patternごとにカウント
      # mut_counts <- Data_tmp %>%
      #   #filter(Hugo_Symbol == "DNMT3A", color_group == "TOCH") %>%
      #   filter(color_group == "LOCH") %>%
      #   group_by(Tumor_Sample_Barcode, mut_pattern) %>%
      #   summarise(Count = n(), .groups = "drop")
      # # 2. 各患者のTOCH変異数を取得
      # dnmt3a_counts <- Data_tmp %>%
      #   filter(color_group == "TOCH") %>%
      #   group_by(Tumor_Sample_Barcode, mut_pattern) %>%
      #   summarise(Count = n(), .groups = "drop") %>% 
      #   group_by(Tumor_Sample_Barcode) %>%
      #   summarise(DNMT3A_TOCH_count = sum(Count), .groups = "drop")
      # # 3. データの結合とmut_patternの割合を計算
      # plot_data <- mut_counts %>%
      #   left_join(dnmt3a_counts, by = "Tumor_Sample_Barcode") %>%
      #   mutate(Proportion = Count / DNMT3A_TOCH_count) %>%
      #   mutate(Proportion = replace_na(Proportion, 0)) %>%
      #   mutate(DNMT3A_TOCH_count = replace_na(DNMT3A_TOCH_count, 0))
      #   # 4. 各TOCH変異数ごとにmut_patternの割合を集計
      # plot_summary <- plot_data %>%
      #   group_by(DNMT3A_TOCH_count, mut_pattern) %>%
      #   summarise(Total_Count = sum(Count), .groups = "drop") %>%
      #   group_by(DNMT3A_TOCH_count) %>%
      #   mutate(Proportion = Total_Count / sum(Total_Count)) %>%
      #   ungroup()
      # # 5. 積み上げ棒グラフの作成（100%スケール）
      # ggplot(plot_summary, aes(x = factor(DNMT3A_TOCH_count), y = Proportion, fill = mut_pattern)) +
      #   geom_bar(stat = "identity", color = "black", position = "fill") +
      #   scale_y_continuous(labels = percent_format(accuracy = 1)) +
      #   labs(
      #     x = "DNMT3A TOCH変異数（患者ごと）",
      #     y = "mut_patternの割合",
      #     fill = "mut_pattern",
      #     title = "DNMT3A TOCH変異数ごとのmut_pattern割合（100%スケール）"
      #   ) +
      #   theme_minimal()
      
      mutation_data = 
      (data.frame(Data_GENIE %>%
                   filter(Variant_Classification == "Splice_Site" &
                            Sample_type == "Blood_cancer" &
                            ((Reference_Allele == "C" & Tumor_Seq_Allele2 == "T") |
                               (Reference_Allele == "G" & Tumor_Seq_Allele2 == "A")))))
      data_maf_ACC = mutation_data %>% dplyr::select(Hugo_Symbol)
      data_maf_ACC$Chromosome = mutation_data$Chromosome
      # data_maf_ACC$Chromosome = str_remove(mutation_data$Chromosome, "chr")
      # data_maf_ACC$Chromosome = paste0("chr", mutation_data$Chromosome)
      data_maf_ACC$Start_Position = as.integer(mutation_data$Start_Position)
      data_maf_ACC$End_Position = as.integer(mutation_data$Start_Position) + nchar(mutation_data$Reference_Allele) - 1
      data_maf_ACC$Strand = "+"
      data_maf_ACC$Variant_Classification = "Missense_Mutation"
      data_maf_ACC$Variant_Type = ifelse(nchar(mutation_data$Reference_Allele) == 1 & nchar(mutation_data$Tumor_Seq_Allele2) == 1, "SNV",
                                         ifelse(nchar(mutation_data$Reference_Allele) > nchar(mutation_data$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf_ACC$Reference_Allele = mutation_data$Reference_Allele
      data_maf_ACC$Tumor_Seq_Allele1 = mutation_data$Reference_Allele
      data_maf_ACC$Tumor_Seq_Allele2 = mutation_data$Tumor_Seq_Allele2
      data_maf_ACC$Tumor_Sample_Barcode = mutation_data$Tumor_Sample_Barcode
      data_maf_ACC$name = mutation_data$name
      data_maf_ACC$Cancername = mutation_data$Cancername
      data_maf_ACC$Age = mutation_data$Age
      data_maf_ACC$pnt_col = mutation_data$pnt_col
      #hg19 <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
      #hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      data_maf_ACC = as.data.frame(data_maf_ACC)
      maf_ACC <- filterSNV(dataSet = data_maf_ACC,
                           seq_colNames = c("Reference_Allele",
                                            "Tumor_Seq_Allele1",
                                            "Tumor_Seq_Allele2"))
      maf_ACC <- attachContext(mutData = maf_ACC,
                               chr_colName = "Chromosome",
                               start_colName = "Start_Position",
                               end_colName = "End_Position",
                               nucl_contextN = 3,
                               BSGenomeDb = hg38)
      maf_ACC <- removeMismatchMut(mutData = maf_ACC,
                                   refMut_colName = "Reference_Allele",
                                   context_colName = "context",
                                   refMut_format = "N")
      mutation_data_signature <- attachMutType(mutData = maf_ACC,
                                               ref_colName = "Reference_Allele",
                                               var_colName = "Tumor_Seq_Allele1",
                                               var2_colName = "Tumor_Seq_Allele2",
                                               context_colName = "context")
      ID_ACC = (mutation_data_signature %>% filter(mutType == "A[C>T]C"))$Tumor_Sample_Barcode
      ID_CCT = (mutation_data_signature %>% filter(mutType == "C[C>T]T"))$Tumor_Sample_Barcode
      data_tmp = Data_GENIE %>%
        dplyr::filter(Sample_type == "Blood_cancer" &
                        CH_hotspot_candidate == 1 &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        # !str_detect(unique_alt, "trunc") &
                        # !str_detect(unique_alt, "fs_NOS") &
                        !str_detect(unique_alt, "\\+1G>A") &
                        !str_detect(unique_alt, "splice")
        )
      data_tmp = data.frame(sort(unique(data_tmp$unique_alt)))
      colnames(data_tmp) = "unique_alt"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% ID_ACC ~ 1,
          TRUE ~ 0))
      if(file.exists("source/volcano_GENIE_ACC.rda")){
        load("source/volcano_GENIE_ACC.rda")
      } else{
        for(i in 1:length(data_tmp$unique_alt)){
          mut_ID = (Data_GENIE %>% dplyr::filter(unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
          Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$unique_alt)))
          }
        }
        save(data_tmp, file = "source/volcano_GENIE_ACC.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^10] = 2^10
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-15] = 10^-15
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = unique_alt,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor splice site A[C>T]C variant",
             subtitle = "Genes with hotspot variants, right: A[C>T]C favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,12),
                           breaks = c(-4,-2,0,2,4,6,8,10),
                           labels = c("~-4","-2","0","2","4","6","8","10~")) +
        scale_y_continuous(limits = c(0,16),
                           breaks = c(0,5,10,15),
                           labels = c("0","5","10","15~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      table_GENIE_ACC = data_tmp
      plots[[120]] = g      
      data_tmp = Data_GENIE %>%
        dplyr::filter(Sample_type == "Blood_cancer" &
                        CH_hotspot_candidate == 1 &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        #!str_detect(unique_alt, "trunc") &
                        #!str_detect(unique_alt, "fs_NOS") &
                        !str_detect(unique_alt, "\\-1G>A") &
                        !str_detect(unique_alt, "splice")
        )
      data_tmp = data.frame(sort(unique(data_tmp$unique_alt)))
      colnames(data_tmp) = "unique_alt"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% ID_CCT ~ 1,
          TRUE ~ 0))
      if(file.exists("source/volcano_GENIE_CCT.rda")){
        load("source/volcano_GENIE_CCT.rda")
      } else{
        for(i in 1:length(data_tmp$unique_alt)){
          mut_ID = (Data_GENIE %>% dplyr::filter(unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
          Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$unique_alt)))
          }
        }
        save(data_tmp, file = "source/volcano_GENIE_CCT.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^10] = 2^10
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-15] = 10^-15
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = unique_alt,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor splice site C[C>T]T variant",
             subtitle = "Genes with hotspot variants, right: C[C>T]T favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,6),
                           breaks = c(-4,-2,0,2,4,6),
                           labels = c("~-4","-2","0","2","4","6~")) +
        scale_y_continuous(limits = c(0,10),
                           breaks = c(0,5,10),
                           labels = c("0","5","10~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      table_GENIE_CCT = data_tmp
      plots[[121]] = g    
      mutation_data = 
        (data.frame(Data_MAF_target %>%
                    filter(non_germline_non_error <= 1 &
                             str_detect(name, ">") & !str_detect(name, "TERT_c\\.-79") &
                             Panel == "F1Liquid CDx" &
                             ((Reference_Allele == "C" & Tumor_Seq_Allele2 == "T") |
                                (Reference_Allele == "G" & Tumor_Seq_Allele2 == "A")))))
      data_maf_ACC = mutation_data %>% dplyr::select(Hugo_Symbol)
      # data_maf_ACC$Chromosome = mutation_data$Chromosome
      # data_maf_ACC$Chromosome = str_remove(mutation_data$Chromosome, "chr")
      data_maf_ACC$Chromosome = paste0("chr", mutation_data$Chromosome)
      data_maf_ACC$Start_Position = as.integer(mutation_data$Start_Position)
      data_maf_ACC$End_Position = as.integer(mutation_data$Start_Position) + nchar(mutation_data$Reference_Allele) - 1
      data_maf_ACC$Strand = "+"
      data_maf_ACC$Variant_Classification = "Missense_Mutation"
      data_maf_ACC$Variant_Type = ifelse(nchar(mutation_data$Reference_Allele) == 1 & nchar(mutation_data$Tumor_Seq_Allele2) == 1, "SNV",
                                         ifelse(nchar(mutation_data$Reference_Allele) > nchar(mutation_data$Tumor_Seq_Allele2), "DEL","INS"))
      data_maf_ACC$Reference_Allele = mutation_data$Reference_Allele
      data_maf_ACC$Tumor_Seq_Allele1 = mutation_data$Reference_Allele
      data_maf_ACC$Tumor_Seq_Allele2 = mutation_data$Tumor_Seq_Allele2
      data_maf_ACC$Tumor_Sample_Barcode = mutation_data$Tumor_Sample_Barcode
      data_maf_ACC$name = mutation_data$name
      data_maf_ACC$Cancername = mutation_data$Cancername
      data_maf_ACC$Age = mutation_data$Age
      data_maf_ACC$pnt_col = mutation_data$pnt_col
      #hg19 <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
      #hg38 <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
      setProgress(detail = "24012")
      data_maf_ACC = as.data.frame(data_maf_ACC)
      maf_ACC <- filterSNV(dataSet = data_maf_ACC,
                           seq_colNames = c("Reference_Allele",
                                            "Tumor_Seq_Allele1",
                                            "Tumor_Seq_Allele2"))
      maf_ACC <- attachContext(mutData = maf_ACC,
                               chr_colName = "Chromosome",
                               start_colName = "Start_Position",
                               end_colName = "End_Position",
                               nucl_contextN = 3,
                               BSGenomeDb = hg38)
      maf_ACC <- removeMismatchMut(mutData = maf_ACC,
                                   refMut_colName = "Reference_Allele",
                                   context_colName = "context",
                                   refMut_format = "N")
      mutation_data_signature <- attachMutType(mutData = maf_ACC,
                                               ref_colName = "Reference_Allele",
                                               var_colName = "Tumor_Seq_Allele1",
                                               var2_colName = "Tumor_Seq_Allele2",
                                               context_colName = "context")
      ID_ACC = (mutation_data_signature %>% filter(mutType == "A[C>T]C"))$Tumor_Sample_Barcode
      ID_CCT = (mutation_data_signature %>% filter(mutType == "C[C>T]T"))$Tumor_Sample_Barcode
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" &
                        non_germline_non_error == 1 &
                        unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        # !str_detect(unique_alt, "trunc") &
                        # !str_detect(unique_alt, "fs_NOS") &
                        !str_detect(unique_alt, "\\+1G>A") &
                        #Hugo_Symbol %in% CH_GENE_NatComms &
                        No_found_in_F1L >= Min_count) %>%
        dplyr::select(unique_position, unique_alt, name, Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, amino.acid.change, Evidence_level, CH_genes, No_found_in_F1, No_found_in_F1L, No_found_in_F1_position, No_found_in_F1L_position, F1_patients, F1L_patients, Variant_detection_rate_F1, Variant_detection_rate_F1L, unique_position) %>%
        dplyr::distinct(unique_alt) %>%
        dplyr::arrange(unique_alt)
      data_tmp = data.frame(sort(unique(data_tmp$unique_alt)))
      colnames(data_tmp) = "unique_alt"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Data_summary %>%
        filter(Panel == "F1Liquid CDx") %>%
        select(Tumor_Sample_Barcode, Age, Cancername)
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% ID_ACC ~ 1,
          TRUE ~ 0))
      if(file.exists("source/volcano_CCAT_ACC.rda")){
        load("source/volcano_CCAT_ACC.rda")
      } else{
        for(i in 1:length(data_tmp$unique_alt)){
          mut_ID = (Data_MAF_target %>% dplyr::filter(non_germline_non_error <= 1 & unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
          Data_ORR = speedglm(favor_liquid ~ mutation + Age + Cancername, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$unique_alt)))
          }
        }
        save(data_tmp, file = "source/volcano_CCAT_ACC.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      # data_tmp$p_value_adj = data_tmp$p_value
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= 'none')
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = unique_alt,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "CCAT, Volcano plot for favor splice site A[C>T]C variant",
             subtitle = "Genes with hotspot variants, right: A[C>T]C favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,4.5),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      table_CCAT_ACC = data_tmp
      plots[[122]] = g      
      data_tmp = Data_MAF_target %>%
        dplyr::filter(Panel == "F1Liquid CDx" &
                        non_germline_non_error == 1 &
                        unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        #!str_detect(unique_alt, "trunc") &
                        #!str_detect(unique_alt, "fs_NOS") &
                        !str_detect(unique_alt, "-1G>A") &
                        Hugo_Symbol %in% CH_GENE_NatComms &
                        No_found_in_F1L >= Min_count) %>%
        dplyr::select(unique_position, unique_alt, name, Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, amino.acid.change, Evidence_level, CH_genes, No_found_in_F1, No_found_in_F1L, No_found_in_F1_position, No_found_in_F1L_position, F1_patients, F1L_patients, Variant_detection_rate_F1, Variant_detection_rate_F1L, unique_position) %>%
        dplyr::distinct(unique_alt) %>%
        dplyr::arrange(unique_alt)
      data_tmp = data.frame(sort(unique(data_tmp$unique_alt)))
      colnames(data_tmp) = "unique_alt"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Data_summary %>%
        filter(Panel == "F1Liquid CDx") %>%
        select(Tumor_Sample_Barcode, Age, Cancername)
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% ID_CCT ~ 1,
          TRUE ~ 0))
      if(file.exists("source/volcano_CCAT_CCT.rda")){
        load("source/volcano_CCAT_CCT.rda")
      } else{
        for(i in 1:length(data_tmp$unique_alt)){
          mut_ID = (Data_MAF_target %>% dplyr::filter(non_germline_non_error <= 1 & unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
          Data_ORR = speedglm(favor_liquid ~ mutation + Age + Cancername, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$unique_alt)))
          }
        }
        save(data_tmp, file = "source/volcano_CCAT_CCT.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^6] = 2^6
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-15] = 10^-3
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = unique_alt,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "CCAT, Volcano plot for favor splice site C[C>T]T variant",
             subtitle = "Genes with hotspot variants, right: C[C>T]T favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,6.5),
                           breaks = c(-4,-2,0,2,4,6),
                           labels = c("~-4","-2","0","2","4","6~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      table_CCAT_CCT = data_tmp
      plots[[123]] = g    
      print(10)
      Summary_GENIE_solid = read_csv("source/GENIE_ALL.csv", show_col_types = FALSE) %>%
        filter(Sample_type != "Blood_cancer")
      ID_Solid_cancer = unique(Summary_GENIE_solid$Tumor_Sample_Barcode)
      if (!file.exists("source/GENIE_solid.tsv")) {
        library(rtracklayer)
        library(GenomicRanges)
        Data_GENIE_solid = read_tsv("source/GENIE_variants.txt", show_col_types = FALSE)
        Data_GENIE_solid = Data_GENIE_solid %>% dplyr::filter(Tumor_Sample_Barcode %in% ID_Solid_cancer)
        Data_GENIE_solid$Chromosome = paste0("chr", Data_GENIE_solid$Chromosome)
        Data_GENIE_solid$Chromosome[Data_GENIE_solid$Chromosome == "chrMT"]= "chrM"
        chain_file <- "source/hg19ToHg38.over.chain"
        if (!file.exists(chain_file)) {
          download.file("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz",
                        destfile = paste0(chain_file, ".gz"))
          system(paste0("gunzip -f ", chain_file, ".gz"))
        }
        gr <- GRanges(seqnames = Data_GENIE_solid$Chromosome,
                      ranges = IRanges(start = Data_GENIE_solid$Start_Position, end = Data_GENIE_solid$End_Position),
                      strand = "*")
        chain <- import.chain(chain_file)
        # lifted_gr <- liftOver(gr, chain)
        # lifted_gr <- unlist(lifted_gr)
        # tmp = lifted_gr@partitioning@end
        # tmp0 = diff(tmp)
        # table_correction = data.frame(tmp0)
        # table_correction$count = 0
        # k = 1
        # for(i in 1:length(tmp0)){
        #   if(tmp0[i] == 1){
        #     k = k + 1
        #   } else if(tmp0[i] == 0){
        #     Data_GENIE_solid = Data_GENIE_solid[-k,]
        #   } else if(tmp0[i] == 2){
        #     lifted_gr = lifted_gr[-(k + 1),]
        #     k = k + 1
        #   } else if(tmp0[i] == 4){
        #     lifted_gr = lifted_gr[-(k + 1),]
        #     lifted_gr = lifted_gr[-(k + 1),]
        #     lifted_gr = lifted_gr[-(k + 1),]
        #     k = k + 1
        #   }
        # }
        lifted_list <- liftOver(gr, chain)          # GRangesList
        nmap <- elementNROWS(lifted_list)           # 各入力が何件に変換されたか
        
        # 0件（liftOverできない）を落とす
        keep <- which(nmap >= 1)
        Data_GENIE_solid <- Data_GENIE_solid[keep, , drop = FALSE]
        lifted_list <- lifted_list[keep]
        
        # 複数件は「最初の1件だけ採用」
        lifted_1 <- endoapply(lifted_list, function(x) x[1])
        
        # 1:1にGRangesへ
        lifted_gr <- unlist(lifted_1, use.names = FALSE)
        Data_GENIE_solid$Chromosome <- as.character(seqnames(lifted_gr))
        Data_GENIE_solid$Start_Position <- start(lifted_gr)
        Data_GENIE_solid$End_Position <- end(lifted_gr)
        Data_GENIE_solid$HGVSp_Short[is.na(Data_GENIE_solid$HGVSp_Short)] = Data_GENIE_solid$HGVSc[is.na(Data_GENIE_solid$HGVSp_Short)]
        Data_GENIE_solid$HGVSp_Short[is.na(Data_GENIE_solid$HGVSp_Short)] = paste0(
          Data_GENIE_solid$Start_Position[is.na(Data_GENIE_solid$HGVSp_Short)],
          Data_GENIE_solid$Reference_Allele[is.na(Data_GENIE_solid$HGVSp_Short)],">",
          Data_GENIE_solid$Tumor_Seq_Allele2[is.na(Data_GENIE_solid$HGVSp_Short)]
        )
        Data_GENIE_solid$gnomAD_AF[is.na(Data_GENIE_solid$gnomAD_AF)] = 0
        Data_GENIE_solid = Data_GENIE_solid %>%
          dplyr::select(
            Hugo_Symbol,  Tumor_Sample_Barcode, HGVSp_Short, Variant_Classification, Chromosome, Start_Position, End_Position,
            Reference_Allele, Tumor_Seq_Allele2, t_alt_count, t_depth, gnomAD_AF) %>%
          dplyr::left_join(Summary_GENIE_solid, by = "Tumor_Sample_Barcode")
        Data_GENIE_solid$VAF = Data_GENIE_solid$t_alt_count / Data_GENIE_solid$t_depth
        Data_GENIE_solid = Data_GENIE_solid[!is.na(Data_GENIE_solid$VAF),]
        Data_GENIE_solid = Data_GENIE_solid[Data_GENIE_solid$VAF != Inf,]
        Data_GENIE_solid = Data_GENIE_solid[Data_GENIE_solid$VAF != -Inf,]
        Data_GENIE_solid$HGVSp_Short = str_remove(Data_GENIE_solid$HGVSp_Short, "p.")
        Data_GENIE_solid = Data_GENIE_solid %>%
          dplyr::select(-t_alt_count, -t_depth)
        colname_data = colnames(Data_GENIE_solid)
        colname_data[colname_data == "HGVSp_Short"] = "amino.acid.change"
        colname_data[colname_data == "gnomAD_AF"] = "MinorAF"
        colnames(Data_GENIE_solid) = colname_data
        write_tsv(Data_GENIE_solid, "source/GENIE_solid.tsv")
      } else {
        Data_GENIE_solid = read_tsv("source/GENIE_solid.tsv", show_col_types = FALSE)
      }
      Data_GENIE_solid$Chromosome = str_remove(Data_GENIE_solid$Chromosome, "chr")
      Data_GENIE_solid$unique_position = paste0(Data_GENIE_solid$Chromosome, "_",
                                                Data_GENIE_solid$Start_Position, "_",
                                                Data_GENIE_solid$Reference_Allele, "_",
                                                Data_GENIE_solid$Tumor_Seq_Allele2
      )
      Data_GENIE_solid$name = paste0(Data_GENIE_solid$Hugo_Symbol, "_",
                                     Data_GENIE_solid$amino.acid.change
      )
      ID_GENIE_F1L = read_csv("source/genie_F1L.csv", show_col_types = F)$`Sample ID`
      Data_tmp = Data_GENIE_solid %>% dplyr::filter(!Tumor_Sample_Barcode %in% ID_GENIE_F1L)
      Data_tmp$Position = as.numeric(unlist(str_extract(Data_tmp$amino.acid.change, "\\d+")))
      Data_tmp = Data_tmp %>% left_join(table_DNMT3A, by = "Position")
      Data_tmp$Destabilizing[is.na(Data_tmp$Destabilizing)] = "Unknown"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "c")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "splice")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "\\*")] = "Truncating"
      Data_tmp$Destabilizing = factor(Data_tmp$Destabilizing)
      lolliplot_gene = input$lolliplot_gene
      if(is.null(lolliplot_gene)){
        lolliplot_gene = "DNMT3A"
      } else if(lolliplot_gene == ""){
        lolliplot_gene = "DNMT3A"
      }
      data_lolliplot = Data_tmp %>%
        dplyr::select(Chromosome,
                      Start_Position,
                      End_Position,
                      Reference_Allele,
                      Tumor_Seq_Allele2,
                      Tumor_Sample_Barcode,
                      Hugo_Symbol,
                      amino.acid.change)
      if(lolliplot_gene == "DNMT3A"){
        data_lolliplot$Variant_Classification = Data_tmp$Destabilizing
      } else{
        data_lolliplot$Variant_Classification = "Missense_Mutation"
      }
      data_lolliplot = data_lolliplot %>%
        dplyr::filter(!str_detect(amino.acid.change, "c.")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "_")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "^\\*")) %>%
        dplyr::filter(Hugo_Symbol == lolliplot_gene) 
      if(length(data_lolliplot$Chromosome) == 0){
        plots[[124]] = gg_empty()
      } else {
        var<-unique(data_lolliplot[!duplicated(data_lolliplot),])#remove duplicates
        var[order(var$Hugo_Symbol,gsub("([A-Z]+)([0-9]+)","\\2",var$amino.acid.change)),]#order
        var.freq<-plyr::count(var,c("Hugo_Symbol","amino.acid.change","Variant_Classification"))
        var.aanum<-unlist(lapply(regmatches(var.freq$amino.acid.change,gregexpr(pattern="*(\\d+)",var.freq$amino.acid.change)),function(x) x[[1]]))
        var.plot.data<-cbind(var.freq,as.numeric(as.character(var.aanum)))#hard way to remove factor
        colnames(var.plot.data)[5]<-"var.aanum"
        var.plot.data = var.plot.data[var.plot.data$var.aanum >=1 & is.finite(var.plot.data$var.aanum) & !is.na(var.plot.data$var.aanum),]
        
        var.plot<-isolate(var.plot.data)
        var.plot<-var.plot[var.plot$Hugo_Symbol==lolliplot_gene,]
        var.plot<-var.plot[order(var.plot$var.aanum),]
        var.plot$amino.acid.change<-as.character(var.plot$amino.acid.change)
        
        p<-ggplot(var.plot,aes(var.aanum,freq,color=Variant_Classification,label=amino.acid.change)) +
          geom_segment(aes(x=var.aanum,y=0,xend=var.aanum,yend=freq),
                       color=ifelse(var.plot$freq>=input$lolliplot_no*0.5,"orange","grey50"),
                       linewidth=ifelse(var.plot$freq>=input$lolliplot_no*0.5,1.2,0.6)) +
          geom_point(size=5) +
          geom_text_repel(nudge_y=max(var.plot$freq) * 0.05,
                          color=ifelse(var.plot$freq>=input$lolliplot_no*0.5,"orange","NA"),
                          size=ifelse(var.plot$freq>=input$lolliplot_no*0.5,4,2),
                          fontface="bold", max.overlaps=Inf) +
          theme_classic()
        proteins_acc<-read.table(file("source/UniProt.txt",
                                      encoding='UTF-8-BOM'),header=T)
        if(length(proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]) == 1){
          proteins_acc<-proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]
          proteins_acc_url<-gsub(" ","%2C",proteins_acc)
          baseurl<-"https://www.ebi.ac.uk/proteins/api/features?offset=0&size=100&accession="
          url<-paste0(baseurl,proteins_acc_url)
           flag_lolliplot = 0
          if(file.exists(paste0("source/lolliplot/", lolliplot_gene, ".rda"))){
            load(paste0("source/lolliplot/", lolliplot_gene, ".rda"))
            flag_lolliplot = 1
          } else {
            prots_feat<-try(GET(url,accept_json()), silent = FALSE)
            if (class(prots_feat) != "try-error"){
              flag_lolliplot = 1
              save(prots_feat, file=paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))
            }
          }
          if (flag_lolliplot == 1){
            prots_feat_red<-(httr::content(prots_feat))
            features_total_plot<-NULL
            for(i in 1:length(prots_feat_red)){ 
              features_temp<-drawProteins::extract_feat_acc(prots_feat_red[[i]])#the extract_feat_acc() function takes features into a data.frame
              features_temp$order<-i # this order is needed for plotting later
              features_total_plot<-rbind(features_total_plot,features_temp)
            }
            plot_start<-0#starts 0
            plot_end<-max(features_total_plot$end)
            if("CHAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="CHAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.1*max(var.plot$freq),ymax=-0.04*max(var.plot$freq)),
                          colour="grey",
                          fill="grey",
                          inherit.aes=F)
            }
            if("DNA_BIND"%in%(unique(features_total_plot$type))) {
              features_total_plot[features_total_plot$type=="DNA_BIND","description"] <- features_total_plot[features_total_plot$type=="DNA_BIND","type"]
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DNA_BIND",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("DOMAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DOMAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("MOTIF"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="MOTIF",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("REPEAT"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="REPEAT",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            p <- p +
              scale_x_continuous(breaks=round(seq(plot_start,plot_end,by=50)),name="Amino acid number")
          }
        }
        p <- p +
          theme(plot.title=element_text(face="bold",size=(15),hjust=0),axis.text.x=element_text(angle=90,hjust=1)) +
          labs(y="Mutation frequency in AACR project GENIE dataset",
               title=paste(nrow(data_lolliplot), " variants in ",lolliplot_gene,", GENIE hematologic cohort",sep="")) +
          scale_y_continuous(breaks=seq(0,max(var.plot$freq*1.05),max(floor(max(var.plot$freq)/100)*10,5)),name="Frequency")
        plots[[124]] = p
      }
      print(10)
      Data_tmp = Data_MAF_target %>%
        filter(non_germline_non_error <= 1 &
                 Panel == "FoundationOne CDx")
      Data_tmp$Position = as.numeric(unlist(str_extract(Data_tmp$amino.acid.change, "\\d+")))
      Data_tmp = Data_tmp %>% left_join(table_DNMT3A, by = "Position")
      Data_tmp$Destabilizing[is.na(Data_tmp$Destabilizing)] = "Unknown"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "c")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "splice")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "\\*")] = "Truncating"
      Data_tmp$Destabilizing = factor(Data_tmp$Destabilizing)
      lolliplot_gene = input$lolliplot_gene
      if(is.null(lolliplot_gene)){
        lolliplot_gene = "DNMT3A"
      } else if(lolliplot_gene == ""){
        lolliplot_gene = "DNMT3A"
      }
      data_lolliplot = Data_tmp %>%
        dplyr::select(Chromosome,
                      Start_Position,
                      End_Position,
                      Reference_Allele,
                      Tumor_Seq_Allele2,
                      Tumor_Sample_Barcode,
                      Hugo_Symbol,
                      amino.acid.change)
      if(lolliplot_gene == "DNMT3A"){
        data_lolliplot$Variant_Classification = Data_tmp$Destabilizing
      } else{
        data_lolliplot$Variant_Classification = "Missense_Mutation"
      }
      data_lolliplot = data_lolliplot %>%
        dplyr::filter(!str_detect(amino.acid.change, "c.")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "_")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "^\\*")) %>%
        dplyr::filter(Hugo_Symbol == lolliplot_gene) 
      if(length(data_lolliplot$Chromosome) == 0){
        plots[[125]] = gg_empty()
      } else {
        var<-unique(data_lolliplot[!duplicated(data_lolliplot),])#remove duplicates
        var[order(var$Hugo_Symbol,gsub("([A-Z]+)([0-9]+)","\\2",var$amino.acid.change)),]#order
        var.freq<-plyr::count(var,c("Hugo_Symbol","amino.acid.change","Variant_Classification"))
        var.aanum<-unlist(lapply(regmatches(var.freq$amino.acid.change,gregexpr(pattern="*(\\d+)",var.freq$amino.acid.change)),function(x) x[[1]]))
        var.plot.data<-cbind(var.freq,as.numeric(as.character(var.aanum)))#hard way to remove factor
        colnames(var.plot.data)[5]<-"var.aanum"
        var.plot.data = var.plot.data[var.plot.data$var.aanum >=1 & is.finite(var.plot.data$var.aanum) & !is.na(var.plot.data$var.aanum),]
        
        var.plot<-isolate(var.plot.data)
        var.plot<-var.plot[var.plot$Hugo_Symbol==lolliplot_gene,]
        var.plot<-var.plot[order(var.plot$var.aanum),]
        var.plot$amino.acid.change<-as.character(var.plot$amino.acid.change)
        
        p<-ggplot(var.plot,aes(var.aanum,freq,color=Variant_Classification,label=amino.acid.change)) +
          geom_segment(aes(x=var.aanum,y=0,xend=var.aanum,yend=freq),
                       color=ifelse(var.plot$freq>=input$lolliplot_no*0.3,"orange","grey50"),
                       linewidth=ifelse(var.plot$freq>=input$lolliplot_no*0.3,1.2,0.6)) +
          geom_point(size=5) +
          geom_text_repel(nudge_y=max(var.plot$freq) * 0.05,
                          color=ifelse(var.plot$freq>=input$lolliplot_no*0.3,"orange","NA"),
                          size=ifelse(var.plot$freq>=input$lolliplot_no*0.3,4,2),
                          fontface="bold", max.overlaps=Inf) +
          theme_classic()
        proteins_acc<-read.table(file("source/UniProt.txt",
                                      encoding='UTF-8-BOM'),header=T)
        if(length(proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]) == 1){
          proteins_acc<-proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]
          proteins_acc_url<-gsub(" ","%2C",proteins_acc)
          baseurl<-"https://www.ebi.ac.uk/proteins/api/features?offset=0&size=100&accession="
          url<-paste0(baseurl,proteins_acc_url)
          flag_lolliplot = 0
          if(file.exists(paste0("source/lolliplot/", lolliplot_gene, ".rda"))){
            load(paste0("source/lolliplot/", lolliplot_gene, ".rda"))
            flag_lolliplot = 1
          } else {
            prots_feat<-try(GET(url,accept_json()), silent = FALSE)
            if (class(prots_feat) != "try-error"){
              flag_lolliplot = 1
              save(prots_feat, file=paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))
            }
          }
          if (flag_lolliplot == 1){
            prots_feat_red<-(httr::content(prots_feat))
            features_total_plot<-NULL
            for(i in 1:length(prots_feat_red)){ 
              features_temp<-drawProteins::extract_feat_acc(prots_feat_red[[i]])#the extract_feat_acc() function takes features into a data.frame
              features_temp$order<-i # this order is needed for plotting later
              features_total_plot<-rbind(features_total_plot,features_temp)
            }
            plot_start<-0#starts 0
            plot_end<-max(features_total_plot$end)
            if("CHAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="CHAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.1*max(var.plot$freq),ymax=-0.04*max(var.plot$freq)),
                          colour="grey",
                          fill="grey",
                          inherit.aes=F)
            }
            if("DNA_BIND"%in%(unique(features_total_plot$type))) {
              features_total_plot[features_total_plot$type=="DNA_BIND","description"] <- features_total_plot[features_total_plot$type=="DNA_BIND","type"]
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DNA_BIND",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("DOMAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DOMAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("MOTIF"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="MOTIF",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("REPEAT"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="REPEAT",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            p <- p +
              scale_x_continuous(breaks=round(seq(plot_start,plot_end,by=50)),name="Amino acid number")
          }
        }
        p <- p +
          theme(plot.title=element_text(face="bold",size=(15),hjust=0),axis.text.x=element_text(angle=90,hjust=1)) +
          labs(y="Mutation frequency in C-CAT dataset",
               title=paste(nrow(data_lolliplot), " variants in ",lolliplot_gene,", C-CAT F1 CDx",sep="")) +
          scale_y_continuous(breaks=seq(0,max(var.plot$freq*1.05),max(floor(max(var.plot$freq)/100)*10,5)),name="Frequency")
        plots[[125]] = p
      }
      print(10)
      Data_tmp = Data_GENIE_solid %>% dplyr::filter(Tumor_Sample_Barcode %in% ID_GENIE_F1L)
      Data_tmp$Position = as.numeric(unlist(str_extract(Data_tmp$amino.acid.change, "\\d+")))
      Data_tmp = Data_tmp %>% left_join(table_DNMT3A, by = "Position")
      Data_tmp$Destabilizing[is.na(Data_tmp$Destabilizing)] = "Unknown"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "c")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "splice")] = "Splice"
      Data_tmp$Destabilizing[str_detect(Data_tmp$name, "\\*")] = "Truncating"
      Data_tmp$Destabilizing = factor(Data_tmp$Destabilizing)
      lolliplot_gene = input$lolliplot_gene
      if(is.null(lolliplot_gene)){
        lolliplot_gene = "DNMT3A"
      } else if(lolliplot_gene == ""){
        lolliplot_gene = "DNMT3A"
      }
      data_lolliplot = Data_tmp %>%
        dplyr::select(Chromosome,
                      Start_Position,
                      End_Position,
                      Reference_Allele,
                      Tumor_Seq_Allele2,
                      Tumor_Sample_Barcode,
                      Hugo_Symbol,
                      amino.acid.change)
      if(lolliplot_gene == "DNMT3A"){
        data_lolliplot$Variant_Classification = Data_tmp$Destabilizing
      } else{
        data_lolliplot$Variant_Classification = "Missense_Mutation"
      }
      data_lolliplot = data_lolliplot %>%
        dplyr::filter(!str_detect(amino.acid.change, "c.")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "_")) %>%
        dplyr::filter(!str_detect(amino.acid.change, "^\\*")) %>%
        dplyr::filter(Hugo_Symbol == lolliplot_gene) 
      setProgress(detail = "24651")
      if(length(data_lolliplot$Chromosome) == 0){
        plots[[126]] = gg_empty()
      } else {
        var<-unique(data_lolliplot[!duplicated(data_lolliplot),])#remove duplicates
        var[order(var$Hugo_Symbol,gsub("([A-Z]+)([0-9]+)","\\2",var$amino.acid.change)),]#order
        var.freq<-plyr::count(var,c("Hugo_Symbol","amino.acid.change","Variant_Classification"))
        var.aanum<-unlist(lapply(regmatches(var.freq$amino.acid.change,gregexpr(pattern="*(\\d+)",var.freq$amino.acid.change)),function(x) x[[1]]))
        var.plot.data<-cbind(var.freq,as.numeric(as.character(var.aanum)))#hard way to remove factor
        colnames(var.plot.data)[5]<-"var.aanum"
        var.plot.data = var.plot.data[var.plot.data$var.aanum >=1 & is.finite(var.plot.data$var.aanum) & !is.na(var.plot.data$var.aanum),]
        
        var.plot<-isolate(var.plot.data)
        var.plot<-var.plot[var.plot$Hugo_Symbol==lolliplot_gene,]
        var.plot<-var.plot[order(var.plot$var.aanum),]
        var.plot$amino.acid.change<-as.character(var.plot$amino.acid.change)
        
        p<-ggplot(var.plot,aes(var.aanum,freq,color=Variant_Classification,label=amino.acid.change)) +
          geom_segment(aes(x=var.aanum,y=0,xend=var.aanum,yend=freq),
                       color=ifelse(var.plot$freq>=input$lolliplot_no*0.06,"orange","grey50"),
                       linewidth=ifelse(var.plot$freq>=input$lolliplot_no*0.06,1.2,0.6)) +
          geom_point(size=5) +
          geom_text_repel(nudge_y=max(var.plot$freq) * 0.05,
                          color=ifelse(var.plot$freq>=input$lolliplot_no*0.06,"orange","NA"),
                          size=ifelse(var.plot$freq>=input$lolliplot_no*0.06,4,2),
                          fontface="bold", max.overlaps=Inf) +
          theme_classic()
        proteins_acc<-read.table(file("source/UniProt.txt",
                                      encoding='UTF-8-BOM'),header=T)
        if(length(proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]) == 1){
          proteins_acc<-proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene,2]
          proteins_acc_url<-gsub(" ","%2C",proteins_acc)
          baseurl<-"https://www.ebi.ac.uk/proteins/api/features?offset=0&size=100&accession="
          url<-paste0(baseurl,proteins_acc_url)
          flag_lolliplot = 0
          if(file.exists(paste0("source/lolliplot/", lolliplot_gene, ".rda"))){
            load(paste0("source/lolliplot/", lolliplot_gene, ".rda"))
            flag_lolliplot = 1
          } else {
            prots_feat<-try(GET(url,accept_json()), silent = FALSE)
            if (class(prots_feat) != "try-error"){
              flag_lolliplot = 1
              save(prots_feat, file=paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))
            }
          }
          if (flag_lolliplot == 1){
            prots_feat_red<-(httr::content(prots_feat))
            features_total_plot<-NULL
            for(i in 1:length(prots_feat_red)){ 
              features_temp<-drawProteins::extract_feat_acc(prots_feat_red[[i]])#the extract_feat_acc() function takes features into a data.frame
              features_temp$order<-i # this order is needed for plotting later
              features_total_plot<-rbind(features_total_plot,features_temp)
            }
            plot_start<-0#starts 0
            plot_end<-max(features_total_plot$end)
            if("CHAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="CHAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.1*max(var.plot$freq),ymax=-0.04*max(var.plot$freq)),
                          colour="grey",
                          fill="grey",
                          inherit.aes=F)
            }
            if("DNA_BIND"%in%(unique(features_total_plot$type))) {
              features_total_plot[features_total_plot$type=="DNA_BIND","description"] <- features_total_plot[features_total_plot$type=="DNA_BIND","type"]
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DNA_BIND",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("DOMAIN"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="DOMAIN",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("MOTIF"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="MOTIF",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            if("REPEAT"%in%(unique(features_total_plot$type))){
              p <- p +
                geom_rect(data=features_total_plot[features_total_plot$type=="REPEAT",],
                          mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                          inherit.aes=F)
            }
            p <- p +
              scale_x_continuous(breaks=round(seq(plot_start,plot_end,by=50)),name="Amino acid number")
          }
        }
        p <- p +
          theme(plot.title=element_text(face="bold",size=(15),hjust=0),axis.text.x=element_text(angle=90,hjust=1)) +
          labs(y="Mutation frequency in GENIE PROV dataset",
               title=paste(nrow(data_lolliplot), " variants in ",lolliplot_gene,", GENIE F1-Liquid CDx",sep="")) +
         scale_y_continuous(breaks=seq(0,max(var.plot$freq*1.05),max(floor(max(var.plot$freq)/100)*10,5)),name="Frequency")
        plots[[126]] = p
      }
      print(10)
      # 1. 全症例リスト（年齢情報含む）
      all_cases <- Data_summary_F1L %>%
        select(Tumor_Sample_Barcode, Age)
      # all_cases$Age = round(all_cases$Age/5)*5
      mutation_info <- Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          TRUE ~ "Others"
        )) %>% 
        filter(color_group != "Others") %>%
        select(Tumor_Sample_Barcode, color_group) %>%
        distinct()
      all_cases = all_cases %>% left_join(mutation_info, by="Tumor_Sample_Barcode")
      all_cases$color_group[is.na(all_cases$color_group)] = "Others"
      all_cases = all_cases %>%
        filter(color_group != "Others")
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      ID_truncation = unique((Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(DNMT3A_comut = case_when(
          str_detect(name, "\\*") ~ "Truncation",
          str_detect(name, "c") ~ "Truncation",
          str_detect(name, "fs") ~ "Truncation",
          TRUE ~ "Other"
        )) %>%
        filter(DNMT3A_comut == "Truncation"))$Tumor_Sample_Barcode)
      mutation_per_case <- Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A" & Tumor_Sample_Barcode %in% ID_truncation) %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          TRUE ~ "Others"
        )) %>% 
        filter(color_group != "Others") %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = n(), .groups = "drop")
      gene_freq = c("Y735C", "R882H", "R882C", "R635W", "I780T")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = gene_freq)
      all_mutations <- expand_grid(all_cases %>% select(-color_group), color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      all_mutations$mutation_count = fun_zero(all_mutations$mutation_count, all_mutations$mutation_count)
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases %>%
        group_by(Age, color_group) %>%
        summarise(total_patients = n(), .groups = "drop")
      Data_DNMT3A_count = 
        Data_MAF_DNMT3A %>%
        dplyr::filter(Hugo_Symbol == "DNMT3A") %>%
        dplyr::mutate(color_group = case_when(
          name == "DNMT3A_Y735C" ~ "Y735C",
          str_detect(name, "R882H") ~ "R882H",
          str_detect(name, "R882C") ~ "R882C",
          str_detect(name, "R635W") ~ "R635W",
          str_detect(name, "I780T") ~ "I780T",
          TRUE ~ "Others"
        ))%>% 
        filter(color_group != "Others")
      # table_count = data.frame(gene_freq)
      # colnames(table_count) = "color_group"
      # table_count$count = 0
      # for(i in 1:length(table_count$count)){
      #   table_count$count[i] = length(unique((Data_DNMT3A_count %>% filter(color_group == gene_freq[i]))$unique_position))
      # }
      # 5. 年齢ごと & `color_group` ごとに変異を持つ症例の割合を計算
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = c("Age", "color_group")) %>% 
        # left_join(table_count, by = "color_group") %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = total_patients,  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        )
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:100)
      # age_range <- tibble(Age = 0:20*5)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      final_data$color_group <- factor(final_data$color_group, levels = gene_freq)
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", alpha = 0.1, span = 1, size = 1.2) +  # loess平滑化
        labs(title = "C-CAT, truncation DNMT3A co-variant Presence Ratio by Age",
             x = "Age",
             y = "Truncation DNMT3A co-variant Presence Ratio (with at least one mutation)") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())
      plots[[127]] = g
      print(10)
      mutation_data = Data_MAF_target %>%
        filter(!is.na(ICI_Effect)) %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
          Hugo_Symbol %in% CH_gene_entropy_3 &
          Panel == "F1Liquid CDx" &
            pnt_col %in% c("A"))
      mutation_data_DNMT3A <- mutation_data %>%
        dplyr::mutate(color_group = case_when(
          ICI_Effect == 1 ~ "Y735C",
          TRUE ~ "I780T"
        ))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% (mutation_data_DNMT3A %>% filter(color_group == "Y735C"))$Tumor_Sample_Barcode ~ 1,
          TRUE ~ 0
        )) %>%
        filter(DNMT3A_735_882 != 2)
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Age, Cancername, TMB, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      Data_summary_tmp = Data_summary_tmp %>%
        mutate(TMB_H = case_when(
          TMB > 10 ~ "High",
          TRUE ~ "Low"
        )) 
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$unique_alt))
      colnames(data_tmp) = "unique_alt"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      if(file.exists("source/ICI_effect_variant.rda")){
        load("source/ICI_effect_variant.rda")
      } else{
        for(i in 1:length(data_tmp$odds_ratio)){
          mut_ID = (mutation_data %>% dplyr::filter(unique_alt == data_tmp$unique_alt[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
          Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age + TMB_H, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$unique_alt)))
          }
        }
        save(data_tmp, file="source/ICI_effect_variant.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = unique_alt,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor ICI objective responce",
             subtitle = "Only liquid-favor variants, right: ICI effective variant",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[128]] = g
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      if(file.exists("source/ICI_effect_gene.rda")){
        load("source/ICI_effect_gene.rda")
      } else{
        for(i in 1:length(data_tmp$odds_ratio)){
          mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
          Data_summary_tmp = Data_summary_tmp %>%
            dplyr::mutate(mutation = case_when(
              Tumor_Sample_Barcode %in% mut_ID ~ 1,
              TRUE ~ 0))
          data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
          data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
          data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
          Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age + TMB_H, data=Data_summary_tmp, family=binomial())
          data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
          if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
            fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                  data_tmp$WT_735[i],
                                  data_tmp$mut_882[i],
                                  data_tmp$WT_882[i]),
                                ncol = 2
            )
            data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
          } else{
            data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
          }
          if(i %% 10 == 0){
            print(paste(i, ",", length(data_tmp$Hugo_Symbol)))
          }
        }
        save(data_tmp, file="source/ICI_effect_gene.rda")
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "Volcano plot for favor ICI objective responce",
             subtitle = "Only liquid-favor variants, right:  ICI effective gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[129]] = g
      print(10)
      # 1. 全症例リスト（年齢情報含む）
      all_cases <- Data_MAF_DNMT3A %>%
        select(Tumor_Sample_Barcode, Age)
      all_cases_color_group <- Data_MAF_DNMT3A %>%
        select(Tumor_Sample_Barcode, Age, color_group)
      all_cases$Age = round(all_cases$Age/5)*5
      all_cases_color_group$Age = round(all_cases$Age/5)*5
      # 2. 変異データを症例ごとに集計し、`color_group` ごとの変異数を計算
      mutation_per_case <- Data_MAF_DNMT3A %>%
        group_by(Tumor_Sample_Barcode, color_group) %>%
        summarise(mutation_count = VAF, .groups = "drop")
      # 3. すべての症例 × すべての color_group の組み合わせを作成（変異なしを考慮）
      color_groups <- tibble(color_group = c("Epigenetic", "DNA_repair", "Splicing", "MAPK", "Other_gene"))
      all_mutations <- expand_grid(all_cases, color_groups) %>%
        left_join(mutation_per_case, by = c("Tumor_Sample_Barcode", "color_group")) %>%
        replace_na(list(mutation_count = 0))  # 変異なしの症例に mutation_count = 0 を設定
      # 4. 各年齢の総症例数を計算
      total_cases <- all_cases_color_group %>%
        group_by(Age, color_group) %>%
        summarise(total_patients = n(), .groups = "drop")
      # 5. 年齢ごと & `color_group` ごとに変異を持つ症例の割合を計算
      mutation_stats <- all_mutations %>%
        group_by(Age, color_group) %>%
        left_join(total_cases, by = c("Age", "color_group")) %>% 
        summarise(
          mean_mutation = sum(mutation_count) / total_patients,  # その年齢での color_group の変異総数
          sample_size = total_patients,  # その年齢での総症例数
          sd_mutation = ifelse(sample_size > 1, sd(mutation_count), 0),
          .groups = "drop"
        )
      # 6. 0歳から100歳までのすべての年齢と `color_group` の組み合わせを補完
      age_range <- tibble(Age = 0:20*5)
      # すべての組み合わせを作成し、データを統合
      complete_data <- expand_grid(age_range, color_groups) %>%
        left_join(mutation_stats, by = c("Age", "color_group")) %>%
        replace_na(list(mutation_count = 0, mean_mutation = 0, sd_mutation = 0, sample_size = 0))
      # 7. 信頼区間の計算（全症例に基づく）
      final_data <- complete_data %>%
        mutate(
          lower = mean_mutation - 1.96 * sd_mutation / sqrt(sample_size),  # 信頼区間の下限
          upper = mean_mutation + 1.96 * sd_mutation / sqrt(sample_size)   # 信頼区間の上限
        ) %>%
        filter(sample_size != 0) %>% distinct()
      # 8. プロット
      g <- ggplot(final_data, aes(x = Age, y = mean_mutation, color = color_group)) +
        geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.2, alpha = 0.3) +  # 信頼区間
        geom_point(alpha = 0.3) +  # 生データを薄くプロット
        geom_smooth(method = "loess", span = 1, size = 1.2, alpha = 0.1) +  # loess平滑化
        labs(title = "C-CAT, Smoothed LD-CH/LE-CH Variant Allele Frequency by Age",
             x = "Age",
             y = "Mean variant allele frequency") +
        theme_minimal() +
        scale_y_continuous(labels = scales::percent_format())      
      plots[[130]] = g
      print(10)
      
      msigPlot(maf.counts_DNMT3A_2,
             sample = "735 and 882",
             main = paste0("C-CAT patients with DNMT3A Y735C mut and R882 mut, ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                           " non-Y735C/882-DNMT3A and ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                           " other hotspot muts in ",
                           length(ID_DNMT3A_735_and_882_2),
                           " pts (mean, ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_735_and_882_2)),
                                  digits=2),
                           " and ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735 and 882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_735_and_882_2)),
                                  digits=2),
                           " muts/pt)"),
             ylim = c(0, 0.30))
      plot_obj <- recordPlot()
      plots[[137]] = ggdraw(plot_obj)
      msigPlot(maf.counts_DNMT3A_2,
             sample = "735",
             main = paste0("C-CAT patients with DNMT3A Y735C mut, ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                           " non-Y735C/882-DNMT3A and ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                           " other hotspot muts in ",
                           length(ID_DNMT3A_735_2),
                           " pts (mean, ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_735_2)),
                                  digits=2),
                           " and ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_735_2)),
                                  digits=2),
                           " muts/pt)"),
             ylim = c(0, 0.20))
      plot_obj <- recordPlot()
      plots[[138]] = ggdraw(plot_obj)
      msigPlot(maf.counts_DNMT3A_2,
             sample = "882",
             main = paste0("C-CAT patients with DNMT3A R882 mut, ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                           " non-Y735C/882-DNMT3A and ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                           " other hotspot muts in ",
                           length(ID_DNMT3A_882_2),
                           " pts (mean, ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_882_2)),
                                  digits=2),
                           " and ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "882" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_882_2)),
                                  digits=2),
                           " muts/pt)"),
             ylim = c(0, 0.20))
      plot_obj <- recordPlot()
      plots[[139]] = ggdraw(plot_obj)
      msigPlot(maf.counts_DNMT3A_2,
             sample = "other mut",
             main = paste0("C-CAT patients with other DNMT3A hotspot mutations, ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                           " non-Y735C/882-DNMT3A and ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                           " other hotspot muts in ",
                           length(ID_DNMT3A_other_2),
                           " pts (mean ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_other_2)),
                                  digits=2),
                           " and ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "other mut" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_other_2)),
                                  digits=2),
                           " muts/pt)"),
             ylim = c(0, 0.20))
      plot_obj <- recordPlot()
      plots[[140]] = ggdraw(plot_obj)
      msigPlot(maf.counts_DNMT3A_2,
             sample = "none",
             main = paste0("C-CAT patients with no DNMT3A hotspot mutations, ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                           " non-Y735C/882-DNMT3A and ",
                           length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                           " other hotspot muts in ",
                           length(ID_DNMT3A_none_2),
                           " pts (mean, ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_none_2)),
                                  digits=2),
                           " and ",
                           format(fun_zero(length((Data_maf_DNMT3A_2 %>% dplyr::filter(DNMT3A_status == "none" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                           length(ID_DNMT3A_none_2)),
                                  digits=2),
                           " muts/pt)"),
             ylim = c(0, 0.25))
      plot_obj <- recordPlot()
      plots[[141]] = ggdraw(plot_obj)
      msigPlot(maf.counts_DNMT3A_3,
               sample = "735",
               main = paste0("C-CAT patients with DNMT3A Y735C mut, ",
                             length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                             " non-Y735C-DNMT3A and ",
                             length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                             " other hotspot muts in ",
                             length(ID_DNMT3A_735_3),
                             " pts (mean, ",
                             format(fun_zero(length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                             length(ID_DNMT3A_735_3)),
                                    digits=2),
                             " and ",
                             format(fun_zero(length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "735" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                             length(ID_DNMT3A_735_3)),
                                    digits=2),
                             " muts/pt)"),
               ylim = c(0, 0.20))
      plot_obj <- recordPlot()
      plots[[142]] = ggdraw(plot_obj)
      setProgress(detail = "25235")
      # 集計値
      other_total <- 8200
      other_actc  <- 897
      y735c_total <- 152
      y735c_actc  <- 28
      rate_other <- other_actc / other_total
      rate_y735c <- y735c_actc  / y735c_total
      RR <- rate_y735c / rate_other
      rate_other
      rate_y735c
      RR
      # exactな率比の検定（2群の発生率比較）
      pt <- poisson.test(
        x = c(y735c_actc, other_actc),
        T = c(y735c_total, other_total),
        r = 1
      )
      pt
      msigPlot(maf.counts_DNMT3A_3,
               sample = "other",
               main = paste0("C-CAT patients without DNMT3A Y735C mut, ",
                             length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "other" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                             " non-Y735C-DNMT3A and ",
                             length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "other" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                             " other hotspot muts in ",
                             length(ID_DNMT3A_none_3),
                             " pts (mean, ",
                             format(fun_zero(length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "other" & Hugo_Symbol == "DNMT3A"))$Hugo_Symbol),
                                             length(ID_DNMT3A_none_3)),
                                    digits=2),
                             " and ",
                             format(fun_zero(length((Data_maf_DNMT3A_3 %>% dplyr::filter(DNMT3A_status == "other" & Hugo_Symbol != "DNMT3A"))$Hugo_Symbol),
                                             length(ID_DNMT3A_none_3)),
                                    digits=2),
                             " muts/pt)"),
               ylim = c(0, 0.20))
      plot_obj <- recordPlot()
      plots[[143]] = ggdraw(plot_obj)
      msigPlot(maf.counts, sample = "A", main = "Mutational signature of LD-CH", ylim = c(0, 0.35))
      plot_obj <- recordPlot()
      plots[[144]] = ggdraw(plot_obj)
      msigPlot(maf.counts, sample = "B", main = "Mutational signature of LE-CH", ylim = c(0, 0.35))
      plot_obj <- recordPlot()
      plots[[145]] = ggdraw(plot_obj)
      msigPlot(maf.counts_DNMT3A_GENIE_all, sample = "all", main = "GENIE hematologic cohort, variants in CH-related genes", ylim = c(0, 0.20))
      plot_obj <- recordPlot()
      plots[[146]] = ggdraw(plot_obj)
      
      if(!is.null(significant_factors_nomo)){
        plot(cal, lwd=2, lty=1,
             cex.lab=1.2, cex.axis=1, cex.main=1.2, cex.sub=0.6,
             xlim=c(0, 1), ylim= c(0, 1),
             xlab="Nomogram-Predicted Probability of LD-CH/LE-CH",
             ylab="Actual LD-CH/LE-CH rate",
             legend=FALSE)
        text(x=.4, y=.0, adj = c(0,0), paste0("LD-CH/LE-CH prediction\n",
                                              "Model lokelihood ratio test P-value: ", format(m2$stats["P"][[1]], nsmall=3),'\n',
                                              "500-time-bootstrapped-Brier score: ", round(val[9,5], digits=3),'\n',
                                              "500-time-bootstrapped-Nagelkerke R2 = ", round(val[2,5], digits=3),'\n',
                                              "500-time-bootstrapped-concordance index = ", round(((val[1,5] + 1) / 2), digits=3),"\n",
                                              "Intercept = ",round(val[3,5], digits=3),"\n",
                                              "Slope = ", round(val[4,5], digits=3)))
        abline(0, 1, lty=3, lwd=2,  col=c("#224444"))
        legend(x=.6, y=.4, legend=c("Apparent", "Bias-corrected", "Ideal"),
               lty=c(3, 1, 2), bty="n")
        plot_obj <- recordPlot()
        plots[[153]] = ggdraw(plot_obj)
        
        plot(log_nomogram)
        text(x=.42, y=0.07, adj = c(0,0), paste0("Nomogram-predicted CH-related hotspot variant rate"))
        plot_obj <- recordPlot()
        plots[[154]] = ggdraw(plot_obj)
        
        auc_val <- round(logistic_data$auc_value, 3)
        ggroc_obj = ggroc(logistic_data$roc_obj)
        plots[[155]] = ggroc_obj +
          geom_abline(intercept = 1, slope = 1, linetype = "dashed", color = "grey") +
          ggtitle("ROC Curve, Nomogram") +
          theme_minimal() +
          annotate("text", x = 0.75, y = 0.45, label = paste("AUC =", auc_val), size = 5)
      }
    
      IDH2_ID = unique((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & Hugo_Symbol == "IDH2" & str_detect(unique_alt, "R172|R140")))$Tumor_Sample_Barcode)
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% IDH2_ID ~ 1,
          TRUE ~ 0
        )) %>%
        filter(!str_detect(name, "IDH2_R172|IDH2_R140"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "C-CAT, volcano plot for favor with or without IDH2 R140/R172",
             subtitle = "Only liquid-favor variants, right: IDH2 R140/R172-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[156]] = g
      
      JAK2_ID = unique((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & Hugo_Symbol == "JAK2" & str_detect(unique_alt, "V617F")))$Tumor_Sample_Barcode)
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% JAK2_ID ~ 1,
          TRUE ~ 0
        )) %>%
        filter(!str_detect(name, "JAK2_V617F"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "C-CAT, volcano plot for favor with or without JAK2 V617F",
             subtitle = "Only liquid-favor variants, right: JAK2 V617F-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[157]] = g
      TET2_ID = unique((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & unique_alt %in% CH_candidates & Hugo_Symbol == "TET2"))$Tumor_Sample_Barcode)
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% TET2_ID ~ 1,
          TRUE ~ 0
        )) %>%
        filter(!str_detect(name, "TET2_"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "C-CAT, volcano plot for favor with or without TET2 hotspot variants",
             subtitle = "Only liquid-favor variants, right: TET2-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[158]] = g
      CBL_ID = unique((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & unique_alt %in% CH_candidates & Hugo_Symbol == "CBL"))$Tumor_Sample_Barcode)
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0
        )) %>%
        filter(!str_detect(name, "CBL_"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "C-CAT, volcano plot for favor with or without CBL hotspot variants",
             subtitle = "Only liquid-favor variants, right: CBL-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[159]] = g
      
      CBL_ID = unique((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & unique_alt %in% CH_candidates & Hugo_Symbol == "DNMT3A"))$Tumor_Sample_Barcode)
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0
        )) %>%
        filter(!str_detect(name, "DNMT3A_"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "C-CAT, volcano plot for favor with or without DNMT3A hotspot variants",
             subtitle = "Only liquid-favor variants, right: DNMT3A-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[160]] = g
      
      CBL_ID = unique((Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & unique_alt %in% CH_candidates & Hugo_Symbol == "KRAS"))$Tumor_Sample_Barcode)
      mutation_data = Data_MAF_target %>%
        dplyr::filter(non_germline_non_error <= 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "F1Liquid CDx" &
                        pnt_col %in% c("A"))
      mutation_data = mutation_data %>%
        dplyr::mutate(DNMT3A_735_882 = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0
        )) %>%
        filter(!str_detect(name, "KRAS_V14I"))
      Data_summary_tmp = mutation_data %>%
        select(Tumor_Sample_Barcode, Hugo_Symbol, Age, Cancername, DNMT3A_735_882) %>%
        distinct(Tumor_Sample_Barcode, .keep_all = T)
      data_tmp = data.frame(unique((mutation_data %>% filter(pnt_col == "A"))$Hugo_Symbol))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & DNMT3A_735_882 == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & DNMT3A_735_882 == 0))$mutation)
        Data_ORR = speedglm(DNMT3A_735_882 ~ mutation + Cancername + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = data_tmp$p_value
      #data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-4] = 10^-4
      threshold_OR = 0.1
      threshold_P = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        labs(title = "C-CAT, volcano plot for favor with or without KRAS V14I",
             subtitle = "Only liquid-favor variants, right: KRAS V14I-favor gene",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 45) #+
      plots[[161]] = g
      setProgress(detail = "25879")
      mutation_data = 
        data.frame(Data_GENIE %>%
                      filter(Sample_type == "Blood_cancer" &
                               CH_hotspot_candidate == 1 &
                               Hugo_Symbol %in% CH_gene_entropy_3))
      CBL_ID = unique((mutation_data %>% dplyr::filter(str_detect(unique_alt, "IDH2_R172|IDH2_R140")))$Tumor_Sample_Barcode)
      mutation_data = mutation_data %>%
        filter(!str_detect(name, "IDH2_R172|IDH2_R140"))
      data_tmp = data.frame(sort(unique(mutation_data$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0))
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
        Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-10] = 10^-10
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor with or without IDH2 R140/R172 variants",
             subtitle = "Genes with hotspot variants, right:IDH2 R140/R172-favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[162]] = g
      mutation_data = 
        data.frame(Data_GENIE %>%
                     filter(Sample_type == "Blood_cancer" &
                              CH_hotspot_candidate == 1 &
                              Hugo_Symbol %in% CH_gene_entropy_3))
      CBL_ID = unique((mutation_data %>% dplyr::filter(str_detect(unique_alt, "JAK2_V617F")))$Tumor_Sample_Barcode)
      mutation_data = mutation_data %>%
        filter(!str_detect(name, "JAK2_V617F"))
      data_tmp = data.frame(sort(unique(mutation_data$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0))
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
        Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-10] = 10^-10
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor with or without JAK2 V617F variants",
             subtitle = "Genes with hotspot variants, right:JAK2 V617F-favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[163]] = g
      mutation_data = 
        data.frame(Data_GENIE %>%
                     filter(Sample_type == "Blood_cancer" &
                              CH_hotspot_candidate == 1 &
                              Hugo_Symbol %in% CH_gene_entropy_3))
      CBL_ID = unique((mutation_data %>% dplyr::filter(str_detect(unique_alt, "CBL_")))$Tumor_Sample_Barcode)
      mutation_data = mutation_data %>%
        filter(!str_detect(name, "CBL_"))
      data_tmp = data.frame(sort(unique(mutation_data$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0))
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
        Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-10] = 10^-10
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor with or without CBL variants",
             subtitle = "Genes with hotspot variants, right:CBL-favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[164]] = g
      mutation_data = 
        data.frame(Data_GENIE %>%
                     filter(Sample_type == "Blood_cancer" &
                              CH_hotspot_candidate == 1 &
                              Hugo_Symbol %in% CH_gene_entropy_3))
      CBL_ID = unique((mutation_data %>% dplyr::filter(str_detect(unique_alt, "TET2_")))$Tumor_Sample_Barcode)
      mutation_data = mutation_data %>%
        filter(!str_detect(name, "TET2_"))
      data_tmp = data.frame(sort(unique(mutation_data$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0))
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
        Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-10] = 10^-10
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      setProgress(detail = "26259")
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor with or without TET2 variants",
             subtitle = "Genes with hotspot variants, right:TET2-favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[165]] = g
      mutation_data = 
        data.frame(Data_GENIE %>%
                     filter(Sample_type == "Blood_cancer" &
                              CH_hotspot_candidate == 1 &
                              Hugo_Symbol %in% CH_gene_entropy_3))
      CBL_ID = unique((mutation_data %>% dplyr::filter(str_detect(unique_alt, "DNMT3A_")))$Tumor_Sample_Barcode)
      mutation_data = mutation_data %>%
        filter(!str_detect(name, "DNMT3A_"))
      data_tmp = data.frame(sort(unique(mutation_data$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0))
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
        Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-10] = 10^-10
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor with or without DNMT3A variants",
             subtitle = "Genes with hotspot variants, right:DNMT3A-favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[166]] = g
      mutation_data = 
        data.frame(Data_GENIE %>%
                     filter(Sample_type == "Blood_cancer" &
                              CH_hotspot_candidate == 1 &
                              Hugo_Symbol %in% CH_gene_entropy_3))
      CBL_ID = unique((mutation_data %>% dplyr::filter(str_detect(unique_alt, "KRAS_V14I")))$Tumor_Sample_Barcode)
      mutation_data = mutation_data %>%
        filter(!str_detect(name, "KRAS_V14I"))
      data_tmp = data.frame(sort(unique(mutation_data$Hugo_Symbol)))
      colnames(data_tmp) = "Hugo_Symbol"
      data_tmp$odds_ratio = 1
      data_tmp$p_value = 1
      data_tmp$mut_735 = 0
      data_tmp$WT_735 = 0
      data_tmp$mut_882 = 0
      data_tmp$WT_882 = 0
      Data_summary_tmp = Summary_GENIE
      Data_summary_tmp = Data_summary_tmp %>%
        dplyr::mutate(favor_liquid = case_when(
          Tumor_Sample_Barcode %in% CBL_ID ~ 1,
          TRUE ~ 0))
      for(i in 1:length(data_tmp$Hugo_Symbol)){
        mut_ID = (mutation_data %>% dplyr::filter(Hugo_Symbol == data_tmp$Hugo_Symbol[i]))$Tumor_Sample_Barcode
        Data_summary_tmp = Data_summary_tmp %>%
          dplyr::mutate(mutation = case_when(
            Tumor_Sample_Barcode %in% mut_ID ~ 1,
            TRUE ~ 0))
        data_tmp$mut_735[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 1))$mutation)
        data_tmp$WT_735[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 1))$mutation)
        data_tmp$mut_882[i] = length((Data_summary_tmp %>% filter(mutation == 1 & favor_liquid == 0))$mutation)
        data_tmp$WT_882[i] = length((Data_summary_tmp %>% filter(mutation == 0 & favor_liquid == 0))$mutation)
        Data_ORR = speedglm(favor_liquid ~ mutation + Age, data=Data_summary_tmp, family=binomial())
        data_tmp$odds_ratio[i] = exp(summary(Data_ORR)$coefficients["mutation",1])
        if(data_tmp$odds_ratio[i] > 1024 | data_tmp$odds_ratio[i] < 1/1024){
          fisher_tmp = matrix(c(data_tmp$mut_735[i],
                                data_tmp$WT_735[i],
                                data_tmp$mut_882[i],
                                data_tmp$WT_882[i]),
                              ncol = 2
          )
          data_tmp$p_value[i] = fisher.test(fisher_tmp,)$p.value
        } else{
          data_tmp$p_value[i] = summary(Data_ORR)$coefficients["mutation",4]
        }
      }
      data_tmp$odds_ratio_adj = data_tmp$odds_ratio
      data_tmp$p_value_adj = p.adjust(data_tmp$p_value, method= "none")
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio > 2^4] = 2^4
      data_tmp$odds_ratio_adj[data_tmp$odds_ratio < 2^-4] = 2^-4
      data_tmp$p_value_adj[data_tmp$p_value_adj < 10^-10] = 10^-10
      threshold_OR = 4
      threshold_OR_TP53 = 4
      threshold_OR_OTHER = 4
      threshold_OR_tumor = 1
      threshold_P = 0.01
      threshold_P_TP53 = 0.01
      threshold_P_OTHER = 0.01
      data_tmp <- data_tmp %>%
        mutate( logpval = -log10(p_value_adj),
                logfold = log2(odds_ratio_adj) ,
                aelabel = Hugo_Symbol,
                pnt_col = ifelse(
                  (logfold > log2(threshold_OR)) &
                    (logpval > -log10(threshold_P)), "A",
                  ifelse(
                    (logfold < -log2(threshold_OR_tumor)# &
                     # (logpval > -log10(threshold_P) & Hugo_Symbol != "TP53" & CH_genes == 1 |
                     #  logpval > -log10(threshold_P_OTHER) & CH_genes == 0 |
                     #  logpval > -log10(threshold_P_TP53) & Hugo_Symbol == "TP53")
                    ), "C",
                    # ifelse(
                    #   logfold >= -log2(threshold_OR_tumor) & logfold < log2(1),"D",
                    # ifelse(
                    #   logfold >= -log2(1) & logfold < log2(2),"C",
                    "B"))#))
        )
      g = ggplot(data_tmp,
                 aes(x=logfold, y=logpval, label = aelabel)) +
        geom_point() +
        # scale_shape_manual(values = c(3,1,4,6,8)) +
        # scale_color_manual(values = c("#000000","#4477FF","#00FF00", "#FF0000", "#8833EE")) +
        labs(title = "GENIE, Volcano plot for favor with or without KRAS V14I variants",
             subtitle = "Genes with hotspot variants, right:KRAS V14I-favor",
             x="log2 odds ratio",
             y="-log10 p-value") +
        theme_minimal() +
        geom_vline(xintercept=0, linetype = "dashed", col="blue") +
        geom_hline(yintercept=-log10(threshold_P), linetype = "dashed", col="red") +
        #geom_hline(yintercept=-log10(threshold_P_TP53), linetype = "dashed", col="blue") +
        # coord_cartesian(xlim = c(-12,12) ,
        #                 ylim = c(0,50)) +
        scale_x_continuous(limits = c(-4.5,4),
                           breaks = c(-4,-2,0,2,4),
                           labels = c("~-4","-2","0","2","4~")) +
        scale_y_continuous(limits = c(0,4),
                           breaks = c(0,2,4),
                           labels = c("0","2","4~")) +
        geom_text_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 15) #+
      #geom_label_repel(aes(label = aelabel), size = 3,vjust = 0.14, max.overlaps = 20) +
      #theme(legend.position = "none")
      plots[[167]] = g
      empty_plot <- patchwork::wrap_elements(grid::grid.rect(gp = grid::gpar(col = NA)))
      
      design <- "
      AAACCC
      BBDDEE
      "
      # レイアウトに図を配置
      layout <- plots[[1]] + plots[[119]] + plots[[45]] + plots[[14]] + plots[[178]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 1. Overview of CH-related hotspots. \n",
                          "(a) Histogram of frequencies by tissue type for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(b) A volcano plot was generated to illustrate which hotspot mutations were more frequently detected in either \n",
                          "the FoundationOne CDx (solid tumor panel) or the FoundationOne Liquid CDx. \n",
                          "The x-axis represents the log2 odds ratio of detection frequency, \n",
                          "while the y-axis represents the -log10 P-value. \n",
                          "Mutations were color-coded based on their type, \n",
                          "and the shape was differentiated depending on whether the mutation belonged to one of the \n",
                          "four major CH-related gene categories; epigenetic regulators (DNMT3A, TET2, ASXL1, IDH2), DNA repair genes \n",
                          "(TP53, CHEK2, ATM, RAD21, MDM4), MAPK (PTPN11, KRAS, CBL), and splicing factors (SF3B1, U2AF1). \n",
                          "(c) Histogram showing the number of CH-related hotspot variants detected per patient. \n",
                          "(d) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(e) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne CDx test."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_1.pdf", plot = layout_with_legend, width = 18, height = 18)
      ggsave("source/figure_1__.pdf", plot = layout_with_legend, width = 16, height = 18)
      ggsave("source/figure_1___.pdf", plot = layout_with_legend, width = 14, height = 18)
      ggsave("source/figure_1____.pdf", plot = layout_with_legend, width = 20, height = 18)
      ggsave("source/figure_1_____.pdf", plot = layout_with_legend, width = 22, height = 18)
      design <- "
      AAACCC
      BBDDEE
      "
      # レイアウトに図を配置
      layout <- plots[[1]] + plots[[119]] + plots[[365]] + plots[[14]] + plots[[178]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 1. Overview of CH-related hotspots. \n",
                          "(a) Histogram of frequencies by tissue type for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(b) A volcano plot was generated to illustrate which hotspot mutations were more frequently detected in either \n",
                          "the FoundationOne CDx (solid tumor panel) or the FoundationOne Liquid CDx. \n",
                          "The x-axis represents the log2 odds ratio of detection frequency, \n",
                          "while the y-axis represents the -log10 P-value. \n",
                          "Mutations were color-coded based on their type, \n",
                          "and the shape was differentiated depending on whether the mutation belonged to one of the \n",
                          "four major CH-related gene categories; epigenetic regulators (DNMT3A, TET2, ASXL1, IDH2), DNA repair genes \n",
                          "(TP53, CHEK2, ATM, RAD21, MDM4), MAPK (PTPN11, KRAS, CBL), and splicing factors (SF3B1, U2AF1). \n",
                          "(c) Histogram showing the number of CH-related hotspot variants detected per patient. \n",
                          "(d) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(e) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne CDx test."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_1_all_genes.pdf", plot = layout_with_legend, width = 18, height = 18)
      design <- "
      AAACCC
      BBDDEE
      "
      # レイアウトに図を配置
      layout <- plots[[1]] + plots[[119]] + plots[[369]] + plots[[14]] + plots[[178]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 1. Overview of CH-related hotspots. \n",
                          "(a) Histogram of frequencies by tissue type for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(b) A volcano plot was generated to illustrate which hotspot mutations were more frequently detected in either \n",
                          "the FoundationOne CDx (solid tumor panel) or the FoundationOne Liquid CDx. \n",
                          "The x-axis represents the log2 odds ratio of detection frequency, \n",
                          "while the y-axis represents the -log10 P-value. \n",
                          "Mutations were color-coded based on their type, \n",
                          "and the shape was differentiated depending on whether the mutation belonged to one of the \n",
                          "four major CH-related gene categories; epigenetic regulators (DNMT3A, TET2, ASXL1, IDH2), DNA repair genes \n",
                          "(TP53, CHEK2, ATM, RAD21, MDM4), MAPK (PTPN11, KRAS, CBL), and splicing factors (SF3B1, U2AF1). \n",
                          "(c) Histogram showing the number of CH-related hotspot variants detected per patient. \n",
                          "(d) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(e) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne CDx test."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_1_non_CH_genes.pdf", plot = layout_with_legend, width = 18, height = 18)
      design <- "
      AAACCC
      BBDDEE
      "
      # レイアウトに図を配置
      layout <- plots[[1]] + plots[[119]] + plots[[368]] + plots[[14]] + plots[[178]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 1. Overview of CH-related hotspots. \n",
                          "(a) Histogram of frequencies by tissue type for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(b) A volcano plot was generated to illustrate which hotspot mutations were more frequently detected in either \n",
                          "the FoundationOne CDx (solid tumor panel) or the FoundationOne Liquid CDx. \n",
                          "The x-axis represents the log2 odds ratio of detection frequency, \n",
                          "while the y-axis represents the -log10 P-value. \n",
                          "Mutations were color-coded based on their type, \n",
                          "and the shape was differentiated depending on whether the mutation belonged to one of the \n",
                          "four major CH-related gene categories; epigenetic regulators (DNMT3A, TET2, ASXL1, IDH2), DNA repair genes \n",
                          "(TP53, CHEK2, ATM, RAD21, MDM4), MAPK (PTPN11, KRAS, CBL), and splicing factors (SF3B1, U2AF1). \n",
                          "(c) Histogram showing the number of CH-related hotspot variants detected per patient. \n",
                          "(d) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(e) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne CDx test."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_1_non_all_CH_genes.pdf", plot = layout_with_legend, width = 18, height = 18)
      design <- "
      AAACCC
      BBDDEE
      "
      # レイアウトに図を配置
      layout <- plots[[1]] + plots[[119]] + plots[[370]] + plots[[14]] + plots[[178]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 1. Overview of CH-related hotspots. \n",
                          "(a) Histogram of frequencies by tissue type for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(b) A volcano plot was generated to illustrate which hotspot mutations were more frequently detected in either \n",
                          "the FoundationOne CDx (solid tumor panel) or the FoundationOne Liquid CDx. \n",
                          "The x-axis represents the log2 odds ratio of detection frequency, \n",
                          "while the y-axis represents the -log10 P-value. \n",
                          "Mutations were color-coded based on their type, \n",
                          "and the shape was differentiated depending on whether the mutation belonged to one of the \n",
                          "four major CH-related gene categories; epigenetic regulators (DNMT3A, TET2, ASXL1, IDH2), DNA repair genes \n",
                          "(TP53, CHEK2, ATM, RAD21, MDM4), MAPK (PTPN11, KRAS, CBL), and splicing factors (SF3B1, U2AF1). \n",
                          "(c) Histogram showing the number of CH-related hotspot variants detected per patient. \n",
                          "(d) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne Liquid CDx test. \n",
                          "(e) Heatmap of hotspot mutation frequencies across tissue types and CH-related genes for patients who underwent the FoundationOne CDx test."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_1_CH_genes_in_all_genes.pdf", plot = layout_with_legend, width = 18, height = 18)
      design <- "
      AAABBB
      CCDDEE
      FFFFFF
      FFFFFF
      GGHHII
      "
      # レイアウトに図を配置
      layout <- plots[[96]] + plots[[253]] + plots[[94]] + plots[[98]] + plots[[99]] + plots[[100]] + plots[[27]] + plots[[28]] + plots[[176]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 2. Association between gene type and age. \n",
                          "(a) Histogram of the frequency of TOCH or LOCH variants per gene in F1-liquid. \n",
                          "(b) Histogram of the frequency of TICH variants per gene in F1-solid. \n",
                          "(c) Moving average curves of age and frequency of individuals with CH-related hotspot variants, all genes. \n",
                          "(d) Moving average curves of age and frequency of individuals with CH-related hotspot variants by gene type. \n",
                          "(e) Moving average curves of age and VAF by gene type. Bars, 95% CI.\n",
                          "(f) The association between clinical information and the detection of TOCH or LOCH in each CH-related \n",
                          "gene mutation was evaluated using a logistic regression model adjusted for age and gender. Bars, 95% CI. \n",
                          "(g, h) CH-related gene mutations from blood cells \n", 
                          "contaminating tumors could be detected in the solid tumor panel. However, \n", 
                          "mutations detected only in the Liquid panel tended to have lower VAFs, \n", 
                          "suggesting that they might not have been detected due to lower sensitivity \n", 
                          "in the solid tumor panel. A possible explanation for this is that these \n", 
                          "mutations may have arisen as second hits. (i) Box plot of VAF by mutation type. ***Wilcoxon test, p_value < 0.001.",
                          "Q-values < 0.05, < 0.01, and < 0.001 were denoted by *, **, and ***, respectively.\n",
                          "The genes analyzed included epigenetic regulators (DNMT3A, TET2, ASXL1, IDH2), \n",
                          "DNA repair genes (TP53, CHEK2, ATM, RAD21, MDM4), MAPK (PTPN11, KRAS, CBL), and splicing factors (SF3B1, U2AF1)."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_2.pdf", plot = layout_with_legend, width = 18, height = 26)
      ggsave("source/figure_2_.pdf", plot = layout_with_legend, width = 24, height = 34)
      ggsave("source/figure_2__.pdf", plot = layout_with_legend, width = 14, height = 26)
      ggsave("source/figure_2___.pdf", plot = layout_with_legend, width = 20, height = 34)
      ggsave("source/figure_2____.pdf", plot = layout_with_legend, width = 11, height = 26)
      ggsave("source/figure_2_____.pdf", plot = layout_with_legend, width = 16, height = 34)
      ggsave("source/figure_2______.pdf", plot = layout_with_legend, width = 22, height = 26)
      ggsave("source/figure_2_______.pdf", plot = layout_with_legend, width = 30, height = 34)
      ggsave("source/figure_2________.pdf", plot = layout_with_legend, width = 24, height = 26)
      design <- "
      AAABBB
      AAABBB
      AAABBB
      CCDDEE
      CCDDEE
      FFGGHH
      FFGGHH
      "
      # レイアウトに図を配置
      layout <- plots[[16]] + plots[[17]] + plots[[131]] + plots[[132]] +plots[[133]] + 
                plots[[134]] + plots[[135]] + plots[[136]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 3_old. Scatter plots showing the relationship between hotspot mutation VAF and age. \n",
                          "(a) Hotspot mutations in CH-related genes detected by FoundationOne Liquid CDx for the \n",
                          "four major gene categories: epigenetic regulators (DNMT3A, TET2, ASXL1, IDH2), \n",
                          "DNA repair genes (TP53, CHEK2, ATM, RAD21, MDM4), MAPK (PTPN11, KRAS, CBL), and splicing factors (SF3B1, U2AF1). \n",
                          "(b) Hotspot mutations in CH-related genes other than these three major categories, detected by FoundationOne Liquid CDx. \n",
                          "(c–e) Hotspot mutations in epigenetic regulators, DNA repair genes, and splicing factors, respectively, as detected by FoundationOne Liquid CDx. \n",
                          "(f–h) Hotspot mutations in epigenetic regulators, DNA repair genes, and splicing factors, respectively, as detected by FoundationOne CDx (solid tumor panel)."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_3_old.pdf", plot = layout_with_legend, width = 18, height = 24)
      design <- "
      AB
      CD
      EE
      "
      # レイアウトに図を配置
      layout <- plots[[4]] + plots[[5]] + plots[[6]] + plots[[7]] + plots[[364]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                        "Figure S1. Scatter plot of VAF and age for variants at any hotspots. \n",
                        "(a) CH-related genes after removing germline variants and errors. \n",
                        "(b) Non-CH-related genes after removing germline variants and errors. \n",
                        "(c) Filtered germline variants and errors in CH-related genes. \n",
                        "(d) Filtered germline variants and errors in non-CH-related genes. \n",
                        "(e) Manually excluded suspected germline variants."))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S1.pdf", plot = layout_with_legend, width = 16, height = 25)
      design <- "
      AAAA
      AAAA
      AAAA
      ####
      BBBB
      BBBB
      BBBB
      ####
      CCCC
      CCCC
      CCCC
      ####
      DDDD
      DDDD
      DDDD
      ####
      EEEE
      EEEE
      EEEE
      "
      # レイアウトに図を配置
      layout <- plots[[137]] + plots[[138]] +plots[[139]] + 
        plots[[140]] + plots[[141]] + plots[[146]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S3_. Mutational signature analysis stratified by the presence of DNMT3A variants, specifically the R882 and Y735C variants \n",
                          "R882 and Y735C variants were excluded from the analysis. \n",
                          "(a) Cases with co-occurring R882 and Y735C variants \n",
                          "(b) Cases with Y735C variants \n",
                          "(c) Cases with R882 variants \n",
                          "(d) Cases with other DNMT3A variants \n",
                          "(e) Cases without DNMT3A variants "))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S3_.pdf", plot = layout_with_legend, width = 18, height = 24)
      design <- "
      AA#BBB
      AA#BBB
      ###BBB
      ######
      CCCCCC
      CCCCCC
      CCCCCC
      ######
      DDEE##
      DDEE##
      FFGG##
      FFGG##
      "
      # レイアウトに図を配置
      layout <-  plots[[155]] +plots[[153]] + plots[[154]] +plots[[149]] +plots[[150]] + 
        plots[[151]] + plots[[152]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 3. A model for predicting whether hotspot mutations in CH-related genes represent true CH-related alterations. \n",
                          "(a) ROC curve for the nomogram. \n",
                          "(b,c) Calibration plot and nomogram. The nomogram was generated using the lrm function from the rms package (R) with a penalty setting of 1. Nagelkerke R² was computed using the blorr package, and possible sampling bias was corrected using 500 bootstrap resamples before estimating the concordance index. \n",
                          "(d) ROC curve for the LightGBM model. \n",
                          "(e) Feature importance for the LightGBM model. \n",
                          "(f) ROC curve for the random forest model. \n",
                          "(g) Feature importance for the random forest model."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_3_new.pdf", plot = layout_with_legend, width = 14, height = 29)
      ggsave("source/figure_3_new_.pdf", plot = layout_with_legend, width = 18, height = 36)
      layout <-  plots[[366]] +plots[[153]] + plots[[154]] +plots[[149]] +plots[[150]] + 
        plots[[151]] + plots[[152]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 3. A model for predicting whether hotspot mutations in CH-related genes represent true CH-related alterations. \n",
                          "(a) ROC curve for the nomogram. \n",
                          "(b,c) Calibration plot and nomogram. The nomogram was generated using the lrm function from the rms package (R) with a penalty setting of 1. Nagelkerke R² was computed using the blorr package, and possible sampling bias was corrected using 500 bootstrap resamples before estimating the concordance index. \n",
                          "(d) ROC curve for the LightGBM model. \n",
                          "(e) Feature importance for the LightGBM model. \n",
                          "(f) ROC curve for the random forest model. \n",
                          "(g) Feature importance for the random forest model."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_3_new__.pdf", plot = layout_with_legend, width = 14, height = 29)
      ggsave("source/figure_3_new___.pdf", plot = layout_with_legend, width = 18, height = 36)
      design <- "
      AB
      CD
      EF
      "
      # レイアウトに図を配置
      layout <-  plots[[156]] +plots[[157]] + plots[[158]] +plots[[159]] + plots[[160]] +plots[[161]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S6. Co-mutation tendency in C-CAT FoundationOne cohort. \n",
                          "(a) IDH2 R140/R172. \n",
                          "(b) JAK2 V617F. \n",
                          "(c) TET2. \n",
                          "(d) CBL. \n",
                          "(e) DNMT3A. \n",
                          "(f) KRAS."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S6.pdf", plot = layout_with_legend, width = 18, height = 28)
      design <- "
      AB
      CD
      EF
      "
      # レイアウトに図を配置
      layout <-  plots[[162]] +plots[[163]] + plots[[165]] +plots[[164]] + plots[[166]] +plots[[167]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S7. Co-mutation tendency in GENIE blood tumor cohort. \n",
                          "(a) IDH2 R140/R172. \n",
                          "(b) JAK2 V617F. \n",
                          "(c) TET2. \n",
                          "(d) CBL. \n",
                          "(e) DNMT3A. \n",
                          "(f) KRAS."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S7.pdf", plot = layout_with_legend, width = 18, height = 28)
      design <- "
      AB
      "
      # レイアウトに図を配置
      layout <-  plots[[115]] + plots[[118]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S8. Variant frequency in CH-related genes in C-CAT cohort. \n",
                          "(a) Heatmap of the frequencies of non-germline mutations detected across tissue types and CH-related genes for patients who underwent the FoundationOne liquid CDx test. \n",
                          "(b) Heatmap of the frequencies of non-germline mutations detected across tissue types and CH-related genes for patients who underwent the FoundationOne CDx test."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S8.pdf", plot = layout_with_legend, width = 18, height = 10)
      # レイアウトに図を配置
      design <- "
      AB
      CC
      CC
      "
      layout <- plots[[187]] + plots[[186]] + plots[[206]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S10. Overall survival after CGP tesing in C-CAT cohort. \n",
                          "Only patients with overall-survival data were included. \n",
                          "(a) Patients who underwent the FoundationOne liquid CDx test. \n",
                          "(b) Patients who underwent the FoundationOne CDx test. \n",
                          "(c) Hazard ratio and 95% confidence intervals estimated with multivariable Cox model for the risk of death. \n",
                          "TICH, Tumor-infiltrative CH. TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S10.pdf", plot = layout_with_legend, width = 18, height = 28)
      design <- "
      AB
      CC
      CC
      "
      # レイアウトに図を配置
      layout <-  plots[[198]] + plots[[212]] + plots[[179]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S11. Histology and mutation type. \n",
                          "(a) Patients with LOCH had more CH variants than those with TOCH only.\n",
                          "(b) TMB and CH status in patients who underwent F1-liquid. \n",
                          "(c) Rate of TICH/TOCH/LOCH were different among cancer types. \n",
                          "TICH, Tumor-infiltrative CH. TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S11.pdf", plot = layout_with_legend, width = 18, height = 28)
      design <- "
      AB
      CC
      CC
      "
      # レイアウトに図を配置
      layout <-  plots[[210]] + plots[[209]] + plots[[211]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S12. Overall survival after the second line treatment in C-CAT cohort. \n",
                          "Only patients who underwent CGP tests just after the second line treatment with overall-survival data were included. \n",
                          "(a) Patients who underwent the FoundationOne liquid CDx test. \n",
                          "(b) Patients who underwent the FoundationOne CDx test. \n",
                          "(c) Hazard ratio and 95% confidence intervals estimated with multivariable Cox model for the risk of death. \n",
                          "TICH, Tumor-infiltrative CH. TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S12.pdf", plot = layout_with_legend, width = 18, height = 28)
      design <- "
      ABC
      DEF
      "
      # レイアウトに図を配置
      layout <-  plots[[216]] + plots[[217]] + plots[[226]] +  plots[[248]] + plots[[249]] + plots[[147]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S13. Liquid-only CH variants were not frequently observed in patients with many TOCH variants. \n",
                          "(a) Patients were grouped by the CH variant counts, and stacked graphs were plotted by the count of LOCH and TOCH variants. \n",
                          "(b) Mean LOCH counts. \n",
                          "(c) Proportional graphs. \n",
                          "(d) Patients were grouped by DNMT3A Y735 variant, and stacked graphs were plotted by the count of LOCH and TOCH variants. \n",
                          "(e) Mean LOCH counts. \n",
                          "(f) Proportional graphs. \n",
                          "TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S13.pdf", plot = layout_with_legend, width = 27, height = 20)
      design <- "
      ABC
      DEF
      GHI
      JKL
      MNO
      "
      # レイアウトに図を配置
      layout <-  plots[[218]] + plots[[219]] + plots[[227]] + plots[[220]] + plots[[221]] +  plots[[228]] + 
        plots[[222]] + plots[[223]] + plots[[229]] + plots[[224]] + plots[[225]] + plots[[230]] + plots[[231]] + plots[[232]] + plots[[233]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S14. Liquid-only CH variants were frequently observed in patients with many CH variants. \n",
                          "(a) Patients were grouped by the DNMT3A CH variant counts, and stacked graphs were plotted by the count of LOCH and TOCH variants. \n",
                          "(b) Mean LOCH counts. \n",
                          "(c) Proportional graphs. \n",
                          "(d, e, f) CHEK2. \n",
                          "(g, h, i) TET2. \n",
                          "(j, k, l) ASXL1. \n",
                          "(m, n, o) ATM. \n",
                          "TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S14.pdf", plot = layout_with_legend, width = 27, height = 40)
      design <- "
      ABC
      DEF
      GHI
      JKL
      MNO
      "
      # レイアウトに図を配置
      layout <-  plots[[234]] + plots[[235]] + plots[[236]] + plots[[237]] + plots[[238]] +  plots[[239]] + 
        plots[[240]] + plots[[241]] + plots[[242]] + plots[[243]] + plots[[244]] + plots[[245]] + plots[[246]] + plots[[247]] + plots[[248]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S15. Liquid-only CH variants were frequently observed in patients with many CH variants. \n",
                          "Each values were derived from each target gene. \n",
                          "(a) Patients were grouped by the DNMT3A CH variant counts, and stacked graphs were plotted by the count of LOCH and TOCH variants. \n",
                          "(b) Mean LOCH counts. \n",
                          "(c) Proportional graphs. \n",
                          "(d, e, f) CHEK2. \n",
                          "(g, h, i) TET2. \n",
                          "(j, k, l) ASXL1. \n",
                          "(m, n, o) ATM. \n",
                          "TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S15.pdf", plot = layout_with_legend, width = 27, height = 40)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[169]] + plots[[168]] + plots[[170]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S16. ICI (Nivo/Pembro) effect and TICH in F1-solid data. \n",
                          "(a) TICH in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S16.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[255]] + plots[[254]] + plots[[256]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S17. ICI (Nivo/Pembro) effect and CH in F1-liquid data. \n",
                          "(a) CH in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S17.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[258]] + plots[[257]] + plots[[259]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S18. ICI (Nivo/Pembro) effect and DDR TICH in F1-solid data. \n",
                          "(a) DDR TICH in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S18.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[261]] + plots[[260]] + plots[[262]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S19. ICI (Nivo/Pembro) effect and DDR CH in F1-liquid data. \n",
                          "(a) DDR CH in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S19.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[264]] + plots[[263]] + plots[[265]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S20. ICI (Nivo/Pembro) effect and DDR SEV in F1-solid data. \n",
                          "(a) DDR SEV in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S20.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[267]] + plots[[266]] + plots[[268]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S21. ICI (Nivo/Pembro) effect and DDR SEV in F1-liquid data. \n",
                          "(a) DDR SEV in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S21.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[270]] + plots[[269]] + plots[[271]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S22. ICI (Nivo/Pembro) effect and Epigenetic SEV in F1-solid data. \n",
                          "(a) Epigenetic SEV in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S22.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[273]] + plots[[272]] + plots[[274]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S23. ICI (Nivo/Pembro) effect and Epigenetic SEV in F1-liquid data. \n",
                          "(a) Epigenetic SEV in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S23.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[276]] + plots[[275]] + plots[[277]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S24. ICI (Nivo/Pembro) effect and Epigenetic TICH in F1-solid data. \n",
                          "(a) Epigenetic TICH in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S24.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A#
      BC
      BC
      "
      # レイアウトに図を配置
      layout <- plots[[279]] + plots[[278]] + plots[[280]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S25. ICI (Nivo/Pembro) effect and Epigenetic CH in F1-liquid data. \n",
                          "(a) Epigenetic CH in any genes and treatment on time with ICI. \n",
                          "(b) Objective response rate and various factors. \n",
                          "(c) Hazard rate of time on treatment of ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S25.pdf", plot = layout_with_legend, width = 18, height = 20)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[358]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S26. ICI (Nivo/Pembro) effect and TICH in F1-solid data. \n",
                          "(a) TICH in any genes and time on treatment with ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S26.pdf", plot = layout_with_legend, width = 8, height = 8)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[357]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 6. ICI (Nivo/Pembro) effect and TICH in F1-solid data. \n",
                          "(a) TICH in any genes and time on treatment with ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_6_new.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[371]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 6. ICI (Nivo/Pembro) effect and TICH in F1-solid data. \n",
                          "(a) TICH in any genes and time on treatment with ICI, TMB-low. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_6_new_1.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[372]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 6. ICI (Nivo/Pembro) effect and TICH in F1-solid data. \n",
                          "(a) TICH in any genes and time on treatment with ICI, TMB-high. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_6_new_2.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[359]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S4. ICI (Nivo/Pembro) effect and DDR TICH in F1-solid data. \n",
                          "(a) TICH in DDR genes and time on treatment with ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S4_new.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[360]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S5. ICI (Nivo/Pembro) effect and LD-CH/LE-CH in F1-liquid data. \n",
                          "(a) TOCH or LOCH in any genes and time on treatment with ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S5_new.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[361]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S6. ICI (Nivo/Pembro) effect and DDR TOCH/LOCH in F1-liquid data. \n",
                          "(a) TOCH or LOCH in DDR genes and time on treatment with ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S6_new.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[362]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S7. ICI (Nivo/Pembro) effect and DDR SEVs in F1 CDx data. \n",
                          "(a) SEVs in DDR genes and time on treatment with ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S7_new.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      A
      "
      # レイアウトに図を配置
      layout <- plots[[363]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S8. ICI (Nivo/Pembro) effect and DDR SEVs in F1-liquid data. \n",
                          "(a) SEVs in DDR genes and time on treatment with ICI. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S8_new.pdf", plot = layout_with_legend, width = 8, height = 12)
      design <- "
      AB
      CD
      EF
      "
      # レイアウトに図を配置
      layout <- plots[[169]] + plots[[258]] + plots[[264]] + plots[[255]] + plots[[261]] + plots[[267]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 5. ICI (Nivo/Pembro) effect and variants. \n",
                          "(a) TICH in any genes and treatment on time with ICI, FoundationOne CDx. \n",
                          "(b) TICH in DDR genes and treatment on time with ICI, FoundationOne CDx. \n",
                          "(c) SEVs in DDR genes and treatment on time with ICI, FoundationOne CDx. \n",
                          "(d) TOCH/LOCH in any genes and treatment on time with ICI, FoundationOne Liquid CDx. \n",
                          "(e) TOCH/LOCH in DDR genes and treatment on time with ICI, FoundationOne Liquid CDx. \n",
                          "(f) SEVs in DDR genes and treatment on time with ICI, FoundationOne Liquid CDx. \n",
                          "TICH, Tumor-infiltrating CH, TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_5_new.pdf", plot = layout_with_legend, width = 18, height = 20)
      # Data_MAF %>% dplyr::filter(Panel == "FoundationOne CDx")
      # Data_MAF %>% dplyr::filter(Panel == "F1Liquid CDx")
      # Data_MAF_target %>% dplyr::filter((No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) & non_germline_non_error == 1 & Panel == "F1Liquid CDx")
      # Data_MAF_target %>% dplyr::filter((No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1) & non_germline_non_error == 1 & Panel == "F1Liquid CDx") %>% dplyr::distinct(unique_alt)
      # Data_MAF_target %>% dplyr::filter(Panel == "F1Liquid CDx" & non_germline_non_error == 1 & Hugo_Symbol %in% CH_GENE_NatComms & (No_found_in_F1L >= Min_count | No_found_in_F1 >= Min_count_F1))
      # Data_MAF_target_table_2 %>% dplyr::distinct(unique_alt)
      # write_excel_csv(Data_MAF_target_table_2 %>% dplyr::distinct(unique_alt, .keep_all = T), "source/TOCHLOCHSelection.csv")
      # Data_MAF_target %>% dplyr::filter(non_germline_non_error == 1) %>% dplyr::filter(unique_alt %in% CH_candidates & Hugo_Symbol %in% CH_gene_entropy_3 & Panel == "F1Liquid CDx")
      # Data_MAF_target %>% dplyr::filter(non_germline_non_error == 1) %>% dplyr::filter(unique_alt %in% CH_candidates & Hugo_Symbol %in% CH_gene_entropy_3 & Panel == "F1Liquid CDx") %>% dplyr::distinct(Hugo_Symbol)
      # Data_MAF_target %>% dplyr::filter(non_germline_non_error == 1) %>% dplyr::filter(unique_alt %in% CH_candidates & Hugo_Symbol %in% CH_gene_entropy_3 & Panel == "F1Liquid CDx") %>% dplyr::distinct(unique_alt)
      # Data_MAF_target %>% dplyr::filter(non_germline_non_error == 1) %>% dplyr::filter(unique_alt %in% CH_candidates & Hugo_Symbol %in% CH_gene_entropy_3 & Panel == "F1Liquid CDx") %>% dplyr::distinct(Tumor_Sample_Barcode)
      # Data_MAF_target %>% dplyr::filter(non_germline_non_error == 1) %>% dplyr::filter(unique_alt %in% CH_candidates & Hugo_Symbol %in% CH_gene_entropy_3 & Panel == "FoundationOne CDx") %>% dplyr::distinct(Tumor_Sample_Barcode)
      Data_MAF_target_ = Data_MAF_target
      id_map <- data.frame(
        Tumor_Sample_Barcode_original = unique(Data_MAF_target_$Tumor_Sample_Barcode),
        Tumor_Sample_Barcode_anon = sprintf("patient_%05d", seq_along(unique(Data_MAF_target_$Tumor_Sample_Barcode)))
      )
      
      # 匿名化を適用
      Data_MAF_target_$Tumor_Sample_Barcode <- id_map$Tumor_Sample_Barcode_anon[
        match(Data_MAF_target_$Tumor_Sample_Barcode, id_map$Tumor_Sample_Barcode_original)
      ]
      data_tmp_uni_detail$Tumor_Sample_Barcode <- id_map$Tumor_Sample_Barcode_anon[
        match(data_tmp_uni_detail$Tumor_Sample_Barcode, id_map$Tumor_Sample_Barcode_original)
      ]
      Data_tmp = Data_MAF_target_ %>%
        dplyr::filter(non_germline_non_error == 1) %>%
        dplyr::filter(unique_alt %in% CH_candidates &
                        Hugo_Symbol %in% CH_gene_entropy_3 &
                        Panel == "FoundationOne CDx")
      Data_tmp = Data_tmp %>%
        left_join(F1L_germline_error, by="Hugo_Symbol") %>%
        dplyr::filter(max_VAF >= 0.02 | is.na(max_VAF))
      table_save_tmp_G = Data_tmp %>% dplyr::arrange(SNV_check) %>%
        dplyr::arrange(Chromosome, Start_Position)
      write_excel_csv(table_save_tmp_G, "source/Table_S5_F1_TICH_1.csv")
      table_save_tmp_H = Data_tmp %>% dplyr::arrange(SNV_check) %>%
        dplyr::distinct(unique_alt, .keep_all = T) %>%
        dplyr::arrange(Chromosome, Start_Position)
      write_excel_csv(table_save_tmp_H, "source/Table_S5_F1_TICH_2.csv")

      write_excel_csv(Data_MAF_target_ %>% dplyr::filter(non_germline_non_error == 1) %>% dplyr::filter(unique_alt %in% CH_candidates & Hugo_Symbol %in% CH_gene_entropy_3 & Panel == "F1Liquid CDx"), "source/TOCHLOCH_individual_variants.csv")
      write_excel_csv(data_tmp_uni_detail, "source/hotspot_variants_somatic_or_CH.csv")
      write_excel_csv(id_map, "source/id_map.csv")
      
      design <- "
      AB
      CC
      CC
      "
      # レイアウトに図を配置
      layout <- plots[[175]] + plots[[174]] + plots[[188]] + 
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S9. Overall survival after the first line treatment in C-CAT cohort. \n",
                          "Only patients who underwent CGP tests just after the first line treatment with overall-survival data were included. \n",
                          "(a) Patients who underwent the FoundationOne liquid CDx test. \n",
                          "(b) Patients who underwent the FoundationOne CDx test. \n",
                          "(c) Hazard ratio and 95% confidence intervals estimated with multivariable Cox model for the risk of death. \n",
                          "TICH, Tumor-infiltrative CH. TOCH, Tumor-observed CH, LOCH, Liquid-only CH."
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S9.pdf", plot = layout_with_legend, width = 18, height = 28)
      design <- "
      AAABBB
      AAABBB
      AAABBB
      ######
      CCCCCC
      CCCCCC
      CCCCCC
      ######
      DDDDDD
      DDDDDD
      DDDDDD
      ######
      EEEEEE
      EEEEEE
      EEEEEE
      ######
      FFFFFF
      FFFFFF
      FFFFFF
      ######
      GGGHHH
      GGGHHH
      GGGHHH
      GGGHHH
      GGGHHH
      GGGHHH
      "
      # レイアウトに図を配置
      layout <- plots[[3]] + plots[[125]] + plots[[142]] + plots[[143]] + plots[[144]] + 
        plots[[145]] + plots[[44]] + plots[[113]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure 4. Characteristics of mutations frequently detected in the Liquid panel. \n",  
                          "(a) Lolliplot showing the frequency of DNMT3A mutations in the \n",  
                          "FoundationOne Liquid CDx panel. The Y735C mutation was highly frequent. \n",  
                          "(b) Frequency of DNMT3A mutations in the FoundationOne CDx solid tumor panel, \n",  
                          "where the Y735C mutation was less frequent. \n",  
                          "(c, d) Mutational signature analysis comparing cases with and without the \n",  
                          "DNMT3A Y735C mutation. Cases with the Y735C mutation exhibited a higher \n",  
                          "frequency of A[C>T]C mutations. \n",  
                          "(e, f) Mutations detected only in the Liquid panel showed a \n",  
                          "higher prevalence of A[C>T]C and lower A[C>T]G mutations. \n",  
                          "(g) Histogram showing the distribution of odds ratios for each CH-related gene \n",  
                          "hotspot in the Liquid panel and solid tumor panel. Mutations detected only in \n",  
                          "the Liquid panel showed a remarkably low frequency of C>T mutations at CpG sites. \n",  
                          "(h) Volcano plot illustrating the relationship between +1G>A mutations at splice \n",  
                          "donor sites and CH-related gene hotspot mutations. The presence of the Y735C \n",  
                          "mutation was correlated with splice site mutations."  ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_4_new.pdf", plot = layout_with_legend, width = 18, height = 40)
      design <- "
      AAA
      AAA
      AAA
      AAA
      BBB
      BBB
      BBB 
      BBB 
      CCC
      CCC
      CCC
      CCC
      "
      # レイアウトに図を配置
      layout <- plots[[126]] + plots[[124]] + plots[[43]]  +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S2-1. Lolliplot showing the distribution of CH-related hotspot variants in DNMT3A detected in the AACR Project GENIE cohort. \n",
                          " DNMT3A Y735C variant was relatively frequently detected in the liquid panel but showed a decreased frequency in the solid tumor panel and hematologic malignancy panels. \n", 
                          "(a) CH-related gene variants detected using FoundationOne Liquid CDx across various solid tumors. \n",
                          "(b) CH-related gene variants detected using solid tumor panels across various solid tumors. \n",
                          "(c) CH-related gene variants detected in hematologic malignancies. \n",
                          "(d) Mutational signature of CH-related gene variants detected in hematologic malignancies. \n",
                          "(e)  A volcano plot was generated to illustrate which hotspot mutations were more frequently detected in either \n",
                          "the FoundationOne CDx (solid tumor panel) or the FoundationOne Liquid CDx. \n",
                          "The x-axis represents the log2 odds ratio of detection frequency, while the y-axis represents the -log10 P-value. \n",
                          "Mutations were color-coded based on their functional type. \n",
                          "(f, g) Evaluation of whether hotspot variants in DNMT3A were more \n", 
                          "frequently detected in the FoundationOne CDx solid tumor panel or the \n", 
                          "FoundationOne Liquid CDx panel, or in the FoundationOne Liquid CDx panel \n", 
                          "versus the AACR project GENIE hematologic malignancies dataset. \n.",
                          "Most hotspot variants were more frequently detected in the Liquid panel, suggesting their advantage in clonal \n", 
                          "haematopoiesis, while their importance in hematologic malignancy formation was considered lower" 
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S2_1_.pdf", plot = layout_with_legend, width = 24, height = 25)
      ggsave("source/figure_S2_1__.pdf", plot = layout_with_legend, width = 16, height = 16)
      design <- "
      ###
      AAA
      AAA
      AAA
      AAA
      ###
      BBC
      BBC
      BBC
      BBC
      BBD
      BBD
      BBD
      BBD
      "
      # レイアウトに図を配置
      layout <- plots[[146]] + plots[[9]] + plots[[64]] + plots[[53]] +
        plot_layout(design = design) +
        plot_annotation(tag_levels = 'a', 
                        caption = paste0(
                          "Figure S2-2. Lolliplot showing the distribution of CH-related hotspot variants in DNMT3A detected in the AACR Project GENIE cohort. \n",
                          " DNMT3A Y735C variant was relatively frequently detected in the liquid panel but showed a decreased frequency in the solid tumor panel and hematologic malignancy panels. \n", 
                          "(a) Mutational signature of CH-related gene variants detected in hematologic malignancies. \n",
                          "(b)  A volcano plot was generated to illustrate which hotspot mutations were more frequently detected in either \n",
                          "the FoundationOne CDx (solid tumor panel) or the FoundationOne Liquid CDx. \n",
                          "The x-axis represents the log2 odds ratio of detection frequency, while the y-axis represents the -log10 P-value. \n",
                          "Mutations were color-coded based on their functional type. \n",
                          "(c, d) Evaluation of whether hotspot variants in DNMT3A were more \n", 
                          "frequently detected in the FoundationOne CDx solid tumor panel or the \n", 
                          "FoundationOne Liquid CDx panel, or in the FoundationOne Liquid CDx panel \n", 
                          "versus the AACR project GENIE hematologic malignancies dataset. \n.",
                          "Most hotspot variants were more frequently detected in the Liquid panel, suggesting their advantage in clonal \n", 
                          "haematopoiesis, while their importance in hematologic malignancy formation was considered lower" 
                        ))
      layout_with_legend <- layout + plot_layout(guides = "collect") & theme(legend.position = "bottom")
      ggsave("source/figure_S2_2_.pdf", plot = layout_with_legend, width = 24, height = 25)
      ggsave("source/figure_S2_2__.pdf", plot = layout_with_legend, width = 16, height = 16)
    })
  })
}
# Run the application 
shinyApp(ui = ui, server = server)
